{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "J5DvgSColT9Y"
   },
   "source": [
    "# Plot drug resistance frequencies"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "s6E0OCU6laVT"
   },
   "source": [
    "## Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "QddUVloxB97i"
   },
   "source": [
    "This notebook creates a table summarising the number of samples with inferred antimalarial drug resistance (DR) in different populations using the data from *Plasmodium falciparum* version 8 (Pf8) project.\n",
    "\n",
    "**This notebook takes 1 minute to run.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "8aHguQMOliOG"
   },
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "fnmE5kdRlrf_"
   },
   "source": [
    "Install and import the malariagen Python package:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "k56UYmO2lkig",
    "outputId": "7a8e23f6-d6a3-47ef-8410-2796f746f969"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n",
      "  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n",
      "  Preparing metadata (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.0/4.0 MB\u001b[0m \u001b[31m19.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
      "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m71.7/71.7 kB\u001b[0m \u001b[31m5.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m775.9/775.9 kB\u001b[0m \u001b[31m42.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m25.9/25.9 MB\u001b[0m \u001b[31m66.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m8.7/8.7 MB\u001b[0m \u001b[31m94.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m210.6/210.6 kB\u001b[0m \u001b[31m15.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m6.3/6.3 MB\u001b[0m \u001b[31m94.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m81.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m7.8/7.8 MB\u001b[0m \u001b[31m82.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m78.3/78.3 kB\u001b[0m \u001b[31m5.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m101.7/101.7 kB\u001b[0m \u001b[31m7.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m8.9/8.9 MB\u001b[0m \u001b[31m91.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m228.0/228.0 kB\u001b[0m \u001b[31m16.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m13.4/13.4 MB\u001b[0m \u001b[31m89.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m60.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25h  Building wheel for malariagen_data (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
      "  Building wheel for dash-cytoscape (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
      "  Building wheel for asciitree (setup.py) ... \u001b[?25l\u001b[?25hdone\n"
     ]
    }
   ],
   "source": [
    "!pip install malariagen_data -q --no-warn-conflicts\n",
    "import malariagen_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "1JGWthtIlop-"
   },
   "source": [
    "Import required python libraries that are installed at colab by default."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "id": "laxWs51glmlv"
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import collections\n",
    "import re\n",
    "import scipy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "oL05c-_klv6I"
   },
   "source": [
    "## Access Pf8 Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "YlPcoczFl16k"
   },
   "source": [
    "We use the malariagen data package to load the release data.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 144
    },
    "id": "JBrrYb6glym5",
    "outputId": "cab9f148-1857-4269-de63-89eafa80bf64"
   },
   "outputs": [
    {
     "data": {
      "application/vnd.google.colaboratory.intrinsic+json": {
       "summary": "{\n  \"name\": \"sample_metadata\",\n  \"rows\": 33325,\n  \"fields\": [\n    {\n      \"column\": \"Sample\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 33325,\n        \"samples\": [\n          \"SPT50302\",\n          \"RCN08686\",\n          \"PD1008-C\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Study\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 99,\n        \"samples\": [\n          \"1108-PF-GAB-BOUYOU-AKOTET\",\n          \"1134-PF-ML-CONWAY\",\n          \"1248-PF-MW-MCCANN-SM\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Country\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 34,\n        \"samples\": [\n          \"Vietnam\",\n          \"Myanmar\",\n          \"Benin\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Admin level 1\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 122,\n        \"samples\": [\n          \"Central\",\n          \"Tanintharyi\",\n          \"Mandalay\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Country latitude\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 10.038786162655637,\n        \"min\": -19.37564815406632,\n        \"max\": 23.83676210254897,\n        \"num_unique_values\": 34,\n        \"samples\": [\n          16.69436455,\n          21.117142514264128,\n          9.654223250770558\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Country longitude\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 47.984959548013094,\n        \"min\": -86.61620049368882,\n        \"max\": 145.25400675267255,\n        \"num_unique_values\": 34,\n        \"samples\": [\n          106.55179596,\n          96.51020132013552,\n          2.339712982208921\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Admin level 1 latitude\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 10.380792971121629,\n        \"min\": -25.52759199095002,\n        \"max\": 26.07383704651308,\n        \"num_unique_values\": 123,\n        \"samples\": [\n          5.560964256901143,\n          12.661493915029125,\n          21.20922330626896\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Admin level 1 longitude\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 49.10306563588118,\n        \"min\": -87.18812587071439,\n        \"max\": 150.84596252,\n        \"num_unique_values\": 123,\n        \"samples\": [\n          -1.1972257865739595,\n          98.74518851348084,\n          96.00349430157443\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Year\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 5.250367741391281,\n        \"min\": 1966.0,\n        \"max\": 2022.0,\n        \"num_unique_values\": 39,\n        \"samples\": [\n          1967.0,\n          1970.0,\n          2013.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"ENA\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 33192,\n        \"samples\": [\n          \"ERR388777\",\n          \"ERR5839312\",\n          \"ERR10893230\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"All samples same case\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 31608,\n        \"samples\": [\n          \"SPT70928\",\n          \"SPT15573\",\n          \"SPT90741\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Population\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 10,\n        \"samples\": [\n          \"AF-C\",\n          \"AF-E\",\n          \"OC-NG\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"% callable\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 28.960883670035095,\n        \"min\": 0.0,\n        \"max\": 95.12,\n        \"num_unique_values\": 7049,\n        \"samples\": [\n          23.9,\n          78.7,\n          45.94\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"QC pass\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          false,\n          true\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Exclusion reason\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 6,\n        \"samples\": [\n          \"Analysis_set\",\n          \"Low_coverage\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Sample type\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 3,\n        \"samples\": [\n          \"gDNA\",\n          \"sWGA\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Sample was in Pf7\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          false,\n          true\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}",
       "type": "dataframe",
       "variable_name": "sample_metadata"
      },
      "text/html": [
       "\n",
       "  <div id=\"df-87348608-f7d3-4f00-9988-c2f489f94d87\" class=\"colab-df-container\">\n",
       "    <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Study</th>\n",
       "      <th>Country</th>\n",
       "      <th>Admin level 1</th>\n",
       "      <th>Country latitude</th>\n",
       "      <th>Country longitude</th>\n",
       "      <th>Admin level 1 latitude</th>\n",
       "      <th>Admin level 1 longitude</th>\n",
       "      <th>Year</th>\n",
       "      <th>ENA</th>\n",
       "      <th>All samples same case</th>\n",
       "      <th>Population</th>\n",
       "      <th>% callable</th>\n",
       "      <th>QC pass</th>\n",
       "      <th>Exclusion reason</th>\n",
       "      <th>Sample type</th>\n",
       "      <th>Sample was in Pf7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FP0008-C</td>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR1081237</td>\n",
       "      <td>FP0008-C</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>82.48</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FP0009-C</td>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR1081238</td>\n",
       "      <td>FP0009-C</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>88.95</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FP0010-CW</td>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR2889621</td>\n",
       "      <td>FP0010-CW</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>87.01</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    <div class=\"colab-df-buttons\">\n",
       "\n",
       "  <div class=\"colab-df-container\">\n",
       "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-87348608-f7d3-4f00-9988-c2f489f94d87')\"\n",
       "            title=\"Convert this dataframe to an interactive table.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
       "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "\n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    .colab-df-buttons div {\n",
       "      margin-bottom: 4px;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "    <script>\n",
       "      const buttonEl =\n",
       "        document.querySelector('#df-87348608-f7d3-4f00-9988-c2f489f94d87 button.colab-df-convert');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      async function convertToInteractive(key) {\n",
       "        const element = document.querySelector('#df-87348608-f7d3-4f00-9988-c2f489f94d87');\n",
       "        const dataTable =\n",
       "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                    [key], {});\n",
       "        if (!dataTable) return;\n",
       "\n",
       "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "          + ' to learn more about interactive tables.';\n",
       "        element.innerHTML = '';\n",
       "        dataTable['output_type'] = 'display_data';\n",
       "        await google.colab.output.renderOutput(dataTable, element);\n",
       "        const docLink = document.createElement('div');\n",
       "        docLink.innerHTML = docLinkHtml;\n",
       "        element.appendChild(docLink);\n",
       "      }\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "\n",
       "<div id=\"df-7cec3fe3-5757-4835-9e44-bb5c944d89fa\">\n",
       "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-7cec3fe3-5757-4835-9e44-bb5c944d89fa')\"\n",
       "            title=\"Suggest charts\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "     width=\"24px\">\n",
       "    <g>\n",
       "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
       "    </g>\n",
       "</svg>\n",
       "  </button>\n",
       "\n",
       "<style>\n",
       "  .colab-df-quickchart {\n",
       "      --bg-color: #E8F0FE;\n",
       "      --fill-color: #1967D2;\n",
       "      --hover-bg-color: #E2EBFA;\n",
       "      --hover-fill-color: #174EA6;\n",
       "      --disabled-fill-color: #AAA;\n",
       "      --disabled-bg-color: #DDD;\n",
       "  }\n",
       "\n",
       "  [theme=dark] .colab-df-quickchart {\n",
       "      --bg-color: #3B4455;\n",
       "      --fill-color: #D2E3FC;\n",
       "      --hover-bg-color: #434B5C;\n",
       "      --hover-fill-color: #FFFFFF;\n",
       "      --disabled-bg-color: #3B4455;\n",
       "      --disabled-fill-color: #666;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart {\n",
       "    background-color: var(--bg-color);\n",
       "    border: none;\n",
       "    border-radius: 50%;\n",
       "    cursor: pointer;\n",
       "    display: none;\n",
       "    fill: var(--fill-color);\n",
       "    height: 32px;\n",
       "    padding: 0;\n",
       "    width: 32px;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart:hover {\n",
       "    background-color: var(--hover-bg-color);\n",
       "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "    fill: var(--button-hover-fill-color);\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart-complete:disabled,\n",
       "  .colab-df-quickchart-complete:disabled:hover {\n",
       "    background-color: var(--disabled-bg-color);\n",
       "    fill: var(--disabled-fill-color);\n",
       "    box-shadow: none;\n",
       "  }\n",
       "\n",
       "  .colab-df-spinner {\n",
       "    border: 2px solid var(--fill-color);\n",
       "    border-color: transparent;\n",
       "    border-bottom-color: var(--fill-color);\n",
       "    animation:\n",
       "      spin 1s steps(1) infinite;\n",
       "  }\n",
       "\n",
       "  @keyframes spin {\n",
       "    0% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "      border-left-color: var(--fill-color);\n",
       "    }\n",
       "    20% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    30% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    40% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    60% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    80% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "    90% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "  }\n",
       "</style>\n",
       "\n",
       "  <script>\n",
       "    async function quickchart(key) {\n",
       "      const quickchartButtonEl =\n",
       "        document.querySelector('#' + key + ' button');\n",
       "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
       "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
       "      try {\n",
       "        const charts = await google.colab.kernel.invokeFunction(\n",
       "            'suggestCharts', [key], {});\n",
       "      } catch (error) {\n",
       "        console.error('Error during call to suggestCharts:', error);\n",
       "      }\n",
       "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
       "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
       "    }\n",
       "    (() => {\n",
       "      let quickchartButtonEl =\n",
       "        document.querySelector('#df-7cec3fe3-5757-4835-9e44-bb5c944d89fa button');\n",
       "      quickchartButtonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "    })();\n",
       "  </script>\n",
       "</div>\n",
       "\n",
       "    </div>\n",
       "  </div>\n"
      ],
      "text/plain": [
       "      Sample              Study     Country   Admin level 1  Country latitude  \\\n",
       "0   FP0008-C  1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
       "1   FP0009-C  1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
       "2  FP0010-CW  1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
       "\n",
       "   Country longitude  Admin level 1 latitude  Admin level 1 longitude    Year  \\\n",
       "0         -10.337093               16.565426                -9.832345  2014.0   \n",
       "1         -10.337093               16.565426                -9.832345  2014.0   \n",
       "2         -10.337093               16.565426                -9.832345  2014.0   \n",
       "\n",
       "          ENA All samples same case Population  % callable  QC pass  \\\n",
       "0  ERR1081237              FP0008-C       AF-W       82.48     True   \n",
       "1  ERR1081238              FP0009-C       AF-W       88.95     True   \n",
       "2  ERR2889621             FP0010-CW       AF-W       87.01     True   \n",
       "\n",
       "  Exclusion reason Sample type  Sample was in Pf7  \n",
       "0     Analysis_set        gDNA               True  \n",
       "1     Analysis_set        gDNA               True  \n",
       "2     Analysis_set        sWGA               True  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "release_data = malariagen_data.Pf8()\n",
    "sample_metadata = release_data.sample_metadata()\n",
    "\n",
    "# take a glance at the metadata dataframe\n",
    "sample_metadata.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "1fXWbVbZH13A"
   },
   "source": [
    "We will only examine the QC pass samples in the analysis of these notebooks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "id": "V_aeZe1hHqMD"
   },
   "outputs": [],
   "source": [
    "# Retain QC pass samples\n",
    "qc_sample_metadata = sample_metadata.loc[sample_metadata['QC pass']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "-Otz5oyEl8UG"
   },
   "source": [
    "## Access DR Classification and Genotype Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "4XwL9dSOBrfx"
   },
   "source": [
    "We can access inferred resistance status classifications of QC-pass Pf8 samples from the Sanger's cloud storage.\n",
    "\n",
    "This dataset includes the samples that are predicted to be resistant to 10 drugs or combinations of drugs and to rapid diagnostic tests (RDT) detection: chloroquine, pyrimethamine, sulfadoxine, mefloquine, artemisinin, piperaquine, sulfadoxine- pyrimethamine for treatment of uncomplicated malaria, sulfadoxine- pyrimethamine for intermittent preventive treatment in pregnancy, artesunate-mefloquine, dihydroartemisinin-piperaquine, *hrp2* and *hrp3* gene deletions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 206
    },
    "id": "MN_SE_R7l2Re",
    "outputId": "705f33f4-28ea-44dd-862e-3e50269e0a6f"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Chloroquine</th>\n",
       "      <th>Pyrimethamine</th>\n",
       "      <th>Sulfadoxine</th>\n",
       "      <th>Mefloquine</th>\n",
       "      <th>Artemisinin</th>\n",
       "      <th>Piperaquine</th>\n",
       "      <th>SP (uncomplicated)</th>\n",
       "      <th>SP (IPTp)</th>\n",
       "      <th>AS-MQ</th>\n",
       "      <th>DHA-PPQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FP0008-C</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FP0009-C</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FP0010-CW</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FP0011-CW</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FP0012-CW</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Sample   Chloroquine Pyrimethamine   Sulfadoxine Mefloquine Artemisinin  \\\n",
       "0   FP0008-C  Undetermined  Undetermined  Undetermined  Sensitive   Sensitive   \n",
       "1   FP0009-C     Resistant     Resistant     Sensitive  Sensitive   Sensitive   \n",
       "2  FP0010-CW  Undetermined     Resistant     Resistant  Sensitive   Sensitive   \n",
       "3  FP0011-CW  Undetermined     Resistant  Undetermined  Sensitive   Sensitive   \n",
       "4  FP0012-CW     Resistant     Resistant     Sensitive  Sensitive   Sensitive   \n",
       "\n",
       "  Piperaquine SP (uncomplicated)  SP (IPTp)      AS-MQ    DHA-PPQ  \n",
       "0   Sensitive          Sensitive  Sensitive  Sensitive  Sensitive  \n",
       "1   Sensitive          Resistant  Sensitive  Sensitive  Sensitive  \n",
       "2   Sensitive          Resistant  Sensitive  Sensitive  Sensitive  \n",
       "3   Sensitive          Resistant  Sensitive  Sensitive  Sensitive  \n",
       "4   Sensitive          Resistant  Sensitive  Sensitive  Sensitive  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the data\n",
    "resistance_classification_fn = pd.read_csv('https://pf8-release.cog.sanger.ac.uk/Pf8_inferred_resistance_status_classification.tsv', sep='\\t')\n",
    "\n",
    "# Rename the first column as 'Sample'\n",
    "resistance_classification_fn = resistance_classification_fn.rename(columns={resistance_classification_fn.columns[0]: 'Sample'})\n",
    "\n",
    "# Print the first rows\n",
    "resistance_classification_fn.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "TrwayUMQAD3G"
   },
   "source": [
    "We will also use the genotypes utilised in drug resistance status classification containing amino acid and copy number genotypes at six loci: *crt*, *dhfr*, *dhps*, *mdr1*, *kelch13*, *plasmepsin* *2-3*. This dataset is also available from the Sanger's cloud storage."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 288
    },
    "id": "tmBxiVvE1cra",
    "outputId": "4bad0791-7281-4845-8f17-3117dca00183"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>crt_72[C]</th>\n",
       "      <th>crt_74[M]</th>\n",
       "      <th>crt_75[N]</th>\n",
       "      <th>crt_76[K]</th>\n",
       "      <th>crt_72-76[CVMNK]</th>\n",
       "      <th>crt_93[T]</th>\n",
       "      <th>crt_97[H]</th>\n",
       "      <th>crt_218[I]</th>\n",
       "      <th>crt_220[A]</th>\n",
       "      <th>...</th>\n",
       "      <th>mdr1_1034[S]</th>\n",
       "      <th>mdr1_1042[N]</th>\n",
       "      <th>mdr1_1226[F]</th>\n",
       "      <th>mdr1_1246[D]</th>\n",
       "      <th>arps10_127-128[VD]</th>\n",
       "      <th>fd_193[D]</th>\n",
       "      <th>mdr2_484[T]</th>\n",
       "      <th>kelch13_349-726_ns_changes</th>\n",
       "      <th>mdr1_dup_call</th>\n",
       "      <th>pm2_dup_call</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FP0008-C</td>\n",
       "      <td>C</td>\n",
       "      <td>I,M</td>\n",
       "      <td>E,N</td>\n",
       "      <td>T,K</td>\n",
       "      <td>CVIET,CVMNK</td>\n",
       "      <td>T</td>\n",
       "      <td>H</td>\n",
       "      <td>I</td>\n",
       "      <td>S,A</td>\n",
       "      <td>...</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>D</td>\n",
       "      <td>VD</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FP0009-C</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>CVIET</td>\n",
       "      <td>T</td>\n",
       "      <td>H</td>\n",
       "      <td>I</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>Y</td>\n",
       "      <td>VD</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FP0010-CW</td>\n",
       "      <td>C</td>\n",
       "      <td>I,M</td>\n",
       "      <td>E,N</td>\n",
       "      <td>T,K</td>\n",
       "      <td>CVIET,CVMNK</td>\n",
       "      <td>T</td>\n",
       "      <td>H</td>\n",
       "      <td>I</td>\n",
       "      <td>S,A</td>\n",
       "      <td>...</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>D</td>\n",
       "      <td>VD</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FP0011-CW</td>\n",
       "      <td>C</td>\n",
       "      <td>I,M</td>\n",
       "      <td>E,N</td>\n",
       "      <td>T,K</td>\n",
       "      <td>CVIET,CVMNK</td>\n",
       "      <td>T</td>\n",
       "      <td>H</td>\n",
       "      <td>I</td>\n",
       "      <td>S,A</td>\n",
       "      <td>...</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>D</td>\n",
       "      <td>VD</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FP0012-CW</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>CVIET</td>\n",
       "      <td>T</td>\n",
       "      <td>H</td>\n",
       "      <td>I</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>D</td>\n",
       "      <td>VD</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Sample crt_72[C] crt_74[M] crt_75[N] crt_76[K] crt_72-76[CVMNK]  \\\n",
       "0   FP0008-C         C       I,M       E,N       T,K      CVIET,CVMNK   \n",
       "1   FP0009-C         C         I         E         T            CVIET   \n",
       "2  FP0010-CW         C       I,M       E,N       T,K      CVIET,CVMNK   \n",
       "3  FP0011-CW         C       I,M       E,N       T,K      CVIET,CVMNK   \n",
       "4  FP0012-CW         C         I         E         T            CVIET   \n",
       "\n",
       "  crt_93[T] crt_97[H] crt_218[I] crt_220[A]  ... mdr1_1034[S] mdr1_1042[N]  \\\n",
       "0         T         H          I        S,A  ...            S            N   \n",
       "1         T         H          I          S  ...            S            N   \n",
       "2         T         H          I        S,A  ...            S            N   \n",
       "3         T         H          I        S,A  ...            S            N   \n",
       "4         T         H          I          S  ...            S            N   \n",
       "\n",
       "  mdr1_1226[F] mdr1_1246[D] arps10_127-128[VD] fd_193[D] mdr2_484[T]  \\\n",
       "0            F            D                 VD         D           T   \n",
       "1            F            Y                 VD         D           T   \n",
       "2            F            D                 VD         D           T   \n",
       "3            F            D                 VD         D           T   \n",
       "4            F            D                 VD         D           T   \n",
       "\n",
       "  kelch13_349-726_ns_changes mdr1_dup_call pm2_dup_call  \n",
       "0                        NaN             0            0  \n",
       "1                        NaN             0            0  \n",
       "2                        NaN             0            0  \n",
       "3                        NaN             0            0  \n",
       "4                        NaN             0            0  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the data\n",
    "drm_calls_fn = pd.read_csv('https://pf8-release.cog.sanger.ac.uk/Pf8_drug_resistance_marker_genotypes.tsv', sep='\\t', )\n",
    "\n",
    "# Rename the first column as 'Sample'\n",
    "drm_calls_fn = drm_calls_fn.rename(columns={drm_calls_fn.columns[0]: 'Sample'})\n",
    "\n",
    "# Print the first rows\n",
    "drm_calls_fn.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "LHpYlsEi0U3t"
   },
   "source": [
    "##  Load all data into single DataFrame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "bTi6Rups8DWk"
   },
   "source": [
    "We can merge these 3 datasets (metadata, drug resistance genotype and classification) to facilitate streamlined analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 614
    },
    "id": "-kDSuhugMTYf",
    "outputId": "9174eb38-474c-42db-e6b5-0622b6a0aa1a"
   },
   "outputs": [
    {
     "data": {
      "application/vnd.google.colaboratory.intrinsic+json": {
       "type": "dataframe"
      },
      "text/html": [
       "\n",
       "  <div id=\"df-55dc2711-14a3-4f53-b436-6f0ad7b8d4e4\" class=\"colab-df-container\">\n",
       "    <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Study</th>\n",
       "      <th>Country</th>\n",
       "      <th>Admin level 1</th>\n",
       "      <th>Country latitude</th>\n",
       "      <th>Country longitude</th>\n",
       "      <th>Admin level 1 latitude</th>\n",
       "      <th>Admin level 1 longitude</th>\n",
       "      <th>Year</th>\n",
       "      <th>ENA</th>\n",
       "      <th>...</th>\n",
       "      <th>mdr1_1034[S]</th>\n",
       "      <th>mdr1_1042[N]</th>\n",
       "      <th>mdr1_1226[F]</th>\n",
       "      <th>mdr1_1246[D]</th>\n",
       "      <th>arps10_127-128[VD]</th>\n",
       "      <th>fd_193[D]</th>\n",
       "      <th>mdr2_484[T]</th>\n",
       "      <th>kelch13_349-726_ns_changes</th>\n",
       "      <th>mdr1_dup_call</th>\n",
       "      <th>pm2_dup_call</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FP0008-C</td>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR1081237</td>\n",
       "      <td>...</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>D</td>\n",
       "      <td>VD</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FP0009-C</td>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR1081238</td>\n",
       "      <td>...</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>Y</td>\n",
       "      <td>VD</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FP0010-CW</td>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR2889621</td>\n",
       "      <td>...</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>D</td>\n",
       "      <td>VD</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FP0011-CW</td>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR2889624</td>\n",
       "      <td>...</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>D</td>\n",
       "      <td>VD</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FP0012-CW</td>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR2889627</td>\n",
       "      <td>...</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>D</td>\n",
       "      <td>VD</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24404</th>\n",
       "      <td>SPT92049</td>\n",
       "      <td>1306-PF-NG-NGWA-SM</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Oyo</td>\n",
       "      <td>9.592268</td>\n",
       "      <td>8.097575</td>\n",
       "      <td>8.143664</td>\n",
       "      <td>3.618846</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>ERR10893258</td>\n",
       "      <td>...</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>-</td>\n",
       "      <td>VD</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24405</th>\n",
       "      <td>SPT92054</td>\n",
       "      <td>1306-PF-NG-NGWA-SM</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Oyo</td>\n",
       "      <td>9.592268</td>\n",
       "      <td>8.097575</td>\n",
       "      <td>8.143664</td>\n",
       "      <td>3.618846</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>ERR10893326</td>\n",
       "      <td>...</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>D</td>\n",
       "      <td>VD</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24406</th>\n",
       "      <td>SPT92057</td>\n",
       "      <td>1306-PF-NG-NGWA-SM</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Oyo</td>\n",
       "      <td>9.592268</td>\n",
       "      <td>8.097575</td>\n",
       "      <td>8.143664</td>\n",
       "      <td>3.618846</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>ERR10893333</td>\n",
       "      <td>...</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>D</td>\n",
       "      <td>VD</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24407</th>\n",
       "      <td>SPT94772</td>\n",
       "      <td>1268-PF-MULTI-PAMGEN-SM</td>\n",
       "      <td>Gambia</td>\n",
       "      <td>Western</td>\n",
       "      <td>13.451482</td>\n",
       "      <td>-15.372910</td>\n",
       "      <td>13.245396</td>\n",
       "      <td>-16.401559</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>ERR10789456</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>VD</td>\n",
       "      <td>D</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24408</th>\n",
       "      <td>SPT94773</td>\n",
       "      <td>1268-PF-MULTI-PAMGEN-SM</td>\n",
       "      <td>Gambia</td>\n",
       "      <td>Western</td>\n",
       "      <td>13.451482</td>\n",
       "      <td>-15.372910</td>\n",
       "      <td>13.245396</td>\n",
       "      <td>-16.401559</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>ERR10789457</td>\n",
       "      <td>...</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>D</td>\n",
       "      <td>VD</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24409 rows × 56 columns</p>\n",
       "</div>\n",
       "    <div class=\"colab-df-buttons\">\n",
       "\n",
       "  <div class=\"colab-df-container\">\n",
       "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-55dc2711-14a3-4f53-b436-6f0ad7b8d4e4')\"\n",
       "            title=\"Convert this dataframe to an interactive table.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
       "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "\n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    .colab-df-buttons div {\n",
       "      margin-bottom: 4px;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "    <script>\n",
       "      const buttonEl =\n",
       "        document.querySelector('#df-55dc2711-14a3-4f53-b436-6f0ad7b8d4e4 button.colab-df-convert');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      async function convertToInteractive(key) {\n",
       "        const element = document.querySelector('#df-55dc2711-14a3-4f53-b436-6f0ad7b8d4e4');\n",
       "        const dataTable =\n",
       "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                    [key], {});\n",
       "        if (!dataTable) return;\n",
       "\n",
       "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "          + ' to learn more about interactive tables.';\n",
       "        element.innerHTML = '';\n",
       "        dataTable['output_type'] = 'display_data';\n",
       "        await google.colab.output.renderOutput(dataTable, element);\n",
       "        const docLink = document.createElement('div');\n",
       "        docLink.innerHTML = docLinkHtml;\n",
       "        element.appendChild(docLink);\n",
       "      }\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "\n",
       "<div id=\"df-b34bfd7d-3117-42b0-89bf-d15791018355\">\n",
       "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-b34bfd7d-3117-42b0-89bf-d15791018355')\"\n",
       "            title=\"Suggest charts\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "     width=\"24px\">\n",
       "    <g>\n",
       "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
       "    </g>\n",
       "</svg>\n",
       "  </button>\n",
       "\n",
       "<style>\n",
       "  .colab-df-quickchart {\n",
       "      --bg-color: #E8F0FE;\n",
       "      --fill-color: #1967D2;\n",
       "      --hover-bg-color: #E2EBFA;\n",
       "      --hover-fill-color: #174EA6;\n",
       "      --disabled-fill-color: #AAA;\n",
       "      --disabled-bg-color: #DDD;\n",
       "  }\n",
       "\n",
       "  [theme=dark] .colab-df-quickchart {\n",
       "      --bg-color: #3B4455;\n",
       "      --fill-color: #D2E3FC;\n",
       "      --hover-bg-color: #434B5C;\n",
       "      --hover-fill-color: #FFFFFF;\n",
       "      --disabled-bg-color: #3B4455;\n",
       "      --disabled-fill-color: #666;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart {\n",
       "    background-color: var(--bg-color);\n",
       "    border: none;\n",
       "    border-radius: 50%;\n",
       "    cursor: pointer;\n",
       "    display: none;\n",
       "    fill: var(--fill-color);\n",
       "    height: 32px;\n",
       "    padding: 0;\n",
       "    width: 32px;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart:hover {\n",
       "    background-color: var(--hover-bg-color);\n",
       "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "    fill: var(--button-hover-fill-color);\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart-complete:disabled,\n",
       "  .colab-df-quickchart-complete:disabled:hover {\n",
       "    background-color: var(--disabled-bg-color);\n",
       "    fill: var(--disabled-fill-color);\n",
       "    box-shadow: none;\n",
       "  }\n",
       "\n",
       "  .colab-df-spinner {\n",
       "    border: 2px solid var(--fill-color);\n",
       "    border-color: transparent;\n",
       "    border-bottom-color: var(--fill-color);\n",
       "    animation:\n",
       "      spin 1s steps(1) infinite;\n",
       "  }\n",
       "\n",
       "  @keyframes spin {\n",
       "    0% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "      border-left-color: var(--fill-color);\n",
       "    }\n",
       "    20% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    30% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    40% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    60% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    80% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "    90% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "  }\n",
       "</style>\n",
       "\n",
       "  <script>\n",
       "    async function quickchart(key) {\n",
       "      const quickchartButtonEl =\n",
       "        document.querySelector('#' + key + ' button');\n",
       "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
       "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
       "      try {\n",
       "        const charts = await google.colab.kernel.invokeFunction(\n",
       "            'suggestCharts', [key], {});\n",
       "      } catch (error) {\n",
       "        console.error('Error during call to suggestCharts:', error);\n",
       "      }\n",
       "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
       "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
       "    }\n",
       "    (() => {\n",
       "      let quickchartButtonEl =\n",
       "        document.querySelector('#df-b34bfd7d-3117-42b0-89bf-d15791018355 button');\n",
       "      quickchartButtonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "    })();\n",
       "  </script>\n",
       "</div>\n",
       "\n",
       "    </div>\n",
       "  </div>\n"
      ],
      "text/plain": [
       "          Sample                    Study     Country   Admin level 1  \\\n",
       "0       FP0008-C        1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi   \n",
       "1       FP0009-C        1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi   \n",
       "2      FP0010-CW        1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi   \n",
       "3      FP0011-CW        1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi   \n",
       "4      FP0012-CW        1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi   \n",
       "...          ...                      ...         ...             ...   \n",
       "24404   SPT92049       1306-PF-NG-NGWA-SM     Nigeria             Oyo   \n",
       "24405   SPT92054       1306-PF-NG-NGWA-SM     Nigeria             Oyo   \n",
       "24406   SPT92057       1306-PF-NG-NGWA-SM     Nigeria             Oyo   \n",
       "24407   SPT94772  1268-PF-MULTI-PAMGEN-SM      Gambia         Western   \n",
       "24408   SPT94773  1268-PF-MULTI-PAMGEN-SM      Gambia         Western   \n",
       "\n",
       "       Country latitude  Country longitude  Admin level 1 latitude  \\\n",
       "0             20.265149         -10.337093               16.565426   \n",
       "1             20.265149         -10.337093               16.565426   \n",
       "2             20.265149         -10.337093               16.565426   \n",
       "3             20.265149         -10.337093               16.565426   \n",
       "4             20.265149         -10.337093               16.565426   \n",
       "...                 ...                ...                     ...   \n",
       "24404          9.592268           8.097575                8.143664   \n",
       "24405          9.592268           8.097575                8.143664   \n",
       "24406          9.592268           8.097575                8.143664   \n",
       "24407         13.451482         -15.372910               13.245396   \n",
       "24408         13.451482         -15.372910               13.245396   \n",
       "\n",
       "       Admin level 1 longitude    Year          ENA  ... mdr1_1034[S]  \\\n",
       "0                    -9.832345  2014.0   ERR1081237  ...            S   \n",
       "1                    -9.832345  2014.0   ERR1081238  ...            S   \n",
       "2                    -9.832345  2014.0   ERR2889621  ...            S   \n",
       "3                    -9.832345  2014.0   ERR2889624  ...            S   \n",
       "4                    -9.832345  2014.0   ERR2889627  ...            S   \n",
       "...                        ...     ...          ...  ...          ...   \n",
       "24404                 3.618846  2019.0  ERR10893258  ...            S   \n",
       "24405                 3.618846  2019.0  ERR10893326  ...            S   \n",
       "24406                 3.618846  2019.0  ERR10893333  ...            S   \n",
       "24407               -16.401559  2017.0  ERR10789456  ...            -   \n",
       "24408               -16.401559  2017.0  ERR10789457  ...            S   \n",
       "\n",
       "      mdr1_1042[N]  mdr1_1226[F]  mdr1_1246[D] arps10_127-128[VD] fd_193[D]  \\\n",
       "0                N             F             D                 VD         D   \n",
       "1                N             F             Y                 VD         D   \n",
       "2                N             F             D                 VD         D   \n",
       "3                N             F             D                 VD         D   \n",
       "4                N             F             D                 VD         D   \n",
       "...            ...           ...           ...                ...       ...   \n",
       "24404            N             F             -                 VD         D   \n",
       "24405            N             F             D                 VD         D   \n",
       "24406            N             F             D                 VD         D   \n",
       "24407            -             -             -                 VD         D   \n",
       "24408            N             F             D                 VD         D   \n",
       "\n",
       "       mdr2_484[T] kelch13_349-726_ns_changes mdr1_dup_call pm2_dup_call  \n",
       "0                T                        NaN             0            0  \n",
       "1                T                        NaN             0            0  \n",
       "2                T                        NaN             0            0  \n",
       "3                T                        NaN             0            0  \n",
       "4                T                        NaN             0            0  \n",
       "...            ...                        ...           ...          ...  \n",
       "24404            T                          -            -1           -1  \n",
       "24405            T                        NaN             0           -1  \n",
       "24406            T                          -            -1           -1  \n",
       "24407            -                          -            -1           -1  \n",
       "24408            T                        NaN             0            0  \n",
       "\n",
       "[24409 rows x 56 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.merge(qc_sample_metadata,drm_calls_fn,on='Sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 242
    },
    "id": "nDJUwazk0YdL",
    "outputId": "c6485d79-9b64-4724-a0f9-eb6a0127c59b"
   },
   "outputs": [
    {
     "data": {
      "application/vnd.google.colaboratory.intrinsic+json": {
       "type": "dataframe",
       "variable_name": "df_all_sample_metadata"
      },
      "text/html": [
       "\n",
       "  <div id=\"df-d9434466-e6cc-488b-9077-25e478f02d2c\" class=\"colab-df-container\">\n",
       "    <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Study</th>\n",
       "      <th>Country</th>\n",
       "      <th>Admin level 1</th>\n",
       "      <th>Country latitude</th>\n",
       "      <th>Country longitude</th>\n",
       "      <th>Admin level 1 latitude</th>\n",
       "      <th>Admin level 1 longitude</th>\n",
       "      <th>Year</th>\n",
       "      <th>ENA</th>\n",
       "      <th>...</th>\n",
       "      <th>Chloroquine</th>\n",
       "      <th>Pyrimethamine</th>\n",
       "      <th>Sulfadoxine</th>\n",
       "      <th>Mefloquine</th>\n",
       "      <th>Artemisinin</th>\n",
       "      <th>Piperaquine</th>\n",
       "      <th>SP (uncomplicated)</th>\n",
       "      <th>SP (IPTp)</th>\n",
       "      <th>AS-MQ</th>\n",
       "      <th>DHA-PPQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FP0008-C</td>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR1081237</td>\n",
       "      <td>...</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FP0009-C</td>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR1081238</td>\n",
       "      <td>...</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FP0010-CW</td>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR2889621</td>\n",
       "      <td>...</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 66 columns</p>\n",
       "</div>\n",
       "    <div class=\"colab-df-buttons\">\n",
       "\n",
       "  <div class=\"colab-df-container\">\n",
       "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-d9434466-e6cc-488b-9077-25e478f02d2c')\"\n",
       "            title=\"Convert this dataframe to an interactive table.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
       "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "\n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    .colab-df-buttons div {\n",
       "      margin-bottom: 4px;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "    <script>\n",
       "      const buttonEl =\n",
       "        document.querySelector('#df-d9434466-e6cc-488b-9077-25e478f02d2c button.colab-df-convert');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      async function convertToInteractive(key) {\n",
       "        const element = document.querySelector('#df-d9434466-e6cc-488b-9077-25e478f02d2c');\n",
       "        const dataTable =\n",
       "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                    [key], {});\n",
       "        if (!dataTable) return;\n",
       "\n",
       "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "          + ' to learn more about interactive tables.';\n",
       "        element.innerHTML = '';\n",
       "        dataTable['output_type'] = 'display_data';\n",
       "        await google.colab.output.renderOutput(dataTable, element);\n",
       "        const docLink = document.createElement('div');\n",
       "        docLink.innerHTML = docLinkHtml;\n",
       "        element.appendChild(docLink);\n",
       "      }\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "\n",
       "<div id=\"df-72297905-68ab-4a2b-b4c0-c9b158743a44\">\n",
       "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-72297905-68ab-4a2b-b4c0-c9b158743a44')\"\n",
       "            title=\"Suggest charts\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "     width=\"24px\">\n",
       "    <g>\n",
       "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
       "    </g>\n",
       "</svg>\n",
       "  </button>\n",
       "\n",
       "<style>\n",
       "  .colab-df-quickchart {\n",
       "      --bg-color: #E8F0FE;\n",
       "      --fill-color: #1967D2;\n",
       "      --hover-bg-color: #E2EBFA;\n",
       "      --hover-fill-color: #174EA6;\n",
       "      --disabled-fill-color: #AAA;\n",
       "      --disabled-bg-color: #DDD;\n",
       "  }\n",
       "\n",
       "  [theme=dark] .colab-df-quickchart {\n",
       "      --bg-color: #3B4455;\n",
       "      --fill-color: #D2E3FC;\n",
       "      --hover-bg-color: #434B5C;\n",
       "      --hover-fill-color: #FFFFFF;\n",
       "      --disabled-bg-color: #3B4455;\n",
       "      --disabled-fill-color: #666;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart {\n",
       "    background-color: var(--bg-color);\n",
       "    border: none;\n",
       "    border-radius: 50%;\n",
       "    cursor: pointer;\n",
       "    display: none;\n",
       "    fill: var(--fill-color);\n",
       "    height: 32px;\n",
       "    padding: 0;\n",
       "    width: 32px;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart:hover {\n",
       "    background-color: var(--hover-bg-color);\n",
       "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "    fill: var(--button-hover-fill-color);\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart-complete:disabled,\n",
       "  .colab-df-quickchart-complete:disabled:hover {\n",
       "    background-color: var(--disabled-bg-color);\n",
       "    fill: var(--disabled-fill-color);\n",
       "    box-shadow: none;\n",
       "  }\n",
       "\n",
       "  .colab-df-spinner {\n",
       "    border: 2px solid var(--fill-color);\n",
       "    border-color: transparent;\n",
       "    border-bottom-color: var(--fill-color);\n",
       "    animation:\n",
       "      spin 1s steps(1) infinite;\n",
       "  }\n",
       "\n",
       "  @keyframes spin {\n",
       "    0% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "      border-left-color: var(--fill-color);\n",
       "    }\n",
       "    20% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    30% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    40% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    60% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    80% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "    90% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "  }\n",
       "</style>\n",
       "\n",
       "  <script>\n",
       "    async function quickchart(key) {\n",
       "      const quickchartButtonEl =\n",
       "        document.querySelector('#' + key + ' button');\n",
       "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
       "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
       "      try {\n",
       "        const charts = await google.colab.kernel.invokeFunction(\n",
       "            'suggestCharts', [key], {});\n",
       "      } catch (error) {\n",
       "        console.error('Error during call to suggestCharts:', error);\n",
       "      }\n",
       "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
       "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
       "    }\n",
       "    (() => {\n",
       "      let quickchartButtonEl =\n",
       "        document.querySelector('#df-72297905-68ab-4a2b-b4c0-c9b158743a44 button');\n",
       "      quickchartButtonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "    })();\n",
       "  </script>\n",
       "</div>\n",
       "\n",
       "    </div>\n",
       "  </div>\n"
      ],
      "text/plain": [
       "      Sample              Study     Country   Admin level 1  Country latitude  \\\n",
       "0   FP0008-C  1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
       "1   FP0009-C  1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
       "2  FP0010-CW  1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
       "\n",
       "   Country longitude  Admin level 1 latitude  Admin level 1 longitude    Year  \\\n",
       "0         -10.337093               16.565426                -9.832345  2014.0   \n",
       "1         -10.337093               16.565426                -9.832345  2014.0   \n",
       "2         -10.337093               16.565426                -9.832345  2014.0   \n",
       "\n",
       "          ENA  ...   Chloroquine Pyrimethamine   Sulfadoxine  Mefloquine  \\\n",
       "0  ERR1081237  ...  Undetermined  Undetermined  Undetermined   Sensitive   \n",
       "1  ERR1081238  ...     Resistant     Resistant     Sensitive   Sensitive   \n",
       "2  ERR2889621  ...  Undetermined     Resistant     Resistant   Sensitive   \n",
       "\n",
       "  Artemisinin Piperaquine  SP (uncomplicated)  SP (IPTp)      AS-MQ    DHA-PPQ  \n",
       "0   Sensitive   Sensitive           Sensitive  Sensitive  Sensitive  Sensitive  \n",
       "1   Sensitive   Sensitive           Resistant  Sensitive  Sensitive  Sensitive  \n",
       "2   Sensitive   Sensitive           Resistant  Sensitive  Sensitive  Sensitive  \n",
       "\n",
       "[3 rows x 66 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge the dataframes on \"Sample\" column\n",
    "df_all_sample_metadata = pd.merge(pd.merge(qc_sample_metadata,drm_calls_fn,on='Sample'),resistance_classification_fn,on='Sample')\n",
    "\n",
    "# Print first 3 rows\n",
    "df_all_sample_metadata.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "CiW7rU1zINHf"
   },
   "source": [
    "Sulfadoxine-Pyrimethamine (SP) is used for the treatment of uncomplicated cases and Intermittent Preventive Treatment in Pregnancy (IPTp). To make these two status clear in the dataset, we rename the columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "id": "gKHJrr_s0MQt"
   },
   "outputs": [],
   "source": [
    "# Define a dictionary of old&new column names\n",
    "column_name_changes = collections.OrderedDict()\n",
    "column_name_changes['SP (uncomplicated)'] = 'SP (treatment)'\n",
    "column_name_changes['SP-IPTp'] = 'SP (IPTp)'\n",
    "\n",
    "# Rename the columns using the dictionary\n",
    "df_all_sample_metadata.rename(columns=column_name_changes, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "cZFSrQkO0d9C"
   },
   "source": [
    "## Table setups"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "xS4YEFRGIsvy"
   },
   "source": [
    "We proceed by defining a dictionary where each key represents a drug along with its associated gene, mutation, column name in the dataset, and sensitivity/resistance indicators.\n",
    "\n",
    "For details on the heuristics employed to map genetic markers to resistance status classification, please refer to [here](https://pf8-release.cog.sanger.ac.uk/Pf8_resistance_classification.pdf)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "id": "NRDsSwjL0fU5"
   },
   "outputs": [],
   "source": [
    "drm_dict = collections.OrderedDict()\n",
    "\n",
    "drm_dict['Chloroquine'] = {\n",
    "    'gene': 'crt',\n",
    "    'mutation': '76T',\n",
    "    'column_name': 'crt_76[K]',\n",
    "    'sensitive': 'K',\n",
    "    'resistant': 'T',\n",
    "}\n",
    "\n",
    "drm_dict['Pyrimethamine'] = {\n",
    "    'gene': 'dhfr',\n",
    "    'mutation': '108N',\n",
    "    'column_name': 'dhfr_108[S]',\n",
    "    'sensitive': 'S',\n",
    "    'resistant': 'N',\n",
    "}\n",
    "\n",
    "drm_dict['Sulfadoxine'] = {\n",
    "    'gene': 'dhps',\n",
    "    'mutation': '437G',\n",
    "    'column_name': 'dhps_437[G]',\n",
    "    'sensitive': 'A',\n",
    "    'resistant': 'G',\n",
    "}\n",
    "\n",
    "drm_dict['Mefloquine'] = {\n",
    "    'gene': 'mdr1',\n",
    "    'mutation': '2+ copies',\n",
    "    'column_name': 'dup_mdr1',\n",
    "    'sensitive': 0,\n",
    "    'resistant': 1,\n",
    "}\n",
    "\n",
    "drm_dict['Artemisinin'] = {\n",
    "    'gene': 'kelch13',\n",
    "    'mutation': 'WHO list',\n",
    "    'column_name': None,\n",
    "}\n",
    "\n",
    "drm_dict['Piperaquine'] = {\n",
    "    'gene': 'plasmepsin 2-3',\n",
    "    'mutation': '2+ copies',\n",
    "    'column_name': 'dup_pm2',\n",
    "    'sensitive': 0,\n",
    "    'resistant': 1,\n",
    "}\n",
    "\n",
    "drm_dict['SP (treatment)'] = {\n",
    "    'gene': 'dhfr',\n",
    "    'mutation': 'triple mutant',\n",
    "    'column_name': None\n",
    "}\n",
    "\n",
    "drm_dict['SP (IPTp)'] = {\n",
    "    'gene': 'dhfr and dhps',\n",
    "    'mutation': 'sextuple mutant',\n",
    "    'column_name': None\n",
    "}\n",
    "\n",
    "drm_dict['AS-MQ'] = {\n",
    "    'gene': 'kelch13 and mdr1',\n",
    "    'mutation': '',\n",
    "    'column_name': None\n",
    "}\n",
    "\n",
    "drm_dict['DHA-PPQ'] = {\n",
    "    'gene': 'kelch13 and plasmepsin 2-3',\n",
    "    'mutation': '',\n",
    "    'column_name': None\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "3cfvT7nFN8NB"
   },
   "source": [
    "To elaborate on population names in the final table, we define a dictionary listing ten populations with the following abbreviations:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "id": "NkEr9h970mRX"
   },
   "outputs": [],
   "source": [
    "populations = collections.OrderedDict()\n",
    "populations['SA']       = \"South America\"\n",
    "populations['AF-W']     = \"Africa - West\"\n",
    "populations['AF-C']     = \"Africa - Central\"\n",
    "populations['AF-NE']    = \"Africa - Northeast\"\n",
    "populations['AF-E']     = \"Africa - East\"\n",
    "populations['AS-S-E']  = \"Asia - South - East\"\n",
    "populations['AS-S-FE']  = \"Asia - South - Far East\"\n",
    "populations['AS-SE-W'] = \"Asia - Southeast - West\"\n",
    "populations['AS-SE-E'] = \"Asia - Southeast - East\"\n",
    "populations['OC-NG']    = \"Oceania - New Guinea\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "8NQ1MCFS7Cit"
   },
   "source": [
    "## Summary of DR sample sizes across populations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "1c2wlN1MazjI"
   },
   "source": [
    "We want to create a table summarizing the number of samples by population.\n",
    "\n",
    "To do this, we will write a function that counts number of samples that are resistant and sensitive for a given drug."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "id": "yVDUp_IwQWuE"
   },
   "outputs": [],
   "source": [
    "def n_agg(x):\n",
    "    \"\"\"\n",
    "    Aggregate function to calculate the count of non-undetermined samples for each drug in a given population.\n",
    "\n",
    "    Parameters:\n",
    "    - x: DataFrame\n",
    "\n",
    "    Returns:\n",
    "    - pd.Series: Counts of sensitive samples that are resistant to each drug.\n",
    "    \"\"\"\n",
    "\n",
    "    # Initialize an ordered dictionary to store drug names and their corresponding counts\n",
    "    names = collections.OrderedDict()\n",
    "\n",
    "    # Iterate through each drug in drm_dict\n",
    "    for drug in drm_dict:\n",
    "        # Count the number of non-undetermined samples for the current drug\n",
    "        n = np.count_nonzero((x[drug] != 'Undetermined'))\n",
    "\n",
    "        # Store the drug name and its count in the dictionary\n",
    "        names[drug] = n\n",
    "\n",
    "    # Return the result as a pandas Series\n",
    "    return pd.Series(names)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "DwKP3U4ycg-M"
   },
   "source": [
    "As an addition to table, we would like to include minimum and maximum drug resistance sample size in the population names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 464
    },
    "id": "VNLkQeY_7HYb",
    "outputId": "441e560d-ebf5-43cc-9761-85ccb1e9c37c"
   },
   "outputs": [
    {
     "data": {
      "application/vnd.google.colaboratory.intrinsic+json": {
       "summary": "{\n  \"name\": \"df_drm_n_table\",\n  \"rows\": 10,\n  \"fields\": [\n    {\n      \"column\": \"Marker\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 10,\n        \"samples\": [\n          \"kelch13 and mdr1 \",\n          \"dhfr 108N\",\n          \"plasmepsin 2-3 2+ copies\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Associated with resistance to\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 10,\n        \"samples\": [\n          \"AS-MQ\",\n          \"Pyrimethamine\",\n          \"Piperaquine\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"South America (n=177-224)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 20,\n        \"min\": 177,\n        \"max\": 224,\n        \"num_unique_values\": 8,\n        \"samples\": [\n          217,\n          182,\n          223\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Africa - West (n=6648-8837)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 654,\n        \"min\": 6648,\n        \"max\": 8837,\n        \"num_unique_values\": 10,\n        \"samples\": [\n          7846,\n          8035,\n          6648\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Africa - Central (n=870-1184)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 105,\n        \"min\": 870,\n        \"max\": 1184,\n        \"num_unique_values\": 10,\n        \"samples\": [\n          1059,\n          1184,\n          870\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Africa - Northeast (n=122-204)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 32,\n        \"min\": 122,\n        \"max\": 204,\n        \"num_unique_values\": 9,\n        \"samples\": [\n          160,\n          200,\n          201\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Africa - East (n=3599-4000)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 146,\n        \"min\": 3599,\n        \"max\": 4000,\n        \"num_unique_values\": 10,\n        \"samples\": [\n          3940,\n          4000,\n          3599\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Asia - South - East (n=81-188)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 37,\n        \"min\": 81,\n        \"max\": 188,\n        \"num_unique_values\": 10,\n        \"samples\": [\n          157,\n          162,\n          81\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Asia - South - Far East (n=1209-1369)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 60,\n        \"min\": 1209,\n        \"max\": 1369,\n        \"num_unique_values\": 9,\n        \"samples\": [\n          1216,\n          1369,\n          1248\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Asia - Southeast - West (n=1665-1884)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 70,\n        \"min\": 1665,\n        \"max\": 1884,\n        \"num_unique_values\": 9,\n        \"samples\": [\n          1795,\n          1884,\n          1837\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Asia - Southeast - East (n=2910-5798)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 859,\n        \"min\": 2910,\n        \"max\": 5798,\n        \"num_unique_values\": 10,\n        \"samples\": [\n          5087,\n          5796,\n          4868\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Oceania - New Guinea (n=293-340)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 16,\n        \"min\": 293,\n        \"max\": 340,\n        \"num_unique_values\": 9,\n        \"samples\": [\n          320,\n          332,\n          338\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}",
       "type": "dataframe",
       "variable_name": "df_drm_n_table"
      },
      "text/html": [
       "\n",
       "  <div id=\"df-f63de62e-3ff3-42bb-88f4-5509dd5a60e8\" class=\"colab-df-container\">\n",
       "    <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Associated with resistance to</th>\n",
       "      <th>South America (n=177-224)</th>\n",
       "      <th>Africa - West (n=6648-8837)</th>\n",
       "      <th>Africa - Central (n=870-1184)</th>\n",
       "      <th>Africa - Northeast (n=122-204)</th>\n",
       "      <th>Africa - East (n=3599-4000)</th>\n",
       "      <th>Asia - South - East (n=81-188)</th>\n",
       "      <th>Asia - South - Far East (n=1209-1369)</th>\n",
       "      <th>Asia - Southeast - West (n=1665-1884)</th>\n",
       "      <th>Asia - Southeast - East (n=2910-5798)</th>\n",
       "      <th>Oceania - New Guinea (n=293-340)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Marker</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>crt 76T</th>\n",
       "      <td>Chloroquine</td>\n",
       "      <td>223</td>\n",
       "      <td>8284</td>\n",
       "      <td>1039</td>\n",
       "      <td>196</td>\n",
       "      <td>3944</td>\n",
       "      <td>178</td>\n",
       "      <td>1330</td>\n",
       "      <td>1877</td>\n",
       "      <td>5798</td>\n",
       "      <td>331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dhfr 108N</th>\n",
       "      <td>Pyrimethamine</td>\n",
       "      <td>217</td>\n",
       "      <td>8035</td>\n",
       "      <td>1184</td>\n",
       "      <td>200</td>\n",
       "      <td>4000</td>\n",
       "      <td>162</td>\n",
       "      <td>1369</td>\n",
       "      <td>1884</td>\n",
       "      <td>5796</td>\n",
       "      <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dhps 437G</th>\n",
       "      <td>Sulfadoxine</td>\n",
       "      <td>224</td>\n",
       "      <td>8007</td>\n",
       "      <td>1155</td>\n",
       "      <td>196</td>\n",
       "      <td>3871</td>\n",
       "      <td>179</td>\n",
       "      <td>1296</td>\n",
       "      <td>1884</td>\n",
       "      <td>5728</td>\n",
       "      <td>331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mdr1 2+ copies</th>\n",
       "      <td>Mefloquine</td>\n",
       "      <td>177</td>\n",
       "      <td>6936</td>\n",
       "      <td>923</td>\n",
       "      <td>125</td>\n",
       "      <td>3707</td>\n",
       "      <td>93</td>\n",
       "      <td>1241</td>\n",
       "      <td>1739</td>\n",
       "      <td>4966</td>\n",
       "      <td>307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kelch13 WHO list</th>\n",
       "      <td>Artemisinin</td>\n",
       "      <td>180</td>\n",
       "      <td>7220</td>\n",
       "      <td>938</td>\n",
       "      <td>151</td>\n",
       "      <td>3712</td>\n",
       "      <td>144</td>\n",
       "      <td>1335</td>\n",
       "      <td>1775</td>\n",
       "      <td>5295</td>\n",
       "      <td>302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>plasmepsin 2-3 2+ copies</th>\n",
       "      <td>Piperaquine</td>\n",
       "      <td>182</td>\n",
       "      <td>6648</td>\n",
       "      <td>870</td>\n",
       "      <td>122</td>\n",
       "      <td>3599</td>\n",
       "      <td>81</td>\n",
       "      <td>1248</td>\n",
       "      <td>1779</td>\n",
       "      <td>4868</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dhfr triple mutant</th>\n",
       "      <td>SP (treatment)</td>\n",
       "      <td>224</td>\n",
       "      <td>7474</td>\n",
       "      <td>1058</td>\n",
       "      <td>201</td>\n",
       "      <td>3646</td>\n",
       "      <td>186</td>\n",
       "      <td>1209</td>\n",
       "      <td>1837</td>\n",
       "      <td>5689</td>\n",
       "      <td>338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dhfr and dhps sextuple mutant</th>\n",
       "      <td>SP (IPTp)</td>\n",
       "      <td>224</td>\n",
       "      <td>8837</td>\n",
       "      <td>1138</td>\n",
       "      <td>204</td>\n",
       "      <td>3980</td>\n",
       "      <td>188</td>\n",
       "      <td>1216</td>\n",
       "      <td>1665</td>\n",
       "      <td>2910</td>\n",
       "      <td>340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kelch13 and mdr1</th>\n",
       "      <td>AS-MQ</td>\n",
       "      <td>199</td>\n",
       "      <td>7846</td>\n",
       "      <td>1059</td>\n",
       "      <td>160</td>\n",
       "      <td>3940</td>\n",
       "      <td>157</td>\n",
       "      <td>1350</td>\n",
       "      <td>1795</td>\n",
       "      <td>5087</td>\n",
       "      <td>320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kelch13 and plasmepsin 2-3</th>\n",
       "      <td>DHA-PPQ</td>\n",
       "      <td>190</td>\n",
       "      <td>7524</td>\n",
       "      <td>981</td>\n",
       "      <td>157</td>\n",
       "      <td>3804</td>\n",
       "      <td>149</td>\n",
       "      <td>1350</td>\n",
       "      <td>1824</td>\n",
       "      <td>4949</td>\n",
       "      <td>313</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    <div class=\"colab-df-buttons\">\n",
       "\n",
       "  <div class=\"colab-df-container\">\n",
       "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-f63de62e-3ff3-42bb-88f4-5509dd5a60e8')\"\n",
       "            title=\"Convert this dataframe to an interactive table.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
       "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "\n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    .colab-df-buttons div {\n",
       "      margin-bottom: 4px;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "    <script>\n",
       "      const buttonEl =\n",
       "        document.querySelector('#df-f63de62e-3ff3-42bb-88f4-5509dd5a60e8 button.colab-df-convert');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      async function convertToInteractive(key) {\n",
       "        const element = document.querySelector('#df-f63de62e-3ff3-42bb-88f4-5509dd5a60e8');\n",
       "        const dataTable =\n",
       "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                    [key], {});\n",
       "        if (!dataTable) return;\n",
       "\n",
       "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "          + ' to learn more about interactive tables.';\n",
       "        element.innerHTML = '';\n",
       "        dataTable['output_type'] = 'display_data';\n",
       "        await google.colab.output.renderOutput(dataTable, element);\n",
       "        const docLink = document.createElement('div');\n",
       "        docLink.innerHTML = docLinkHtml;\n",
       "        element.appendChild(docLink);\n",
       "      }\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "\n",
       "<div id=\"df-aabe1650-1de4-4d7f-b245-8ee458530b2e\">\n",
       "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-aabe1650-1de4-4d7f-b245-8ee458530b2e')\"\n",
       "            title=\"Suggest charts\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "     width=\"24px\">\n",
       "    <g>\n",
       "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
       "    </g>\n",
       "</svg>\n",
       "  </button>\n",
       "\n",
       "<style>\n",
       "  .colab-df-quickchart {\n",
       "      --bg-color: #E8F0FE;\n",
       "      --fill-color: #1967D2;\n",
       "      --hover-bg-color: #E2EBFA;\n",
       "      --hover-fill-color: #174EA6;\n",
       "      --disabled-fill-color: #AAA;\n",
       "      --disabled-bg-color: #DDD;\n",
       "  }\n",
       "\n",
       "  [theme=dark] .colab-df-quickchart {\n",
       "      --bg-color: #3B4455;\n",
       "      --fill-color: #D2E3FC;\n",
       "      --hover-bg-color: #434B5C;\n",
       "      --hover-fill-color: #FFFFFF;\n",
       "      --disabled-bg-color: #3B4455;\n",
       "      --disabled-fill-color: #666;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart {\n",
       "    background-color: var(--bg-color);\n",
       "    border: none;\n",
       "    border-radius: 50%;\n",
       "    cursor: pointer;\n",
       "    display: none;\n",
       "    fill: var(--fill-color);\n",
       "    height: 32px;\n",
       "    padding: 0;\n",
       "    width: 32px;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart:hover {\n",
       "    background-color: var(--hover-bg-color);\n",
       "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "    fill: var(--button-hover-fill-color);\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart-complete:disabled,\n",
       "  .colab-df-quickchart-complete:disabled:hover {\n",
       "    background-color: var(--disabled-bg-color);\n",
       "    fill: var(--disabled-fill-color);\n",
       "    box-shadow: none;\n",
       "  }\n",
       "\n",
       "  .colab-df-spinner {\n",
       "    border: 2px solid var(--fill-color);\n",
       "    border-color: transparent;\n",
       "    border-bottom-color: var(--fill-color);\n",
       "    animation:\n",
       "      spin 1s steps(1) infinite;\n",
       "  }\n",
       "\n",
       "  @keyframes spin {\n",
       "    0% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "      border-left-color: var(--fill-color);\n",
       "    }\n",
       "    20% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    30% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    40% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    60% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    80% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "    90% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "  }\n",
       "</style>\n",
       "\n",
       "  <script>\n",
       "    async function quickchart(key) {\n",
       "      const quickchartButtonEl =\n",
       "        document.querySelector('#' + key + ' button');\n",
       "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
       "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
       "      try {\n",
       "        const charts = await google.colab.kernel.invokeFunction(\n",
       "            'suggestCharts', [key], {});\n",
       "      } catch (error) {\n",
       "        console.error('Error during call to suggestCharts:', error);\n",
       "      }\n",
       "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
       "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
       "    }\n",
       "    (() => {\n",
       "      let quickchartButtonEl =\n",
       "        document.querySelector('#df-aabe1650-1de4-4d7f-b245-8ee458530b2e button');\n",
       "      quickchartButtonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "    })();\n",
       "  </script>\n",
       "</div>\n",
       "\n",
       "  <div id=\"id_6cb66345-cb18-4301-bc6c-742f08fc91a6\">\n",
       "    <style>\n",
       "      .colab-df-generate {\n",
       "        background-color: #E8F0FE;\n",
       "        border: none;\n",
       "        border-radius: 50%;\n",
       "        cursor: pointer;\n",
       "        display: none;\n",
       "        fill: #1967D2;\n",
       "        height: 32px;\n",
       "        padding: 0 0 0 0;\n",
       "        width: 32px;\n",
       "      }\n",
       "\n",
       "      .colab-df-generate:hover {\n",
       "        background-color: #E2EBFA;\n",
       "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "        fill: #174EA6;\n",
       "      }\n",
       "\n",
       "      [theme=dark] .colab-df-generate {\n",
       "        background-color: #3B4455;\n",
       "        fill: #D2E3FC;\n",
       "      }\n",
       "\n",
       "      [theme=dark] .colab-df-generate:hover {\n",
       "        background-color: #434B5C;\n",
       "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "        fill: #FFFFFF;\n",
       "      }\n",
       "    </style>\n",
       "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('df_drm_n_table')\"\n",
       "            title=\"Generate code using this dataframe.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "    <script>\n",
       "      (() => {\n",
       "      const buttonEl =\n",
       "        document.querySelector('#id_6cb66345-cb18-4301-bc6c-742f08fc91a6 button.colab-df-generate');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      buttonEl.onclick = () => {\n",
       "        google.colab.notebook.generateWithVariable('df_drm_n_table');\n",
       "      }\n",
       "      })();\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "    </div>\n",
       "  </div>\n"
      ],
      "text/plain": [
       "                              Associated with resistance to  \\\n",
       "Marker                                                        \n",
       "crt 76T                                         Chloroquine   \n",
       "dhfr 108N                                     Pyrimethamine   \n",
       "dhps 437G                                       Sulfadoxine   \n",
       "mdr1 2+ copies                                   Mefloquine   \n",
       "kelch13 WHO list                                Artemisinin   \n",
       "plasmepsin 2-3 2+ copies                        Piperaquine   \n",
       "dhfr triple mutant                           SP (treatment)   \n",
       "dhfr and dhps sextuple mutant                     SP (IPTp)   \n",
       "kelch13 and mdr1                                      AS-MQ   \n",
       "kelch13 and plasmepsin 2-3                          DHA-PPQ   \n",
       "\n",
       "                               South America (n=177-224)  \\\n",
       "Marker                                                     \n",
       "crt 76T                                              223   \n",
       "dhfr 108N                                            217   \n",
       "dhps 437G                                            224   \n",
       "mdr1 2+ copies                                       177   \n",
       "kelch13 WHO list                                     180   \n",
       "plasmepsin 2-3 2+ copies                             182   \n",
       "dhfr triple mutant                                   224   \n",
       "dhfr and dhps sextuple mutant                        224   \n",
       "kelch13 and mdr1                                     199   \n",
       "kelch13 and plasmepsin 2-3                           190   \n",
       "\n",
       "                               Africa - West (n=6648-8837)  \\\n",
       "Marker                                                       \n",
       "crt 76T                                               8284   \n",
       "dhfr 108N                                             8035   \n",
       "dhps 437G                                             8007   \n",
       "mdr1 2+ copies                                        6936   \n",
       "kelch13 WHO list                                      7220   \n",
       "plasmepsin 2-3 2+ copies                              6648   \n",
       "dhfr triple mutant                                    7474   \n",
       "dhfr and dhps sextuple mutant                         8837   \n",
       "kelch13 and mdr1                                      7846   \n",
       "kelch13 and plasmepsin 2-3                            7524   \n",
       "\n",
       "                               Africa - Central (n=870-1184)  \\\n",
       "Marker                                                         \n",
       "crt 76T                                                 1039   \n",
       "dhfr 108N                                               1184   \n",
       "dhps 437G                                               1155   \n",
       "mdr1 2+ copies                                           923   \n",
       "kelch13 WHO list                                         938   \n",
       "plasmepsin 2-3 2+ copies                                 870   \n",
       "dhfr triple mutant                                      1058   \n",
       "dhfr and dhps sextuple mutant                           1138   \n",
       "kelch13 and mdr1                                        1059   \n",
       "kelch13 and plasmepsin 2-3                               981   \n",
       "\n",
       "                               Africa - Northeast (n=122-204)  \\\n",
       "Marker                                                          \n",
       "crt 76T                                                   196   \n",
       "dhfr 108N                                                 200   \n",
       "dhps 437G                                                 196   \n",
       "mdr1 2+ copies                                            125   \n",
       "kelch13 WHO list                                          151   \n",
       "plasmepsin 2-3 2+ copies                                  122   \n",
       "dhfr triple mutant                                        201   \n",
       "dhfr and dhps sextuple mutant                             204   \n",
       "kelch13 and mdr1                                          160   \n",
       "kelch13 and plasmepsin 2-3                                157   \n",
       "\n",
       "                               Africa - East (n=3599-4000)  \\\n",
       "Marker                                                       \n",
       "crt 76T                                               3944   \n",
       "dhfr 108N                                             4000   \n",
       "dhps 437G                                             3871   \n",
       "mdr1 2+ copies                                        3707   \n",
       "kelch13 WHO list                                      3712   \n",
       "plasmepsin 2-3 2+ copies                              3599   \n",
       "dhfr triple mutant                                    3646   \n",
       "dhfr and dhps sextuple mutant                         3980   \n",
       "kelch13 and mdr1                                      3940   \n",
       "kelch13 and plasmepsin 2-3                            3804   \n",
       "\n",
       "                               Asia - South - East (n=81-188)  \\\n",
       "Marker                                                          \n",
       "crt 76T                                                   178   \n",
       "dhfr 108N                                                 162   \n",
       "dhps 437G                                                 179   \n",
       "mdr1 2+ copies                                             93   \n",
       "kelch13 WHO list                                          144   \n",
       "plasmepsin 2-3 2+ copies                                   81   \n",
       "dhfr triple mutant                                        186   \n",
       "dhfr and dhps sextuple mutant                             188   \n",
       "kelch13 and mdr1                                          157   \n",
       "kelch13 and plasmepsin 2-3                                149   \n",
       "\n",
       "                               Asia - South - Far East (n=1209-1369)  \\\n",
       "Marker                                                                 \n",
       "crt 76T                                                         1330   \n",
       "dhfr 108N                                                       1369   \n",
       "dhps 437G                                                       1296   \n",
       "mdr1 2+ copies                                                  1241   \n",
       "kelch13 WHO list                                                1335   \n",
       "plasmepsin 2-3 2+ copies                                        1248   \n",
       "dhfr triple mutant                                              1209   \n",
       "dhfr and dhps sextuple mutant                                   1216   \n",
       "kelch13 and mdr1                                                1350   \n",
       "kelch13 and plasmepsin 2-3                                      1350   \n",
       "\n",
       "                               Asia - Southeast - West (n=1665-1884)  \\\n",
       "Marker                                                                 \n",
       "crt 76T                                                         1877   \n",
       "dhfr 108N                                                       1884   \n",
       "dhps 437G                                                       1884   \n",
       "mdr1 2+ copies                                                  1739   \n",
       "kelch13 WHO list                                                1775   \n",
       "plasmepsin 2-3 2+ copies                                        1779   \n",
       "dhfr triple mutant                                              1837   \n",
       "dhfr and dhps sextuple mutant                                   1665   \n",
       "kelch13 and mdr1                                                1795   \n",
       "kelch13 and plasmepsin 2-3                                      1824   \n",
       "\n",
       "                               Asia - Southeast - East (n=2910-5798)  \\\n",
       "Marker                                                                 \n",
       "crt 76T                                                         5798   \n",
       "dhfr 108N                                                       5796   \n",
       "dhps 437G                                                       5728   \n",
       "mdr1 2+ copies                                                  4966   \n",
       "kelch13 WHO list                                                5295   \n",
       "plasmepsin 2-3 2+ copies                                        4868   \n",
       "dhfr triple mutant                                              5689   \n",
       "dhfr and dhps sextuple mutant                                   2910   \n",
       "kelch13 and mdr1                                                5087   \n",
       "kelch13 and plasmepsin 2-3                                      4949   \n",
       "\n",
       "                               Oceania - New Guinea (n=293-340)  \n",
       "Marker                                                           \n",
       "crt 76T                                                     331  \n",
       "dhfr 108N                                                   332  \n",
       "dhps 437G                                                   331  \n",
       "mdr1 2+ copies                                              307  \n",
       "kelch13 WHO list                                            302  \n",
       "plasmepsin 2-3 2+ copies                                    293  \n",
       "dhfr triple mutant                                          338  \n",
       "dhfr and dhps sextuple mutant                               340  \n",
       "kelch13 and mdr1                                            320  \n",
       "kelch13 and plasmepsin 2-3                                  313  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate counts of non-undetermined samples for each drug in each population\n",
    "df_drm_n_table = (\n",
    "    df_all_sample_metadata\n",
    "    .groupby('Population')\n",
    "    .apply(n_agg, include_groups=False)\n",
    "    .rename_axis(None)\n",
    "    .transpose()\n",
    "    .loc[:, populations.keys()]\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "# Calculate the minimum and maximum sample sizes across populations\n",
    "min_n = df_drm_n_table.min()\n",
    "max_n = df_drm_n_table.max()\n",
    "\n",
    "# Customize the index to include gene and mutation information for each drug\n",
    "df_drm_n_table.index = [\"%s %s\" % (drm_dict[drug]['gene'], drm_dict[drug]['mutation']) for drug in drm_dict]\n",
    "df_drm_n_table.index.names = ['Marker']\n",
    "\n",
    "# Rename the columns to indicate their association with drug resistance and sample counts\n",
    "df_drm_n_table.rename(columns={'index': 'Associated with resistance to'}, inplace=True)\n",
    "\n",
    "# Update column names to include population names and corresponding sample size ranges\n",
    "for population in populations:\n",
    "    new_column_name = f'{populations[population]} (n={min_n[population]}-{max_n[population]})'\n",
    "    df_drm_n_table.rename(columns={population: new_column_name}, inplace=True)\n",
    "\n",
    "# Display the final drug resistance count table\n",
    "df_drm_n_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Kkoy6lwGouuN"
   },
   "source": [
    "**Table 1.** Numbers of samples used to determine proportions in Table 2."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "g_ey7FCwQuAA"
   },
   "source": [
    "To save the table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "id": "t4QP1cf_QwjC"
   },
   "outputs": [],
   "source": [
    "df_drm_n_table.to_excel(\"DRM_table_sample_numbers.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "-2YbaP-KnTYb"
   },
   "source": [
    "## Summary of DR frequencies across populations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "EsHY5DKnYpoX"
   },
   "source": [
    "Now, we would like to calculate drug resistance proportions in each population.\n",
    "\n",
    "We can easily adjust the function to count number of samples that are resistant and sensitive for a given drug, and calculate the proportion.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "id": "amTK8om50oEP"
   },
   "outputs": [],
   "source": [
    "def proportion_agg(x):\n",
    "    \"\"\"\n",
    "    Aggregate function to calculate the proportion of resistance for each drug in a given population.\n",
    "\n",
    "    Parameters:\n",
    "    - x: DataFrame\n",
    "\n",
    "    Returns:\n",
    "    - pd.Series: Proportions of resistance for each drug.\n",
    "    \"\"\"\n",
    "\n",
    "    # Initialize an ordered dictionary to store drug names and their corresponding proportions\n",
    "    names = collections.OrderedDict()\n",
    "\n",
    "    # Iterate through each drug in drm_dict\n",
    "    for drug in drm_dict:\n",
    "        # Count the number of non-undetermined samples for the current drug\n",
    "        n = np.count_nonzero((x[drug] != 'Undetermined'))\n",
    "\n",
    "        # Check if there are no non-undetermined samples for the drug\n",
    "        if n == 0:\n",
    "            proportion = np.nan  # Set proportion to NaN to avoid division by zero\n",
    "        else:\n",
    "            # Calculate the proportion of resistant samples for the drug\n",
    "            proportion = round(np.count_nonzero(\n",
    "                (x[drug] == 'Resistant')\n",
    "            ) / np.count_nonzero(\n",
    "                (x[drug] != 'Undetermined')\n",
    "            )*100)\n",
    "\n",
    "        # Store the drug name and its proportion in the dictionary\n",
    "        names[drug] = proportion\n",
    "\n",
    "    # Return the result as a pandas Series\n",
    "    return pd.Series(names)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "g17tr7u9aWH1"
   },
   "source": [
    "Let's apply this function and create a table summarizing drug resistance proportions by population.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 464
    },
    "id": "lSG0_DSxPPy5",
    "outputId": "b918e36e-5798-4cc0-dd63-cf8ed3064148"
   },
   "outputs": [
    {
     "data": {
      "application/vnd.google.colaboratory.intrinsic+json": {
       "summary": "{\n  \"name\": \"df_drm_table\",\n  \"rows\": 10,\n  \"fields\": [\n    {\n      \"column\": \"Marker\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 10,\n        \"samples\": [\n          \"kelch13 and mdr1 \",\n          \"dhfr 108N\",\n          \"plasmepsin 2-3 2+ copies\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Associated with resistance to\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 10,\n        \"samples\": [\n          \"AS-MQ\",\n          \"Pyrimethamine\",\n          \"Piperaquine\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"South America (n=177-224)\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 4,\n        \"samples\": [\n          \"74%\",\n          \"0%\",\n          \"100%\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Africa - West (n=6648-8837)\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"89%\",\n          \"79%\",\n          \"81%\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Africa - Central (n=870-1184)\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 6,\n        \"samples\": [\n          \"35%\",\n          \"99%\",\n          \"5%\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Africa - Northeast (n=122-204)\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 6,\n        \"samples\": [\n          \"48%\",\n          \"98%\",\n          \"2%\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Africa - East (n=3599-4000)\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 6,\n        \"samples\": [\n          \"14%\",\n          \"97%\",\n          \"4%\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Asia - South - East (n=81-188)\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"62%\",\n          \"2%\",\n          \"9%\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Asia - South - Far East (n=1209-1369)\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 6,\n        \"samples\": [\n          \"94%\",\n          \"100%\",\n          \"13%\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Asia - Southeast - West (n=1665-1884)\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 8,\n        \"samples\": [\n          \"100%\",\n          \"86%\",\n          \"99%\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Asia - Southeast - East (n=2910-5798)\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 10,\n        \"samples\": [\n          \"3%\",\n          \"99%\",\n          \"43%\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Oceania - New Guinea (n=293-340)\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"99%\",\n          \"0%\",\n          \"69%\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}",
       "type": "dataframe",
       "variable_name": "df_drm_table"
      },
      "text/html": [
       "\n",
       "  <div id=\"df-53c37be1-c09f-4f91-bbb2-de3c4ee83162\" class=\"colab-df-container\">\n",
       "    <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Associated with resistance to</th>\n",
       "      <th>South America (n=177-224)</th>\n",
       "      <th>Africa - West (n=6648-8837)</th>\n",
       "      <th>Africa - Central (n=870-1184)</th>\n",
       "      <th>Africa - Northeast (n=122-204)</th>\n",
       "      <th>Africa - East (n=3599-4000)</th>\n",
       "      <th>Asia - South - East (n=81-188)</th>\n",
       "      <th>Asia - South - Far East (n=1209-1369)</th>\n",
       "      <th>Asia - Southeast - West (n=1665-1884)</th>\n",
       "      <th>Asia - Southeast - East (n=2910-5798)</th>\n",
       "      <th>Oceania - New Guinea (n=293-340)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Marker</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>crt 76T</th>\n",
       "      <td>Chloroquine</td>\n",
       "      <td>100%</td>\n",
       "      <td>26%</td>\n",
       "      <td>35%</td>\n",
       "      <td>48%</td>\n",
       "      <td>14%</td>\n",
       "      <td>33%</td>\n",
       "      <td>94%</td>\n",
       "      <td>99%</td>\n",
       "      <td>96%</td>\n",
       "      <td>96%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dhfr 108N</th>\n",
       "      <td>Pyrimethamine</td>\n",
       "      <td>74%</td>\n",
       "      <td>89%</td>\n",
       "      <td>99%</td>\n",
       "      <td>98%</td>\n",
       "      <td>97%</td>\n",
       "      <td>62%</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>99%</td>\n",
       "      <td>99%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dhps 437G</th>\n",
       "      <td>Sulfadoxine</td>\n",
       "      <td>71%</td>\n",
       "      <td>81%</td>\n",
       "      <td>96%</td>\n",
       "      <td>79%</td>\n",
       "      <td>86%</td>\n",
       "      <td>9%</td>\n",
       "      <td>89%</td>\n",
       "      <td>100%</td>\n",
       "      <td>83%</td>\n",
       "      <td>69%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mdr1 2+ copies</th>\n",
       "      <td>Mefloquine</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>30%</td>\n",
       "      <td>4%</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kelch13 WHO list</th>\n",
       "      <td>Artemisinin</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>36%</td>\n",
       "      <td>62%</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>plasmepsin 2-3 2+ copies</th>\n",
       "      <td>Piperaquine</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>43%</td>\n",
       "      <td>0%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dhfr triple mutant</th>\n",
       "      <td>SP (treatment)</td>\n",
       "      <td>0%</td>\n",
       "      <td>79%</td>\n",
       "      <td>88%</td>\n",
       "      <td>51%</td>\n",
       "      <td>83%</td>\n",
       "      <td>2%</td>\n",
       "      <td>46%</td>\n",
       "      <td>86%</td>\n",
       "      <td>87%</td>\n",
       "      <td>0%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dhfr and dhps sextuple mutant</th>\n",
       "      <td>SP (IPTp)</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>5%</td>\n",
       "      <td>2%</td>\n",
       "      <td>4%</td>\n",
       "      <td>0%</td>\n",
       "      <td>13%</td>\n",
       "      <td>79%</td>\n",
       "      <td>12%</td>\n",
       "      <td>0%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kelch13 and mdr1</th>\n",
       "      <td>AS-MQ</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>10%</td>\n",
       "      <td>3%</td>\n",
       "      <td>0%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kelch13 and plasmepsin 2-3</th>\n",
       "      <td>DHA-PPQ</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td>39%</td>\n",
       "      <td>0%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    <div class=\"colab-df-buttons\">\n",
       "\n",
       "  <div class=\"colab-df-container\">\n",
       "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-53c37be1-c09f-4f91-bbb2-de3c4ee83162')\"\n",
       "            title=\"Convert this dataframe to an interactive table.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
       "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "\n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    .colab-df-buttons div {\n",
       "      margin-bottom: 4px;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "    <script>\n",
       "      const buttonEl =\n",
       "        document.querySelector('#df-53c37be1-c09f-4f91-bbb2-de3c4ee83162 button.colab-df-convert');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      async function convertToInteractive(key) {\n",
       "        const element = document.querySelector('#df-53c37be1-c09f-4f91-bbb2-de3c4ee83162');\n",
       "        const dataTable =\n",
       "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                    [key], {});\n",
       "        if (!dataTable) return;\n",
       "\n",
       "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "          + ' to learn more about interactive tables.';\n",
       "        element.innerHTML = '';\n",
       "        dataTable['output_type'] = 'display_data';\n",
       "        await google.colab.output.renderOutput(dataTable, element);\n",
       "        const docLink = document.createElement('div');\n",
       "        docLink.innerHTML = docLinkHtml;\n",
       "        element.appendChild(docLink);\n",
       "      }\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "\n",
       "<div id=\"df-254c29df-16d7-442c-a848-62d834a8fb27\">\n",
       "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-254c29df-16d7-442c-a848-62d834a8fb27')\"\n",
       "            title=\"Suggest charts\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "     width=\"24px\">\n",
       "    <g>\n",
       "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
       "    </g>\n",
       "</svg>\n",
       "  </button>\n",
       "\n",
       "<style>\n",
       "  .colab-df-quickchart {\n",
       "      --bg-color: #E8F0FE;\n",
       "      --fill-color: #1967D2;\n",
       "      --hover-bg-color: #E2EBFA;\n",
       "      --hover-fill-color: #174EA6;\n",
       "      --disabled-fill-color: #AAA;\n",
       "      --disabled-bg-color: #DDD;\n",
       "  }\n",
       "\n",
       "  [theme=dark] .colab-df-quickchart {\n",
       "      --bg-color: #3B4455;\n",
       "      --fill-color: #D2E3FC;\n",
       "      --hover-bg-color: #434B5C;\n",
       "      --hover-fill-color: #FFFFFF;\n",
       "      --disabled-bg-color: #3B4455;\n",
       "      --disabled-fill-color: #666;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart {\n",
       "    background-color: var(--bg-color);\n",
       "    border: none;\n",
       "    border-radius: 50%;\n",
       "    cursor: pointer;\n",
       "    display: none;\n",
       "    fill: var(--fill-color);\n",
       "    height: 32px;\n",
       "    padding: 0;\n",
       "    width: 32px;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart:hover {\n",
       "    background-color: var(--hover-bg-color);\n",
       "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "    fill: var(--button-hover-fill-color);\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart-complete:disabled,\n",
       "  .colab-df-quickchart-complete:disabled:hover {\n",
       "    background-color: var(--disabled-bg-color);\n",
       "    fill: var(--disabled-fill-color);\n",
       "    box-shadow: none;\n",
       "  }\n",
       "\n",
       "  .colab-df-spinner {\n",
       "    border: 2px solid var(--fill-color);\n",
       "    border-color: transparent;\n",
       "    border-bottom-color: var(--fill-color);\n",
       "    animation:\n",
       "      spin 1s steps(1) infinite;\n",
       "  }\n",
       "\n",
       "  @keyframes spin {\n",
       "    0% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "      border-left-color: var(--fill-color);\n",
       "    }\n",
       "    20% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    30% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    40% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    60% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    80% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "    90% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "  }\n",
       "</style>\n",
       "\n",
       "  <script>\n",
       "    async function quickchart(key) {\n",
       "      const quickchartButtonEl =\n",
       "        document.querySelector('#' + key + ' button');\n",
       "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
       "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
       "      try {\n",
       "        const charts = await google.colab.kernel.invokeFunction(\n",
       "            'suggestCharts', [key], {});\n",
       "      } catch (error) {\n",
       "        console.error('Error during call to suggestCharts:', error);\n",
       "      }\n",
       "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
       "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
       "    }\n",
       "    (() => {\n",
       "      let quickchartButtonEl =\n",
       "        document.querySelector('#df-254c29df-16d7-442c-a848-62d834a8fb27 button');\n",
       "      quickchartButtonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "    })();\n",
       "  </script>\n",
       "</div>\n",
       "\n",
       "  <div id=\"id_1de2157e-1c45-4857-b591-a89d54563799\">\n",
       "    <style>\n",
       "      .colab-df-generate {\n",
       "        background-color: #E8F0FE;\n",
       "        border: none;\n",
       "        border-radius: 50%;\n",
       "        cursor: pointer;\n",
       "        display: none;\n",
       "        fill: #1967D2;\n",
       "        height: 32px;\n",
       "        padding: 0 0 0 0;\n",
       "        width: 32px;\n",
       "      }\n",
       "\n",
       "      .colab-df-generate:hover {\n",
       "        background-color: #E2EBFA;\n",
       "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "        fill: #174EA6;\n",
       "      }\n",
       "\n",
       "      [theme=dark] .colab-df-generate {\n",
       "        background-color: #3B4455;\n",
       "        fill: #D2E3FC;\n",
       "      }\n",
       "\n",
       "      [theme=dark] .colab-df-generate:hover {\n",
       "        background-color: #434B5C;\n",
       "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "        fill: #FFFFFF;\n",
       "      }\n",
       "    </style>\n",
       "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('df_drm_table')\"\n",
       "            title=\"Generate code using this dataframe.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "    <script>\n",
       "      (() => {\n",
       "      const buttonEl =\n",
       "        document.querySelector('#id_1de2157e-1c45-4857-b591-a89d54563799 button.colab-df-generate');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      buttonEl.onclick = () => {\n",
       "        google.colab.notebook.generateWithVariable('df_drm_table');\n",
       "      }\n",
       "      })();\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "    </div>\n",
       "  </div>\n"
      ],
      "text/plain": [
       "                              Associated with resistance to  \\\n",
       "Marker                                                        \n",
       "crt 76T                                         Chloroquine   \n",
       "dhfr 108N                                     Pyrimethamine   \n",
       "dhps 437G                                       Sulfadoxine   \n",
       "mdr1 2+ copies                                   Mefloquine   \n",
       "kelch13 WHO list                                Artemisinin   \n",
       "plasmepsin 2-3 2+ copies                        Piperaquine   \n",
       "dhfr triple mutant                           SP (treatment)   \n",
       "dhfr and dhps sextuple mutant                     SP (IPTp)   \n",
       "kelch13 and mdr1                                      AS-MQ   \n",
       "kelch13 and plasmepsin 2-3                          DHA-PPQ   \n",
       "\n",
       "                              South America (n=177-224)  \\\n",
       "Marker                                                    \n",
       "crt 76T                                            100%   \n",
       "dhfr 108N                                           74%   \n",
       "dhps 437G                                           71%   \n",
       "mdr1 2+ copies                                       0%   \n",
       "kelch13 WHO list                                     0%   \n",
       "plasmepsin 2-3 2+ copies                             0%   \n",
       "dhfr triple mutant                                   0%   \n",
       "dhfr and dhps sextuple mutant                        0%   \n",
       "kelch13 and mdr1                                     0%   \n",
       "kelch13 and plasmepsin 2-3                           0%   \n",
       "\n",
       "                              Africa - West (n=6648-8837)  \\\n",
       "Marker                                                      \n",
       "crt 76T                                               26%   \n",
       "dhfr 108N                                             89%   \n",
       "dhps 437G                                             81%   \n",
       "mdr1 2+ copies                                         0%   \n",
       "kelch13 WHO list                                       0%   \n",
       "plasmepsin 2-3 2+ copies                               0%   \n",
       "dhfr triple mutant                                    79%   \n",
       "dhfr and dhps sextuple mutant                          0%   \n",
       "kelch13 and mdr1                                       0%   \n",
       "kelch13 and plasmepsin 2-3                             0%   \n",
       "\n",
       "                              Africa - Central (n=870-1184)  \\\n",
       "Marker                                                        \n",
       "crt 76T                                                 35%   \n",
       "dhfr 108N                                               99%   \n",
       "dhps 437G                                               96%   \n",
       "mdr1 2+ copies                                           0%   \n",
       "kelch13 WHO list                                         0%   \n",
       "plasmepsin 2-3 2+ copies                                 0%   \n",
       "dhfr triple mutant                                      88%   \n",
       "dhfr and dhps sextuple mutant                            5%   \n",
       "kelch13 and mdr1                                         0%   \n",
       "kelch13 and plasmepsin 2-3                               0%   \n",
       "\n",
       "                              Africa - Northeast (n=122-204)  \\\n",
       "Marker                                                         \n",
       "crt 76T                                                  48%   \n",
       "dhfr 108N                                                98%   \n",
       "dhps 437G                                                79%   \n",
       "mdr1 2+ copies                                            0%   \n",
       "kelch13 WHO list                                          0%   \n",
       "plasmepsin 2-3 2+ copies                                  0%   \n",
       "dhfr triple mutant                                       51%   \n",
       "dhfr and dhps sextuple mutant                             2%   \n",
       "kelch13 and mdr1                                          0%   \n",
       "kelch13 and plasmepsin 2-3                                0%   \n",
       "\n",
       "                              Africa - East (n=3599-4000)  \\\n",
       "Marker                                                      \n",
       "crt 76T                                               14%   \n",
       "dhfr 108N                                             97%   \n",
       "dhps 437G                                             86%   \n",
       "mdr1 2+ copies                                         0%   \n",
       "kelch13 WHO list                                       0%   \n",
       "plasmepsin 2-3 2+ copies                               0%   \n",
       "dhfr triple mutant                                    83%   \n",
       "dhfr and dhps sextuple mutant                          4%   \n",
       "kelch13 and mdr1                                       0%   \n",
       "kelch13 and plasmepsin 2-3                             0%   \n",
       "\n",
       "                              Asia - South - East (n=81-188)  \\\n",
       "Marker                                                         \n",
       "crt 76T                                                  33%   \n",
       "dhfr 108N                                                62%   \n",
       "dhps 437G                                                 9%   \n",
       "mdr1 2+ copies                                            0%   \n",
       "kelch13 WHO list                                          0%   \n",
       "plasmepsin 2-3 2+ copies                                  0%   \n",
       "dhfr triple mutant                                        2%   \n",
       "dhfr and dhps sextuple mutant                             0%   \n",
       "kelch13 and mdr1                                          0%   \n",
       "kelch13 and plasmepsin 2-3                                0%   \n",
       "\n",
       "                              Asia - South - Far East (n=1209-1369)  \\\n",
       "Marker                                                                \n",
       "crt 76T                                                         94%   \n",
       "dhfr 108N                                                      100%   \n",
       "dhps 437G                                                       89%   \n",
       "mdr1 2+ copies                                                   0%   \n",
       "kelch13 WHO list                                                 0%   \n",
       "plasmepsin 2-3 2+ copies                                         0%   \n",
       "dhfr triple mutant                                              46%   \n",
       "dhfr and dhps sextuple mutant                                   13%   \n",
       "kelch13 and mdr1                                                 0%   \n",
       "kelch13 and plasmepsin 2-3                                       0%   \n",
       "\n",
       "                              Asia - Southeast - West (n=1665-1884)  \\\n",
       "Marker                                                                \n",
       "crt 76T                                                         99%   \n",
       "dhfr 108N                                                      100%   \n",
       "dhps 437G                                                      100%   \n",
       "mdr1 2+ copies                                                  30%   \n",
       "kelch13 WHO list                                                36%   \n",
       "plasmepsin 2-3 2+ copies                                         0%   \n",
       "dhfr triple mutant                                              86%   \n",
       "dhfr and dhps sextuple mutant                                   79%   \n",
       "kelch13 and mdr1                                                10%   \n",
       "kelch13 and plasmepsin 2-3                                       0%   \n",
       "\n",
       "                              Asia - Southeast - East (n=2910-5798)  \\\n",
       "Marker                                                                \n",
       "crt 76T                                                         96%   \n",
       "dhfr 108N                                                       99%   \n",
       "dhps 437G                                                       83%   \n",
       "mdr1 2+ copies                                                   4%   \n",
       "kelch13 WHO list                                                62%   \n",
       "plasmepsin 2-3 2+ copies                                        43%   \n",
       "dhfr triple mutant                                              87%   \n",
       "dhfr and dhps sextuple mutant                                   12%   \n",
       "kelch13 and mdr1                                                 3%   \n",
       "kelch13 and plasmepsin 2-3                                      39%   \n",
       "\n",
       "                              Oceania - New Guinea (n=293-340)  \n",
       "Marker                                                          \n",
       "crt 76T                                                    96%  \n",
       "dhfr 108N                                                  99%  \n",
       "dhps 437G                                                  69%  \n",
       "mdr1 2+ copies                                              1%  \n",
       "kelch13 WHO list                                            1%  \n",
       "plasmepsin 2-3 2+ copies                                    0%  \n",
       "dhfr triple mutant                                          0%  \n",
       "dhfr and dhps sextuple mutant                               0%  \n",
       "kelch13 and mdr1                                            0%  \n",
       "kelch13 and plasmepsin 2-3                                  0%  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a table summarizing drug resistance proportions by population\n",
    "\n",
    "# Group the DataFrame by 'Population' and apply the 'proportion_agg' function to calculate resistance proportions\n",
    "df_drm_table = (\n",
    "    df_all_sample_metadata\n",
    "    .groupby('Population')\n",
    "    .apply(proportion_agg, include_groups=False)\n",
    "    .rename_axis(None)\n",
    "    .transpose()\n",
    "    .loc[:, populations.keys()]\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "# Customize the index to include gene and mutation information for each drug\n",
    "df_drm_table.index = [\"%s %s\" % (drm_dict[drug]['gene'], drm_dict[drug]['mutation']) for drug in drm_dict]\n",
    "df_drm_table.index.names = ['Marker']\n",
    "\n",
    "# Rename the columns to indicate their association with drug resistance\n",
    "df_drm_table.rename(columns={'index': 'Associated with resistance to'}, inplace=True)\n",
    "\n",
    "# Update column names to include population names and corresponding sample size ranges\n",
    "for population in populations:\n",
    "    new_column_name = f'{populations[population]} (n={min_n[population]}-{max_n[population]})'\n",
    "    df_drm_table.rename(columns={population: new_column_name}, inplace=True)\n",
    "\n",
    "# Add '%' symbol to all values in the table\n",
    "df_drm_table = df_drm_table.map(lambda x: f\"{int(x)}%\" if isinstance(x, (int, float)) else x)\n",
    "\n",
    "# Display the final drug resistance table\n",
    "df_drm_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MoMG3wxmjtoo"
   },
   "source": [
    "**Table 2.** Frequency of different sets of polymorphisms associated with drug resistance in samples from different geographical regions.\n",
    "All samples were classified into different types of drug resistance based on published genetic markers, and represent best attempt based on the available data. Each type of resistance was considered to be either present, absent or unknown for a given sample. For each resistance type, the table reports: the genetic markers considered; the drug they are associated with; the proportion of samples in each major sub-population classified as resistant out of the samples where the type was not unknown. The number of samples classified as either resistant or not resistant varies for each type of resistance considered (e.g. due to different levels of genomic accessibility); numbers in brackets report the minimum and maximum number analysed while the exact numbers considered are reported in Table 1 of this notebook. SP: sulfadoxine-pyrimethamine; treatment: SP used for the clinical treatment of uncomplicated malaria; IPTp: SP used for intermittent preventive treatment in pregnancy; AS-MQ: artesunate + mefloquine combination therapy; DHA-PPQ: dihydroartemisinin + piperaquine combination therapy. *dhfr* triple mutant refers to having all three of 51I, 59R and 108N in *dhfr*. *dhfr* and *dhps* sextuple mutant refers to having all five of 51I, 59R and 108N in *dhfr* and 437G and 540E in *dhps*, plus one of *dhfr*:164L, *dhps*:581G, *dhps*:613S or *dhps*:613T. Full details of the rules used to infer resistance status from genetic markers can be found [here](https://pf8-release.cog.sanger.ac.uk/Pf8_resistance_classification.pdf)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "lJnlyrhnQJqz"
   },
   "source": [
    "To save the table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "id": "jmN7Fuxo0qyy"
   },
   "outputs": [],
   "source": [
    "df_drm_table.to_excel(\"DRM_table.xlsx\")"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
