
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Geographic patterns of population differentiation and gene flow &#8212; MalariaGEN parasite data user guide</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]], "processRefs": false, "processEnvironments": false}})</script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Neighbour Joining Tree for sWGA vs. gDNA" href="neighbour-joining-tree-swga.html" />
    <link rel="prev" title="Principal Coordinate Analysis" href="PCoA.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">MalariaGEN parasite data user guide</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../landing-page.html">
   MalariaGEN parasite data user guide
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Pf7
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../Data_access.html">
   Pf7 Data Access
  </a>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../examples.html">
   Pf7 Analysis Examples
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="igv-notebook.html">
     Visualizing Genomic Regions of Pf7 Samples with IGV Plots
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sample-summary-table.html">
     Summarise Samples by Administrative Division
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sample-collection-over-time.html">
     Visualising sample collections over time
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="barplot-samples-by-country.html">
     Barplot of Samples by Country
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="variant-summary-table.html">
     Summarise Genetic Variants
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="global-sample-map.html">
     Plot Global Map of Samples
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="drug-resistance-summary.html">
     Drug resistance frequencies across Pf7
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="chloroquine-resistance-map.html">
     Plot Map of Chloroquine Resistance
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="summarise_hrp2_hrp3_deletions.html">
     <em>
      hrp2
     </em>
     and
     <em>
      hrp3
     </em>
     deletions: summary of deletion breakpoints
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="visualise_hrp2_hrp3_deletions.html">
     Visualise HRP Deletion Breakpoints
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="csp-cterm-analysis.html">
     Analyse csp C-Terminal Variation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="PCoA.html">
     Principal Coordinate Analysis
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Geographic patterns of population differentiation and gene flow
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="neighbour-joining-tree-swga.html">
     Neighbour Joining Tree for sWGA vs. gDNA
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="proportion-eba175.html">
     Plot
     <em>
      eba175
     </em>
     C Allele Proportions
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../api.html">
   Pf7 API
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Pv4
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../pv4/Data_access.html">
   Pv4 Data Access
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../pv4/api.html">
   Pv4 API
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/pf7/analysis_examples/geographic-genetic-differentiation.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
                onclick="printPdf(this)" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/malariagen/parasite-data"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/malariagen/parasite-data/issues/new?title=Issue%20on%20page%20%2Fpf7/analysis_examples/geographic-genetic-differentiation.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/malariagen/parasite-data/master?urlpath=lab/tree/docs/pf7/analysis_examples/geographic-genetic-differentiation.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        <a class="colab-button" href="https://colab.research.google.com/github/malariagen/parasite-data/blob/master/docs/pf7/analysis_examples/geographic-genetic-differentiation.ipynb"><button type="button" class="btn btn-secondary topbarbtn"
                title="Launch Colab" data-toggle="tooltip" data-placement="left"><img class="colab-button-logo"
                    src="../../_static/images/logo_colab.png"
                    alt="Interact on Colab">Colab</button></a>
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#access-pf7-data">
   Access Pf7 Data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#figure-preparation-defining-geographic-regions">
   Figure Preparation: Defining Geographic Regions
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#calculating-geographic-distance">
   Calculating Geographic Distance
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#access-to-mean-f-st-values-between-all-major-sub-subpopulations">
   Access to mean
   <span class="math notranslate nohighlight">
    \(F_{ST}\)
   </span>
   values between all major sub-subpopulations
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#plotting-the-genetic-geographic-distance-between-major-sub-populations">
   Plotting the genetic-geographic distance between major sub-populations
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#save-the-figure">
     Save the figure
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Geographic patterns of population differentiation and gene flow</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#access-pf7-data">
   Access Pf7 Data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#figure-preparation-defining-geographic-regions">
   Figure Preparation: Defining Geographic Regions
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#calculating-geographic-distance">
   Calculating Geographic Distance
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#access-to-mean-f-st-values-between-all-major-sub-subpopulations">
   Access to mean
   <span class="math notranslate nohighlight">
    \(F_{ST}\)
   </span>
   values between all major sub-subpopulations
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#plotting-the-genetic-geographic-distance-between-major-sub-populations">
   Plotting the genetic-geographic distance between major sub-populations
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#save-the-figure">
     Save the figure
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            
              <div>
                
  <div class="section" id="geographic-patterns-of-population-differentiation-and-gene-flow">
<h1>Geographic patterns of population differentiation and gene flow<a class="headerlink" href="#geographic-patterns-of-population-differentiation-and-gene-flow" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>This notebook recreates <a class="reference external" href="https://figshare.com/ndownloader/files/38478530">Supplementary Table 6 and 7 from the Pf7 paper</a> - tables which depict the genetic distance (<span class="math notranslate nohighlight">\(F_{ST}\)</span>) and geographic distance between subpopulations. It contains two panels, A) which compares African subpopulations and B) which compares Asian Southeast West and Asian Southeast East populations with all other subpopulations.</p>
<p><strong>This notebook should take approximately 2 minutes to run.</strong></p>
</div>
<div class="section" id="setup">
<h2>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">¶</a></h2>
<p>Install and import the malariagen Python package:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>!pip install -q --no-warn-conflicts malariagen_data
import malariagen_data
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">148.5/148.5 kB</span> <span class=" -Color -Color-Red">3.1 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">3.1/3.1 MB</span> <span class=" -Color -Color-Red">41.4 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">52.8/52.8 MB</span> <span class=" -Color -Color-Red">16.2 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">10.2/10.2 MB</span> <span class=" -Color -Color-Red">83.0 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">4.0/4.0 MB</span> <span class=" -Color -Color-Red">75.5 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">302.5/302.5 kB</span> <span class=" -Color -Color-Red">21.9 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">138.7/138.7 kB</span> <span class=" -Color -Color-Red">8.4 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">20.9/20.9 MB</span> <span class=" -Color -Color-Red">30.3 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">8.1/8.1 MB</span> <span class=" -Color -Color-Red">46.1 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">206.9/206.9 kB</span> <span class=" -Color -Color-Red">3.1 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
?25h  Preparing metadata (setup.py) ... ?25l?25hdone
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">7.7/7.7 MB</span> <span class=" -Color -Color-Red">53.3 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">1.6/1.6 MB</span> <span class=" -Color -Color-Red">28.1 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
?25h  Building wheel for asciitree (setup.py) ... ?25l?25hdone
</pre></div>
</div>
</div>
</div>
<p>Import required python libraries that are installed at colab by default.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># required for data formatting</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">Any</span><span class="p">,</span> <span class="n">List</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="c1"># required to calculate great circle distance</span>
<span class="kn">from</span> <span class="nn">math</span> <span class="kn">import</span> <span class="n">radians</span><span class="p">,</span> <span class="n">cos</span><span class="p">,</span> <span class="n">sin</span><span class="p">,</span> <span class="n">asin</span><span class="p">,</span> <span class="n">sqrt</span>
<span class="c1"># required for plotting</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">matplotlib.lines</span> <span class="kn">import</span> <span class="n">Line2D</span>
<span class="kn">import</span> <span class="nn">matplotlib.patches</span> <span class="k">as</span> <span class="nn">mpatches</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">matplotlib.ticker</span> <span class="k">as</span> <span class="nn">ticker</span>
<span class="c1"># required for saving the figure</span>
<span class="kn">from</span> <span class="nn">google.colab</span> <span class="kn">import</span> <span class="n">drive</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="access-pf7-data">
<h2>Access Pf7 Data<a class="headerlink" href="#access-pf7-data" title="Permalink to this headline">¶</a></h2>
<p>We use the malariagen data package to load the release data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">release_data</span> <span class="o">=</span> <span class="n">malariagen_data</span><span class="o">.</span><span class="n">Pf7</span><span class="p">()</span>
<span class="n">df_samples</span> <span class="o">=</span> <span class="n">release_data</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="figure-preparation-defining-geographic-regions">
<h2>Figure Preparation: Defining Geographic Regions<a class="headerlink" href="#figure-preparation-defining-geographic-regions" title="Permalink to this headline">¶</a></h2>
<p>In Pf7, countries are grouped into ten major sub-populations based on their geographic and genetic characteristics.</p>
<p><code class="docutils literal notranslate"><span class="pre">df_samples</span></code> dataframe has a <code class="docutils literal notranslate"><span class="pre">Population</span></code> column that contains abbreviated names for each sub-population to which we can assign a specific colour to be used in the figure.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">georegions</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">([</span>
    <span class="p">(</span><span class="s1">&#39;SA&#39;</span><span class="p">,</span><span class="nb">dict</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="s1">&#39;South America&#39;</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;#4daf4a&#39;</span><span class="p">)),</span>
    <span class="p">(</span><span class="s1">&#39;all_africa&#39;</span><span class="p">,</span><span class="nb">dict</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="s1">&#39;Africa&#39;</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;#faedcb&#39;</span><span class="p">)),</span>
    <span class="p">(</span><span class="s1">&#39;AF-W&#39;</span><span class="p">,</span><span class="nb">dict</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="s1">&#39;West Africa&#39;</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;#e31a1c&#39;</span><span class="p">)),</span>
    <span class="p">(</span><span class="s1">&#39;AF-C&#39;</span><span class="p">,</span><span class="nb">dict</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="s1">&#39;Central Africa&#39;</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;#fd8d3c&#39;</span><span class="p">)),</span>
    <span class="p">(</span><span class="s1">&#39;AF-NE&#39;</span><span class="p">,</span><span class="nb">dict</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="s1">&#39;Northeast Africa&#39;</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;#bb8129&#39;</span><span class="p">)),</span>
    <span class="p">(</span><span class="s1">&#39;AF-E&#39;</span><span class="p">,</span><span class="nb">dict</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="s1">&#39;East Africa&#39;</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;#fecc5c&#39;</span><span class="p">)),</span>

    <span class="p">(</span><span class="s1">&#39;AS-S-FE&#39;</span><span class="p">,</span><span class="nb">dict</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="s1">&#39;Asia - South - West&#39;</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;#dfc0eb&#39;</span><span class="p">)),</span>
    <span class="p">(</span><span class="s1">&#39;AS-S-E&#39;</span><span class="p">,</span><span class="nb">dict</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="s1">&#39;Asia - South - East&#39;</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;#984ea3&#39;</span><span class="p">)),</span>

    <span class="p">(</span><span class="s1">&#39;AS-SE-W&#39;</span><span class="p">,</span><span class="nb">dict</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="s1">&#39;Asia - Southeast - West&#39;</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;#9ecae1&#39;</span><span class="p">)),</span>
    <span class="p">(</span><span class="s1">&#39;AS-SE-E&#39;</span><span class="p">,</span><span class="nb">dict</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="s1">&#39;Asia - Southeast - East&#39;</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;#3182bd&#39;</span><span class="p">)),</span>

    <span class="p">(</span><span class="s1">&#39;OC-NG&#39;</span><span class="p">,</span><span class="nb">dict</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="s1">&#39;Oceania - New Guinea&#39;</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;#f781bf&#39;</span><span class="p">))])</span>
</pre></div>
</div>
</div>
</div>
<p>We would like to plot African and Asian sub-populations separately to do a region-wise comparison.</p>
<p>We can do this by mapping the continent name codes using subpopulation name codes existing in the <code class="docutils literal notranslate"><span class="pre">Population</span></code> column of <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Match continent name codes with subpopulations: AM (America), AF (Africa), AS (Asia), OC (Oceania)</span>
<span class="n">continent_codes</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;SA&#39;</span><span class="p">:</span> <span class="s1">&#39;AM&#39;</span><span class="p">,</span> <span class="s1">&#39;AF-W&#39;</span><span class="p">:</span> <span class="s1">&#39;AF&#39;</span><span class="p">,</span> <span class="s1">&#39;AF-C&#39;</span><span class="p">:</span> <span class="s1">&#39;AF&#39;</span><span class="p">,</span> <span class="s1">&#39;AF-NE&#39;</span><span class="p">:</span> <span class="s1">&#39;AF&#39;</span><span class="p">,</span> <span class="s1">&#39;AF-E&#39;</span><span class="p">:</span> <span class="s1">&#39;AF&#39;</span><span class="p">,</span>
                  <span class="s1">&#39;AS-S-FE&#39;</span><span class="p">:</span> <span class="s1">&#39;AS&#39;</span><span class="p">,</span> <span class="s1">&#39;AS-S-E&#39;</span><span class="p">:</span> <span class="s1">&#39;AS&#39;</span><span class="p">,</span> <span class="s1">&#39;AS-SE-W&#39;</span><span class="p">:</span> <span class="s1">&#39;AS&#39;</span><span class="p">,</span> <span class="s1">&#39;AS-SE-E&#39;</span><span class="p">:</span> <span class="s1">&#39;AS&#39;</span><span class="p">,</span> <span class="s1">&#39;OC-NG&#39;</span><span class="p">:</span> <span class="s1">&#39;OC&#39;</span><span class="p">}</span>
<span class="c1"># Map the continent name codes using population name codes</span>
<span class="n">df_samples</span><span class="p">[</span><span class="s1">&#39;ContinentCode&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_samples</span><span class="p">[</span><span class="s1">&#39;Population&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">continent_codes</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now that we added continents, we can find out how many countries are assigned in each continent.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Count the country in each continent</span>
<span class="n">df_samples</span><span class="p">[</span><span class="s1">&#39;ContinentCode&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">(</span><span class="n">dropna</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AF     11290
AS      8848
OC       384
AM       182
NaN      160
Name: ContinentCode, dtype: int64
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="calculating-geographic-distance">
<h2>Calculating Geographic Distance<a class="headerlink" href="#calculating-geographic-distance" title="Permalink to this headline">¶</a></h2>
<p>We can calculate the geographic distance between between two points (on our case, populations) on the Earth’s surface using their latitude and longitude coordinates.</p>
<p>We will create a custom function for this task.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a function</span>
<span class="k">def</span> <span class="nf">haversine</span><span class="p">(</span><span class="n">lon1</span><span class="p">,</span> <span class="n">lat1</span><span class="p">,</span> <span class="n">lon2</span><span class="p">,</span> <span class="n">lat2</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate the great circle distance between two points on the earth (specified in decimal degrees)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Convert decimal degrees to radians</span>
    <span class="n">lon1</span><span class="p">,</span> <span class="n">lat1</span><span class="p">,</span> <span class="n">lon2</span><span class="p">,</span> <span class="n">lat2</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">radians</span><span class="p">,</span> <span class="p">[</span><span class="n">lon1</span><span class="p">,</span> <span class="n">lat1</span><span class="p">,</span> <span class="n">lon2</span><span class="p">,</span> <span class="n">lat2</span><span class="p">])</span>
    <span class="c1"># Haversine formula</span>
    <span class="n">dlon</span> <span class="o">=</span> <span class="n">lon2</span> <span class="o">-</span> <span class="n">lon1</span>
    <span class="n">dlat</span> <span class="o">=</span> <span class="n">lat2</span> <span class="o">-</span> <span class="n">lat1</span>
    <span class="n">a</span> <span class="o">=</span> <span class="n">sin</span><span class="p">(</span><span class="n">dlat</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span><span class="o">**</span><span class="mi">2</span> <span class="o">+</span> <span class="n">cos</span><span class="p">(</span><span class="n">lat1</span><span class="p">)</span> <span class="o">*</span> <span class="n">cos</span><span class="p">(</span><span class="n">lat2</span><span class="p">)</span> <span class="o">*</span> <span class="n">sin</span><span class="p">(</span><span class="n">dlon</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span><span class="o">**</span><span class="mi">2</span>
    <span class="n">c</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">asin</span><span class="p">(</span><span class="n">sqrt</span><span class="p">(</span><span class="n">a</span><span class="p">))</span>
    <span class="n">km</span> <span class="o">=</span> <span class="mi">6367</span> <span class="o">*</span> <span class="n">c</span>
    <span class="k">return</span> <span class="n">km</span>
</pre></div>
</div>
</div>
</div>
<p>Now, it is time to calculate the geographic distances between populations. To step-by-step do that:</p>
<ul class="simple">
<li><p>we will create pair population pairs</p></li>
<li><p>calculates distances between them</p></li>
<li><p>stores the results in a structured DataFrame</p></li>
</ul>
<p>Only QC-pass samples will be included in the calculation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Filter samples based on quality control criteria</span>
<span class="n">filtered_samples</span> <span class="o">=</span> <span class="n">df_samples</span><span class="p">[</span><span class="n">df_samples</span><span class="p">[</span><span class="s1">&#39;QC pass&#39;</span><span class="p">]]</span>

<span class="c1"># Define aggregation parameters for latitude and longitude</span>
<span class="n">aggregation</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;Country latitude&#39;</span><span class="p">:</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span>
        <span class="s1">&#39;Country longitude&#39;</span><span class="p">:</span> <span class="s1">&#39;mean&#39;</span>
    <span class="p">}</span>
<span class="c1"># Group samples by the &#39;Population&#39; column and perform aggregation</span>
<span class="n">gc</span> <span class="o">=</span> <span class="n">filtered_samples</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Population&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="n">aggregation</span><span class="p">)</span>

<span class="c1"># Create a list of population pairs, excluding self-pairs</span>
<span class="n">pop_pairs_list</span> <span class="o">=</span> <span class="p">[[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">gc</span><span class="o">.</span><span class="n">index</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">gc</span><span class="o">.</span><span class="n">index</span> <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="n">j</span><span class="p">]</span>

<span class="c1"># Create a MultiIndex to represent the pairs</span>
<span class="n">geo_index</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">MultiIndex</span><span class="o">.</span><span class="n">from_tuples</span><span class="p">(</span><span class="n">pop_pairs_list</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;first&#39;</span><span class="p">,</span> <span class="s1">&#39;second&#39;</span><span class="p">])</span>

<span class="c1"># Initialize a DataFrame to store distances</span>
<span class="n">geo_distances</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;distance&#39;</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="n">geo_index</span><span class="p">)</span>

<span class="c1"># Calculate distances and populate the &#39;distance&#39; column in geo_distances</span>
<span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">geo_distances</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="n">geo_distances</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">f</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">distance</span> <span class="o">=</span> <span class="n">haversine</span><span class="p">(</span>
        <span class="n">lat1</span><span class="o">=</span><span class="n">gc</span><span class="p">[</span><span class="s1">&#39;Country latitude&#39;</span><span class="p">][</span><span class="n">f</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]],</span>
        <span class="n">lon1</span><span class="o">=</span><span class="n">gc</span><span class="p">[</span><span class="s1">&#39;Country longitude&#39;</span><span class="p">][</span><span class="n">f</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]],</span>
        <span class="n">lat2</span><span class="o">=</span><span class="n">gc</span><span class="p">[</span><span class="s1">&#39;Country latitude&#39;</span><span class="p">][</span><span class="n">f</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]],</span>
        <span class="n">lon2</span><span class="o">=</span><span class="n">gc</span><span class="p">[</span><span class="s1">&#39;Country longitude&#39;</span><span class="p">][</span><span class="n">f</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]]</span>
    <span class="p">)</span>
<span class="c1"># Print the results</span>
<span class="n">geo_distances</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-bdb416b7-be6e-4c5b-8469-e97b1f24afa2" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>distance</th>
    </tr>
    <tr>
      <th>first</th>
      <th>second</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="5" valign="top">AF-C</th>
      <th>AF-E</th>
      <td>1408.618195</td>
    </tr>
    <tr>
      <th>AF-NE</th>
      <td>1717.731717</td>
    </tr>
    <tr>
      <th>AF-W</th>
      <td>3354.887193</td>
    </tr>
    <tr>
      <th>AS-S-E</th>
      <td>6700.558535</td>
    </tr>
    <tr>
      <th>AS-S-FE</th>
      <td>7724.625897</td>
    </tr>
    <tr>
      <th>...</th>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th rowspan="5" valign="top">SA</th>
      <th>AS-S-E</th>
      <td>15945.593911</td>
    </tr>
    <tr>
      <th>AS-S-FE</th>
      <td>16574.510674</td>
    </tr>
    <tr>
      <th>AS-SE-E</th>
      <td>17994.185511</td>
    </tr>
    <tr>
      <th>AS-SE-W</th>
      <td>17558.854817</td>
    </tr>
    <tr>
      <th>OC-NG</th>
      <td>16823.02993</td>
    </tr>
  </tbody>
</table>
<p>90 rows × 1 columns</p>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-bdb416b7-be6e-4c5b-8469-e97b1f24afa2')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-bdb416b7-be6e-4c5b-8469-e97b1f24afa2 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-bdb416b7-be6e-4c5b-8469-e97b1f24afa2');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-e9642980-ebe4-4be7-80c8-bf69d3fa9df8">
  <button class="colab-df-quickchart" onclick="quickchart('df-e9642980-ebe4-4be7-80c8-bf69d3fa9df8')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-e9642980-ebe4-4be7-80c8-bf69d3fa9df8 button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
</div>
<div class="section" id="access-to-mean-f-st-values-between-all-major-sub-subpopulations">
<h2>Access to mean <span class="math notranslate nohighlight">\(F_{ST}\)</span> values between all major sub-subpopulations<a class="headerlink" href="#access-to-mean-f-st-values-between-all-major-sub-subpopulations" title="Permalink to this headline">¶</a></h2>
<p><span class="math notranslate nohighlight">\(F_{ST}\)</span>, also known as the Fixation Index, is a measure of how genetic differences within a population compare to genetic differences between populations. <span class="math notranslate nohighlight">\(F_{ST}\)</span> = 0 means there is no difference between populations, conversely, <span class="math notranslate nohighlight">\(F_{ST}\)</span> = 1 indicates there is complete genetic differentiation between populations. For a deeper understanding on <span class="math notranslate nohighlight">\(F_{ST}\)</span> statistics, consider referring to the following resources:</p>
<ul class="simple">
<li><p><strong><a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1205159/">Estimation of Levels of Gene Flow from DNA Sequence Data</a></strong> by <em>Hudson et al.</em></p></li>
<li><p><strong><a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3759727/">Estimating and interpreting FST: The impact of rare variants</a></strong> by <em>Bhatia et al.</em></p></li>
<li><p><strong><a class="reference external" href="https://www.youtube.com/watch?v=I8RCOI7n4XI">A short video about <span class="math notranslate nohighlight">\(F_{ST}\)</span> statistics</a></strong> by <em>Sarah Tishkoff</em> from U. Pennsylvania</p></li>
</ul>
<p>In this notebook, we will leverage pre-calculated mean <span class="math notranslate nohighlight">\(F_{ST}\)</span> values between all major sub-subpopulations, which were previously computed for Pf7 using <a class="reference external" href="https://scikit-allel.readthedocs.io/en/stable/stats/fst.html">Hudson’s method as implemented in scikit-allel v1.2.0</a></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Our website is undergoing some maintenance, please use this temporary URL</span>
<span class="c1"># Try this URL</span>
<span class="n">fst</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;https://www.malariagen.net/wp-content/uploads/2024/01/fst_populations_all_snps_pf7.txt&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># But if it does not work, try this URL (uncomment the command)</span>
<span class="c1">#fst = pd.read_csv(&#39;https://www.malariagen.net/sites/default/files/fst_populations_all_snps_pf7.txt&#39;, sep=&#39;\t&#39;, index_col=0)</span>

<span class="c1"># Rename subpopulation names in accordance with Pf7</span>
<span class="n">fst</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;AS-SA-W&quot;</span><span class="p">:</span> <span class="s2">&quot;AS-S-FE&quot;</span><span class="p">,</span> <span class="s2">&quot;AS-SA-E&quot;</span><span class="p">:</span> <span class="s2">&quot;AS-S-E&quot;</span><span class="p">,</span> <span class="s1">&#39;AS-SEA-W&#39;</span><span class="p">:</span> <span class="s1">&#39;AS-SE-W&#39;</span><span class="p">,</span> <span class="s1">&#39;AS-SEA-E&#39;</span> <span class="p">:</span> <span class="s1">&#39;AS-SE-E&#39;</span><span class="p">},</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Rename the index values with the new subpopulation names</span>
<span class="n">fst</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">fst</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<span class="n">fst</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-1f5c5037-82bb-4fca-b162-ddebc5edece1" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>SA</th>
      <th>AF-W</th>
      <th>AF-C</th>
      <th>AF-NE</th>
      <th>AF-E</th>
      <th>AS-S-FE</th>
      <th>AS-S-E</th>
      <th>AS-SE-W</th>
      <th>AS-SE-E</th>
      <th>OC-NG</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>SA</th>
      <td>0.0</td>
      <td>0.210957</td>
      <td>0.216355</td>
      <td>0.221376</td>
      <td>0.215139</td>
      <td>0.257292</td>
      <td>0.273814</td>
      <td>0.348240</td>
      <td>0.376699</td>
      <td>0.345133</td>
    </tr>
    <tr>
      <th>AF-W</th>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.015034</td>
      <td>0.030750</td>
      <td>0.021481</td>
      <td>0.089009</td>
      <td>0.111180</td>
      <td>0.200027</td>
      <td>0.229990</td>
      <td>0.201879</td>
    </tr>
    <tr>
      <th>AF-C</th>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.024246</td>
      <td>0.009104</td>
      <td>0.084730</td>
      <td>0.106478</td>
      <td>0.193681</td>
      <td>0.224552</td>
      <td>0.196541</td>
    </tr>
    <tr>
      <th>AF-NE</th>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.020444</td>
      <td>0.087930</td>
      <td>0.107703</td>
      <td>0.187708</td>
      <td>0.219425</td>
      <td>0.192452</td>
    </tr>
    <tr>
      <th>AF-E</th>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.081217</td>
      <td>0.101255</td>
      <td>0.185666</td>
      <td>0.215896</td>
      <td>0.189025</td>
    </tr>
    <tr>
      <th>AS-S-FE</th>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.031117</td>
      <td>0.115080</td>
      <td>0.151120</td>
      <td>0.145261</td>
    </tr>
    <tr>
      <th>AS-S-E</th>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.064807</td>
      <td>0.102968</td>
      <td>0.121421</td>
    </tr>
    <tr>
      <th>AS-SE-W</th>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.070569</td>
      <td>0.134227</td>
    </tr>
    <tr>
      <th>AS-SE-E</th>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.159498</td>
    </tr>
    <tr>
      <th>OC-NG</th>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-1f5c5037-82bb-4fca-b162-ddebc5edece1')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-1f5c5037-82bb-4fca-b162-ddebc5edece1 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-1f5c5037-82bb-4fca-b162-ddebc5edece1');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-81326c69-e52c-4782-9b33-97d32c48b4cc">
  <button class="colab-df-quickchart" onclick="quickchart('df-81326c69-e52c-4782-9b33-97d32c48b4cc')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-81326c69-e52c-4782-9b33-97d32c48b4cc button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
<p>This is an upper triangle matrix, which contains many 0 values that we can replace with the true values.</p>
<p>At the end, we will have a symmetric matrix, and only self-pairs will be 0.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># We can pretty easily make into a symmetric matrix using the handy numpy function triu along with a transpose</span>
<span class="n">arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">triu</span><span class="p">(</span><span class="n">fst</span><span class="p">)</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">triu</span><span class="p">(</span><span class="n">fst</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
<span class="c1"># Get the population names in a list for symmetric dataframe</span>
<span class="n">pop_names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">fst</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="c1"># Create the symmetric matrix</span>
<span class="n">fst_symmetric</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">arr</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="n">columns</span><span class="o">=</span><span class="n">pop_names</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">pop_names</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
<span class="n">fst_symmetric</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-2ceb5cfb-9057-4dd5-a11e-ecc4af3feae1" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>SA</th>
      <th>AF-W</th>
      <th>AF-C</th>
      <th>AF-NE</th>
      <th>AF-E</th>
      <th>AS-S-FE</th>
      <th>AS-S-E</th>
      <th>AS-SE-W</th>
      <th>AS-SE-E</th>
      <th>OC-NG</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>SA</th>
      <td>NaN</td>
      <td>0.210957</td>
      <td>0.216355</td>
      <td>0.221376</td>
      <td>0.215139</td>
      <td>0.257292</td>
      <td>0.273814</td>
      <td>0.348240</td>
      <td>0.376699</td>
      <td>0.345133</td>
    </tr>
    <tr>
      <th>AF-W</th>
      <td>0.210957</td>
      <td>NaN</td>
      <td>0.015034</td>
      <td>0.030750</td>
      <td>0.021481</td>
      <td>0.089009</td>
      <td>0.111180</td>
      <td>0.200027</td>
      <td>0.229990</td>
      <td>0.201879</td>
    </tr>
    <tr>
      <th>AF-C</th>
      <td>0.216355</td>
      <td>0.015034</td>
      <td>NaN</td>
      <td>0.024246</td>
      <td>0.009104</td>
      <td>0.084730</td>
      <td>0.106478</td>
      <td>0.193681</td>
      <td>0.224552</td>
      <td>0.196541</td>
    </tr>
    <tr>
      <th>AF-NE</th>
      <td>0.221376</td>
      <td>0.030750</td>
      <td>0.024246</td>
      <td>NaN</td>
      <td>0.020444</td>
      <td>0.087930</td>
      <td>0.107703</td>
      <td>0.187708</td>
      <td>0.219425</td>
      <td>0.192452</td>
    </tr>
    <tr>
      <th>AF-E</th>
      <td>0.215139</td>
      <td>0.021481</td>
      <td>0.009104</td>
      <td>0.020444</td>
      <td>NaN</td>
      <td>0.081217</td>
      <td>0.101255</td>
      <td>0.185666</td>
      <td>0.215896</td>
      <td>0.189025</td>
    </tr>
    <tr>
      <th>AS-S-FE</th>
      <td>0.257292</td>
      <td>0.089009</td>
      <td>0.084730</td>
      <td>0.087930</td>
      <td>0.081217</td>
      <td>NaN</td>
      <td>0.031117</td>
      <td>0.115080</td>
      <td>0.151120</td>
      <td>0.145261</td>
    </tr>
    <tr>
      <th>AS-S-E</th>
      <td>0.273814</td>
      <td>0.111180</td>
      <td>0.106478</td>
      <td>0.107703</td>
      <td>0.101255</td>
      <td>0.031117</td>
      <td>NaN</td>
      <td>0.064807</td>
      <td>0.102968</td>
      <td>0.121421</td>
    </tr>
    <tr>
      <th>AS-SE-W</th>
      <td>0.348240</td>
      <td>0.200027</td>
      <td>0.193681</td>
      <td>0.187708</td>
      <td>0.185666</td>
      <td>0.115080</td>
      <td>0.064807</td>
      <td>NaN</td>
      <td>0.070569</td>
      <td>0.134227</td>
    </tr>
    <tr>
      <th>AS-SE-E</th>
      <td>0.376699</td>
      <td>0.229990</td>
      <td>0.224552</td>
      <td>0.219425</td>
      <td>0.215896</td>
      <td>0.151120</td>
      <td>0.102968</td>
      <td>0.070569</td>
      <td>NaN</td>
      <td>0.159498</td>
    </tr>
    <tr>
      <th>OC-NG</th>
      <td>0.345133</td>
      <td>0.201879</td>
      <td>0.196541</td>
      <td>0.192452</td>
      <td>0.189025</td>
      <td>0.145261</td>
      <td>0.121421</td>
      <td>0.134227</td>
      <td>0.159498</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-2ceb5cfb-9057-4dd5-a11e-ecc4af3feae1')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-2ceb5cfb-9057-4dd5-a11e-ecc4af3feae1 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-2ceb5cfb-9057-4dd5-a11e-ecc4af3feae1');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-3ebfe95e-c2e3-4b4d-9daa-853363687de6">
  <button class="colab-df-quickchart" onclick="quickchart('df-3ebfe95e-c2e3-4b4d-9daa-853363687de6')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-3ebfe95e-c2e3-4b4d-9daa-853363687de6 button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
<p>Now, both geographic and genetic distance data are ready for plotting.</p>
</div>
<div class="section" id="plotting-the-genetic-geographic-distance-between-major-sub-populations">
<h2>Plotting the genetic-geographic distance between major sub-populations<a class="headerlink" href="#plotting-the-genetic-geographic-distance-between-major-sub-populations" title="Permalink to this headline">¶</a></h2>
<p>We will start creating Subplot 1 which shows the geo-genetic distances between African subpopulations and Subplot 2 which shows Asian Southeast West and Asian Southeast East populations with all other subpopulations.</p>
<p>Briefly, after adjusting various figure parameters, we will iterate through combinations of pairs of populations and retrieve their corresponding geographic distance and Fst values.</p>
<p>For visual clarity, we join pairs of markers in Subplot 2 with connecting lines.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Adjust font, label and tick sizes</span>
<span class="n">rcParams</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span>
<span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;font.size&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">8</span>
<span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;axes.labelsize&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">14</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;xtick.labelsize&#39;</span><span class="p">]</span><span class="o">=</span><span class="mi">12</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;ytick.labelsize&#39;</span><span class="p">]</span><span class="o">=</span><span class="mi">12</span>

<span class="c1"># Create 2 subplots, one for Africa (ax) and one for Asia&amp;others (ax1)</span>
<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span><span class="n">ax</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">7</span><span class="p">))</span>

<span class="c1"># Customize subplot appearance</span>
<span class="c1"># Remove the top and intersecting spines</span>
<span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;top&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;top&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

<span class="c1">### PLOT 1 - African populations</span>

<span class="c1"># Set axis limits and labels</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">20001</span><span class="p">])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.4</span><span class="p">])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">20001</span><span class="p">,</span><span class="mi">5000</span><span class="p">))</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.41</span><span class="p">,</span><span class="mf">0.1</span><span class="p">))</span>
<span class="n">ax</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_major_formatter</span><span class="p">(</span><span class="n">ticker</span><span class="o">.</span><span class="n">StrMethodFormatter</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{x:,.0f}</span><span class="s1">&#39;</span><span class="p">))</span>

<span class="n">af_pop_names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AF-W&#39;</span><span class="p">,</span><span class="s1">&#39;AF-C&#39;</span><span class="p">,</span><span class="s1">&#39;AF-NE&#39;</span><span class="p">,</span><span class="s1">&#39;AF-E&#39;</span><span class="p">]</span>

<span class="c1"># Iterate through all combinations of population pairs</span>
<span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span><span class="n">pop_names</span><span class="p">,</span> <span class="n">r</span><span class="o">=</span><span class="mi">2</span><span class="p">)):</span>

    <span class="c1"># Check if both populations are African populations</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">pop</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">af_pop_names</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">pop</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="n">af_pop_names</span><span class="p">):</span>
        <span class="c1"># Save values for plotting</span>
        <span class="n">fst_val</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pop</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">pop</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
        <span class="n">geo_distance</span><span class="o">=</span> <span class="n">geo_distances</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pop</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">pop</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>


        <span class="n">marker_style</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span> <span class="n">c</span><span class="o">=</span><span class="n">georegions</span><span class="p">[</span><span class="n">pop</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="s1">&#39;c&#39;</span><span class="p">],</span><span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span>
                    <span class="n">markersize</span><span class="o">=</span><span class="mi">13</span><span class="p">,</span> <span class="n">markerfacecoloralt</span><span class="o">=</span><span class="n">georegions</span><span class="p">[</span><span class="n">pop</span><span class="p">[</span><span class="mi">1</span><span class="p">]][</span><span class="s1">&#39;c&#39;</span><span class="p">],</span> <span class="n">markeredgecolor</span><span class="o">=</span><span class="s1">&#39;#333333&#39;</span><span class="p">)</span>

        <span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">geo_distance</span><span class="p">,</span> <span class="n">fst_val</span><span class="p">,</span>
                <span class="n">fillstyle</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">marker_style</span><span class="p">)</span>

        <span class="n">ax1</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">georegions</span><span class="p">[</span><span class="n">pop</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="s1">&#39;n&#39;</span><span class="p">],</span>  <span class="p">(</span><span class="n">geo_distance</span><span class="o">-</span><span class="mi">145</span><span class="p">,</span> <span class="n">fst_val</span><span class="o">-</span><span class="mf">0.00007</span><span class="p">),</span><span class="n">horizontalalignment</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">georegions</span><span class="p">[</span><span class="n">pop</span><span class="p">[</span><span class="mi">1</span><span class="p">]][</span><span class="s1">&#39;n&#39;</span><span class="p">],</span>  <span class="p">(</span><span class="n">geo_distance</span><span class="o">+</span><span class="mi">145</span><span class="p">,</span> <span class="n">fst_val</span><span class="o">-</span><span class="mf">0.00007</span><span class="p">),</span><span class="n">horizontalalignment</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">)</span>

<span class="c1">### PLOT 2 - Asian and other populations</span>

<span class="c1"># Store fst and geo values to use after plotting</span>
<span class="n">fst_values</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">geo_values</span> <span class="o">=</span> <span class="p">[]</span>

<span class="c1"># Set axis limits and labels</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">5200</span><span class="p">])</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.04</span><span class="p">])</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.041</span><span class="p">,</span><span class="mf">0.01</span><span class="p">))</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Genetic distance ($F_</span><span class="si">{ST}</span><span class="s1">$)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">17</span><span class="p">,</span><span class="n">labelpad</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

<span class="c1"># Iterate through all combinations for population</span>
<span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span><span class="n">pop_names</span><span class="p">,</span> <span class="n">r</span><span class="o">=</span><span class="mi">2</span><span class="p">)):</span>

    <span class="k">if</span> <span class="p">(</span><span class="s1">&#39;AS-SE-E&#39;</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">)</span> <span class="o">^</span> <span class="p">(</span><span class="s1">&#39;AS-SE-W&#39;</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">):</span>

        <span class="n">fst_val</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pop</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">pop</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
        <span class="n">geo_distance</span><span class="o">=</span> <span class="n">geo_distances</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pop</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">pop</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>

        <span class="k">if</span> <span class="s1">&#39;AS-SE-E&#39;</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">:</span>
            <span class="n">fst_values</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fst_val</span><span class="p">)</span>
            <span class="n">geo_values</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">geo_distance</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

            <span class="n">marker_style</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">georegions</span><span class="p">[</span><span class="s1">&#39;AS-SE-E&#39;</span><span class="p">][</span><span class="s1">&#39;c&#39;</span><span class="p">],</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span>
                        <span class="n">markersize</span><span class="o">=</span><span class="mi">13</span><span class="p">)</span>

        <span class="k">elif</span> <span class="s1">&#39;AS-SE-W&#39;</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">:</span>
            <span class="n">fst_values</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fst_val</span><span class="p">)</span>
            <span class="n">geo_values</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">geo_distance</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

            <span class="n">marker_style</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">georegions</span><span class="p">[</span><span class="s1">&#39;AS-SE-W&#39;</span><span class="p">][</span><span class="s1">&#39;c&#39;</span><span class="p">],</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span>
                        <span class="n">markersize</span><span class="o">=</span><span class="mi">13</span><span class="p">)</span>

            <span class="n">secondary_pop</span> <span class="o">=</span> <span class="p">[</span><span class="n">pop</span><span class="p">[</span><span class="n">n</span><span class="p">]</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pop</span><span class="p">))</span> <span class="k">if</span> <span class="p">(</span><span class="n">pop</span><span class="p">[</span><span class="n">n</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;AS-SE-W&#39;</span><span class="p">)]</span>

            <span class="c1"># Align labels to markers</span>
            <span class="k">if</span> <span class="n">secondary_pop</span> <span class="o">==</span> <span class="p">[</span><span class="s1">&#39;AF-NE&#39;</span><span class="p">]:</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">secondary_pop</span><span class="p">),</span>  <span class="p">(</span><span class="n">geo_distance</span><span class="o">-</span><span class="mi">10</span><span class="p">,</span> <span class="n">fst_val</span><span class="o">+</span><span class="mf">0.009</span><span class="p">),</span><span class="n">horizontalalignment</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">secondary_pop</span> <span class="o">==</span> <span class="p">[</span><span class="s1">&#39;OC-NG&#39;</span><span class="p">]:</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">secondary_pop</span><span class="p">),</span>  <span class="p">(</span><span class="n">geo_distance</span><span class="o">+</span><span class="mi">1300</span><span class="p">,</span> <span class="n">fst_val</span><span class="o">+</span><span class="mf">0.009</span><span class="p">),</span><span class="n">horizontalalignment</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">secondary_pop</span> <span class="o">==</span> <span class="p">[</span><span class="s1">&#39;SA&#39;</span><span class="p">]:</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">secondary_pop</span><span class="p">),</span>  <span class="p">(</span><span class="n">geo_distance</span><span class="o">+</span><span class="mi">800</span><span class="p">,</span> <span class="n">fst_val</span><span class="o">+</span><span class="mf">0.007</span><span class="p">),</span><span class="n">horizontalalignment</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">secondary_pop</span><span class="p">),</span> <span class="p">(</span><span class="n">geo_distance</span><span class="o">+</span><span class="mi">400</span><span class="p">,</span> <span class="n">fst_val</span><span class="o">+</span><span class="mf">0.015</span><span class="p">))</span>

        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">geo_distance</span><span class="p">,</span> <span class="n">fst_val</span><span class="p">,</span> <span class="o">**</span><span class="n">marker_style</span><span class="p">)</span>

<span class="c1"># Create legend for population colors</span>
<span class="n">pop1</span> <span class="o">=</span> <span class="n">mpatches</span><span class="o">.</span><span class="n">Patch</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">georegions</span><span class="p">[</span><span class="s1">&#39;AS-SE-W&#39;</span><span class="p">][</span><span class="s1">&#39;c&#39;</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Asia - Southeast - West&#39;</span><span class="p">)</span>
<span class="n">pop2</span> <span class="o">=</span> <span class="n">mpatches</span><span class="o">.</span><span class="n">Patch</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">georegions</span><span class="p">[</span><span class="s1">&#39;AS-SE-E&#39;</span><span class="p">][</span><span class="s1">&#39;c&#39;</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Asia - Southeast - East&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="p">[</span><span class="n">pop1</span><span class="p">,</span><span class="n">pop2</span><span class="p">])</span>

<span class="c1"># Join pairs of markers with connecting lines</span>
<span class="n">connect_markers</span> <span class="o">=</span> <span class="p">[</span><span class="n">n</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">fst_values</span><span class="p">))</span> <span class="k">if</span> <span class="n">n</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">==</span> <span class="mi">0</span><span class="p">]</span>
<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">connect_markers</span><span class="p">:</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">geo_values</span><span class="p">[</span><span class="n">n</span><span class="p">],</span><span class="n">geo_values</span><span class="p">[</span><span class="n">n</span><span class="o">+</span><span class="mi">1</span><span class="p">]],[</span><span class="n">fst_values</span><span class="p">[</span><span class="n">n</span><span class="p">],</span><span class="n">fst_values</span><span class="p">[</span><span class="n">n</span><span class="o">+</span><span class="mi">1</span><span class="p">]],</span><span class="s1">&#39;-k&#39;</span><span class="p">,</span><span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">.5</span><span class="p">,</span><span class="n">zorder</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="c1"># Add subplots to the figure</span>
<span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">111</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="c1"># Make the extra tick labels invisible</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">labelcolor</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="c1"># Set common x-axis label</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Geographic distance (great-circle distance in km)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">17</span><span class="p">,</span> <span class="n">labelpad</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Text(0.5, 0, &#39;Geographic distance (great-circle distance in km)&#39;)
</pre></div>
</div>
<img alt="../../_images/geographic-genetic-differentiation_32_1.png" src="../../_images/geographic-genetic-differentiation_32_1.png" />
</div>
</div>
<p><strong>Figure Legend. Geographic patterns of population differentiation and gene flow.</strong>
Each point represents one pairwise comparison between two regional parasite sub- populations. The x-axis reports the geographic separation between the two sub- populations, measured as great-circle distance between the centre of mass of each sub- population and without taking into account natural barriers. The y-axis reports the genetic differentiation between the two sub-populations, measured as average genome-wide FST. These two figures show that the sub-populations from northeast Africa and the eastern part of SE Asia are more genetically distinct from other sub-populations than might be expected due to their geographic separation. (A) Comparison of African sub-populations. Points are coloured based on the two sub-populations they represent. The distance from northeast Africa (AF-NE) to other African sub-populations is generally greater than that between the other African sub-populations. For example the central African sub-population (AF-C) is closer genetically to the west African sub-population (AF-W) than it is to the northeast African sub-population, despite being closer geographically to the latter. (B) Comparison of two SE Asian sub-populations against all other sub-populations. Compared to all other sub- populations, the sub-population from the eastern part of SE Asia (AS-SEA-E) generally has a greater genetic distance than that from the western part of SE Asia, and this difference in genetic distances is more than might be expected due to the extra geographic distance.</p>
<div class="section" id="save-the-figure">
<h3>Save the figure<a class="headerlink" href="#save-the-figure" title="Permalink to this headline">¶</a></h3>
<p>We can output this to a location in Google Drive.</p>
<p>First we need to connect Google Drive by running the following:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># You will need to authorise Google Colab access to Google Drive</span>
<span class="n">drive</span><span class="o">.</span><span class="n">mount</span><span class="p">(</span><span class="s1">&#39;/content/drive&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Mounted at /content/drive
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># This will send the file to your Google Drive, where you can download it from if needed</span>
<span class="c1"># Change the file path if you wish to send the file to a specific location</span>
<span class="c1"># Change the file name if you wish to call it something else</span>

<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;/content/drive/My Drive/GeneticGeographicDistance.pdf&#39;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;/content/drive/My Drive/GeneticGeographicDistance.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">240</span><span class="p">)</span> <span class="c1"># increase the dpi for higher resolution</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./pf7/analysis_examples"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="PCoA.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Principal Coordinate Analysis</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="neighbour-joining-tree-swga.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Neighbour Joining Tree for sWGA vs. gDNA</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By MalariaGEN<br/>
    
        &copy; Copyright 2023.<br/>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>