
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Neighbour Joining Tree for sWGA vs. gDNA &#8212; MalariaGEN parasite data user guide</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Plot eba175 C Allele Proportions" href="proportion-eba175.html" />
    <link rel="prev" title="Geographic patterns of population differentiation and gene flow" href="geographic-genetic-differentiation.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">MalariaGEN parasite data user guide</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../landing-page.html">
   MalariaGEN parasite data user guide
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Pf7
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../Data_access.html">
   Pf7 Data Access
  </a>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../examples.html">
   Pf7 Analysis Examples
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="useful-how-tos.html">
     Useful Genomics How-Tos
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
    <label for="toctree-checkbox-2">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="igv-notebook.html">
       Visualizing Genomic Regions of Pf7 Samples with IGV Plots
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sample-summary-table.html">
     Summarise Samples by Administrative Division
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sample-collection-over-time.html">
     Visualising sample collections over time
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="barplot-samples-by-country.html">
     Barplot of Samples by Country
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="variant-summary-table.html">
     Summarise Genetic Variants
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="global-sample-map.html">
     Plot Global Map of Samples
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="drug-resistance-summary.html">
     Drug Resistance Frequencies Across Pf7
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="chloroquine-resistance-map.html">
     Plot Map of Chloroquine Resistance
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="summarise_hrp2_hrp3_deletions.html">
     Summarise
     <em>
      hrp2
     </em>
     and
     <em>
      hrp3
     </em>
     Deletions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="visualise_hrp2_hrp3_deletions.html">
     Visualise
     <em>
      hrp2
     </em>
     and
     <em>
      hrp3
     </em>
     Deletions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="csp-cterm-analysis.html">
     Analyse csp C-Terminal Variation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="PCoA.html">
     Principal Coordinate Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="geographic-genetic-differentiation.html">
     Geographic patterns of population differentiation and gene flow
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Neighbour Joining Tree for sWGA vs. gDNA
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="proportion-eba175.html">
     Plot
     <em>
      eba175
     </em>
     C Allele Proportions
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../api.html">
   Pf7 API
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Pv4
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../pv4/Data_access.html">
   Pv4 Data Access
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../pv4/api.html">
   Pv4 API
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/pf7/analysis_examples/neighbour-joining-tree-swga.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
                onclick="printPdf(this)" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/malariagen/parasite-data"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/malariagen/parasite-data/issues/new?title=Issue%20on%20page%20%2Fpf7/analysis_examples/neighbour-joining-tree-swga.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/malariagen/parasite-data/master?urlpath=lab/tree/docs/pf7/analysis_examples/neighbour-joining-tree-swga.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        <a class="colab-button" href="https://colab.research.google.com/github/malariagen/parasite-data/blob/master/docs/pf7/analysis_examples/neighbour-joining-tree-swga.ipynb"><button type="button" class="btn btn-secondary topbarbtn"
                title="Launch Colab" data-toggle="tooltip" data-placement="left"><img class="colab-button-logo"
                    src="../../_static/images/logo_colab.png"
                    alt="Interact on Colab">Colab</button></a>
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#neighbour-joining-trees">
     Neighbour Joining Trees
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#activate-r-magic">
   Activate R magic
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#access-the-data">
   Access the data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#create-a-dataframe-for-the-sample-data-in-r">
   Create a dataframe for the sample data in R
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#begin-the-njt-process">
   Begin the NJT Process
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#a-ghana-upper-east-2015">
   A) Ghana, Upper East, 2015
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#create-a-nj-tree-object">
     Create a NJ tree object
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#plotting-functions-for-the-njt">
     Plotting Functions for the NJT
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#create-a-dataframe-with-sample-type-colour-assignments">
     Create a dataframe with sample type + colour assignments
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#put-all-the-functions-into-action">
     Put all the functions into action
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#plot-the-njt">
     Plot the NJT
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#write-the-figure-to-a-file">
     Write the figure to a file
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#continue-on-to-make-the-rest-of-the-figure">
   Continue on to make the rest of the figure
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#b-kenya-kilifi-2007-2012">
   B) Kenya, Kilifi, 2007-2012
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#redfine-pass-samples">
     Redfine pass_samples
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#remake-the-tree-object">
     Remake the tree object
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#redefine-groups-for-our-current-pass-samples-variable">
     Redefine ‘groups’ for our current ‘pass_samples’ variable
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#plot-the-new-njt">
     Plot the new NJT
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     Write the figure to a file
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#c-cambodia-pursat-2016">
   C) Cambodia, Pursat, 2016
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id2">
     Write the figure to a file
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#d-vietnam-binh-phuoc-2016">
   D) Vietnam, Binh Phuoc, 2016
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id3">
     Write the figure to a file
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Neighbour Joining Tree for sWGA vs. gDNA</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#neighbour-joining-trees">
     Neighbour Joining Trees
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#activate-r-magic">
   Activate R magic
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#access-the-data">
   Access the data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#create-a-dataframe-for-the-sample-data-in-r">
   Create a dataframe for the sample data in R
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#begin-the-njt-process">
   Begin the NJT Process
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#a-ghana-upper-east-2015">
   A) Ghana, Upper East, 2015
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#create-a-nj-tree-object">
     Create a NJ tree object
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#plotting-functions-for-the-njt">
     Plotting Functions for the NJT
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#create-a-dataframe-with-sample-type-colour-assignments">
     Create a dataframe with sample type + colour assignments
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#put-all-the-functions-into-action">
     Put all the functions into action
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#plot-the-njt">
     Plot the NJT
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#write-the-figure-to-a-file">
     Write the figure to a file
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#continue-on-to-make-the-rest-of-the-figure">
   Continue on to make the rest of the figure
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#b-kenya-kilifi-2007-2012">
   B) Kenya, Kilifi, 2007-2012
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#redfine-pass-samples">
     Redfine pass_samples
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#remake-the-tree-object">
     Remake the tree object
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#redefine-groups-for-our-current-pass-samples-variable">
     Redefine ‘groups’ for our current ‘pass_samples’ variable
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#plot-the-new-njt">
     Plot the new NJT
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     Write the figure to a file
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#c-cambodia-pursat-2016">
   C) Cambodia, Pursat, 2016
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id2">
     Write the figure to a file
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#d-vietnam-binh-phuoc-2016">
   D) Vietnam, Binh Phuoc, 2016
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id3">
     Write the figure to a file
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            
              <div>
                
  <div class="section" id="neighbour-joining-tree-for-swga-vs-gdna">
<h1>Neighbour Joining Tree for sWGA vs. gDNA<a class="headerlink" href="#neighbour-joining-tree-for-swga-vs-gdna" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>This notebook recreates <a class="reference external" href="https://figshare.com/ndownloader/files/38478530">Supplementary Figure 3</a> from the <a class="reference external" href="https://wellcomeopenresearch.org/articles/8-22">Pf7 paper</a> - a neighbour-joining tree (NJT) comparing selective whole-genome amplification <strong>(sWGA)</strong> with whole genome sequencing <strong>(WGS)</strong> samples.</p>
<p>The tree demonstrates the lack of bias due to amplification/sequencing methods in Pf7.</p>
<p>Unlike most of the other example notebooks in the Pf7 data resource, this notebook runs using predominantly R code rather than Python. The main reason for this is that the package used to build the NJT, ape, has better options for customising and visualising the tree than currently available Python libraries.</p>
<p><strong>This notebook should take approximately five minutes to run.</strong></p>
<div class="section" id="neighbour-joining-trees">
<h3>Neighbour Joining Trees<a class="headerlink" href="#neighbour-joining-trees" title="Permalink to this headline">¶</a></h3>
<p>A neighbour joining tree, or NJT, is a type of distance-based tree used to visualise population structure for a set of samples. A tree will plot more closely related samples on nearby ‘branches’, while more distantly related samples appear farther apart in a tree.</p>
<p>A NJT makes use of a distance matrix, which contains information on how near/distant samples are in an evolutionary sense. In the case of the Pf7 data, this distance matrix was calculated using high-quality bi-allelic single nucleotide polymorphisms (SNPs) from across the Pf genome. A NJT is created by joining the closest samples (or <strong>neighbours</strong>) over and over until the final tree is built.</p>
<p>For this use case, a NJT is a good choice because it is less computationally expensive than other trees such as Maximum Likelihood while still being a robust method for analysing genetic structure. You can read the original paper on the method <a class="reference external" href="https://academic.oup.com/mbe/article/4/4/406/1029664">here</a> and a more user-friendly guide to the concept <a class="reference external" href="https://www.tenderisthebyte.com/blog/2022/08/31/neighbor-joining-trees/#:~:text=The%20method%20was%20first%20introduced,via%20an%20iterative%20clustering%20process.">here</a></p>
</div>
</div>
<div class="section" id="activate-r-magic">
<h2>Activate R magic<a class="headerlink" href="#activate-r-magic" title="Permalink to this headline">¶</a></h2>
<p>This notebook is running from a Python kernel so that we can load the MalariaGEN python package with a “pip install” command.</p>
<p>As mentioned, we need the rest of the notebook to run with R. To enable this, we activate <strong>R magic commands</strong>. This means that anytime a code cell starts with <strong>“%%R”</strong>, it will run R code within the cell.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># activate R magic</span>
<span class="o">%</span><span class="n">load_ext</span> <span class="n">rpy2</span><span class="o">.</span><span class="n">ipython</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="setup">
<h2>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">¶</a></h2>
<p>Install the malariagen package.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>!pip install -q --no-warn-conflicts malariagen_data
</pre></div>
</div>
</div>
</div>
<p>Load the required Python libraries</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">malariagen_data</span>
<span class="kn">from</span> <span class="nn">ftplib</span> <span class="kn">import</span> <span class="n">FTP</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">rpy2.robjects</span> <span class="k">as</span> <span class="nn">robjects</span>
<span class="kn">from</span> <span class="nn">rpy2.robjects</span> <span class="kn">import</span> <span class="n">pandas2ri</span>
<span class="kn">from</span> <span class="nn">google.colab</span> <span class="kn">import</span> <span class="n">drive</span>

<span class="c1"># Suppress FutureWarnings</span>
<span class="c1"># This surpresses a warning from pandas2ri.py when it uses iteritems() that occurs at a few points in the notebook</span>
<span class="c1"># The warning states that iteritems will be replaced by items() in the future</span>
<span class="c1"># It appears to be a bug in rpy2 that we can ignore for now</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="n">action</span><span class="o">=</span><span class="s1">&#39;ignore&#39;</span><span class="p">,</span> <span class="n">category</span><span class="o">=</span><span class="ne">FutureWarning</span><span class="p">)</span> <span class="c1"># comment out this code if you would like to see the warnings</span>
</pre></div>
</div>
</div>
</div>
<p>Install the required R packages</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># This code block surpresses a long list of &#39;warnings&#39; from rpy2 when installing R packages</span>
<span class="c1"># The &#39;warnings&#39; apppear to be empty or contain progress messages for the installation</span>
<span class="c1"># For this reason, we surpress them to keep the notebook neater.</span>
<span class="c1"># If you want to see them, comment out this block.</span>
<span class="kn">from</span> <span class="nn">rpy2.rinterface_lib.callbacks</span> <span class="kn">import</span> <span class="n">logger</span> <span class="k">as</span> <span class="n">rpy2_logger</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="n">rpy2_logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">ERROR</span><span class="p">)</span>   <span class="c1"># will display errors, but not warnings</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">install</span><span class="o">.</span><span class="n">packages</span><span class="p">(</span><span class="s2">&quot;ape&quot;</span><span class="p">,</span> <span class="n">lib</span> <span class="o">=</span> <span class="s1">&#39;/usr/local/lib/R/site-library&#39;</span><span class="p">)</span>
<span class="n">install</span><span class="o">.</span><span class="n">packages</span><span class="p">(</span><span class="s2">&quot;RcppCNPy&quot;</span><span class="p">,</span> <span class="n">lib</span> <span class="o">=</span> <span class="s1">&#39;/usr/local/lib/R/site-library&#39;</span><span class="p">)</span>
<span class="n">install</span><span class="o">.</span><span class="n">packages</span><span class="p">(</span><span class="s2">&quot;reticulate&quot;</span><span class="p">,</span> <span class="n">lib</span> <span class="o">=</span> <span class="s1">&#39;/usr/local/lib/R/site-library&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Reset the logging level to its default value to show future warnings and errors</span>
<span class="n">rpy2_logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">NOTSET</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Load the required R packages</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">library</span><span class="p">(</span><span class="n">ape</span><span class="p">)</span>
<span class="n">library</span><span class="p">(</span><span class="n">RcppCNPy</span><span class="p">)</span>
<span class="n">library</span><span class="p">(</span><span class="n">reticulate</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="access-the-data">
<h2>Access the data<a class="headerlink" href="#access-the-data" title="Permalink to this headline">¶</a></h2>
<p>We need to load the Pf7 metadata…</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load the Pf7 metadata</span>
<span class="n">release_data</span> <span class="o">=</span> <span class="n">malariagen_data</span><span class="o">.</span><span class="n">Pf7</span><span class="p">()</span>
<span class="n">sample_metadata</span> <span class="o">=</span> <span class="n">release_data</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>…And the distance matrix (genetic distances between all 20,864 samples).</p>
<p>This available from a FTP (file transfer protocol) server. We log in to this server and make a copy of the distance matrix in our Google Colab space.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load the distance matrix</span>
<span class="c1"># Replace these values with your FTP server details</span>
<span class="n">ftp_host</span> <span class="o">=</span> <span class="s1">&#39;ngs.sanger.ac.uk&#39;</span>
<span class="n">ftp_file_path</span> <span class="o">=</span> <span class="s1">&#39;/production/malaria/Resource/34/Pf7_genetic_distance_matrix.npy&#39;</span>

<span class="c1"># Create an FTP instance and connect to the server</span>
<span class="n">ftp</span> <span class="o">=</span> <span class="n">FTP</span><span class="p">(</span><span class="n">ftp_host</span><span class="p">)</span>
<span class="n">ftp</span><span class="o">.</span><span class="n">login</span><span class="p">()</span>

<span class="c1"># Change the current directory to the location of the file</span>
<span class="n">ftp</span><span class="o">.</span><span class="n">cwd</span><span class="p">(</span><span class="s1">&#39;/production/malaria/Resource/34/&#39;</span><span class="p">)</span>  <span class="c1"># Change to the root directory or your desired directory</span>

<span class="c1"># Download the file to Colab</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;Pf7_genetic_distance_matrix.npy&#39;</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
    <span class="n">ftp</span><span class="o">.</span><span class="n">retrbinary</span><span class="p">(</span><span class="s1">&#39;RETR &#39;</span> <span class="o">+</span> <span class="n">ftp_file_path</span><span class="p">,</span> <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">)</span>

<span class="c1"># Close the FTP connection</span>
<span class="n">ftp</span><span class="o">.</span><span class="n">quit</span><span class="p">()</span>

<span class="c1"># Check if the file is downloaded successfully</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="s1">&#39;Pf7_genetic_distance_matrix.npy&#39;</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;File downloaded successfully.&#39;</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;File download failed.&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>File downloaded successfully.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">dist_fn</span> <span class="o">&lt;-</span> <span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="s1">&#39;/content/Pf7_genetic_distance_matrix.npy&#39;</span><span class="p">)</span> <span class="c1"># The path to our copy of the distance matrix</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">np</span> <span class="o">&lt;-</span> <span class="n">import</span><span class="p">(</span><span class="s2">&quot;numpy&quot;</span><span class="p">)</span> <span class="c1"># Import the python library numpy within an R environment</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%R
mean_distance_matrix = np$load(dist_fn) # Define the distance matrix
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="create-a-dataframe-for-the-sample-data-in-r">
<h2>Create a dataframe for the sample data in R<a class="headerlink" href="#create-a-dataframe-for-the-sample-data-in-r" title="Permalink to this headline">¶</a></h2>
<p>Until now we have been working in Python and R to read in the sample data (a python package) and the distance matrix (using Python and R commands).</p>
<p>We will need the sample data to be R-readable to continue to tree building.</p>
<p>In the following steps we use a conversion tool “rpy2.pandas2ri” to convert the pandas datframe sample_metadata to an R dataframe.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pandas2ri</span><span class="o">.</span><span class="n">activate</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">r_dataframe</span> <span class="o">=</span> <span class="n">pandas2ri</span><span class="o">.</span><span class="n">py2rpy</span><span class="p">(</span><span class="n">sample_metadata</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">robjects</span><span class="o">.</span><span class="n">globalenv</span><span class="p">[</span><span class="s1">&#39;r_dataframe&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">r_dataframe</span> <span class="c1">#Add the R dataframe to our R environment</span>
<span class="o">%</span><span class="n">R</span> <span class="n">df_samples</span> <span class="o">&lt;-</span> <span class="n">r_dataframe</span> <span class="c1"># change the dataframe name to something more informative</span>
<span class="o">%</span><span class="n">R</span> <span class="n">head</span><span class="p">(</span><span class="n">df_samples</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-25704f53-b6da-4b7d-839d-a3cfcca57e9a" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Sample</th>
      <th>Study</th>
      <th>Country</th>
      <th>Admin level 1</th>
      <th>Country latitude</th>
      <th>Country longitude</th>
      <th>Admin level 1 latitude</th>
      <th>Admin level 1 longitude</th>
      <th>Year</th>
      <th>ENA</th>
      <th>All samples same case</th>
      <th>Population</th>
      <th>% callable</th>
      <th>QC pass</th>
      <th>Exclusion reason</th>
      <th>Sample type</th>
      <th>Sample was in Pf6</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>FP0008-C</td>
      <td>1147-PF-MR-CONWAY</td>
      <td>Mauritania</td>
      <td>Hodh el Gharbi</td>
      <td>20.265149</td>
      <td>-10.337093</td>
      <td>16.565426</td>
      <td>-9.832345</td>
      <td>2014.0</td>
      <td>ERR1081237</td>
      <td>FP0008-C</td>
      <td>AF-W</td>
      <td>82.16</td>
      <td>1</td>
      <td>Analysis_set</td>
      <td>gDNA</td>
      <td>1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>FP0009-C</td>
      <td>1147-PF-MR-CONWAY</td>
      <td>Mauritania</td>
      <td>Hodh el Gharbi</td>
      <td>20.265149</td>
      <td>-10.337093</td>
      <td>16.565426</td>
      <td>-9.832345</td>
      <td>2014.0</td>
      <td>ERR1081238</td>
      <td>FP0009-C</td>
      <td>AF-W</td>
      <td>88.85</td>
      <td>1</td>
      <td>Analysis_set</td>
      <td>gDNA</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>FP0010-CW</td>
      <td>1147-PF-MR-CONWAY</td>
      <td>Mauritania</td>
      <td>Hodh el Gharbi</td>
      <td>20.265149</td>
      <td>-10.337093</td>
      <td>16.565426</td>
      <td>-9.832345</td>
      <td>2014.0</td>
      <td>ERR2889621</td>
      <td>FP0010-CW</td>
      <td>AF-W</td>
      <td>86.46</td>
      <td>1</td>
      <td>Analysis_set</td>
      <td>sWGA</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-25704f53-b6da-4b7d-839d-a3cfcca57e9a')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-25704f53-b6da-4b7d-839d-a3cfcca57e9a button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-25704f53-b6da-4b7d-839d-a3cfcca57e9a');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-956ea4cb-4dba-486b-8804-6369beaa7b7c">
  <button class="colab-df-quickchart" onclick="quickchart('df-956ea4cb-4dba-486b-8804-6369beaa7b7c')"
            title="Suggest charts."
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-956ea4cb-4dba-486b-8804-6369beaa7b7c button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
</div>
<div class="section" id="begin-the-njt-process">
<h2>Begin the NJT Process<a class="headerlink" href="#begin-the-njt-process" title="Permalink to this headline">¶</a></h2>
<p>Set a colour for each sample type. Here we create a new column in df_samples, ‘population_colour’, which will contain the HEX code for blue if the sample type is gDNA, or red if the sample type is sWGA.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%R
df_samples$population_colour &lt;- ifelse(df_samples$`Sample type` == &quot;gDNA&quot;, &quot;#3182bd&quot;,
                                       ifelse(df_samples$`Sample type` == &quot;sWGA&quot;, &quot;#e31a1c&quot;, &quot;NA&quot;))
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="a-ghana-upper-east-2015">
<h2>A) Ghana, Upper East, 2015<a class="headerlink" href="#a-ghana-upper-east-2015" title="Permalink to this headline">¶</a></h2>
<p>To generate the first NJT of the figure, we must subset the metadata and the distance matrix to include only samples which meet the following criteria:</p>
<ul class="simple">
<li><p>Passed Quality Control (QC), denoted as ‘Analysis set’</p></li>
<li><p>From Ghana, Upper East administrative division</p></li>
<li><p>From 2015</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%R
# We combine the criteria using &quot;&amp;&quot; to create a logical vector &#39;pass_samples&#39;
# where each element refers to a row in the df_samples dataframe
# and evaluates to TRUE if all the conditions are met for that row, or FALSE otherwise.
pass_samples &lt;- (
    df_samples$`Exclusion reason` == &#39;Analysis_set&#39; &amp; # We use backticks (`) around column names with spaces &#39;Exclusion reason&#39;, to ensure they are correctly recognized as column names.
    df_samples$Country == &#39;Ghana&#39; &amp;
    df_samples$`Admin level 1` == &#39;Upper East&#39; &amp;
    df_samples$Year == 2015
)
summary(pass_samples) # We have 225 samples that satisfy the criteria
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>   Mode   FALSE    TRUE 
logical   20639     225 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span># How many samples do we have of each sample type?
%%R
aggregate(Sample ~ Country+`Sample type`, df_samples[pass_samples, ], FUN=length)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  Country Sample type Sample
1   Ghana        gDNA     61
2   Ghana        sWGA    164
</pre></div>
</div>
</div>
</div>
<div class="section" id="create-a-nj-tree-object">
<h3>Create a NJ tree object<a class="headerlink" href="#create-a-nj-tree-object" title="Permalink to this headline">¶</a></h3>
<p>We use the ‘nj’ command from ape on a subset of the distance matrix, defined as the ‘TRUE’ component of the logical vector we created earlier.</p>
<p>This should be a matrix of 225 x 225 samples.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">nj</span><span class="p">(</span><span class="n">mean_distance_matrix</span><span class="p">[</span><span class="n">pass_samples</span><span class="p">,</span> <span class="n">pass_samples</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="plotting-functions-for-the-njt">
<h3>Plotting Functions for the NJT<a class="headerlink" href="#plotting-functions-for-the-njt" title="Permalink to this headline">¶</a></h3>
<p>Below are a series of functions which plot the NJT.</p>
<ol class="simple">
<li><p><strong>computeEdgeGroupCounts</strong>: This function creates a matrix which tracks the counts of groups associated with edges of the NJT. Edge is another word for branch within a tree. It requires a ‘phylotree’ object and a list of labels for groups ‘labelgroups’.</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%R
computeEdgeGroupCounts &lt;- function(phylotree, labelgroups) {

  labels &lt;- phylotree$tip.label
  num_tips &lt;- length(labels)
  edge_names &lt;- unique(sort(c(phylotree$edge)))

  # This matrix will keep track of the group counts for each edge.
  edge_group_counts &lt;- matrix(0, nrow=length(edge_names), ncol=length(unique(sort(labelgroups))))
  rownames(edge_group_counts) &lt;- edge_names
  colnames(edge_group_counts) &lt;- unique(labelgroups)

  # Init the leaf branches.
  sapply(1:num_tips, function(l) {
    edge_group_counts[as.character(l), as.character(labelgroups[phylotree$tip.label[l]])] &lt;&lt;- 1
  })

  # Sort edges by the value of the descendent
  # The first segment will contain the leaves whereas the second the branches (closer to leaves first).
  # We need to do this because leaves are numbered 1:num_tips and the branches CLOSER to the leaves
  # with higher numbers.
  edges &lt;- phylotree$edge[order(phylotree$edge[,2]),]
  branches &lt;- edges[num_tips:nrow(edges),]
  edges[num_tips:nrow(edges),] &lt;- branches[order(branches[,1],decreasing=T),]
  invisible(apply(edges, 1, function(edge) {
    # Check if we are connecting a leaf.
    if(edge[2] &lt;= num_tips) {
      e &lt;- as.character(edge[1])
      g &lt;- as.character(labelgroups[phylotree$tip.label[edge[2]]])
      edge_group_counts[e,g] &lt;&lt;- edge_group_counts[e,g] + 1
    }
    else {
      e1 &lt;- as.character(edge[1])
      e2 &lt;- as.character(edge[2])
      edge_group_counts[e1,] &lt;&lt;- edge_group_counts[e1,] + edge_group_counts[e2,]
    }
  }))
  return(edge_group_counts)
}
</pre></div>
</div>
</div>
</div>
<ol class="simple">
<li><p><strong>assignMajorityGroupColorToEdges</strong>: This function will colour the branches/edges of the tree according to which group makes up most of the samples associated with that branch/edge. It uses the edge_group_counts from the previous function, as well as the phylotree object, and a vector of group colours. In cases where there is an equal split of groups for a branch/edge, the colour assigned is grey.</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%R
assignMajorityGroupColorToEdges &lt;- function(phylotree, edge_group_counts, groupcolors, equality_color=&quot;gray&quot;) {
  edge_colors &lt;- apply(phylotree$edge, 1, function(branch) { # Make an empty vector
    e &lt;- as.character(branch[2])
    major_group_index &lt;- which.max(edge_group_counts[e,]) # Identify the group with the max count
    if(all(edge_group_counts[e,] == edge_group_counts[e,major_group_index])) # Assign grey if all counts equal
      return(equality_color)
    else
      return(groupcolors[colnames(edge_group_counts)[major_group_index]]) # Assign majority group colour if not
  })
  return(edge_colors)
}
</pre></div>
</div>
</div>
</div>
<ol class="simple">
<li><p><strong>assignGroupColorToUniqueEdges</strong>: This function is very similar to the above function. It differs in that rather than checking if group counts are equal, it checks if the 2nd highest group count is negative. This enables the function to identify edges with a clear majority group.</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%R
assignGroupColorToUniqueEdges &lt;- function(phylotree, edge_group_counts, groupcolors, equality_color=&quot;gray&quot;) {
  edge_colors &lt;- apply(phylotree$edge, 1, function(branch) { # Make an empty vector
    e &lt;- as.character(branch[2])
    major_group_index &lt;- which.max(edge_group_counts[e,]) # Identify the group with the max count
    if(sort(-edge_group_counts[e,])[2] &lt; 0) # Check if the 2nd highest count is negative
      return(equality_color) # Return grey if so
    else
      return(groupcolors[colnames(edge_group_counts)[major_group_index]]) # Return majority group colour if not
  })
  return(edge_colors)
}
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="create-a-dataframe-with-sample-type-colour-assignments">
<h3>Create a dataframe with sample type + colour assignments<a class="headerlink" href="#create-a-dataframe-with-sample-type-colour-assignments" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%R
# Create df_pops_with_colours by removing duplicated rows based on sample_type and population_colour columns
df_pops_with_colours &lt;- df_samples[!duplicated(df_samples[, c(&#39;Sample type&#39;, &#39;population_colour&#39;)]), c(&#39;Sample type&#39;, &#39;population_colour&#39;)]

# Display the resulting dataframe
df_pops_with_colours
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>    Sample type population_colour
0          gDNA           #3182bd
2          sWGA           #e31a1c
322         MDA                NA
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%R
colourings_population &lt;- df_pops_with_colours$population_colour # create a vector of colour codes
names(colourings_population) &lt;- df_pops_with_colours$`Sample type` # give each element in that vector a name - assigned from the corresponding sample type
colourings_population
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>     gDNA      sWGA       MDA 
&quot;#3182bd&quot; &quot;#e31a1c&quot;      &quot;NA&quot; 
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="put-all-the-functions-into-action">
<h3>Put all the functions into action<a class="headerlink" href="#put-all-the-functions-into-action" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%R
groups = as.factor(df_samples[pass_samples, &#39;Sample type&#39;])
names(groups) = tree[3]$&#39;tip.label&#39;
counts = computeEdgeGroupCounts(tree, groups)
edge_colors = unlist(assignGroupColorToUniqueEdges(
    tree, counts, groupcolors=colourings_population, equality_color=&#39;gray&#39;
))
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="plot-the-njt">
<h3>Plot the NJT<a class="headerlink" href="#plot-the-njt" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">plot</span><span class="o">.</span><span class="n">phylo</span><span class="p">(</span>
    <span class="n">tree</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;unrooted&quot;</span><span class="p">,</span>
    <span class="n">tip</span><span class="o">=</span><span class="n">df_samples</span><span class="p">[</span><span class="n">pass_samples</span><span class="p">,</span> <span class="s1">&#39;population_colour&#39;</span><span class="p">],</span>
    <span class="n">edge</span><span class="o">.</span><span class="n">color</span><span class="o">=</span><span class="n">unlist</span><span class="p">(</span><span class="n">edge_colors</span><span class="p">),</span>
    <span class="n">show</span><span class="o">.</span><span class="n">tip</span><span class="o">.</span><span class="n">label</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/neighbour-joining-tree-swga_45_0.png" src="../../_images/neighbour-joining-tree-swga_45_0.png" />
</div>
</div>
<p><strong>Figure legend</strong>: Lack of bias in population structure due to use of sWGA from Ghana, Upper East, in 2015. Genome-wide unrooted neighbour-joining tree showing population structure where one subset of samples were sequenced from genomic DNA (gDNA, n = 61) material (shown as blue lines), and a second subset of samples were sequenced using sWGA material (n = 164, shown in red).</p>
</div>
<div class="section" id="write-the-figure-to-a-file">
<h3>Write the figure to a file<a class="headerlink" href="#write-the-figure-to-a-file" title="Permalink to this headline">¶</a></h3>
<p>We can output this to a location in Google Drive</p>
<p>First we need to connect Google Drive by running the following:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># You will need to authorise Google Colab access to Google Drive</span>
<span class="n">drive</span><span class="o">.</span><span class="n">mount</span><span class="p">(</span><span class="s1">&#39;/content/drive&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(&quot;/content/drive&quot;, force_remount=True).
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="c1"># This will send the file to your Google Drive, where you can download it from if needed</span>
<span class="c1"># Change the file path if you wish to send the file to a specific location</span>
<span class="c1"># Change the file name if you wish to call it something else</span>
<span class="n">pdf</span><span class="p">(</span><span class="s1">&#39;/content/drive/My Drive/Ghana_UpperEast_2015.pdf&#39;</span><span class="p">)</span>
<span class="n">plot</span><span class="o">.</span><span class="n">phylo</span><span class="p">(</span>
    <span class="n">tree</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;unrooted&quot;</span><span class="p">,</span>
    <span class="n">tip</span><span class="o">=</span><span class="n">df_samples</span><span class="p">[</span><span class="n">pass_samples</span><span class="p">,</span> <span class="s1">&#39;population_colour&#39;</span><span class="p">],</span>
    <span class="n">edge</span><span class="o">.</span><span class="n">color</span><span class="o">=</span><span class="n">unlist</span><span class="p">(</span><span class="n">edge_colors</span><span class="p">),</span>
    <span class="n">show</span><span class="o">.</span><span class="n">tip</span><span class="o">.</span><span class="n">label</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] 2
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="continue-on-to-make-the-rest-of-the-figure">
<h2>Continue on to make the rest of the figure<a class="headerlink" href="#continue-on-to-make-the-rest-of-the-figure" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="b-kenya-kilifi-2007-2012">
<h2>B) Kenya, Kilifi, 2007-2012<a class="headerlink" href="#b-kenya-kilifi-2007-2012" title="Permalink to this headline">¶</a></h2>
<p>For the following three plots, we only need to re-run a section of the above code, in order to:</p>
<ul class="simple">
<li><p>Redefine pass_samples as our new subset of interest</p></li>
<li><p>Remake the phylotree object ‘tree’ for our new subset</p></li>
<li><p>Redefine ‘groups’ for the tree in relation to the new ‘pass_samples’</p></li>
<li><p>Replot the tree</p></li>
</ul>
<p>This is because all the plotting functions and the sampletype:colour association dataframe above can be used for any NJT plot. The code blocks that change the plot are the four listed above.</p>
<div class="section" id="redfine-pass-samples">
<h3>Redfine pass_samples<a class="headerlink" href="#redfine-pass-samples" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%R
# We combine the criteria using &quot;&amp;&quot; to create a logical vector &#39;pass_samples&#39;
# where each element refers to a row in the df_samples dataframe
# and evaluates to TRUE if all the conditions are met for that row, or FALSE otherwise.
pass_samples &lt;- (
    df_samples$`Exclusion reason` == &#39;Analysis_set&#39; &amp; # We use backticks (`) around column names with spaces &#39;Exclusion reason&#39;, to ensure they are correctly recognized as column names.
    df_samples$Country == &#39;Kenya&#39; &amp;
    df_samples$`Admin level 1` == &#39;Kilifi&#39; &amp;
    df_samples$Year &gt;= 2007 &amp;
    df_samples$Year &lt;= 2012 # Note we have a range of years included here
)
summary(pass_samples) # We have 373 samples that satisfy the criteria
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>   Mode   FALSE    TRUE 
logical   20491     373 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span># How many samples do we have of each sample type?
%%R
aggregate(Sample ~ Country+`Sample type`, df_samples[pass_samples, ], FUN=length)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  Country Sample type Sample
1   Kenya        gDNA    151
2   Kenya        sWGA    222
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="remake-the-tree-object">
<h3>Remake the tree object<a class="headerlink" href="#remake-the-tree-object" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">nj</span><span class="p">(</span><span class="n">mean_distance_matrix</span><span class="p">[</span><span class="n">pass_samples</span><span class="p">,</span> <span class="n">pass_samples</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="redefine-groups-for-our-current-pass-samples-variable">
<h3>Redefine ‘groups’ for our current ‘pass_samples’ variable<a class="headerlink" href="#redefine-groups-for-our-current-pass-samples-variable" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%R
groups = as.factor(df_samples[pass_samples, &#39;Sample type&#39;])
names(groups) = tree[3]$&#39;tip.label&#39;
counts = computeEdgeGroupCounts(tree, groups)
edge_colors = unlist(assignGroupColorToUniqueEdges(
    tree, counts, groupcolors=colourings_population, equality_color=&#39;gray&#39;
))
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="plot-the-new-njt">
<h3>Plot the new NJT<a class="headerlink" href="#plot-the-new-njt" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">plot</span><span class="o">.</span><span class="n">phylo</span><span class="p">(</span>
    <span class="n">tree</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;unrooted&quot;</span><span class="p">,</span>
    <span class="n">tip</span><span class="o">=</span><span class="n">df_samples</span><span class="p">[</span><span class="n">pass_samples</span><span class="p">,</span> <span class="s1">&#39;population_colour&#39;</span><span class="p">],</span>
    <span class="n">edge</span><span class="o">.</span><span class="n">color</span><span class="o">=</span><span class="n">unlist</span><span class="p">(</span><span class="n">edge_colors</span><span class="p">),</span>
    <span class="n">show</span><span class="o">.</span><span class="n">tip</span><span class="o">.</span><span class="n">label</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/neighbour-joining-tree-swga_60_0.png" src="../../_images/neighbour-joining-tree-swga_60_0.png" />
</div>
</div>
<p><strong>Figure legend</strong>: Lack of bias in population structure due to use of sWGA from Kenya, Kilifi, between 2007 - 2012. Genome-wide unrooted neighbour-joining tree showing population structure where one subset of samples were sequenced from genomic DNA (gDNA, n = 151) material (shown as blue lines), and a second subset of samples were sequenced using sWGA material (n = 222, shown in red).</p>
</div>
<div class="section" id="id1">
<h3>Write the figure to a file<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p>If you have already mounted your Google Drive for the above plot, you do not need to run it again.</p>
<p>If you need to mount it again, uncomment the following code:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># drive.mount(&#39;/content/drive&#39;)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="c1"># This will send the file to your Google Drive, where you can download it from if needed</span>
<span class="c1"># Change the file path if you wish to send the file to a specific location</span>
<span class="c1"># Change the file name if you wish to call it something else</span>
<span class="n">pdf</span><span class="p">(</span><span class="s1">&#39;/content/drive/My Drive/Kenya_Kilifi_2007_2012.pdf&#39;</span><span class="p">)</span>
<span class="n">plot</span><span class="o">.</span><span class="n">phylo</span><span class="p">(</span>
    <span class="n">tree</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;unrooted&quot;</span><span class="p">,</span>
    <span class="n">tip</span><span class="o">=</span><span class="n">df_samples</span><span class="p">[</span><span class="n">pass_samples</span><span class="p">,</span> <span class="s1">&#39;population_colour&#39;</span><span class="p">],</span>
    <span class="n">edge</span><span class="o">.</span><span class="n">color</span><span class="o">=</span><span class="n">unlist</span><span class="p">(</span><span class="n">edge_colors</span><span class="p">),</span>
    <span class="n">show</span><span class="o">.</span><span class="n">tip</span><span class="o">.</span><span class="n">label</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] 2
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="c-cambodia-pursat-2016">
<h2>C) Cambodia, Pursat, 2016<a class="headerlink" href="#c-cambodia-pursat-2016" title="Permalink to this headline">¶</a></h2>
<p>We will now repeat the above steps to generate a NJT in exactly the same way. This time we will look at samples from Cambodia, Pursat, in 2016.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%R
# Generate a logical vector for this tree
pass_samples &lt;- (
    df_samples$`Exclusion reason` == &#39;Analysis_set&#39; &amp; # We use backticks (`) around column names with spaces &#39;Exclusion reason&#39;, to ensure they are correctly recognized as column names.
    df_samples$Country == &#39;Cambodia&#39; &amp;
    df_samples$`Admin level 1` == &#39;Pursat&#39; &amp;
    df_samples$Year == 2016
)
summary(pass_samples) # We have 52 samples that satisfy the criteria
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>   Mode   FALSE    TRUE 
logical   20812      52 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span># How many samples do we have of each sample type?
%%R
aggregate(Sample ~ Country+`Sample type`, df_samples[pass_samples, ], FUN=length)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>   Country Sample type Sample
1 Cambodia        gDNA     30
2 Cambodia        sWGA     22
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">nj</span><span class="p">(</span><span class="n">mean_distance_matrix</span><span class="p">[</span><span class="n">pass_samples</span><span class="p">,</span> <span class="n">pass_samples</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%R
groups = as.factor(df_samples[pass_samples, &#39;Sample type&#39;])
names(groups) = tree[3]$&#39;tip.label&#39;
counts = computeEdgeGroupCounts(tree, groups)
edge_colors = unlist(assignGroupColorToUniqueEdges(
    tree, counts, groupcolors=colourings_population, equality_color=&#39;gray&#39;
))
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">plot</span><span class="o">.</span><span class="n">phylo</span><span class="p">(</span>
    <span class="n">tree</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;unrooted&quot;</span><span class="p">,</span>
    <span class="n">tip</span><span class="o">=</span><span class="n">df_samples</span><span class="p">[</span><span class="n">pass_samples</span><span class="p">,</span> <span class="s1">&#39;population_colour&#39;</span><span class="p">],</span>
    <span class="n">edge</span><span class="o">.</span><span class="n">color</span><span class="o">=</span><span class="n">unlist</span><span class="p">(</span><span class="n">edge_colors</span><span class="p">),</span>
    <span class="n">show</span><span class="o">.</span><span class="n">tip</span><span class="o">.</span><span class="n">label</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/neighbour-joining-tree-swga_70_0.png" src="../../_images/neighbour-joining-tree-swga_70_0.png" />
</div>
</div>
<p><strong>Figure legend</strong>: Lack of bias in population structure due to use of sWGA from Cambodia, Pursat, in 2016. Genome-wide unrooted neighbour-joining tree showing population structure where one subset of samples were sequenced from genomic DNA (gDNA, n = 30) material (shown as blue lines), and a second subset of samples were sequenced using sWGA material (n = 22, shown in red).</p>
<div class="section" id="id2">
<h3>Write the figure to a file<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<p>If you have already mounted your Google Drive for the above plot, you do not need to run it again.</p>
<p>If you need to mount it again, uncomment the following code:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># drive.mount(&#39;/content/drive&#39;)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="c1"># This will send the file to your Google Drive, where you can download it from if needed</span>
<span class="c1"># Change the file path if you wish to send the file to a specific location</span>
<span class="c1"># Change the file name if you wish to call it something else</span>
<span class="n">pdf</span><span class="p">(</span><span class="s1">&#39;/content/drive/My Drive/Cambodia_Pursat_2016.pdf&#39;</span><span class="p">)</span>
<span class="n">plot</span><span class="o">.</span><span class="n">phylo</span><span class="p">(</span>
    <span class="n">tree</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;unrooted&quot;</span><span class="p">,</span>
    <span class="n">tip</span><span class="o">=</span><span class="n">df_samples</span><span class="p">[</span><span class="n">pass_samples</span><span class="p">,</span> <span class="s1">&#39;population_colour&#39;</span><span class="p">],</span>
    <span class="n">edge</span><span class="o">.</span><span class="n">color</span><span class="o">=</span><span class="n">unlist</span><span class="p">(</span><span class="n">edge_colors</span><span class="p">),</span>
    <span class="n">show</span><span class="o">.</span><span class="n">tip</span><span class="o">.</span><span class="n">label</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] 2
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="d-vietnam-binh-phuoc-2016">
<h2>D) Vietnam, Binh Phuoc, 2016<a class="headerlink" href="#d-vietnam-binh-phuoc-2016" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%R
# Generate a logical vector for this tree
pass_samples &lt;- (
    df_samples$`Exclusion reason` == &#39;Analysis_set&#39; &amp; # We use backticks (`) around column names with spaces &#39;Exclusion reason&#39;, to ensure they are correctly recognized as column names.
    df_samples$Country == &#39;Vietnam&#39; &amp;
    df_samples$`Admin level 1` == &#39;Binh Phuoc&#39; &amp;
    df_samples$Year == 2016
)
summary(pass_samples) # We have 61 samples that satisfy the criteria
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>   Mode   FALSE    TRUE 
logical   20803      61 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span># How many samples do we have of each sample type?
%%R
aggregate(Sample ~ Country+`Sample type`, df_samples[pass_samples, ], FUN=length)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  Country Sample type Sample
1 Vietnam        gDNA     39
2 Vietnam        sWGA     22
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">nj</span><span class="p">(</span><span class="n">mean_distance_matrix</span><span class="p">[</span><span class="n">pass_samples</span><span class="p">,</span> <span class="n">pass_samples</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%R
groups = as.factor(df_samples[pass_samples, &#39;Sample type&#39;])
names(groups) = tree[3]$&#39;tip.label&#39;
counts = computeEdgeGroupCounts(tree, groups)
edge_colors = unlist(assignGroupColorToUniqueEdges(
    tree, counts, groupcolors=colourings_population, equality_color=&#39;gray&#39;
))
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">plot</span><span class="o">.</span><span class="n">phylo</span><span class="p">(</span>
    <span class="n">tree</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;unrooted&quot;</span><span class="p">,</span>
    <span class="n">tip</span><span class="o">=</span><span class="n">df_samples</span><span class="p">[</span><span class="n">pass_samples</span><span class="p">,</span> <span class="s1">&#39;population_colour&#39;</span><span class="p">],</span>
    <span class="n">edge</span><span class="o">.</span><span class="n">color</span><span class="o">=</span><span class="n">unlist</span><span class="p">(</span><span class="n">edge_colors</span><span class="p">),</span>
    <span class="n">show</span><span class="o">.</span><span class="n">tip</span><span class="o">.</span><span class="n">label</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/neighbour-joining-tree-swga_80_0.png" src="../../_images/neighbour-joining-tree-swga_80_0.png" />
</div>
</div>
<p><strong>Figure legend</strong>: Lack of bias in population structure due to use of sWGA from Vietnam, Binh Phuoc, in 2016. Genome-wide unrooted neighbour-joining tree showing population structure where one subset of samples were sequenced from genomic DNA (gDNA, n = 39) material (shown as blue lines), and a second subset of samples were sequenced using sWGA material (n = 22, shown in red).</p>
<div class="section" id="id3">
<h3>Write the figure to a file<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<p>If you have already mounted your Google Drive for the above plot, you do not need to run it again.</p>
<p>If you need to mount it again, uncomment the following code:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># drive.mount(&#39;/content/drive&#39;)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="c1"># This will send the file to your Google Drive, where you can download it from if needed</span>
<span class="c1"># Change the file path if you wish to send the file to a specific location</span>
<span class="c1"># Change the file name if you wish to call it something else</span>
<span class="n">pdf</span><span class="p">(</span><span class="s1">&#39;/content/drive/My Drive/Vietnam_BinhPhuoc_2016.pdf&#39;</span><span class="p">)</span>
<span class="n">plot</span><span class="o">.</span><span class="n">phylo</span><span class="p">(</span>
    <span class="n">tree</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;unrooted&quot;</span><span class="p">,</span>
    <span class="n">tip</span><span class="o">=</span><span class="n">df_samples</span><span class="p">[</span><span class="n">pass_samples</span><span class="p">,</span> <span class="s1">&#39;population_colour&#39;</span><span class="p">],</span>
    <span class="n">edge</span><span class="o">.</span><span class="n">color</span><span class="o">=</span><span class="n">unlist</span><span class="p">(</span><span class="n">edge_colors</span><span class="p">),</span>
    <span class="n">show</span><span class="o">.</span><span class="n">tip</span><span class="o">.</span><span class="n">label</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] 2
</pre></div>
</div>
</div>
</div>
<p><strong>Notebook Complete!</strong>
Now you should have four PDF files corresponding to the four panels of supplementary Figure 3 in the Pf7 paper.</p>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./pf7/analysis_examples"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="geographic-genetic-differentiation.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Geographic patterns of population differentiation and gene flow</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="proportion-eba175.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Plot <em>eba175</em> C Allele Proportions</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By MalariaGEN<br/>
    
        &copy; Copyright 2023.<br/>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>