
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Visualise HRP Deletion Breakpoints &#8212; MalariaGEN parasite data user guide</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Analyse csp C-Terminal Variation" href="csp-cterm-analysis.html" />
    <link rel="prev" title="hrp2 and hrp3 deletions: summary of deletion breakpoints" href="summarise_hrp2_hrp3_deletions.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">MalariaGEN parasite data user guide</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../landing-page.html">
   MalariaGEN parasite data user guide
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Pf7
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../Data_access.html">
   Pf7 Data Access
  </a>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../examples.html">
   Pf7 Analysis Examples
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="igv-notebook.html">
     Visualizing Genomic Regions of Pf7 Samples with IGV Plots
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sample-summary-table.html">
     Summarise Samples by Administrative Division
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sample-collection-over-time.html">
     Visualising sample collections over time
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="barplot-samples-by-country.html">
     Barplot of Samples by Country
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="variant-summary-table.html">
     Summarise Genetic Variants
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="global-sample-map.html">
     Plot Global Map of Samples
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="drug-resistance-summary.html">
     Drug resistance frequencies across Pf7
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="chloroquine-resistance-map.html">
     Plot Map of Chloroquine Resistance
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="summarise_hrp2_hrp3_deletions.html">
     <em>
      hrp2
     </em>
     and
     <em>
      hrp3
     </em>
     deletions: summary of deletion breakpoints
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Visualise HRP Deletion Breakpoints
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="csp-cterm-analysis.html">
     Analyse csp C-Terminal Variation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="PCoA.html">
     Principal Coordinate Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="geographic-genetic-differentiation.html">
     Geographic patterns of population differentiation and gene flow
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="neighbour-joining-tree-swga.html">
     Neighbour Joining Tree for sWGA vs. gDNA
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="proportion-eba175.html">
     Plot
     <em>
      eba175
     </em>
     C Allele Proportions
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../api.html">
   Pf7 API
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Pv4
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../pv4/Data_access.html">
   Pv4 Data Access
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../pv4/api.html">
   Pv4 API
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/pf7/analysis_examples/visualise_hrp2_hrp3_deletions.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
                onclick="printPdf(this)" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/malariagen/parasite-data"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/malariagen/parasite-data/issues/new?title=Issue%20on%20page%20%2Fpf7/analysis_examples/visualise_hrp2_hrp3_deletions.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/malariagen/parasite-data/master?urlpath=lab/tree/docs/pf7/analysis_examples/visualise_hrp2_hrp3_deletions.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        <a class="colab-button" href="https://colab.research.google.com/github/malariagen/parasite-data/blob/master/docs/pf7/analysis_examples/visualise_hrp2_hrp3_deletions.ipynb"><button type="button" class="btn btn-secondary topbarbtn"
                title="Launch Colab" data-toggle="tooltip" data-placement="left"><img class="colab-button-logo"
                    src="../../_static/images/logo_colab.png"
                    alt="Interact on Colab">Colab</button></a>
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#access-pf7-data">
   Access Pf7 Data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#hrp2-hrp3-deletions">
   <em>
    hrp2
   </em>
   &amp;
   <em>
    hrp3
   </em>
   Deletions
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#d7-reference-annotation">
   3D7 Reference Annotation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#figure-preparation">
   Figure Preparation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#create-figure">
   Create figure
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#save-figure">
     Save Figure
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Visualise HRP Deletion Breakpoints</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#access-pf7-data">
   Access Pf7 Data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#hrp2-hrp3-deletions">
   <em>
    hrp2
   </em>
   &amp;
   <em>
    hrp3
   </em>
   Deletions
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#d7-reference-annotation">
   3D7 Reference Annotation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#figure-preparation">
   Figure Preparation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#create-figure">
   Create figure
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#save-figure">
     Save Figure
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            
              <div>
                
  <div class="section" id="visualise-hrp-deletion-breakpoints">
<h1>Visualise HRP Deletion Breakpoints<a class="headerlink" href="#visualise-hrp-deletion-breakpoints" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>This notebook will create a public notebook for the generation of Figure 5 from the main <a class="reference external" href="https://wellcomeopenresearch.org/articles/8-22">Pf7 paper</a> - a figure of HRP deletion breakpoints.</p>
<p><em>hrp2</em> and <em>hrp3</em> are genes located in subtelomeric regions of the genome with very high levels of natural variation. Deletion in those genes can cause failure of rapid diagnostic tests and is therefore important to monitor.</p>
<p>Deletion is a genetic event in which a segment of DNA is entirely removed or missing. In this context, ‘breakpoints’ denote specific locations on the chromosome where such deletions take place.</p>
<p><strong>This notebook should take approximately two minutes to run.</strong></p>
</div>
<div class="section" id="setup">
<h2>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">¶</a></h2>
<p>Install and import the malariagen Python package:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>!pip install -q --no-warn-conflicts malariagen_data
import malariagen_data
</pre></div>
</div>
</div>
</div>
<p>Import required python libraries that are installed at colab by default.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">google.colab</span> <span class="kn">import</span> <span class="n">drive</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="access-pf7-data">
<h2>Access Pf7 Data<a class="headerlink" href="#access-pf7-data" title="Permalink to this headline">¶</a></h2>
<p>We use the malariagen data package to load the release data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">release_data</span> <span class="o">=</span> <span class="n">malariagen_data</span><span class="o">.</span><span class="n">Pf7</span><span class="p">()</span>
<span class="n">df_samples</span> <span class="o">=</span> <span class="n">release_data</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="hrp2-hrp3-deletions">
<h2><em>hrp2</em> &amp; <em>hrp3</em> Deletions<a class="headerlink" href="#hrp2-hrp3-deletions" title="Permalink to this headline">¶</a></h2>
<p>We additionally require list of deletion and breakpoint locations within the histidine-rich protein genes II and III (HRP-2 and -3) across 16,203 QC-pass samples. We can access this data from <a class="reference external" href="http://malariagen.net/resource/34">MalariaGEN web page</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Our website is undergoing some maintenance, please use this temporary URL</span>
<span class="c1"># Try this URL</span>
<span class="n">hrp_calls_fn</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;https://www.malariagen.net/wp-content/uploads/2024/01/hrp_calls_pf7.tsv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="c1"># But if it does not work, try this URL (uncomment the command)</span>
<span class="c1">#hrp_calls_fn = pd.read_csv(&#39;https://www.malariagen.net/sites/default/files/hrp_calls_pf7.tsv&#39;, sep=&#39;\t&#39;)</span>

<span class="c1"># Print the shape and first rows</span>
<span class="nb">print</span><span class="p">(</span><span class="n">hrp_calls_fn</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">hrp_calls_fn</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(16203, 7)
</pre></div>
</div>
<div class="output text_html">
  <div id="df-f9ef75f4-360e-4a59-b035-84db66010b8f" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Sample</th>
      <th>HRP2</th>
      <th>HRP3</th>
      <th>HRP2_breakpoint</th>
      <th>HRP3_breakpoint</th>
      <th>HRP2_deletion_type</th>
      <th>HRP3_deletion_type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>FP0008-C</td>
      <td>nodel</td>
      <td>nodel</td>
      <td>-</td>
      <td>-</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>FP0009-C</td>
      <td>nodel</td>
      <td>nodel</td>
      <td>-</td>
      <td>-</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>FP0010-CW</td>
      <td>uncallable</td>
      <td>uncallable</td>
      <td>-</td>
      <td>-</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>FP0011-CW</td>
      <td>uncallable</td>
      <td>uncallable</td>
      <td>-</td>
      <td>-</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>FP0012-CW</td>
      <td>nodel</td>
      <td>nodel</td>
      <td>-</td>
      <td>-</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-f9ef75f4-360e-4a59-b035-84db66010b8f')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-f9ef75f4-360e-4a59-b035-84db66010b8f button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-f9ef75f4-360e-4a59-b035-84db66010b8f');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-77bbb0f7-dad8-4452-a0b3-45e993e71841">
  <button class="colab-df-quickchart" onclick="quickchart('df-77bbb0f7-dad8-4452-a0b3-45e993e71841')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-77bbb0f7-dad8-4452-a0b3-45e993e71841 button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
<p>Now, let’s merge <code class="docutils literal notranslate"><span class="pre">hrp_calls_fn</span></code> with <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> which contains various metadata of Pf7 samples.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Merge df_samples with hrp_calls_fn</span>
<span class="n">df_samples</span> <span class="o">=</span>  <span class="n">df_samples</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">hrp_calls_fn</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="d7-reference-annotation">
<h2>3D7 Reference Annotation<a class="headerlink" href="#d7-reference-annotation" title="Permalink to this headline">¶</a></h2>
<p>Another relevant dataset in this notebook is the 3D7 reference annotation which is
used for genome annotations in Pf7. It can be accessed at the following FTP link: ftp://ngs.sanger.ac.uk/production/malaria/Resource/34/Pfalciparum_replace_Pf3D7_MIT_v3_with_Pf_M76611.gff.</p>
<p>We will use the <code class="docutils literal notranslate"><span class="pre">wget</span></code> package to download this file. After running the following cell, the file should appear in the list of files on the left.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>!wget ftp://ngs.sanger.ac.uk/production/malaria/Resource/34/Pfalciparum_replace_Pf3D7_MIT_v3_with_Pf_M76611.gff
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>--2024-02-08 11:27:43--  ftp://ngs.sanger.ac.uk/production/malaria/Resource/34/Pfalciparum_replace_Pf3D7_MIT_v3_with_Pf_M76611.gff
           =&gt; ‘Pfalciparum_replace_Pf3D7_MIT_v3_with_Pf_M76611.gff.1’
Resolving ngs.sanger.ac.uk (ngs.sanger.ac.uk)... 193.62.203.221
Connecting to ngs.sanger.ac.uk (ngs.sanger.ac.uk)|193.62.203.221|:21... connected.
Logging in as anonymous ... Logged in!
==&gt; SYST ... done.    ==&gt; PWD ... done.
==&gt; TYPE I ... done.  ==&gt; CWD (1) /production/malaria/Resource/34 ... done.
==&gt; SIZE Pfalciparum_replace_Pf3D7_MIT_v3_with_Pf_M76611.gff ... 21035550
==&gt; PASV ... done.    ==&gt; RETR Pfalciparum_replace_Pf3D7_MIT_v3_with_Pf_M76611.gff ... done.
Length: 21035550 (20M) (unauthoritative)

Pfalciparum_replace 100%[===================&gt;]  20.06M  18.2MB/s    in 1.1s    

2024-02-08 11:27:45 (18.2 MB/s) - ‘Pfalciparum_replace_Pf3D7_MIT_v3_with_Pf_M76611.gff.1’ saved [21035550]
</pre></div>
</div>
</div>
</div>
<p>This file is in GFF file format defined by <a class="reference external" href="https://www.ensembl.org/info/website/upload/gff.html">Ensembl</a>.</p>
<p>Since the file is tab-separated, we can proceed to read the data into a Pandas dataframe once again.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Read the file column names.</span>
<span class="n">df_gff</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;/content/Pfalciparum_replace_Pf3D7_MIT_v3_with_Pf_M76611.gff&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;db&#39;</span><span class="p">,</span> <span class="s1">&#39;type&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">,</span> <span class="s1">&#39;u1&#39;</span><span class="p">,</span> <span class="s1">&#39;strand&#39;</span><span class="p">,</span> <span class="s1">&#39;u2&#39;</span><span class="p">,</span> <span class="s1">&#39;id&#39;</span><span class="p">],</span> <span class="n">comment</span><span class="o">=</span><span class="s1">&#39;#&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="c1"># Print the shape (rows and columns) of df_gff</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df_gff</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="c1"># Show the first 5 rows</span>
<span class="n">df_gff</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(40713, 9)
</pre></div>
</div>
<div class="output text_html">
  <div id="df-205d883b-a947-42a2-881e-18c54c4c0997" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>db</th>
      <th>type</th>
      <th>start</th>
      <th>end</th>
      <th>u1</th>
      <th>strand</th>
      <th>u2</th>
      <th>id</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Pf3D7_01_v3</td>
      <td>chado</td>
      <td>repeat_region</td>
      <td>1</td>
      <td>360</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=Pfalciparum_REP_20;comment=telomeric repeat...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Pf3D7_01_v3</td>
      <td>chado</td>
      <td>repeat_region</td>
      <td>361</td>
      <td>1418</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=Pfalciparum_REP_15;comment=14bp repeat</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Pf3D7_01_v3</td>
      <td>chado</td>
      <td>repeat_region</td>
      <td>2160</td>
      <td>3858</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=Pfalciparum_REP_35;comment=65bp repeat</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Pf3D7_01_v3</td>
      <td>chado</td>
      <td>repeat_region</td>
      <td>8856</td>
      <td>9021</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=Pfalciparum_REP_5;comment=25bp repeat</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Pf3D7_01_v3</td>
      <td>chado</td>
      <td>repeat_region</td>
      <td>9313</td>
      <td>9529</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=Pfalciparum_REP_25;comment=26bp repeat</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-205d883b-a947-42a2-881e-18c54c4c0997')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-205d883b-a947-42a2-881e-18c54c4c0997 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-205d883b-a947-42a2-881e-18c54c4c0997');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-6d072828-7bcf-4956-b693-f4b22088d8f6">
  <button class="colab-df-quickchart" onclick="quickchart('df-6d072828-7bcf-4956-b693-f4b22088d8f6')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-6d072828-7bcf-4956-b693-f4b22088d8f6 button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
</div>
<div class="section" id="figure-preparation">
<h2>Figure Preparation<a class="headerlink" href="#figure-preparation" title="Permalink to this headline">¶</a></h2>
<p>We need to find the start and end positions of chromosomes to draw gene annotations in the figure that we are going to create.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Find start and end positions for each chromosome by grouping chromosome coordinates</span>
<span class="n">df_chroms</span> <span class="o">=</span> <span class="n">df_gff</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;chrom&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span><span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="s1">&#39;max&#39;</span><span class="p">})</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="c1"># Set &#39;chrom&#39; as the index</span>
<span class="n">df_chroms</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">df_chroms</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-d2e10693-814f-4c8e-98f6-4364eba938e8" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>start</th>
      <th>end</th>
    </tr>
    <tr>
      <th>chrom</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Pf3D7_01_v3</th>
      <td>1</td>
      <td>640851</td>
    </tr>
    <tr>
      <th>Pf3D7_02_v3</th>
      <td>1</td>
      <td>947102</td>
    </tr>
    <tr>
      <th>Pf3D7_03_v3</th>
      <td>1</td>
      <td>1067971</td>
    </tr>
    <tr>
      <th>Pf3D7_04_v3</th>
      <td>1</td>
      <td>1200490</td>
    </tr>
    <tr>
      <th>Pf3D7_05_v3</th>
      <td>1</td>
      <td>1343557</td>
    </tr>
    <tr>
      <th>Pf3D7_06_v3</th>
      <td>1</td>
      <td>1418242</td>
    </tr>
    <tr>
      <th>Pf3D7_07_v3</th>
      <td>1</td>
      <td>1445207</td>
    </tr>
    <tr>
      <th>Pf3D7_08_v3</th>
      <td>1</td>
      <td>1472805</td>
    </tr>
    <tr>
      <th>Pf3D7_09_v3</th>
      <td>1</td>
      <td>1541735</td>
    </tr>
    <tr>
      <th>Pf3D7_10_v3</th>
      <td>1</td>
      <td>1687656</td>
    </tr>
    <tr>
      <th>Pf3D7_11_v3</th>
      <td>1</td>
      <td>2038340</td>
    </tr>
    <tr>
      <th>Pf3D7_12_v3</th>
      <td>1</td>
      <td>2260569</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3</th>
      <td>1</td>
      <td>2925236</td>
    </tr>
    <tr>
      <th>Pf3D7_14_v3</th>
      <td>1</td>
      <td>3291936</td>
    </tr>
    <tr>
      <th>Pf3D7_API_v3</th>
      <td>1</td>
      <td>34225</td>
    </tr>
    <tr>
      <th>Pf_M76611</th>
      <td>3</td>
      <td>5954</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-d2e10693-814f-4c8e-98f6-4364eba938e8')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-d2e10693-814f-4c8e-98f6-4364eba938e8 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-d2e10693-814f-4c8e-98f6-4364eba938e8');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-076b6ae1-efd7-48f5-a57d-182398cf0995">
  <button class="colab-df-quickchart" onclick="quickchart('df-076b6ae1-efd7-48f5-a57d-182398cf0995')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-076b6ae1-efd7-48f5-a57d-182398cf0995 button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
<p>The next question is: How many samples have deletions in each country?</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># This function returns samples with deletion for each country.</span>

<span class="k">def</span> <span class="nf">breakpoint_agg</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
    <span class="n">names</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">()</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;Countries&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="n">countries</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># Loop over each country</span>
    <span class="c1"># Count non-zero samples (with deletion)</span>
    <span class="k">for</span> <span class="n">country</span> <span class="ow">in</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Country&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
        <span class="n">countries</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">country</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;Country&#39;</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">country</span><span class="p">)</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
    <span class="c1"># Join together country name and number</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;Countries&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">countries</span><span class="p">)</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;Samples with deletion&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">names</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We will apply <code class="docutils literal notranslate"><span class="pre">breakpoint_agg</span></code> function to <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> separately for <em>hrp2</em> and <em>hrp3</em>.</p>
<p>Additionally, we will find genomic coordinates at the edges for mapping in the figure.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Group samples by Deletion type and HRP3_breakpoint</span>
<span class="c1"># Apply breakpoint_agg to count samples with deletion in each country</span>
<span class="n">df_hrp2</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">df_samples</span><span class="p">[</span>
        <span class="n">df_samples</span><span class="p">[</span><span class="s1">&#39;QC pass&#39;</span><span class="p">]</span>
        <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_samples</span><span class="p">[</span><span class="s1">&#39;HRP2&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;del&#39;</span> <span class="p">)</span>
<span class="p">]</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">Gene</span><span class="o">=</span><span class="s1">&#39;$hrp2$&#39;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;HRP2_deletion_type&#39;</span><span class="p">:</span> <span class="s1">&#39;Deletion type&#39;</span><span class="p">})</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Deletion type&#39;</span><span class="p">,</span> <span class="s1">&#39;HRP2_breakpoint&#39;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">breakpoint_agg</span><span class="p">)</span>
    <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="p">)</span>
<span class="c1"># Seperate coordinate value from chromosome</span>
<span class="n">df_hrp2</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_hrp2</span><span class="p">[</span><span class="s1">&#39;HRP2_breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]))</span>

<span class="c1"># Print min and max coordinates</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;HRP2 min breakpoint = </span><span class="si">{</span><span class="n">df_hrp2</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;HRP2 max breakpoint = </span><span class="si">{</span><span class="n">df_hrp2</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">df_hrp2</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>HRP2 min breakpoint = 1373732
HRP2 max breakpoint = 1374986
</pre></div>
</div>
<div class="output text_html">
  <div id="df-a421c564-dedc-4552-8b31-1d33a3cb8bdc" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Deletion type</th>
      <th>HRP2_breakpoint</th>
      <th>Countries</th>
      <th>Samples with deletion</th>
      <th>breakpoint</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Telomere healing</td>
      <td>Pf3D7_08_v3:1373732</td>
      <td>Cambodia (1)</td>
      <td>1</td>
      <td>1373732</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Telomere healing</td>
      <td>Pf3D7_08_v3:1374280</td>
      <td>Sudan (1)</td>
      <td>1</td>
      <td>1374280</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Telomere healing</td>
      <td>Pf3D7_08_v3:1374462</td>
      <td>Indonesia (5)</td>
      <td>5</td>
      <td>1374462</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Telomere healing</td>
      <td>Pf3D7_08_v3:1374932</td>
      <td>Peru (2)</td>
      <td>2</td>
      <td>1374932</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Telomere healing</td>
      <td>Pf3D7_08_v3:1374986</td>
      <td>Peru (6)</td>
      <td>6</td>
      <td>1374986</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-a421c564-dedc-4552-8b31-1d33a3cb8bdc')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-a421c564-dedc-4552-8b31-1d33a3cb8bdc button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-a421c564-dedc-4552-8b31-1d33a3cb8bdc');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-4f86f84b-5cab-4de7-81a8-2790117ca263">
  <button class="colab-df-quickchart" onclick="quickchart('df-4f86f84b-5cab-4de7-81a8-2790117ca263')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-4f86f84b-5cab-4de7-81a8-2790117ca263 button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
<p>A repeat of the same look-up for <em>hrp3</em>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Group samples by Deletion type and HRP3_breakpoint</span>
<span class="c1"># Apply breakpoint_agg to count samples with deletion in each country</span>
<span class="n">df_hrp3</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">df_samples</span><span class="p">[</span>
        <span class="n">df_samples</span><span class="p">[</span><span class="s1">&#39;QC pass&#39;</span><span class="p">]</span>
        <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_samples</span><span class="p">[</span><span class="s1">&#39;HRP3&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;del&#39;</span> <span class="p">)</span>
    <span class="p">]</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;HRP3_deletion_type&#39;</span><span class="p">:</span> <span class="s1">&#39;Deletion type&#39;</span><span class="p">})</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Deletion type&#39;</span><span class="p">,</span> <span class="s1">&#39;HRP3_breakpoint&#39;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">breakpoint_agg</span><span class="p">)</span>
    <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="p">)</span>
<span class="c1"># Seperate coordinate value from chromosome</span>
<span class="n">df_hrp3</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_hrp3</span><span class="p">[</span><span class="s1">&#39;HRP3_breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>

<span class="c1"># Print min and max coordinates</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;HRP3 min breakpoint = </span><span class="si">{</span><span class="n">df_hrp3</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;HRP3 max breakpoint = </span><span class="si">{</span><span class="n">df_hrp3</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">df_hrp3</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>HRP3 min breakpoint = 2800004-2807159
HRP3 max breakpoint = 2841120
</pre></div>
</div>
<div class="output text_html">
  <div id="df-41b27b72-1dd9-4679-8653-60bc7a08b399" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Deletion type</th>
      <th>HRP3_breakpoint</th>
      <th>Countries</th>
      <th>Samples with deletion</th>
      <th>breakpoint</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Chromosome 11 recombination</td>
      <td>Pf3D7_13_v3:2800004-2807159</td>
      <td>Thailand (1), Ghana (1), Indonesia (40), Peru ...</td>
      <td>151</td>
      <td>2800004-2807159</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Chromosome 5 recombination</td>
      <td>Pf3D7_13_v3:2835587-2835612</td>
      <td>Cambodia (20), Vietnam (1)</td>
      <td>21</td>
      <td>2835587-2835612</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2811525</td>
      <td>India (1)</td>
      <td>1</td>
      <td>2811525</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2812344</td>
      <td>Sudan (1)</td>
      <td>1</td>
      <td>2812344</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2815249</td>
      <td>Tanzania (1)</td>
      <td>1</td>
      <td>2815249</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2822480</td>
      <td>Ghana (1)</td>
      <td>1</td>
      <td>2822480</td>
    </tr>
    <tr>
      <th>6</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2823645</td>
      <td>Kenya (1)</td>
      <td>1</td>
      <td>2823645</td>
    </tr>
    <tr>
      <th>7</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2830952</td>
      <td>Cambodia (7)</td>
      <td>7</td>
      <td>2830952</td>
    </tr>
    <tr>
      <th>8</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2832080</td>
      <td>Democratic Republic of the Congo (1)</td>
      <td>1</td>
      <td>2832080</td>
    </tr>
    <tr>
      <th>9</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2834604</td>
      <td>Vietnam (1)</td>
      <td>1</td>
      <td>2834604</td>
    </tr>
    <tr>
      <th>10</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2835532</td>
      <td>Thailand (1)</td>
      <td>1</td>
      <td>2835532</td>
    </tr>
    <tr>
      <th>11</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2837145</td>
      <td>Vietnam (7)</td>
      <td>7</td>
      <td>2837145</td>
    </tr>
    <tr>
      <th>12</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2837392</td>
      <td>Cambodia (2), Laos (1)</td>
      <td>3</td>
      <td>2837392</td>
    </tr>
    <tr>
      <th>13</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2838654</td>
      <td>Indonesia (2)</td>
      <td>2</td>
      <td>2838654</td>
    </tr>
    <tr>
      <th>14</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2841024</td>
      <td>Thailand (1)</td>
      <td>1</td>
      <td>2841024</td>
    </tr>
    <tr>
      <th>15</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2841120</td>
      <td>Indonesia (1)</td>
      <td>1</td>
      <td>2841120</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-41b27b72-1dd9-4679-8653-60bc7a08b399')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-41b27b72-1dd9-4679-8653-60bc7a08b399 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-41b27b72-1dd9-4679-8653-60bc7a08b399');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-42d1a3ad-33b5-4b07-ae02-4090d9828b9d">
  <button class="colab-df-quickchart" onclick="quickchart('df-42d1a3ad-33b5-4b07-ae02-4090d9828b9d')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-42d1a3ad-33b5-4b07-ae02-4090d9828b9d button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
<p>It seems that Chromosome 11 recombination breakpoints are observed in multiple countries which motivates us to print the full list of countries.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Locate the first row by using the index</span>
<span class="n">df_hrp3</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;Countries&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;Thailand (1), Ghana (1), Indonesia (40), Peru (15), Bangladesh (1), Vietnam (2), Colombia (50), Ethiopia (9), Senegal (6), Laos (14), Cambodia (3), Sudan (5), Mali (1), Gambia (3)&#39;
</pre></div>
</div>
</div>
</div>
<p>Given the fact that many of these events result in the deletion of other genes in addition to <em>hrp2</em> and <em>hrp3</em>.</p>
<p>We could have a look at which genes are present within the range of breakpoints before including them in the plot.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Genes in hrp2 breakpoints</span>
<span class="n">df_gff</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Pf3D7_08_v3&#39;</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">1375500</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">1364000</span> <span class="p">)</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-0ba11189-c458-4b68-969b-6cecf9456833" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>db</th>
      <th>type</th>
      <th>start</th>
      <th>end</th>
      <th>u1</th>
      <th>strand</th>
      <th>u2</th>
      <th>id</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>15534</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>three_prime_UTR</td>
      <td>1364640</td>
      <td>1365466</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_0831700.1:3UTR;Parent=PF3D7_0831700.1</td>
    </tr>
    <tr>
      <th>15535</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>gene</td>
      <td>1364640</td>
      <td>1369862</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_0831700;Name=HSP70x;previous_systemat...</td>
    </tr>
    <tr>
      <th>15536</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>mRNA</td>
      <td>1364640</td>
      <td>1369862</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_0831700.1;comment=HSP70-x can be knoc...</td>
    </tr>
    <tr>
      <th>15537</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>1365467</td>
      <td>1367506</td>
      <td>.</td>
      <td>-</td>
      <td>0</td>
      <td>ID=PF3D7_0831700.1:exon:1;Parent=PF3D7_0831700.1</td>
    </tr>
    <tr>
      <th>15538</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>polypeptide</td>
      <td>1365467</td>
      <td>1367506</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_0831700.1:pep;Derives_from=PF3D7_0831...</td>
    </tr>
    <tr>
      <th>15539</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>five_prime_UTR</td>
      <td>1367507</td>
      <td>1367640</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_0831700.1:5UTR:1;Parent=PF3D7_0831700.1</td>
    </tr>
    <tr>
      <th>15540</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>five_prime_UTR</td>
      <td>1368649</td>
      <td>1369862</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_0831700.1:5UTR;Parent=PF3D7_0831700.1</td>
    </tr>
    <tr>
      <th>15541</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>pseudogenic_exon</td>
      <td>1371847</td>
      <td>1372100</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_0831750.1:exon:2;Parent=PF3D7_0831750.1</td>
    </tr>
    <tr>
      <th>15542</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>pseudogene</td>
      <td>1371847</td>
      <td>1372720</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_0831750</td>
    </tr>
    <tr>
      <th>15543</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>pseudogenic_transcript</td>
      <td>1371847</td>
      <td>1372720</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_0831750.1;Parent=PF3D7_0831750;Dbxref...</td>
    </tr>
    <tr>
      <th>15544</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>polypeptide</td>
      <td>1371847</td>
      <td>1372720</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_0831750.1:pep;Derives_from=PF3D7_0831...</td>
    </tr>
    <tr>
      <th>15545</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>pseudogenic_exon</td>
      <td>1372103</td>
      <td>1372223</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_0831750.1:exon:3;Parent=PF3D7_0831750.1</td>
    </tr>
    <tr>
      <th>15546</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>pseudogenic_exon</td>
      <td>1372225</td>
      <td>1372291</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_0831750.1:exon:4;Parent=PF3D7_0831750.1</td>
    </tr>
    <tr>
      <th>15547</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>pseudogenic_exon</td>
      <td>1372294</td>
      <td>1372577</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_0831750.1:exon:5;Parent=PF3D7_0831750.1</td>
    </tr>
    <tr>
      <th>15548</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>pseudogenic_exon</td>
      <td>1372579</td>
      <td>1372667</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_0831750.1:exon:6;Parent=PF3D7_0831750.1</td>
    </tr>
    <tr>
      <th>15549</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>pseudogenic_exon</td>
      <td>1372669</td>
      <td>1372720</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_0831750.1:exon:7;Parent=PF3D7_0831750.1</td>
    </tr>
    <tr>
      <th>15550</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>three_prime_UTR</td>
      <td>1373212</td>
      <td>1374235</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_0831800.1:3UTR;Parent=PF3D7_0831800.1</td>
    </tr>
    <tr>
      <th>15551</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>gene</td>
      <td>1373212</td>
      <td>1376988</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_0831800;Name=HRP2;previous_systematic...</td>
    </tr>
    <tr>
      <th>15552</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>mRNA</td>
      <td>1373212</td>
      <td>1376988</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_0831800.1;comment=HRP2%2C which is on...</td>
    </tr>
    <tr>
      <th>15553</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>1374236</td>
      <td>1375084</td>
      <td>.</td>
      <td>-</td>
      <td>0</td>
      <td>ID=PF3D7_0831800.1:exon:3;Parent=PF3D7_0831800.1</td>
    </tr>
    <tr>
      <th>15554</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>polypeptide</td>
      <td>1374236</td>
      <td>1375299</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_0831800.1:pep;Derives_from=PF3D7_0831...</td>
    </tr>
    <tr>
      <th>15555</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>1375231</td>
      <td>1375299</td>
      <td>.</td>
      <td>-</td>
      <td>0</td>
      <td>ID=PF3D7_0831800.1:exon:2;Parent=PF3D7_0831800.1</td>
    </tr>
    <tr>
      <th>15556</th>
      <td>Pf3D7_08_v3</td>
      <td>chado</td>
      <td>five_prime_UTR</td>
      <td>1375300</td>
      <td>1376988</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_0831800.1:5UTR;Parent=PF3D7_0831800.1</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-0ba11189-c458-4b68-969b-6cecf9456833')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-0ba11189-c458-4b68-969b-6cecf9456833 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-0ba11189-c458-4b68-969b-6cecf9456833');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-d3756752-486e-476e-96ce-fd7009c5dac2">
  <button class="colab-df-quickchart" onclick="quickchart('df-d3756752-486e-476e-96ce-fd7009c5dac2')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-d3756752-486e-476e-96ce-fd7009c5dac2 button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Genes in hrp3 breakpoints</span>
<span class="n">pd</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">display</span><span class="o">.</span><span class="n">max_rows</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">df_gff</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Pf3D7_13_v3&#39;</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">2845000</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">2795000</span> <span class="p">)</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-bc742f21-ce0e-46ad-8893-1d2903892399" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>db</th>
      <th>type</th>
      <th>start</th>
      <th>end</th>
      <th>u1</th>
      <th>strand</th>
      <th>u2</th>
      <th>id</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>34303</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>gene</td>
      <td>2796119</td>
      <td>2797144</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1370800;previous_systematic_id=PF13TR...</td>
    </tr>
    <tr>
      <th>34304</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>ncRNA</td>
      <td>2796119</td>
      <td>2797144</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1370800.1;comment=Barrell%2C March 20...</td>
    </tr>
    <tr>
      <th>34305</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2796119</td>
      <td>2797144</td>
      <td>.</td>
      <td>+</td>
      <td>0</td>
      <td>ID=PF3D7_1370800.1:exon:1;Parent=PF3D7_1370800.1</td>
    </tr>
    <tr>
      <th>34306</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>gene</td>
      <td>2797507</td>
      <td>2798103</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1370900;previous_systematic_id=PF13TR...</td>
    </tr>
    <tr>
      <th>34307</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>ncRNA</td>
      <td>2797507</td>
      <td>2798103</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1370900.1;comment=Barrell%2C March 20...</td>
    </tr>
    <tr>
      <th>34308</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2797507</td>
      <td>2798103</td>
      <td>.</td>
      <td>+</td>
      <td>0</td>
      <td>ID=PF3D7_1370900.1:exon:1;Parent=PF3D7_1370900.1</td>
    </tr>
    <tr>
      <th>34309</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>gene</td>
      <td>2800004</td>
      <td>2802154</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371000;previous_systematic_id=MAL13_...</td>
    </tr>
    <tr>
      <th>34310</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>rRNA</td>
      <td>2800004</td>
      <td>2802154</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371000.1;literature=PMID:17901154;Pa...</td>
    </tr>
    <tr>
      <th>34311</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2800004</td>
      <td>2802154</td>
      <td>.</td>
      <td>+</td>
      <td>0</td>
      <td>ID=PF3D7_1371000.1:exon:3;Parent=PF3D7_1371000.1</td>
    </tr>
    <tr>
      <th>34312</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>gene</td>
      <td>2802527</td>
      <td>2802686</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371200;previous_systematic_id=MAL13_...</td>
    </tr>
    <tr>
      <th>34313</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>rRNA</td>
      <td>2802527</td>
      <td>2802686</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371200.1;literature=PMID:17901154;Pa...</td>
    </tr>
    <tr>
      <th>34314</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2802527</td>
      <td>2802686</td>
      <td>.</td>
      <td>+</td>
      <td>0</td>
      <td>ID=PF3D7_1371200.1:exon:1;Parent=PF3D7_1371200.1</td>
    </tr>
    <tr>
      <th>34315</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>gene</td>
      <td>2802945</td>
      <td>2807159</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371300;previous_systematic_id=MAL13_...</td>
    </tr>
    <tr>
      <th>34316</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>rRNA</td>
      <td>2802945</td>
      <td>2807159</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371300.1;literature=PMID:17901154;Pa...</td>
    </tr>
    <tr>
      <th>34317</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2802945</td>
      <td>2807159</td>
      <td>.</td>
      <td>+</td>
      <td>0</td>
      <td>ID=PF3D7_1371300.1:exon:3;Parent=PF3D7_1371300.1</td>
    </tr>
    <tr>
      <th>34318</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>three_prime_UTR</td>
      <td>2808200</td>
      <td>2808562</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_1371500.1:3UTR;Parent=PF3D7_1371500.1</td>
    </tr>
    <tr>
      <th>34319</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>gene</td>
      <td>2808200</td>
      <td>2810256</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_1371500;previous_systematic_id=MAL13P...</td>
    </tr>
    <tr>
      <th>34320</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>mRNA</td>
      <td>2808200</td>
      <td>2810256</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_1371500.1;Parent=PF3D7_1371500;Dbxref...</td>
    </tr>
    <tr>
      <th>34321</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2808563</td>
      <td>2809000</td>
      <td>.</td>
      <td>-</td>
      <td>0</td>
      <td>ID=PF3D7_1371500.1:exon:1;Parent=PF3D7_1371500.1</td>
    </tr>
    <tr>
      <th>34322</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>polypeptide</td>
      <td>2808563</td>
      <td>2809222</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_1371500.1:1:pep;Derives_from=PF3D7_13...</td>
    </tr>
    <tr>
      <th>34323</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2809154</td>
      <td>2809222</td>
      <td>.</td>
      <td>-</td>
      <td>0</td>
      <td>ID=PF3D7_1371500.1:exon:2;Parent=PF3D7_1371500.1</td>
    </tr>
    <tr>
      <th>34324</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>five_prime_UTR</td>
      <td>2809223</td>
      <td>2810256</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_1371500.1:5UTR;Parent=PF3D7_1371500.1</td>
    </tr>
    <tr>
      <th>34325</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>five_prime_UTR</td>
      <td>2811277</td>
      <td>2821077</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371700.1:5UTR;Parent=PF3D7_1371700.1</td>
    </tr>
    <tr>
      <th>34326</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>gene</td>
      <td>2811277</td>
      <td>2824446</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371700;previous_systematic_id=MAL13P...</td>
    </tr>
    <tr>
      <th>34327</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>mRNA</td>
      <td>2811277</td>
      <td>2824446</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371700.1;literature=PMID:15752424,PM...</td>
    </tr>
    <tr>
      <th>34328</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>pseudogenic_exon</td>
      <td>2811706</td>
      <td>2812263</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371600.1:exon:2;Parent=PF3D7_1371600.1</td>
    </tr>
    <tr>
      <th>34329</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>pseudogene</td>
      <td>2811706</td>
      <td>2820270</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371600;Name=EBL1;previous_systematic...</td>
    </tr>
    <tr>
      <th>34330</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>pseudogenic_transcript</td>
      <td>2811706</td>
      <td>2820270</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371600.1;literature=PMID:15760663,PM...</td>
    </tr>
    <tr>
      <th>34331</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>polypeptide</td>
      <td>2811706</td>
      <td>2820270</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371600.1:pep;Derives_from=PF3D7_1371...</td>
    </tr>
    <tr>
      <th>34332</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>pseudogenic_exon</td>
      <td>2812266</td>
      <td>2819628</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371600.1:exon:3;Parent=PF3D7_1371600.1</td>
    </tr>
    <tr>
      <th>34333</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>pseudogenic_exon</td>
      <td>2819764</td>
      <td>2819851</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371600.1:exon:1;Parent=PF3D7_1371600.1</td>
    </tr>
    <tr>
      <th>34334</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>pseudogenic_exon</td>
      <td>2820015</td>
      <td>2820088</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371600.1:exon:4;Parent=PF3D7_1371600.1</td>
    </tr>
    <tr>
      <th>34335</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>pseudogenic_exon</td>
      <td>2820227</td>
      <td>2820270</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371600.1:exon:5;Parent=PF3D7_1371600.1</td>
    </tr>
    <tr>
      <th>34336</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2821078</td>
      <td>2821173</td>
      <td>.</td>
      <td>+</td>
      <td>0</td>
      <td>ID=PF3D7_1371700.1:exon:2;Parent=PF3D7_1371700.1</td>
    </tr>
    <tr>
      <th>34337</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>polypeptide</td>
      <td>2821078</td>
      <td>2823292</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371700.1:pep;Derives_from=PF3D7_1371...</td>
    </tr>
    <tr>
      <th>34338</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2821278</td>
      <td>2822786</td>
      <td>.</td>
      <td>+</td>
      <td>0</td>
      <td>ID=PF3D7_1371700.1:exon:3;Parent=PF3D7_1371700.1</td>
    </tr>
    <tr>
      <th>34339</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2823212</td>
      <td>2823292</td>
      <td>.</td>
      <td>+</td>
      <td>0</td>
      <td>ID=PF3D7_1371700.1:exon:4;Parent=PF3D7_1371700.1</td>
    </tr>
    <tr>
      <th>34340</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>three_prime_UTR</td>
      <td>2823251</td>
      <td>2824301</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_1371800.1:3UTR;Parent=PF3D7_1371800.1</td>
    </tr>
    <tr>
      <th>34341</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>gene</td>
      <td>2823251</td>
      <td>2825967</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_1371800;previous_systematic_id=MAL13P...</td>
    </tr>
    <tr>
      <th>34342</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>mRNA</td>
      <td>2823251</td>
      <td>2825967</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_1371800.1;Parent=PF3D7_1371800;Dbxref...</td>
    </tr>
    <tr>
      <th>34343</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>three_prime_UTR</td>
      <td>2823293</td>
      <td>2824446</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371700.1:3UTR;Parent=PF3D7_1371700.1</td>
    </tr>
    <tr>
      <th>34344</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2824302</td>
      <td>2825552</td>
      <td>.</td>
      <td>-</td>
      <td>0</td>
      <td>ID=PF3D7_1371800.1:exon:3;Parent=PF3D7_1371800.1</td>
    </tr>
    <tr>
      <th>34345</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>polypeptide</td>
      <td>2824302</td>
      <td>2825852</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_1371800.1:pep;Derives_from=PF3D7_1371...</td>
    </tr>
    <tr>
      <th>34346</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2825781</td>
      <td>2825852</td>
      <td>.</td>
      <td>-</td>
      <td>0</td>
      <td>ID=PF3D7_1371800.1:exon:2;Parent=PF3D7_1371800.1</td>
    </tr>
    <tr>
      <th>34347</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>five_prime_UTR</td>
      <td>2825853</td>
      <td>2825967</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_1371800.1:5UTR;Parent=PF3D7_1371800.1</td>
    </tr>
    <tr>
      <th>34348</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>five_prime_UTR</td>
      <td>2829530</td>
      <td>2829855</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371900.1:5UTR;Parent=PF3D7_1371900.1</td>
    </tr>
    <tr>
      <th>34349</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>gene</td>
      <td>2829530</td>
      <td>2830830</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371900;previous_systematic_id=MAL13P...</td>
    </tr>
    <tr>
      <th>34350</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>mRNA</td>
      <td>2829530</td>
      <td>2830830</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371900.1;Parent=PF3D7_1371900;Dbxref...</td>
    </tr>
    <tr>
      <th>34351</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2829856</td>
      <td>2829927</td>
      <td>.</td>
      <td>+</td>
      <td>0</td>
      <td>ID=PF3D7_1371900.1:exon:1;Parent=PF3D7_1371900.1</td>
    </tr>
    <tr>
      <th>34352</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>polypeptide</td>
      <td>2829856</td>
      <td>2830669</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371900.1:pep;Derives_from=PF3D7_1371...</td>
    </tr>
    <tr>
      <th>34353</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2830109</td>
      <td>2830669</td>
      <td>.</td>
      <td>+</td>
      <td>0</td>
      <td>ID=PF3D7_1371900.1:exon:2;Parent=PF3D7_1371900.1</td>
    </tr>
    <tr>
      <th>34354</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>three_prime_UTR</td>
      <td>2830670</td>
      <td>2830830</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1371900.1:3UTR;Parent=PF3D7_1371900.1</td>
    </tr>
    <tr>
      <th>34355</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>five_prime_UTR</td>
      <td>2832623</td>
      <td>2832951</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1372000.1:5UTR;Parent=PF3D7_1372000.1</td>
    </tr>
    <tr>
      <th>34356</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>gene</td>
      <td>2832623</td>
      <td>2835439</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1372000;previous_systematic_id=MAL13P...</td>
    </tr>
    <tr>
      <th>34357</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>mRNA</td>
      <td>2832623</td>
      <td>2835439</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1372000.1;comment=PHIST domain protei...</td>
    </tr>
    <tr>
      <th>34358</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2832952</td>
      <td>2833086</td>
      <td>.</td>
      <td>+</td>
      <td>0</td>
      <td>ID=PF3D7_1372000.1:exon:1;Parent=PF3D7_1372000.1</td>
    </tr>
    <tr>
      <th>34359</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>polypeptide</td>
      <td>2832952</td>
      <td>2834322</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1372000.1:pep;Derives_from=PF3D7_1372...</td>
    </tr>
    <tr>
      <th>34360</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2833204</td>
      <td>2834322</td>
      <td>.</td>
      <td>+</td>
      <td>0</td>
      <td>ID=PF3D7_1372000.1:exon:2;Parent=PF3D7_1372000.1</td>
    </tr>
    <tr>
      <th>34361</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>three_prime_UTR</td>
      <td>2834323</td>
      <td>2835439</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1372000.1:3UTR;Parent=PF3D7_1372000.1</td>
    </tr>
    <tr>
      <th>34362</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>five_prime_UTR</td>
      <td>2835756</td>
      <td>2837052</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1372100.1:5UTR;Parent=PF3D7_1372100.1</td>
    </tr>
    <tr>
      <th>34363</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>gene</td>
      <td>2835756</td>
      <td>2839580</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1372100;Name=GEXP04;previous_systemat...</td>
    </tr>
    <tr>
      <th>34364</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>mRNA</td>
      <td>2835756</td>
      <td>2839580</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1372100.1;comment=Barrell%2C October ...</td>
    </tr>
    <tr>
      <th>34365</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2837053</td>
      <td>2837136</td>
      <td>.</td>
      <td>+</td>
      <td>0</td>
      <td>ID=PF3D7_1372100.1:exon:1;Parent=PF3D7_1372100.1</td>
    </tr>
    <tr>
      <th>34366</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>polypeptide</td>
      <td>2837053</td>
      <td>2839058</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1372100.1:pep;Derives_from=PF3D7_1372...</td>
    </tr>
    <tr>
      <th>34367</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2837313</td>
      <td>2839058</td>
      <td>.</td>
      <td>+</td>
      <td>0</td>
      <td>ID=PF3D7_1372100.1:exon:2;Parent=PF3D7_1372100.1</td>
    </tr>
    <tr>
      <th>34368</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>three_prime_UTR</td>
      <td>2839059</td>
      <td>2839580</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1372100.1:3UTR;Parent=PF3D7_1372100.1</td>
    </tr>
    <tr>
      <th>34369</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>three_prime_UTR</td>
      <td>2840236</td>
      <td>2840726</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_1372200.1:3UTR;Parent=PF3D7_1372200.1</td>
    </tr>
    <tr>
      <th>34370</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>gene</td>
      <td>2840236</td>
      <td>2842840</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_1372200;Name=HRPIII;previous_systemat...</td>
    </tr>
    <tr>
      <th>34371</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>mRNA</td>
      <td>2840236</td>
      <td>2842840</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_1372200.1;comment=sub-telomeric gene ...</td>
    </tr>
    <tr>
      <th>34372</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2840727</td>
      <td>2841485</td>
      <td>.</td>
      <td>-</td>
      <td>0</td>
      <td>ID=PF3D7_1372200.1:exon:1;Parent=PF3D7_1372200.1</td>
    </tr>
    <tr>
      <th>34373</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>polypeptide</td>
      <td>2840727</td>
      <td>2841703</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_1372200.1:pep;Derives_from=PF3D7_1372...</td>
    </tr>
    <tr>
      <th>34374</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>CDS</td>
      <td>2841635</td>
      <td>2841703</td>
      <td>.</td>
      <td>-</td>
      <td>0</td>
      <td>ID=PF3D7_1372200.1:exon:2;Parent=PF3D7_1372200.1</td>
    </tr>
    <tr>
      <th>34375</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>five_prime_UTR</td>
      <td>2841704</td>
      <td>2842840</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ID=PF3D7_1372200.1:5UTR;Parent=PF3D7_1372200.1</td>
    </tr>
    <tr>
      <th>34376</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>five_prime_UTR</td>
      <td>2843157</td>
      <td>2845766</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1372300.1:5UTR;Parent=PF3D7_1372300.1</td>
    </tr>
    <tr>
      <th>34377</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>gene</td>
      <td>2843157</td>
      <td>2847557</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1372300</td>
    </tr>
    <tr>
      <th>34378</th>
      <td>Pf3D7_13_v3</td>
      <td>chado</td>
      <td>mRNA</td>
      <td>2843157</td>
      <td>2847557</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ID=PF3D7_1372300.1;Parent=PF3D7_1372300;Dbxref...</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-bc742f21-ce0e-46ad-8893-1d2903892399')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-bc742f21-ce0e-46ad-8893-1d2903892399 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-bc742f21-ce0e-46ad-8893-1d2903892399');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-04ab90de-c277-4038-9eb8-9adaac62b279">
  <button class="colab-df-quickchart" onclick="quickchart('df-04ab90de-c277-4038-9eb8-9adaac62b279')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-04ab90de-c277-4038-9eb8-9adaac62b279 button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
</div>
<div class="section" id="create-figure">
<h2>Create figure<a class="headerlink" href="#create-figure" title="Permalink to this headline">¶</a></h2>
<p>This intricate figure serves to map deletion breakpoints in <em>hrp2</em> and <em>hrp3</em> across various countries. The x-axis of the figure displays the deletion breakpoints, while the y-axis shows the countries along with the number of breakpoints they exhibit.</p>
<p>The figure comprises two sections: <em>hrp2</em> (positioned at the top, subplots 1-5) and <em>hrp3</em> (located at the bottom, subplots 6-11), consisting of a total of 11 subplots.</p>
<p>We start by defining distinct colour codes for each genomic region that we will annotate.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a dictionary with distinct colour codes</span>
<span class="n">figure_colours</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">()</span>
<span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_8_13&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;#377eb8&#39;</span>
<span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_11&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;#984ea3&#39;</span>
<span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_5&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;#ff7f00&#39;</span>
<span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;similar_sequence&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;#ffff33&#39;</span>
<span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;hrp_genes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;#e41a1c&#39;</span>
<span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;rrna_genes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;#4daf4a&#39;</span>
<span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;other_genes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;black&#39;</span>
<span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;pseudogenes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;grey&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Full figure</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">11</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="mi">9</span><span class="p">),</span> <span class="n">gridspec_kw</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;height_ratios&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">12</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">32</span><span class="p">,</span> <span class="mi">18</span><span class="p">,</span> <span class="mi">12</span><span class="p">,</span> <span class="mi">12</span><span class="p">],</span> <span class="s1">&#39;hspace&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">})</span>

<span class="c1">### HRP2</span>

<span class="c1"># Set the minimum and maximum positions for hrp2</span>
<span class="n">min_pos</span> <span class="o">=</span> <span class="mi">1371500</span>
<span class="n">max_pos</span> <span class="o">=</span> <span class="mi">1375500</span>

<span class="c1">## Subplot 1: Title and Chromosome Region</span>

<span class="c1"># Set the title</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Pf3D7_08_v3&#39;</span><span class="p">)</span>
<span class="c1"># Configure the x-axis properties</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
<span class="c1"># Configure the y-axis properties</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="c1"># Set the x-axis limits to match the genomic region of &#39;Pf3D7_08_v3&#39;</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_08_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_08_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">]))</span>
<span class="c1"># Define the background colour of the subplot</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_facecolor</span><span class="p">(</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_8_13&#39;</span><span class="p">])</span>
<span class="c1"># Add a vertical span between &#39;min_pos&#39; and &#39;max_pos&#39; with a white background</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">axvspan</span><span class="p">(</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>

<span class="c1">## Subplot 2: Chromosome Boundaries</span>

<span class="c1"># Configure x and y axis ticks</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="c1"># Set the x-axis limits to match the genomic region of &#39;Pf3D7_08_v3&#39;</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_08_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_08_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">]))</span>
<span class="c1"># Define the y-axis limits</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="c1"># Plot horizontal lines marking the start and end positions of &#39;Pf3D7_08_v3&#39;</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_08_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">]),</span> <span class="n">min_pos</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_08_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">]),</span> <span class="n">max_pos</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
<span class="c1"># Hide the left and right spines for a cleaner appearance</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;left&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># Subplot 3: Deletion Breakpoints by Country</span>

<span class="c1"># Create lists for y-axis labels and positions</span>
<span class="n">ylabels</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">yposes</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># Loop through the sorted df_hrp2</span>
<span class="k">for</span> <span class="n">ypos</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df_hrp2</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="n">yposes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ypos</span><span class="p">)</span>
    <span class="c1"># Plot horizontal lines connecting min_pos to each hrp2 deletion breakpoint</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]),</span> <span class="p">(</span><span class="n">ypos</span><span class="p">,</span> <span class="n">ypos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_8_13&#39;</span><span class="p">])</span>
    <span class="c1"># Plot circular markers for each breakpoint position</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">20</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="mi">101</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="mi">40</span><span class="o">*</span><span class="mi">2</span><span class="p">),</span> <span class="p">[</span><span class="n">ypos</span><span class="p">]</span><span class="o">*</span><span class="mi">3</span><span class="p">,</span> <span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_8_13&#39;</span><span class="p">])</span>
    <span class="c1"># Add country names to the y-labels list</span>
    <span class="n">ylabels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Countries&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="c1"># Set y-axis ticks and labels to correspond to country names</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">yposes</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">ylabels</span><span class="p">)</span>
<span class="c1"># Set x-axis limits to cover the specified region between min_pos and max_pos</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">)</span>
<span class="c1"># Adjust y-axis limits to include space for data visualization</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">yposes</span><span class="p">)</span> <span class="o">-</span> <span class="mf">0.5</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">yposes</span><span class="p">)</span> <span class="o">+</span> <span class="mf">0.5</span><span class="p">)</span>
<span class="c1"># Hide x-axis ticks for a cleaner appearance</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>

<span class="c1">## Subplot 4: Genes on the x-axis</span>

<span class="c1"># Set the x-axis limits to cover the specified genomic region between min_pos and max_pos</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">)</span>
<span class="c1"># Set the y-axis limits to display a region between 0 and 1</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="c1"># Iterate through df_gff to process different gene annotations</span>
<span class="k">for</span> <span class="n">ix</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df_gff</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">(</span><span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Pf3D7_08_v3&#39;</span><span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span><span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_pos</span><span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span><span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_pos</span><span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span><span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;CDS&#39;</span><span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="c1"># Determine the color based on the genomic position</span>
    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">1373212</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">1376988</span><span class="p">:</span>
        <span class="n">color</span> <span class="o">=</span> <span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;hrp_genes&#39;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">color</span> <span class="o">=</span> <span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;other_genes&#39;</span><span class="p">]</span>
    <span class="c1"># Plot CDS features with specified line properties and color</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]),</span> <span class="p">(</span><span class="mf">0.75</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">)</span>

<span class="c1"># Iterate through polypeptide</span>
<span class="k">for</span> <span class="n">ix</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df_gff</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">(</span><span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Pf3D7_08_v3&#39;</span><span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span><span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_pos</span><span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span><span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_pos</span><span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span><span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;polypeptide&#39;</span><span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">1373212</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">1376988</span><span class="p">:</span>
        <span class="n">color</span> <span class="o">=</span> <span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;hrp_genes&#39;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">color</span> <span class="o">=</span> <span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;other_genes&#39;</span><span class="p">]</span>
    <span class="c1"># Plot polypeptide features with different line properties and color</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]),</span> <span class="p">(</span><span class="mf">0.75</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">)</span>

<span class="c1"># Iterate through pseudogene</span>
<span class="k">for</span> <span class="n">ix</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df_gff</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">(</span><span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Pf3D7_08_v3&#39;</span><span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span><span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_pos</span><span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span><span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_pos</span><span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span><span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;pseudogene&#39;</span><span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="c1"># Plot pseudogene features with specified line properties and color</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]),</span> <span class="p">(</span><span class="mf">0.75</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;pseudogenes&#39;</span><span class="p">])</span>

<span class="c1"># Add text labels for specific genomic regions</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="mi">1375299</span> <span class="o">+</span> <span class="mi">1374236</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="s1">&#39;$</span><span class="si">{hrp2}</span><span class="s1">$&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="mi">1371847</span> <span class="o">+</span> <span class="mi">1372720</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="s1">&#39;PF3D7_0831750 (pseudogene)&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>

<span class="c1"># Set y-axis ticks, labels, and x-axis ticks to maintain an organized representation</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Genes&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>

<span class="c1">## Subplot 5: Ticks for Genomic Coordinates</span>
<span class="c1"># Set the x-axis limits to cover the specified genomic region between &#39;min_pos&#39; and &#39;max_pos&#39;</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">)</span>
<span class="c1"># Hide the left and right spines to create a cleaner appearance</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;left&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="c1"># Configure the x-axis ticks at specific positions</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([</span><span class="mi">1372000</span><span class="p">,</span> <span class="mi">1373000</span><span class="p">,</span> <span class="mi">1374000</span><span class="p">,</span> <span class="mi">1375000</span><span class="p">])</span>
<span class="c1"># Label the x-axis ticks with corresponding values</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">([</span><span class="s2">&quot;1.372 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;1.373 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;1.374 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;1.375 Mbp&quot;</span><span class="p">])</span>
<span class="c1"># Remove y-axis ticks to maintain a clean look</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="c1"># Position x-axis ticks at the top</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">tick_top</span><span class="p">()</span>
<span class="c1"># Adjust the direction of x-axis ticks and add padding</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">direction</span><span class="o">=</span><span class="s2">&quot;in&quot;</span><span class="p">,</span> <span class="n">pad</span><span class="o">=-</span><span class="mi">20</span><span class="p">)</span>

<span class="c1">### HRP3</span>

<span class="c1"># Set the minimum and maximum positions for hrp3</span>
<span class="n">min_pos</span> <span class="o">=</span> <span class="mi">2795000</span>
<span class="n">max_pos</span> <span class="o">=</span> <span class="mi">2845000</span>

<span class="c1">## Subplot 6: Title and Chromosome Region</span>

<span class="c1"># Set the title</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Pf3D7_13_v3&#39;</span><span class="p">)</span>
<span class="c1"># Hide the x-axis ticks and labels</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
<span class="c1"># Hide the y-axis ticks</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="c1"># Set the x-axis limits to cover the specified genomic region between the start and end positions of chromosome</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_13_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_13_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">]))</span>
<span class="c1"># Customize the background color of the subplot</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">set_facecolor</span><span class="p">(</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_8_13&#39;</span><span class="p">])</span>
<span class="c1"># Create a white span on the plot to mark the genomic region of interest</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">axvspan</span><span class="p">(</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>

<span class="c1">## Subplot 7: Chromosome Boundaries</span>

<span class="c1"># Hide the x-axis ticks and labels</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="c1"># Set the x-axis limits to cover the specified genomic region between the start and end positions of &#39;Pf3D7_13_v3&#39;</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_13_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_13_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">]))</span>
<span class="c1"># Set the y-axis limits to cover a range from 0 to 1</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="c1"># Create lines to mark the start and end positions of the chromosome region with black color</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_13_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">]),</span> <span class="n">min_pos</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_13_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">]),</span> <span class="n">max_pos</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
<span class="c1"># Hide the left and right spines to create a cleaner appearance</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;left&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

<span class="c1">## Subplot 8: Deletion Breakpoints by Country</span>

<span class="c1"># Initialize lists for labels and positions</span>
<span class="n">ylabels</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">yposes</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># Iterate through df_hrp3 sorted by breakpoints</span>
<span class="k">for</span> <span class="n">ypos</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df_hrp3</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="n">yposes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ypos</span><span class="p">)</span>

    <span class="c1"># Check the deletion type and apply different plotting styles accordingly</span>
    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Deletion type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Telomere healing&#39;</span><span class="p">:</span>
        <span class="nb">breakpoint</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">])</span>

        <span class="c1"># Plot the deletion breakpoint and add circular markers</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">min_pos</span><span class="p">,</span> <span class="nb">breakpoint</span><span class="p">),</span> <span class="p">(</span><span class="n">ypos</span><span class="p">,</span> <span class="n">ypos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_8_13&#39;</span><span class="p">])</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">])</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">20</span><span class="o">*</span><span class="mi">25</span><span class="p">,</span> <span class="mi">101</span><span class="o">*</span><span class="mi">25</span><span class="p">,</span> <span class="mi">40</span><span class="o">*</span><span class="mi">25</span><span class="p">),</span> <span class="p">[</span><span class="n">ypos</span><span class="p">]</span><span class="o">*</span><span class="mi">3</span><span class="p">,</span> <span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_8_13&#39;</span><span class="p">])</span>
        <span class="n">ylabels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Countries&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Deletion type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Chromosome 5 recombination&#39;</span><span class="p">:</span>
        <span class="n">breakpoint_start</span><span class="p">,</span> <span class="n">breakpoint_end</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)]</span>

        <span class="c1"># Plot the breakpoints for Chromosome 5 recombination</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">breakpoint_start</span><span class="p">),</span> <span class="p">(</span><span class="n">ypos</span><span class="p">,</span> <span class="n">ypos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_8_13&#39;</span><span class="p">])</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">breakpoint_end</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">),</span> <span class="p">(</span><span class="n">ypos</span><span class="p">,</span> <span class="n">ypos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_5&#39;</span><span class="p">])</span>
        <span class="n">ylabels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Countries&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Deletion type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Chromosome 11 recombination&#39;</span><span class="p">:</span>
        <span class="n">breakpoint_start</span><span class="p">,</span> <span class="n">breakpoint_end</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)]</span>

        <span class="c1"># Plot the breakpoints for Chromosome 11 recombination</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">breakpoint_start</span><span class="p">),</span> <span class="p">(</span><span class="n">ypos</span><span class="p">,</span> <span class="n">ypos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_8_13&#39;</span><span class="p">])</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">breakpoint_end</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">),</span> <span class="p">(</span><span class="n">ypos</span><span class="p">,</span> <span class="n">ypos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_11&#39;</span><span class="p">])</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">breakpoint_start</span><span class="p">,</span> <span class="n">breakpoint_end</span><span class="p">),</span> <span class="p">(</span><span class="n">ypos</span><span class="p">,</span> <span class="n">ypos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;similar_sequence&#39;</span><span class="p">])</span>
        <span class="n">ylabels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;14 countries (151)&quot;</span><span class="p">)</span>

<span class="c1"># Set the y-axis ticks and labels</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">yposes</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">ylabels</span><span class="p">)</span>
<span class="c1"># Set x-axis limits, y-axis limits, and remove x-axis ticks</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">yposes</span><span class="p">)</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">yposes</span><span class="p">)</span><span class="o">+</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>

<span class="c1">## Subplot 9: Genes on the x-axis</span>

<span class="n">bar_pos</span> <span class="o">=</span> <span class="mf">0.9</span>
<span class="n">text_pos</span> <span class="o">=</span> <span class="mf">0.8</span>

<span class="c1"># Set the x-axis and y-axis limits</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="c1"># Iterate through GFF data filtering for relevant features</span>
<span class="c1"># Plot CDS features in distinct colors, depending on their location</span>
<span class="k">for</span> <span class="n">ix</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df_gff</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Pf3D7_13_v3&#39;</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_pos</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_pos</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;CDS&#39;</span> <span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">2840727</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">2841703</span><span class="p">:</span>
        <span class="n">color</span> <span class="o">=</span> <span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;hrp_genes&#39;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">color</span> <span class="o">=</span> <span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;other_genes&#39;</span><span class="p">]</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]),</span> <span class="p">(</span><span class="n">bar_pos</span><span class="p">,</span> <span class="n">bar_pos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">)</span>

<span class="c1"># Plot polypeptide features with similar style, with specific colors</span>
<span class="k">for</span> <span class="n">ix</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df_gff</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Pf3D7_13_v3&#39;</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_pos</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_pos</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;polypeptide&#39;</span> <span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">2840727</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">2841703</span><span class="p">:</span>
        <span class="n">color</span> <span class="o">=</span> <span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;hrp_genes&#39;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">color</span> <span class="o">=</span> <span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;other_genes&#39;</span><span class="p">]</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]),</span> <span class="p">(</span><span class="n">bar_pos</span><span class="p">,</span> <span class="n">bar_pos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">)</span>

<span class="c1"># Plot rRNA and pseudogene features</span>
<span class="k">for</span> <span class="n">ix</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df_gff</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Pf3D7_13_v3&#39;</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_pos</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_pos</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;rRNA&#39;</span> <span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]),</span> <span class="p">(</span><span class="n">bar_pos</span><span class="p">,</span> <span class="n">bar_pos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;rrna_genes&#39;</span><span class="p">])</span>
<span class="c1"># Annotate specific gene positions with labels</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="mi">2796119</span> <span class="o">+</span> <span class="mi">2797144</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;PF3D7_1370800&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="mi">2797507</span> <span class="o">+</span> <span class="mi">2798103</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;PF3D7_1370900&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="mi">2800004</span> <span class="o">+</span> <span class="mi">2802154</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;18S rRNA&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="mi">2802527</span> <span class="o">+</span> <span class="mi">2802686</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;5.8S rRNA&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="mi">2802945</span> <span class="o">+</span> <span class="mi">2807159</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;28S rRNA&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="mi">2808200</span> <span class="o">+</span> <span class="mi">2810256</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;PF3D7_1371500&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="mi">2811706</span> <span class="o">+</span> <span class="mi">2820270</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;$</span><span class="si">{ebl1}</span><span class="s1">$</span><span class="se">\n</span><span class="s1">(pseudogene)&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="mi">2821078</span> <span class="o">+</span> <span class="mi">2823292</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;PF3D7_1371700&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="mi">2824302</span> <span class="o">+</span> <span class="mi">2825852</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;PF3D7_1371800&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="mi">2829856</span> <span class="o">+</span> <span class="mi">2830669</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;PF3D7_1371900&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="mi">2832952</span> <span class="o">+</span> <span class="mi">2834322</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;PF3D7_1372000&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="mi">2837053</span> <span class="o">+</span> <span class="mi">2839058</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;$</span><span class="si">{gexp04}</span><span class="s1">$&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="mi">2840727</span> <span class="o">+</span> <span class="mi">2841703</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;$</span><span class="si">{hrp3}</span><span class="s1">$&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="c1"># Set y-axis ticks, customize x-axis labels, and add x-axis label and y-axis label</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">([</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="s2">&quot;2.80 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;2.81 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;2.82 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;2.83 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;2.84 Mbp&quot;</span><span class="p">])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Genomic coordinates&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Genes&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>

<span class="c1">## Subplot 10: Ticks for Genomic Coordinates</span>

<span class="c1"># Set the x-axis limits</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">)</span>
<span class="c1"># Hide left and right spines to make the plot clean</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;left&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="c1"># Define custom tick positions on the x-axis</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([</span><span class="mi">2800000</span><span class="p">,</span> <span class="mi">2810000</span><span class="p">,</span> <span class="mi">2820000</span><span class="p">,</span> <span class="mi">2830000</span><span class="p">,</span> <span class="mi">2840000</span><span class="p">])</span>
<span class="c1"># Set custom tick labels corresponding to genomic positions</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">([</span><span class="s2">&quot;2.80 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;2.81 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;2.82 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;2.83 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;2.84 Mbp&quot;</span><span class="p">])</span>
<span class="c1"># Hide the y-axis ticks</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="c1"># Move the x-axis ticks to the top</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">tick_top</span><span class="p">()</span>
<span class="c1"># Adjust the direction and padding of x-axis ticks</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">direction</span><span class="o">=</span><span class="s2">&quot;in&quot;</span><span class="p">,</span> <span class="n">pad</span><span class="o">=-</span><span class="mi">20</span><span class="p">)</span>
<span class="c1"># Add a label to the x-axis indicating &#39;Genomic coordinates&#39; at the center of the x-axis</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="n">min_pos</span> <span class="o">+</span> <span class="n">max_pos</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s1">&#39;Genomic coordinates&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>

<span class="c1">## Subplot 11: Legend</span>

<span class="c1"># Hide left, right, and bottom spines</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;left&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;bottom&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="c1"># Hide both x and y-axis ticks</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="c1"># Create legend elements with color bars and labels</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">),</span> <span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_5&#39;</span><span class="p">])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.12</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s1">&#39;Chr 5 sequence&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">),</span> <span class="p">(</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_11&#39;</span><span class="p">])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.12</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="s1">&#39;Chr 11 sequence&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="mf">0.38</span><span class="p">,</span> <span class="mf">0.48</span><span class="p">),</span> <span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;similar_sequence&#39;</span><span class="p">])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.50</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s1">&#39;Highly similar sequence on chr 13 and chr 11&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mf">0.405</span><span class="p">,</span> <span class="mf">0.456</span><span class="p">,</span> <span class="mf">0.025</span><span class="p">),</span> <span class="p">[</span><span class="mf">0.25</span><span class="p">]</span><span class="o">*</span><span class="mi">3</span><span class="p">,</span> <span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_8_13&#39;</span><span class="p">])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.50</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="s1">&#39;Telomeric repeat sequence&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>
<span class="c1"># Set the limits for this subplot</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

<span class="c1"># Ensure the figure layout is tidy</span>
<span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;ipython-input-16-28c0075a88af&gt;:278: UserWarning: FixedFormatter should only be used together with FixedLocator
  axs[8].set_xticklabels([&quot;&quot;, &quot;2.80 Mbp&quot;, &quot;2.81 Mbp&quot;, &quot;2.82 Mbp&quot;, &quot;2.83 Mbp&quot;, &quot;2.84 Mbp&quot;])
</pre></div>
</div>
<img alt="../../_images/visualise_hrp2_hrp3_deletions_39_1.png" src="../../_images/visualise_hrp2_hrp3_deletions_39_1.png" />
</div>
</div>
<p><strong>Figure Legend: HRP deletion breakpoints.</strong> We see five different breakpoints resulting in the deletion of hrp2. Four of these are within exon 2 of the gene whereas the fifth is found between hrp2 and the pseudogene PF3D7_0831750. For all five events we see evidence of telomeric healing from reads that contain part Pf3D7_08_v3 sequence and part telomeric repeat sequence (GGGTTCA/GGGTTTA). We see 16 different breakpoints resulting in the deletion of hrp3. For fourteen of these we see evidence of telomeric healing. Note that many of these events result in the deletion of other genes in addition to hrp3. For twenty samples from Cambodia and a single sample from Vietnam we see evidence of a recombination with chromosome 5 which results in a hybrid chromosome comprising mostly chromosome 13 sequence but a small inverted section of an internal portion of chromosome 5 containing the gene mdr1. We also see evidence of a recombination with chromosome 11 which results in a hybrid chromosome comprising mostly chromosome 13 sequence but also a section of the 3’ end of chromosome 11. This is the most common deletion type, being seen in 151 samples from 14 different countries. Because the recombination occurs between highly similar sequences of a set of three orthologous ribosomal RNA genes found on both chromosomes, it is not possible to identify the exact breakpoints.</p>
<div class="section" id="save-figure">
<h3>Save Figure<a class="headerlink" href="#save-figure" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># You will need to authorise Google Colab access to Google Drive</span>
<span class="n">drive</span><span class="o">.</span><span class="n">mount</span><span class="p">(</span><span class="s1">&#39;/content/drive&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># This will send the file to your Google Drive, where you can download it from if needed</span>
<span class="c1"># Change the file path if you wish to send the file to a specific location</span>
<span class="c1"># Change the file name if you wish to call it something else</span>

<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;/content/drive/My Drive/HRP_Deletions_Figure.pdf&#39;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;/content/drive/My Drive/HRP_Deletions_Figure.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">480</span><span class="p">)</span> <span class="c1"># increase the dpi for higher resolution</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./pf7/analysis_examples"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="summarise_hrp2_hrp3_deletions.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title"><em>hrp2</em> and <em>hrp3</em> deletions: summary of deletion breakpoints</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="csp-cterm-analysis.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Analyse csp C-Terminal Variation</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By MalariaGEN<br/>
    
        &copy; Copyright 2023.<br/>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>