{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "J5DvgSColT9Y"
      },
      "source": [
        "# Drug Resistance Frequencies Across Pf7"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "s6E0OCU6laVT"
      },
      "source": [
        "## Introduction"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "QddUVloxB97i"
      },
      "source": [
        "This notebook recreates main Table 2, Supplementary Table 4 from the [Pf7 paper](https://wellcomeopenresearch.org/articles/8-22/v1).\n",
        "\n",
        "These tables summarise the number of samples associated with malarial drug resistance (DR) in different populations.\n",
        "\n",
        "**This notebook takes 1 minute to run.**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "8aHguQMOliOG"
      },
      "source": [
        "## Setup"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "fnmE5kdRlrf_"
      },
      "source": [
        "Install and import the malariagen Python package:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "k56UYmO2lkig",
        "outputId": "d6fb1df0-6bff-40c6-8b55-65576687433a"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m148.5/148.5 kB\u001b[0m \u001b[31m3.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m11.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m52.8/52.8 MB\u001b[0m \u001b[31m12.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m10.2/10.2 MB\u001b[0m \u001b[31m52.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.0/4.0 MB\u001b[0m \u001b[31m45.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m302.5/302.5 kB\u001b[0m \u001b[31m21.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m138.7/138.7 kB\u001b[0m \u001b[31m13.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m20.9/20.9 MB\u001b[0m \u001b[31m14.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m8.1/8.1 MB\u001b[0m \u001b[31m24.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m206.9/206.9 kB\u001b[0m \u001b[31m12.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m7.7/7.7 MB\u001b[0m \u001b[31m86.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m68.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Building wheel for asciitree (setup.py) ... \u001b[?25l\u001b[?25hdone\n"
          ]
        }
      ],
      "source": [
        "!pip install -q --no-warn-conflicts malariagen_data\n",
        "import malariagen_data"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "1JGWthtIlop-"
      },
      "source": [
        "Import required python libraries that are installed at colab by default."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "id": "laxWs51glmlv"
      },
      "outputs": [],
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import collections\n",
        "import re\n",
        "import scipy"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "oL05c-_klv6I"
      },
      "source": [
        "## Access Pf7 Data"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "YlPcoczFl16k"
      },
      "source": [
        "We use the malariagen data package to load the release data.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 213
        },
        "id": "JBrrYb6glym5",
        "outputId": "8798b909-dd9d-4556-a4a8-3ac7caa73600"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-6ad0ff0e-d34c-439a-aef3-e034675ada45\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Sample</th>\n",
              "      <th>Study</th>\n",
              "      <th>Country</th>\n",
              "      <th>Admin level 1</th>\n",
              "      <th>Country latitude</th>\n",
              "      <th>Country longitude</th>\n",
              "      <th>Admin level 1 latitude</th>\n",
              "      <th>Admin level 1 longitude</th>\n",
              "      <th>Year</th>\n",
              "      <th>ENA</th>\n",
              "      <th>All samples same case</th>\n",
              "      <th>Population</th>\n",
              "      <th>% callable</th>\n",
              "      <th>QC pass</th>\n",
              "      <th>Exclusion reason</th>\n",
              "      <th>Sample type</th>\n",
              "      <th>Sample was in Pf6</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FP0008-C</td>\n",
              "      <td>1147-PF-MR-CONWAY</td>\n",
              "      <td>Mauritania</td>\n",
              "      <td>Hodh el Gharbi</td>\n",
              "      <td>20.265149</td>\n",
              "      <td>-10.337093</td>\n",
              "      <td>16.565426</td>\n",
              "      <td>-9.832345</td>\n",
              "      <td>2014.0</td>\n",
              "      <td>ERR1081237</td>\n",
              "      <td>FP0008-C</td>\n",
              "      <td>AF-W</td>\n",
              "      <td>82.16</td>\n",
              "      <td>True</td>\n",
              "      <td>Analysis_set</td>\n",
              "      <td>gDNA</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>FP0009-C</td>\n",
              "      <td>1147-PF-MR-CONWAY</td>\n",
              "      <td>Mauritania</td>\n",
              "      <td>Hodh el Gharbi</td>\n",
              "      <td>20.265149</td>\n",
              "      <td>-10.337093</td>\n",
              "      <td>16.565426</td>\n",
              "      <td>-9.832345</td>\n",
              "      <td>2014.0</td>\n",
              "      <td>ERR1081238</td>\n",
              "      <td>FP0009-C</td>\n",
              "      <td>AF-W</td>\n",
              "      <td>88.85</td>\n",
              "      <td>True</td>\n",
              "      <td>Analysis_set</td>\n",
              "      <td>gDNA</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>FP0010-CW</td>\n",
              "      <td>1147-PF-MR-CONWAY</td>\n",
              "      <td>Mauritania</td>\n",
              "      <td>Hodh el Gharbi</td>\n",
              "      <td>20.265149</td>\n",
              "      <td>-10.337093</td>\n",
              "      <td>16.565426</td>\n",
              "      <td>-9.832345</td>\n",
              "      <td>2014.0</td>\n",
              "      <td>ERR2889621</td>\n",
              "      <td>FP0010-CW</td>\n",
              "      <td>AF-W</td>\n",
              "      <td>86.46</td>\n",
              "      <td>True</td>\n",
              "      <td>Analysis_set</td>\n",
              "      <td>sWGA</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6ad0ff0e-d34c-439a-aef3-e034675ada45')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-6ad0ff0e-d34c-439a-aef3-e034675ada45 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-6ad0ff0e-d34c-439a-aef3-e034675ada45');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-5600c7e7-f758-44f9-9e43-52967d609a5e\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-5600c7e7-f758-44f9-9e43-52967d609a5e')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-5600c7e7-f758-44f9-9e43-52967d609a5e button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "      Sample              Study     Country   Admin level 1  Country latitude  \\\n",
              "0   FP0008-C  1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
              "1   FP0009-C  1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
              "2  FP0010-CW  1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
              "\n",
              "   Country longitude  Admin level 1 latitude  Admin level 1 longitude    Year  \\\n",
              "0         -10.337093               16.565426                -9.832345  2014.0   \n",
              "1         -10.337093               16.565426                -9.832345  2014.0   \n",
              "2         -10.337093               16.565426                -9.832345  2014.0   \n",
              "\n",
              "          ENA All samples same case Population  % callable  QC pass  \\\n",
              "0  ERR1081237              FP0008-C       AF-W       82.16     True   \n",
              "1  ERR1081238              FP0009-C       AF-W       88.85     True   \n",
              "2  ERR2889621             FP0010-CW       AF-W       86.46     True   \n",
              "\n",
              "  Exclusion reason Sample type  Sample was in Pf6  \n",
              "0     Analysis_set        gDNA               True  \n",
              "1     Analysis_set        gDNA               True  \n",
              "2     Analysis_set        sWGA              False  "
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "release_data = malariagen_data.Pf7()\n",
        "sample_metadata = release_data.sample_metadata()\n",
        "# take a glance at the metadata dataframe\n",
        "sample_metadata.head(3)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "1fXWbVbZH13A"
      },
      "source": [
        "We will only examine the QC pass samples in the analysis of these notebooks."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "id": "V_aeZe1hHqMD"
      },
      "outputs": [],
      "source": [
        "# Retain QC pass samples\n",
        "qc_sample_metadata = sample_metadata.loc[sample_metadata['QC pass']]"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-Otz5oyEl8UG"
      },
      "source": [
        "## Access DR Classification and Genotype Data"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4XwL9dSOBrfx"
      },
      "source": [
        "We can access inferred resistance status classifications of QC-pass Pf7 samples from the [MalariaGEN web resources](https://www.malariagen.net/resource/34).\n",
        "\n",
        "This dataset includes the samples that are resistant to 10 drugs or combinations of drugs and to rapid diagnostic tests (RDT) detection: chloroquine, pyrimethamine, sulfadoxine, mefloquine, artemisinin, piperaquine, sulfadoxine- pyrimethamine for treatment of uncomplicated malaria, sulfadoxine- pyrimethamine for intermittent preventive treatment in pregnancy, artesunate-mefloquine, dihydroartemisinin-piperaquine, hrp2 and hrp3 gene deletions."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "MN_SE_R7l2Re",
        "outputId": "7ff5ba10-c26d-4104-fa6b-0856f02eb5a1"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-38870aa6-62a0-4d8c-8a42-71ef8a11a782\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Sample</th>\n",
              "      <th>Chloroquine</th>\n",
              "      <th>Pyrimethamine</th>\n",
              "      <th>Sulfadoxine</th>\n",
              "      <th>Mefloquine</th>\n",
              "      <th>Artemisinin</th>\n",
              "      <th>Piperaquine</th>\n",
              "      <th>SP (uncomplicated)</th>\n",
              "      <th>SP (IPTp)</th>\n",
              "      <th>AS-MQ</th>\n",
              "      <th>DHA-PPQ</th>\n",
              "      <th>HRP2</th>\n",
              "      <th>HRP3</th>\n",
              "      <th>HRP2 and HRP3</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FP0008-C</td>\n",
              "      <td>Undetermined</td>\n",
              "      <td>Undetermined</td>\n",
              "      <td>Undetermined</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>nodel</td>\n",
              "      <td>nodel</td>\n",
              "      <td>nodel</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>FP0009-C</td>\n",
              "      <td>Resistant</td>\n",
              "      <td>Resistant</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Resistant</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>nodel</td>\n",
              "      <td>nodel</td>\n",
              "      <td>nodel</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>FP0010-CW</td>\n",
              "      <td>Undetermined</td>\n",
              "      <td>Resistant</td>\n",
              "      <td>Resistant</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Resistant</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>uncallable</td>\n",
              "      <td>uncallable</td>\n",
              "      <td>uncallable</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>FP0011-CW</td>\n",
              "      <td>Undetermined</td>\n",
              "      <td>Resistant</td>\n",
              "      <td>Undetermined</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Undetermined</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Resistant</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>uncallable</td>\n",
              "      <td>uncallable</td>\n",
              "      <td>uncallable</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>FP0012-CW</td>\n",
              "      <td>Resistant</td>\n",
              "      <td>Resistant</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Resistant</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>nodel</td>\n",
              "      <td>nodel</td>\n",
              "      <td>nodel</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-38870aa6-62a0-4d8c-8a42-71ef8a11a782')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-38870aa6-62a0-4d8c-8a42-71ef8a11a782 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-38870aa6-62a0-4d8c-8a42-71ef8a11a782');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-49b17cf3-4fa2-446a-9880-09b2416a2fe0\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-49b17cf3-4fa2-446a-9880-09b2416a2fe0')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-49b17cf3-4fa2-446a-9880-09b2416a2fe0 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "      Sample   Chloroquine Pyrimethamine   Sulfadoxine Mefloquine  \\\n",
              "0   FP0008-C  Undetermined  Undetermined  Undetermined  Sensitive   \n",
              "1   FP0009-C     Resistant     Resistant     Sensitive  Sensitive   \n",
              "2  FP0010-CW  Undetermined     Resistant     Resistant  Sensitive   \n",
              "3  FP0011-CW  Undetermined     Resistant  Undetermined  Sensitive   \n",
              "4  FP0012-CW     Resistant     Resistant     Sensitive  Sensitive   \n",
              "\n",
              "    Artemisinin Piperaquine SP (uncomplicated)  SP (IPTp)      AS-MQ  \\\n",
              "0     Sensitive   Sensitive          Sensitive  Sensitive  Sensitive   \n",
              "1     Sensitive   Sensitive          Resistant  Sensitive  Sensitive   \n",
              "2     Sensitive   Sensitive          Resistant  Sensitive  Sensitive   \n",
              "3  Undetermined   Sensitive          Resistant  Sensitive  Sensitive   \n",
              "4     Sensitive   Sensitive          Resistant  Sensitive  Sensitive   \n",
              "\n",
              "     DHA-PPQ        HRP2        HRP3 HRP2 and HRP3  \n",
              "0  Sensitive       nodel       nodel         nodel  \n",
              "1  Sensitive       nodel       nodel         nodel  \n",
              "2  Sensitive  uncallable  uncallable    uncallable  \n",
              "3  Sensitive  uncallable  uncallable    uncallable  \n",
              "4  Sensitive       nodel       nodel         nodel  "
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Read the data from the MalariaGEN.net resources website\n",
        "resistance_classification_fn = pd.read_csv('https://www.malariagen.net/wp-content/uploads/2023/11/Pf7_inferred_resistance_status_classification.txt', sep='\\t')\n",
        "\n",
        "# Print the first rows\n",
        "resistance_classification_fn.head()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TrwayUMQAD3G"
      },
      "source": [
        "We will also use the genotypes utilised in drug resistance status classification containing amino acid and copy number genotypes at six loci: crt, dhfr, dhps, mdr1, kelch13, plasmepsin 2-3. This dataset is also available witin the MalariaGEN web resources."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "tmBxiVvE1cra",
        "outputId": "dc486cea-3142-44ae-c1a8-b6a342038e75"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-4ffd28a9-adf1-4506-8f03-539e702e7e90\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Sample</th>\n",
              "      <th>crt_76[K]</th>\n",
              "      <th>crt_72-76[CVMNK]</th>\n",
              "      <th>dhfr_51[N]</th>\n",
              "      <th>dhfr_59[C]</th>\n",
              "      <th>dhfr_108[S]</th>\n",
              "      <th>dhfr_164[I]</th>\n",
              "      <th>dhps_437[G]</th>\n",
              "      <th>dhps_540[K]</th>\n",
              "      <th>dhps_581[A]</th>\n",
              "      <th>dhps_613[A]</th>\n",
              "      <th>kelch13_349-726_ns_changes</th>\n",
              "      <th>mdr1_dup_call</th>\n",
              "      <th>mdr1_breakpoint</th>\n",
              "      <th>pm2_dup_call</th>\n",
              "      <th>pm2_breakpoint</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FP0008-C</td>\n",
              "      <td>T,K</td>\n",
              "      <td>CVIET,CVMNK</td>\n",
              "      <td>N</td>\n",
              "      <td>C</td>\n",
              "      <td>S,N</td>\n",
              "      <td>I</td>\n",
              "      <td>A,G</td>\n",
              "      <td>K</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>FP0009-C</td>\n",
              "      <td>T</td>\n",
              "      <td>CVIET</td>\n",
              "      <td>I</td>\n",
              "      <td>R</td>\n",
              "      <td>N</td>\n",
              "      <td>I</td>\n",
              "      <td>A</td>\n",
              "      <td>K</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>FP0010-CW</td>\n",
              "      <td>T,K</td>\n",
              "      <td>CVIET,CVMNK</td>\n",
              "      <td>I</td>\n",
              "      <td>R</td>\n",
              "      <td>N</td>\n",
              "      <td>I</td>\n",
              "      <td>G</td>\n",
              "      <td>K</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>FP0011-CW</td>\n",
              "      <td>T,K</td>\n",
              "      <td>CVIET,CVMNK</td>\n",
              "      <td>I</td>\n",
              "      <td>R</td>\n",
              "      <td>N</td>\n",
              "      <td>I</td>\n",
              "      <td>G,A</td>\n",
              "      <td>K</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>-</td>\n",
              "      <td>0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>FP0012-CW</td>\n",
              "      <td>T</td>\n",
              "      <td>CVIET</td>\n",
              "      <td>I</td>\n",
              "      <td>R</td>\n",
              "      <td>N</td>\n",
              "      <td>I</td>\n",
              "      <td>A</td>\n",
              "      <td>K</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-4ffd28a9-adf1-4506-8f03-539e702e7e90')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-4ffd28a9-adf1-4506-8f03-539e702e7e90 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-4ffd28a9-adf1-4506-8f03-539e702e7e90');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-c394ca1c-c83d-4107-aad9-3b9ea8960127\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-c394ca1c-c83d-4107-aad9-3b9ea8960127')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-c394ca1c-c83d-4107-aad9-3b9ea8960127 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "      Sample crt_76[K] crt_72-76[CVMNK] dhfr_51[N] dhfr_59[C] dhfr_108[S]  \\\n",
              "0   FP0008-C       T,K      CVIET,CVMNK          N          C         S,N   \n",
              "1   FP0009-C         T            CVIET          I          R           N   \n",
              "2  FP0010-CW       T,K      CVIET,CVMNK          I          R           N   \n",
              "3  FP0011-CW       T,K      CVIET,CVMNK          I          R           N   \n",
              "4  FP0012-CW         T            CVIET          I          R           N   \n",
              "\n",
              "  dhfr_164[I] dhps_437[G] dhps_540[K] dhps_581[A] dhps_613[A]  \\\n",
              "0           I         A,G           K           A           A   \n",
              "1           I           A           K           A           A   \n",
              "2           I           G           K           A           A   \n",
              "3           I         G,A           K           A           A   \n",
              "4           I           A           K           A           A   \n",
              "\n",
              "  kelch13_349-726_ns_changes  mdr1_dup_call mdr1_breakpoint  pm2_dup_call  \\\n",
              "0                        NaN              0             NaN             0   \n",
              "1                        NaN              0             NaN             0   \n",
              "2                        NaN              0             NaN             0   \n",
              "3                          -              0             NaN             0   \n",
              "4                        NaN              0             NaN             0   \n",
              "\n",
              "  pm2_breakpoint  \n",
              "0            NaN  \n",
              "1            NaN  \n",
              "2            NaN  \n",
              "3            NaN  \n",
              "4            NaN  "
            ]
          },
          "execution_count": 8,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Read the data\n",
        "drm_calls_fn = pd.read_csv('https://www.malariagen.net/wp-content/uploads/2023/11/Pf7_drug_resistance_marker_genotypes.txt', sep='\\t', )\n",
        "\n",
        "# Rename the first column as 'Sample'\n",
        "drm_calls_fn = drm_calls_fn.rename(columns={drm_calls_fn.columns[0]: 'Sample'})\n",
        "\n",
        "# Print the first rows\n",
        "drm_calls_fn.head()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "LHpYlsEi0U3t"
      },
      "source": [
        "##  Load all data into single DataFrame"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bTi6Rups8DWk"
      },
      "source": [
        "We can merge these 3 datasets (metadata, drug resistance genotype and classification) to facilitate streamlined analysis."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 260
        },
        "id": "nDJUwazk0YdL",
        "outputId": "508a7723-fec8-4ee8-9c08-4bf59a938e1f"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-560c49ae-f7e6-4c0f-83db-15b558a102d7\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Sample</th>\n",
              "      <th>Study</th>\n",
              "      <th>Country</th>\n",
              "      <th>Admin level 1</th>\n",
              "      <th>Country latitude</th>\n",
              "      <th>Country longitude</th>\n",
              "      <th>Admin level 1 latitude</th>\n",
              "      <th>Admin level 1 longitude</th>\n",
              "      <th>Year</th>\n",
              "      <th>ENA</th>\n",
              "      <th>...</th>\n",
              "      <th>Mefloquine</th>\n",
              "      <th>Artemisinin</th>\n",
              "      <th>Piperaquine</th>\n",
              "      <th>SP (uncomplicated)</th>\n",
              "      <th>SP (IPTp)</th>\n",
              "      <th>AS-MQ</th>\n",
              "      <th>DHA-PPQ</th>\n",
              "      <th>HRP2</th>\n",
              "      <th>HRP3</th>\n",
              "      <th>HRP2 and HRP3</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FP0008-C</td>\n",
              "      <td>1147-PF-MR-CONWAY</td>\n",
              "      <td>Mauritania</td>\n",
              "      <td>Hodh el Gharbi</td>\n",
              "      <td>20.265149</td>\n",
              "      <td>-10.337093</td>\n",
              "      <td>16.565426</td>\n",
              "      <td>-9.832345</td>\n",
              "      <td>2014.0</td>\n",
              "      <td>ERR1081237</td>\n",
              "      <td>...</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>nodel</td>\n",
              "      <td>nodel</td>\n",
              "      <td>nodel</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>FP0009-C</td>\n",
              "      <td>1147-PF-MR-CONWAY</td>\n",
              "      <td>Mauritania</td>\n",
              "      <td>Hodh el Gharbi</td>\n",
              "      <td>20.265149</td>\n",
              "      <td>-10.337093</td>\n",
              "      <td>16.565426</td>\n",
              "      <td>-9.832345</td>\n",
              "      <td>2014.0</td>\n",
              "      <td>ERR1081238</td>\n",
              "      <td>...</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Resistant</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>nodel</td>\n",
              "      <td>nodel</td>\n",
              "      <td>nodel</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>FP0010-CW</td>\n",
              "      <td>1147-PF-MR-CONWAY</td>\n",
              "      <td>Mauritania</td>\n",
              "      <td>Hodh el Gharbi</td>\n",
              "      <td>20.265149</td>\n",
              "      <td>-10.337093</td>\n",
              "      <td>16.565426</td>\n",
              "      <td>-9.832345</td>\n",
              "      <td>2014.0</td>\n",
              "      <td>ERR2889621</td>\n",
              "      <td>...</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Resistant</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>Sensitive</td>\n",
              "      <td>uncallable</td>\n",
              "      <td>uncallable</td>\n",
              "      <td>uncallable</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>3 rows × 45 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-560c49ae-f7e6-4c0f-83db-15b558a102d7')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-560c49ae-f7e6-4c0f-83db-15b558a102d7 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-560c49ae-f7e6-4c0f-83db-15b558a102d7');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-13f20574-642d-4dbc-ac70-9342e0e7332e\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-13f20574-642d-4dbc-ac70-9342e0e7332e')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-13f20574-642d-4dbc-ac70-9342e0e7332e button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "      Sample              Study     Country   Admin level 1  Country latitude  \\\n",
              "0   FP0008-C  1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
              "1   FP0009-C  1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
              "2  FP0010-CW  1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
              "\n",
              "   Country longitude  Admin level 1 latitude  Admin level 1 longitude    Year  \\\n",
              "0         -10.337093               16.565426                -9.832345  2014.0   \n",
              "1         -10.337093               16.565426                -9.832345  2014.0   \n",
              "2         -10.337093               16.565426                -9.832345  2014.0   \n",
              "\n",
              "          ENA  ... Mefloquine Artemisinin  Piperaquine  SP (uncomplicated)  \\\n",
              "0  ERR1081237  ...  Sensitive   Sensitive    Sensitive           Sensitive   \n",
              "1  ERR1081238  ...  Sensitive   Sensitive    Sensitive           Resistant   \n",
              "2  ERR2889621  ...  Sensitive   Sensitive    Sensitive           Resistant   \n",
              "\n",
              "   SP (IPTp)      AS-MQ    DHA-PPQ        HRP2        HRP3 HRP2 and HRP3  \n",
              "0  Sensitive  Sensitive  Sensitive       nodel       nodel         nodel  \n",
              "1  Sensitive  Sensitive  Sensitive       nodel       nodel         nodel  \n",
              "2  Sensitive  Sensitive  Sensitive  uncallable  uncallable    uncallable  \n",
              "\n",
              "[3 rows x 45 columns]"
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Merge the dataframes on \"Sample\" column\n",
        "df_all_sample_metadata = pd.merge(pd.merge(qc_sample_metadata,drm_calls_fn,on='Sample'),resistance_classification_fn,on='Sample')\n",
        "\n",
        "# Print first 3 rows\n",
        "df_all_sample_metadata.head(3)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "CiW7rU1zINHf"
      },
      "source": [
        "Sulfadoxine-Pyrimethamine (SP) is used for the treatment of uncomplicated cases and Intermittent Preventive Treatment in Pregnancy (IPTp). To make these two status clear in the dataset, we rename the columns."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "gKHJrr_s0MQt"
      },
      "outputs": [],
      "source": [
        "# Define a dictionary of old&new column names\n",
        "column_name_changes = collections.OrderedDict()\n",
        "column_name_changes['SP (uncomplicated)'] = 'SP (treatment)'\n",
        "column_name_changes['SP-IPTp'] = 'SP (IPTp)'\n",
        "\n",
        "# Rename the columns using the dictionary\n",
        "df_all_sample_metadata.rename(columns=column_name_changes, inplace=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cZFSrQkO0d9C"
      },
      "source": [
        "## Table setups"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xS4YEFRGIsvy"
      },
      "source": [
        "We proceed by defining a dictionary where each key represents a drug along with its associated gene, mutation, column name in the dataset, and sensitivity/resistance indicators.\n",
        "\n",
        "For details on the heuristics employed to map genetic markers to resistance status classification, please refer to [here](https://www.malariagen.net/sites/default/files/Pf7_resistance_classification.pdf)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "id": "NRDsSwjL0fU5"
      },
      "outputs": [],
      "source": [
        "drm_dict = collections.OrderedDict()\n",
        "\n",
        "drm_dict['Chloroquine'] = {\n",
        "    'gene': 'crt',\n",
        "    'mutation': '76T',\n",
        "    'column_name': 'crt_76[K]',\n",
        "    'sensitive': 'K',\n",
        "    'resistant': 'T',\n",
        "}\n",
        "\n",
        "drm_dict['Pyrimethamine'] = {\n",
        "    'gene': 'dhfr',\n",
        "    'mutation': '108N',\n",
        "    'column_name': 'dhfr_108[S]',\n",
        "    'sensitive': 'S',\n",
        "    'resistant': 'N',\n",
        "}\n",
        "\n",
        "drm_dict['Sulfadoxine'] = {\n",
        "    'gene': 'dhps',\n",
        "    'mutation': '437G',\n",
        "    'column_name': 'dhps_437[G]',\n",
        "    'sensitive': 'A',\n",
        "    'resistant': 'G',\n",
        "}\n",
        "\n",
        "drm_dict['Mefloquine'] = {\n",
        "    'gene': 'mdr1',\n",
        "    'mutation': '2+ copies',\n",
        "    'column_name': 'dup_mdr1',\n",
        "    'sensitive': 0,\n",
        "    'resistant': 1,\n",
        "}\n",
        "\n",
        "drm_dict['Artemisinin'] = {\n",
        "    'gene': 'kelch13',\n",
        "    'mutation': 'WHO list',\n",
        "    'column_name': None,\n",
        "}\n",
        "\n",
        "drm_dict['Piperaquine'] = {\n",
        "    'gene': 'plasmepsin 2-3',\n",
        "    'mutation': '2+ copies',\n",
        "    'column_name': 'dup_pm2',\n",
        "    'sensitive': 0,\n",
        "    'resistant': 1,\n",
        "}\n",
        "\n",
        "drm_dict['SP (treatment)'] = {\n",
        "    'gene': 'dhfr',\n",
        "    'mutation': 'triple mutant',\n",
        "    'column_name': None\n",
        "}\n",
        "\n",
        "drm_dict['SP (IPTp)'] = {\n",
        "    'gene': 'dhfr and dhps',\n",
        "    'mutation': 'sextuple mutant',\n",
        "    'column_name': None\n",
        "}\n",
        "\n",
        "drm_dict['AS-MQ'] = {\n",
        "    'gene': 'kelch13 and mdr1',\n",
        "    'mutation': '',\n",
        "    'column_name': None\n",
        "}\n",
        "\n",
        "drm_dict['DHA-PPQ'] = {\n",
        "    'gene': 'kelch13 and plasmepsin 2-3',\n",
        "    'mutation': '',\n",
        "    'column_name': None\n",
        "}"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3cfvT7nFN8NB"
      },
      "source": [
        "To elaborate on population names in the final table, we define a dictionary listing ten populations in Pf7 with the following abbreviations:\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {
        "id": "NkEr9h970mRX"
      },
      "outputs": [],
      "source": [
        "populations = collections.OrderedDict()\n",
        "populations['SA']       = \"South America\"\n",
        "populations['AF-W']     = \"Africa - West\"\n",
        "populations['AF-C']     = \"Africa - Central\"\n",
        "populations['AF-NE']    = \"Africa - Northeast\"\n",
        "populations['AF-E']     = \"Africa - East\"\n",
        "populations['AS-S-E']  = \"Asia - South - East\"\n",
        "populations['AS-S-FE']  = \"Asia - South - Far East\"\n",
        "populations['AS-SE-W'] = \"Asia - Southeast - West\"\n",
        "populations['AS-SE-E'] = \"Asia - Southeast - East\"\n",
        "populations['OC-NG']    = \"Oceania - New Guinea\""
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "8NQ1MCFS7Cit"
      },
      "source": [
        "## Summary of drug resistance sample sizes across populations"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "1c2wlN1MazjI"
      },
      "source": [
        "We want to create a table summarizing the number of samples by population.\n",
        "\n",
        "To do this, we will write a function that counts number of samples that are resistant and sensitive for a given drug."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {
        "id": "yVDUp_IwQWuE"
      },
      "outputs": [],
      "source": [
        "def n_agg(x):\n",
        "    \"\"\"\n",
        "    Aggregate function to calculate the count of non-undetermined samples for each drug in a given population.\n",
        "\n",
        "    Parameters:\n",
        "    - x: DataFrame\n",
        "\n",
        "    Returns:\n",
        "    - pd.Series: Counts of sensitive samples that are resistant to each drug.\n",
        "    \"\"\"\n",
        "\n",
        "    # Initialize an ordered dictionary to store drug names and their corresponding counts\n",
        "    names = collections.OrderedDict()\n",
        "\n",
        "    # Iterate through each drug in drm_dict\n",
        "    for drug in drm_dict:\n",
        "        # Count the number of non-undetermined samples for the current drug\n",
        "        n = np.count_nonzero((x[drug] != 'Undetermined'))\n",
        "\n",
        "        # Store the drug name and its count in the dictionary\n",
        "        names[drug] = n\n",
        "\n",
        "    # Return the result as a pandas Series\n",
        "    return pd.Series(names)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "DwKP3U4ycg-M"
      },
      "source": [
        "As an addition to table, we would like to include minimum and maximum drug resistance sample size in the population names."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 498
        },
        "id": "VNLkQeY_7HYb",
        "outputId": "a1fc8374-66e8-4cd0-b6b4-ca32ee00648a"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-098a925e-27f6-4ee6-88e1-159474fcc78e\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Associated with resistance to</th>\n",
              "      <th>South America (n=154-158)</th>\n",
              "      <th>Africa - West (n=5234-6233)</th>\n",
              "      <th>Africa - Central (n=397-520)</th>\n",
              "      <th>Africa - Northeast (n=120-170)</th>\n",
              "      <th>Africa - East (n=1373-1532)</th>\n",
              "      <th>Asia - South - East (n=164-233)</th>\n",
              "      <th>Asia - South - Far East (n=1212-1369)</th>\n",
              "      <th>Asia - Southeast - West (n=1657-1876)</th>\n",
              "      <th>Asia - Southeast - East (n=2059-3684)</th>\n",
              "      <th>Oceania - New Guinea (n=298-341)</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Marker</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>crt 76T</th>\n",
              "      <td>Chloroquine</td>\n",
              "      <td>155</td>\n",
              "      <td>5660</td>\n",
              "      <td>397</td>\n",
              "      <td>157</td>\n",
              "      <td>1388</td>\n",
              "      <td>217</td>\n",
              "      <td>1326</td>\n",
              "      <td>1871</td>\n",
              "      <td>3665</td>\n",
              "      <td>333</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>dhfr 108N</th>\n",
              "      <td>Pyrimethamine</td>\n",
              "      <td>154</td>\n",
              "      <td>5589</td>\n",
              "      <td>517</td>\n",
              "      <td>170</td>\n",
              "      <td>1476</td>\n",
              "      <td>201</td>\n",
              "      <td>1369</td>\n",
              "      <td>1876</td>\n",
              "      <td>3684</td>\n",
              "      <td>333</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>dhps 437G</th>\n",
              "      <td>Sulfadoxine</td>\n",
              "      <td>154</td>\n",
              "      <td>5529</td>\n",
              "      <td>501</td>\n",
              "      <td>162</td>\n",
              "      <td>1424</td>\n",
              "      <td>220</td>\n",
              "      <td>1291</td>\n",
              "      <td>1875</td>\n",
              "      <td>3609</td>\n",
              "      <td>333</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>mdr1 2+ copies</th>\n",
              "      <td>Mefloquine</td>\n",
              "      <td>158</td>\n",
              "      <td>5515</td>\n",
              "      <td>478</td>\n",
              "      <td>123</td>\n",
              "      <td>1509</td>\n",
              "      <td>164</td>\n",
              "      <td>1268</td>\n",
              "      <td>1782</td>\n",
              "      <td>3461</td>\n",
              "      <td>314</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>kelch13 WHO list</th>\n",
              "      <td>Artemisinin</td>\n",
              "      <td>158</td>\n",
              "      <td>5595</td>\n",
              "      <td>505</td>\n",
              "      <td>144</td>\n",
              "      <td>1513</td>\n",
              "      <td>189</td>\n",
              "      <td>1341</td>\n",
              "      <td>1768</td>\n",
              "      <td>3475</td>\n",
              "      <td>305</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>plasmepsin 2-3 2+ copies</th>\n",
              "      <td>Piperaquine</td>\n",
              "      <td>158</td>\n",
              "      <td>5464</td>\n",
              "      <td>519</td>\n",
              "      <td>120</td>\n",
              "      <td>1512</td>\n",
              "      <td>172</td>\n",
              "      <td>1272</td>\n",
              "      <td>1790</td>\n",
              "      <td>3428</td>\n",
              "      <td>298</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>dhfr triple mutant</th>\n",
              "      <td>SP (treatment)</td>\n",
              "      <td>158</td>\n",
              "      <td>5234</td>\n",
              "      <td>440</td>\n",
              "      <td>167</td>\n",
              "      <td>1373</td>\n",
              "      <td>230</td>\n",
              "      <td>1212</td>\n",
              "      <td>1833</td>\n",
              "      <td>3619</td>\n",
              "      <td>339</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>dhfr and dhps sextuple mutant</th>\n",
              "      <td>SP (IPTp)</td>\n",
              "      <td>158</td>\n",
              "      <td>6233</td>\n",
              "      <td>510</td>\n",
              "      <td>170</td>\n",
              "      <td>1446</td>\n",
              "      <td>233</td>\n",
              "      <td>1217</td>\n",
              "      <td>1657</td>\n",
              "      <td>2059</td>\n",
              "      <td>341</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>kelch13 and mdr1</th>\n",
              "      <td>AS-MQ</td>\n",
              "      <td>158</td>\n",
              "      <td>5915</td>\n",
              "      <td>519</td>\n",
              "      <td>150</td>\n",
              "      <td>1532</td>\n",
              "      <td>203</td>\n",
              "      <td>1354</td>\n",
              "      <td>1798</td>\n",
              "      <td>3495</td>\n",
              "      <td>324</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>kelch13 and plasmepsin 2-3</th>\n",
              "      <td>DHA-PPQ</td>\n",
              "      <td>158</td>\n",
              "      <td>5823</td>\n",
              "      <td>520</td>\n",
              "      <td>148</td>\n",
              "      <td>1526</td>\n",
              "      <td>199</td>\n",
              "      <td>1355</td>\n",
              "      <td>1829</td>\n",
              "      <td>3442</td>\n",
              "      <td>316</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-098a925e-27f6-4ee6-88e1-159474fcc78e')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-098a925e-27f6-4ee6-88e1-159474fcc78e button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-098a925e-27f6-4ee6-88e1-159474fcc78e');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-b9959a65-8f23-4671-bc69-2c8748213392\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-b9959a65-8f23-4671-bc69-2c8748213392')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-b9959a65-8f23-4671-bc69-2c8748213392 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "                              Associated with resistance to  \\\n",
              "Marker                                                        \n",
              "crt 76T                                         Chloroquine   \n",
              "dhfr 108N                                     Pyrimethamine   \n",
              "dhps 437G                                       Sulfadoxine   \n",
              "mdr1 2+ copies                                   Mefloquine   \n",
              "kelch13 WHO list                                Artemisinin   \n",
              "plasmepsin 2-3 2+ copies                        Piperaquine   \n",
              "dhfr triple mutant                           SP (treatment)   \n",
              "dhfr and dhps sextuple mutant                     SP (IPTp)   \n",
              "kelch13 and mdr1                                      AS-MQ   \n",
              "kelch13 and plasmepsin 2-3                          DHA-PPQ   \n",
              "\n",
              "                               South America (n=154-158)  \\\n",
              "Marker                                                     \n",
              "crt 76T                                              155   \n",
              "dhfr 108N                                            154   \n",
              "dhps 437G                                            154   \n",
              "mdr1 2+ copies                                       158   \n",
              "kelch13 WHO list                                     158   \n",
              "plasmepsin 2-3 2+ copies                             158   \n",
              "dhfr triple mutant                                   158   \n",
              "dhfr and dhps sextuple mutant                        158   \n",
              "kelch13 and mdr1                                     158   \n",
              "kelch13 and plasmepsin 2-3                           158   \n",
              "\n",
              "                               Africa - West (n=5234-6233)  \\\n",
              "Marker                                                       \n",
              "crt 76T                                               5660   \n",
              "dhfr 108N                                             5589   \n",
              "dhps 437G                                             5529   \n",
              "mdr1 2+ copies                                        5515   \n",
              "kelch13 WHO list                                      5595   \n",
              "plasmepsin 2-3 2+ copies                              5464   \n",
              "dhfr triple mutant                                    5234   \n",
              "dhfr and dhps sextuple mutant                         6233   \n",
              "kelch13 and mdr1                                      5915   \n",
              "kelch13 and plasmepsin 2-3                            5823   \n",
              "\n",
              "                               Africa - Central (n=397-520)  \\\n",
              "Marker                                                        \n",
              "crt 76T                                                 397   \n",
              "dhfr 108N                                               517   \n",
              "dhps 437G                                               501   \n",
              "mdr1 2+ copies                                          478   \n",
              "kelch13 WHO list                                        505   \n",
              "plasmepsin 2-3 2+ copies                                519   \n",
              "dhfr triple mutant                                      440   \n",
              "dhfr and dhps sextuple mutant                           510   \n",
              "kelch13 and mdr1                                        519   \n",
              "kelch13 and plasmepsin 2-3                              520   \n",
              "\n",
              "                               Africa - Northeast (n=120-170)  \\\n",
              "Marker                                                          \n",
              "crt 76T                                                   157   \n",
              "dhfr 108N                                                 170   \n",
              "dhps 437G                                                 162   \n",
              "mdr1 2+ copies                                            123   \n",
              "kelch13 WHO list                                          144   \n",
              "plasmepsin 2-3 2+ copies                                  120   \n",
              "dhfr triple mutant                                        167   \n",
              "dhfr and dhps sextuple mutant                             170   \n",
              "kelch13 and mdr1                                          150   \n",
              "kelch13 and plasmepsin 2-3                                148   \n",
              "\n",
              "                               Africa - East (n=1373-1532)  \\\n",
              "Marker                                                       \n",
              "crt 76T                                               1388   \n",
              "dhfr 108N                                             1476   \n",
              "dhps 437G                                             1424   \n",
              "mdr1 2+ copies                                        1509   \n",
              "kelch13 WHO list                                      1513   \n",
              "plasmepsin 2-3 2+ copies                              1512   \n",
              "dhfr triple mutant                                    1373   \n",
              "dhfr and dhps sextuple mutant                         1446   \n",
              "kelch13 and mdr1                                      1532   \n",
              "kelch13 and plasmepsin 2-3                            1526   \n",
              "\n",
              "                               Asia - South - East (n=164-233)  \\\n",
              "Marker                                                           \n",
              "crt 76T                                                    217   \n",
              "dhfr 108N                                                  201   \n",
              "dhps 437G                                                  220   \n",
              "mdr1 2+ copies                                             164   \n",
              "kelch13 WHO list                                           189   \n",
              "plasmepsin 2-3 2+ copies                                   172   \n",
              "dhfr triple mutant                                         230   \n",
              "dhfr and dhps sextuple mutant                              233   \n",
              "kelch13 and mdr1                                           203   \n",
              "kelch13 and plasmepsin 2-3                                 199   \n",
              "\n",
              "                               Asia - South - Far East (n=1212-1369)  \\\n",
              "Marker                                                                 \n",
              "crt 76T                                                         1326   \n",
              "dhfr 108N                                                       1369   \n",
              "dhps 437G                                                       1291   \n",
              "mdr1 2+ copies                                                  1268   \n",
              "kelch13 WHO list                                                1341   \n",
              "plasmepsin 2-3 2+ copies                                        1272   \n",
              "dhfr triple mutant                                              1212   \n",
              "dhfr and dhps sextuple mutant                                   1217   \n",
              "kelch13 and mdr1                                                1354   \n",
              "kelch13 and plasmepsin 2-3                                      1355   \n",
              "\n",
              "                               Asia - Southeast - West (n=1657-1876)  \\\n",
              "Marker                                                                 \n",
              "crt 76T                                                         1871   \n",
              "dhfr 108N                                                       1876   \n",
              "dhps 437G                                                       1875   \n",
              "mdr1 2+ copies                                                  1782   \n",
              "kelch13 WHO list                                                1768   \n",
              "plasmepsin 2-3 2+ copies                                        1790   \n",
              "dhfr triple mutant                                              1833   \n",
              "dhfr and dhps sextuple mutant                                   1657   \n",
              "kelch13 and mdr1                                                1798   \n",
              "kelch13 and plasmepsin 2-3                                      1829   \n",
              "\n",
              "                               Asia - Southeast - East (n=2059-3684)  \\\n",
              "Marker                                                                 \n",
              "crt 76T                                                         3665   \n",
              "dhfr 108N                                                       3684   \n",
              "dhps 437G                                                       3609   \n",
              "mdr1 2+ copies                                                  3461   \n",
              "kelch13 WHO list                                                3475   \n",
              "plasmepsin 2-3 2+ copies                                        3428   \n",
              "dhfr triple mutant                                              3619   \n",
              "dhfr and dhps sextuple mutant                                   2059   \n",
              "kelch13 and mdr1                                                3495   \n",
              "kelch13 and plasmepsin 2-3                                      3442   \n",
              "\n",
              "                               Oceania - New Guinea (n=298-341)  \n",
              "Marker                                                           \n",
              "crt 76T                                                     333  \n",
              "dhfr 108N                                                   333  \n",
              "dhps 437G                                                   333  \n",
              "mdr1 2+ copies                                              314  \n",
              "kelch13 WHO list                                            305  \n",
              "plasmepsin 2-3 2+ copies                                    298  \n",
              "dhfr triple mutant                                          339  \n",
              "dhfr and dhps sextuple mutant                               341  \n",
              "kelch13 and mdr1                                            324  \n",
              "kelch13 and plasmepsin 2-3                                  316  "
            ]
          },
          "execution_count": 14,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Calculate counts of non-undetermined samples for each drug in each population\n",
        "df_drm_n_table = (\n",
        "    df_all_sample_metadata\n",
        "    .groupby('Population')\n",
        "    .apply(n_agg)\n",
        "    .rename_axis(None)\n",
        "    .transpose()\n",
        "    .loc[:, populations.keys()]\n",
        "    .reset_index()\n",
        ")\n",
        "\n",
        "# Calculate the minimum and maximum sample sizes across populations\n",
        "min_n = df_drm_n_table.min()\n",
        "max_n = df_drm_n_table.max()\n",
        "\n",
        "# Customize the index to include gene and mutation information for each drug\n",
        "df_drm_n_table.index = [\"%s %s\" % (drm_dict[drug]['gene'], drm_dict[drug]['mutation']) for drug in drm_dict]\n",
        "df_drm_n_table.index.names = ['Marker']\n",
        "\n",
        "# Rename the columns to indicate their association with drug resistance and sample counts\n",
        "df_drm_n_table.rename(columns={'index': 'Associated with resistance to'}, inplace=True)\n",
        "\n",
        "# Update column names to include population names and corresponding sample size ranges\n",
        "for population in populations:\n",
        "    new_column_name = f'{populations[population]} (n={min_n[population]}-{max_n[population]})'\n",
        "    df_drm_n_table.rename(columns={population: new_column_name}, inplace=True)\n",
        "\n",
        "# Display the final drug resistance count table\n",
        "df_drm_n_table"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Kkoy6lwGouuN"
      },
      "source": [
        "**Supplementary Table.** Numbers of samples used to determine proportions in the main table."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "g_ey7FCwQuAA"
      },
      "source": [
        "To save the table:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "t4QP1cf_QwjC"
      },
      "outputs": [],
      "source": [
        "df_drm_n_table.to_excel(\"DRM_table_sample_numbers.xlsx\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-2YbaP-KnTYb"
      },
      "source": [
        "## Summary of drug resistance frequencies across populations"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "EsHY5DKnYpoX"
      },
      "source": [
        "Now, we would like to calculate drug resistance proportions in each population.\n",
        "\n",
        "We can easily adjust the function to count number of samples that are resistant and sensitive for a given drug, and calculate the proportion.  "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {
        "id": "amTK8om50oEP"
      },
      "outputs": [],
      "source": [
        "def proportion_agg(x):\n",
        "    \"\"\"\n",
        "    Aggregate function to calculate the proportion of resistance for each drug in a given population.\n",
        "\n",
        "    Parameters:\n",
        "    - x: DataFrame\n",
        "\n",
        "    Returns:\n",
        "    - pd.Series: Proportions of resistance for each drug.\n",
        "    \"\"\"\n",
        "\n",
        "    # Initialize an ordered dictionary to store drug names and their corresponding proportions\n",
        "    names = collections.OrderedDict()\n",
        "\n",
        "    # Iterate through each drug in drm_dict\n",
        "    for drug in drm_dict:\n",
        "        # Count the number of non-undetermined samples for the current drug\n",
        "        n = np.count_nonzero((x[drug] != 'Undetermined'))\n",
        "\n",
        "        # Check if there are no non-undetermined samples for the drug\n",
        "        if n == 0:\n",
        "            proportion = np.nan  # Set proportion to NaN to avoid division by zero\n",
        "        else:\n",
        "            # Calculate the proportion of resistant samples for the drug\n",
        "            proportion = round(np.count_nonzero(\n",
        "                (x[drug] == 'Resistant')\n",
        "            ) / np.count_nonzero(\n",
        "                (x[drug] != 'Undetermined')\n",
        "            )*100)\n",
        "\n",
        "        # Store the drug name and its proportion in the dictionary\n",
        "        names[drug] = proportion\n",
        "\n",
        "    # Return the result as a pandas Series\n",
        "    return pd.Series(names)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "g17tr7u9aWH1"
      },
      "source": [
        "Let's apply this function and create a table summarizing drug resistance proportions by population.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 498
        },
        "id": "lSG0_DSxPPy5",
        "outputId": "cc358416-b104-444a-b35d-879c0fe57755"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-dfbf7d45-fc41-4113-91a8-56aed315dd22\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Associated with resistance to</th>\n",
              "      <th>South America (n=154-158)</th>\n",
              "      <th>Africa - West (n=5234-6233)</th>\n",
              "      <th>Africa - Central (n=397-520)</th>\n",
              "      <th>Africa - Northeast (n=120-170)</th>\n",
              "      <th>Africa - East (n=1373-1532)</th>\n",
              "      <th>Asia - South - East (n=164-233)</th>\n",
              "      <th>Asia - South - Far East (n=1212-1369)</th>\n",
              "      <th>Asia - Southeast - West (n=1657-1876)</th>\n",
              "      <th>Asia - Southeast - East (n=2059-3684)</th>\n",
              "      <th>Oceania - New Guinea (n=298-341)</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Marker</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>crt 76T</th>\n",
              "      <td>Chloroquine</td>\n",
              "      <td>100%</td>\n",
              "      <td>29%</td>\n",
              "      <td>61%</td>\n",
              "      <td>40%</td>\n",
              "      <td>24%</td>\n",
              "      <td>31%</td>\n",
              "      <td>94%</td>\n",
              "      <td>99%</td>\n",
              "      <td>95%</td>\n",
              "      <td>96%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>dhfr 108N</th>\n",
              "      <td>Pyrimethamine</td>\n",
              "      <td>64%</td>\n",
              "      <td>87%</td>\n",
              "      <td>100%</td>\n",
              "      <td>98%</td>\n",
              "      <td>96%</td>\n",
              "      <td>64%</td>\n",
              "      <td>100%</td>\n",
              "      <td>100%</td>\n",
              "      <td>99%</td>\n",
              "      <td>99%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>dhps 437G</th>\n",
              "      <td>Sulfadoxine</td>\n",
              "      <td>60%</td>\n",
              "      <td>78%</td>\n",
              "      <td>97%</td>\n",
              "      <td>82%</td>\n",
              "      <td>83%</td>\n",
              "      <td>8%</td>\n",
              "      <td>89%</td>\n",
              "      <td>100%</td>\n",
              "      <td>83%</td>\n",
              "      <td>69%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>mdr1 2+ copies</th>\n",
              "      <td>Mefloquine</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>29%</td>\n",
              "      <td>5%</td>\n",
              "      <td>1%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>kelch13 WHO list</th>\n",
              "      <td>Artemisinin</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>36%</td>\n",
              "      <td>58%</td>\n",
              "      <td>1%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>plasmepsin 2-3 2+ copies</th>\n",
              "      <td>Piperaquine</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>37%</td>\n",
              "      <td>0%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>dhfr triple mutant</th>\n",
              "      <td>SP (treatment)</td>\n",
              "      <td>0%</td>\n",
              "      <td>77%</td>\n",
              "      <td>85%</td>\n",
              "      <td>61%</td>\n",
              "      <td>80%</td>\n",
              "      <td>1%</td>\n",
              "      <td>46%</td>\n",
              "      <td>86%</td>\n",
              "      <td>88%</td>\n",
              "      <td>0%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>dhfr and dhps sextuple mutant</th>\n",
              "      <td>SP (IPTp)</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>2%</td>\n",
              "      <td>2%</td>\n",
              "      <td>9%</td>\n",
              "      <td>0%</td>\n",
              "      <td>13%</td>\n",
              "      <td>79%</td>\n",
              "      <td>14%</td>\n",
              "      <td>0%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>kelch13 and mdr1</th>\n",
              "      <td>AS-MQ</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>10%</td>\n",
              "      <td>4%</td>\n",
              "      <td>0%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>kelch13 and plasmepsin 2-3</th>\n",
              "      <td>DHA-PPQ</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>0%</td>\n",
              "      <td>34%</td>\n",
              "      <td>0%</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-dfbf7d45-fc41-4113-91a8-56aed315dd22')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-dfbf7d45-fc41-4113-91a8-56aed315dd22 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-dfbf7d45-fc41-4113-91a8-56aed315dd22');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-d372ed46-6132-437c-aaf7-5e37099df70c\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-d372ed46-6132-437c-aaf7-5e37099df70c')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-d372ed46-6132-437c-aaf7-5e37099df70c button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "                              Associated with resistance to  \\\n",
              "Marker                                                        \n",
              "crt 76T                                         Chloroquine   \n",
              "dhfr 108N                                     Pyrimethamine   \n",
              "dhps 437G                                       Sulfadoxine   \n",
              "mdr1 2+ copies                                   Mefloquine   \n",
              "kelch13 WHO list                                Artemisinin   \n",
              "plasmepsin 2-3 2+ copies                        Piperaquine   \n",
              "dhfr triple mutant                           SP (treatment)   \n",
              "dhfr and dhps sextuple mutant                     SP (IPTp)   \n",
              "kelch13 and mdr1                                      AS-MQ   \n",
              "kelch13 and plasmepsin 2-3                          DHA-PPQ   \n",
              "\n",
              "                              South America (n=154-158)  \\\n",
              "Marker                                                    \n",
              "crt 76T                                            100%   \n",
              "dhfr 108N                                           64%   \n",
              "dhps 437G                                           60%   \n",
              "mdr1 2+ copies                                       0%   \n",
              "kelch13 WHO list                                     0%   \n",
              "plasmepsin 2-3 2+ copies                             0%   \n",
              "dhfr triple mutant                                   0%   \n",
              "dhfr and dhps sextuple mutant                        0%   \n",
              "kelch13 and mdr1                                     0%   \n",
              "kelch13 and plasmepsin 2-3                           0%   \n",
              "\n",
              "                              Africa - West (n=5234-6233)  \\\n",
              "Marker                                                      \n",
              "crt 76T                                               29%   \n",
              "dhfr 108N                                             87%   \n",
              "dhps 437G                                             78%   \n",
              "mdr1 2+ copies                                         0%   \n",
              "kelch13 WHO list                                       0%   \n",
              "plasmepsin 2-3 2+ copies                               0%   \n",
              "dhfr triple mutant                                    77%   \n",
              "dhfr and dhps sextuple mutant                          0%   \n",
              "kelch13 and mdr1                                       0%   \n",
              "kelch13 and plasmepsin 2-3                             0%   \n",
              "\n",
              "                              Africa - Central (n=397-520)  \\\n",
              "Marker                                                       \n",
              "crt 76T                                                61%   \n",
              "dhfr 108N                                             100%   \n",
              "dhps 437G                                              97%   \n",
              "mdr1 2+ copies                                          0%   \n",
              "kelch13 WHO list                                        0%   \n",
              "plasmepsin 2-3 2+ copies                                0%   \n",
              "dhfr triple mutant                                     85%   \n",
              "dhfr and dhps sextuple mutant                           2%   \n",
              "kelch13 and mdr1                                        0%   \n",
              "kelch13 and plasmepsin 2-3                              0%   \n",
              "\n",
              "                              Africa - Northeast (n=120-170)  \\\n",
              "Marker                                                         \n",
              "crt 76T                                                  40%   \n",
              "dhfr 108N                                                98%   \n",
              "dhps 437G                                                82%   \n",
              "mdr1 2+ copies                                            0%   \n",
              "kelch13 WHO list                                          0%   \n",
              "plasmepsin 2-3 2+ copies                                  0%   \n",
              "dhfr triple mutant                                       61%   \n",
              "dhfr and dhps sextuple mutant                             2%   \n",
              "kelch13 and mdr1                                          0%   \n",
              "kelch13 and plasmepsin 2-3                                0%   \n",
              "\n",
              "                              Africa - East (n=1373-1532)  \\\n",
              "Marker                                                      \n",
              "crt 76T                                               24%   \n",
              "dhfr 108N                                             96%   \n",
              "dhps 437G                                             83%   \n",
              "mdr1 2+ copies                                         0%   \n",
              "kelch13 WHO list                                       0%   \n",
              "plasmepsin 2-3 2+ copies                               0%   \n",
              "dhfr triple mutant                                    80%   \n",
              "dhfr and dhps sextuple mutant                          9%   \n",
              "kelch13 and mdr1                                       0%   \n",
              "kelch13 and plasmepsin 2-3                             0%   \n",
              "\n",
              "                              Asia - South - East (n=164-233)  \\\n",
              "Marker                                                          \n",
              "crt 76T                                                   31%   \n",
              "dhfr 108N                                                 64%   \n",
              "dhps 437G                                                  8%   \n",
              "mdr1 2+ copies                                             0%   \n",
              "kelch13 WHO list                                           0%   \n",
              "plasmepsin 2-3 2+ copies                                   0%   \n",
              "dhfr triple mutant                                         1%   \n",
              "dhfr and dhps sextuple mutant                              0%   \n",
              "kelch13 and mdr1                                           0%   \n",
              "kelch13 and plasmepsin 2-3                                 0%   \n",
              "\n",
              "                              Asia - South - Far East (n=1212-1369)  \\\n",
              "Marker                                                                \n",
              "crt 76T                                                         94%   \n",
              "dhfr 108N                                                      100%   \n",
              "dhps 437G                                                       89%   \n",
              "mdr1 2+ copies                                                   0%   \n",
              "kelch13 WHO list                                                 0%   \n",
              "plasmepsin 2-3 2+ copies                                         0%   \n",
              "dhfr triple mutant                                              46%   \n",
              "dhfr and dhps sextuple mutant                                   13%   \n",
              "kelch13 and mdr1                                                 0%   \n",
              "kelch13 and plasmepsin 2-3                                       0%   \n",
              "\n",
              "                              Asia - Southeast - West (n=1657-1876)  \\\n",
              "Marker                                                                \n",
              "crt 76T                                                         99%   \n",
              "dhfr 108N                                                      100%   \n",
              "dhps 437G                                                      100%   \n",
              "mdr1 2+ copies                                                  29%   \n",
              "kelch13 WHO list                                                36%   \n",
              "plasmepsin 2-3 2+ copies                                         0%   \n",
              "dhfr triple mutant                                              86%   \n",
              "dhfr and dhps sextuple mutant                                   79%   \n",
              "kelch13 and mdr1                                                10%   \n",
              "kelch13 and plasmepsin 2-3                                       0%   \n",
              "\n",
              "                              Asia - Southeast - East (n=2059-3684)  \\\n",
              "Marker                                                                \n",
              "crt 76T                                                         95%   \n",
              "dhfr 108N                                                       99%   \n",
              "dhps 437G                                                       83%   \n",
              "mdr1 2+ copies                                                   5%   \n",
              "kelch13 WHO list                                                58%   \n",
              "plasmepsin 2-3 2+ copies                                        37%   \n",
              "dhfr triple mutant                                              88%   \n",
              "dhfr and dhps sextuple mutant                                   14%   \n",
              "kelch13 and mdr1                                                 4%   \n",
              "kelch13 and plasmepsin 2-3                                      34%   \n",
              "\n",
              "                              Oceania - New Guinea (n=298-341)  \n",
              "Marker                                                          \n",
              "crt 76T                                                    96%  \n",
              "dhfr 108N                                                  99%  \n",
              "dhps 437G                                                  69%  \n",
              "mdr1 2+ copies                                              1%  \n",
              "kelch13 WHO list                                            1%  \n",
              "plasmepsin 2-3 2+ copies                                    0%  \n",
              "dhfr triple mutant                                          0%  \n",
              "dhfr and dhps sextuple mutant                               0%  \n",
              "kelch13 and mdr1                                            0%  \n",
              "kelch13 and plasmepsin 2-3                                  0%  "
            ]
          },
          "execution_count": 16,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Create a table summarizing drug resistance proportions by population\n",
        "\n",
        "# Group the DataFrame by 'Population' and apply the 'proportion_agg' function to calculate resistance proportions\n",
        "df_drm_table = (\n",
        "    df_all_sample_metadata\n",
        "    .groupby('Population')\n",
        "    .apply(proportion_agg)\n",
        "    .rename_axis(None)\n",
        "    .transpose()\n",
        "    .loc[:, populations.keys()]\n",
        "    .reset_index()\n",
        ")\n",
        "\n",
        "# Customize the index to include gene and mutation information for each drug\n",
        "df_drm_table.index = [\"%s %s\" % (drm_dict[drug]['gene'], drm_dict[drug]['mutation']) for drug in drm_dict]\n",
        "df_drm_table.index.names = ['Marker']\n",
        "\n",
        "# Rename the columns to indicate their association with drug resistance\n",
        "df_drm_table.rename(columns={'index': 'Associated with resistance to'}, inplace=True)\n",
        "\n",
        "# Update column names to include population names and corresponding sample size ranges\n",
        "for population in populations:\n",
        "    new_column_name = f'{populations[population]} (n={min_n[population]}-{max_n[population]})'\n",
        "    df_drm_table.rename(columns={population: new_column_name}, inplace=True)\n",
        "\n",
        "# Add '%' symbol to all values in the table\n",
        "df_drm_table = df_drm_table.applymap(lambda x: f\"{int(x)}%\" if isinstance(x, (int, float)) else x)\n",
        "\n",
        "# Display the final drug resistance table\n",
        "df_drm_table"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MoMG3wxmjtoo"
      },
      "source": [
        "**Table Legend.** Frequency of different sets of polymorphisms associated with drug resistance in samples from different geographical regions.\n",
        "All samples were classified into different types of drug resistance based on published genetic markers, and represent best attempt based on the available data. Each type of resistance was considered to be either present, absent or unknown for a given sample. For each resistance type, the table reports: the genetic markers considered; the drug they are associated with; the proportion of samples in each major sub-population classified as resistant out of the samples where the type was not unknown. The number of samples classified as either resistant or not resistant varies for each type of resistance considered (e.g. due to different levels of genomic accessibility); numbers in brackets report the minimum and maximum number analysed while the exact numbers considered are reported in Supplementary Table 4. SP: sulfadoxine-pyrimethamine; treatment: SP used for the clinical treatment of uncomplicated malaria; IPTp: SP used for intermittent preventive treatment in pregnancy; AS-MQ: artesunate + mefloquine combination therapy; DHA-PPQ: dihydroartemisinin + piperaquine combination therapy. dhfr triple mutant refers to having all three of 51I, 59R and 108N in dhfr. dhfr and dhps sextuple mutant refers to having all five of 51I, 59R and 108N in dhfr and 437G and 540E in dhps, plus one of dhfr:164L, dhps:581G, dhps:613S or dhps:613T. Full details of the rules used to infer resistance status from genetic markers can be found on the resource page at https://www.malariagen.net/resource/34."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lJnlyrhnQJqz"
      },
      "source": [
        "To save the table:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "jmN7Fuxo0qyy"
      },
      "outputs": [],
      "source": [
        "df_drm_table.to_excel(\"Main_DRM_table.xlsx\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "sJDJFjpr7rVr"
      },
      "source": [
        "## Summary of drug resistance frequencies between time periods across populations"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nYwo_gAPedrx"
      },
      "source": [
        "We are interested to compare the drug resistance between periods of 2001-2014 and 2015-2019."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_Qgv-zwJF3JL"
      },
      "source": [
        "To compare drug resistance frequencies between 2001-2014 and 2015-2019, a new column (`Year group`) has been added to categorize samples accordingly."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {
        "id": "fTuY6sGkIMXH"
      },
      "outputs": [],
      "source": [
        "# Create a new column\n",
        "# Catagorize samples collected in 2001-2014\n",
        "df_all_sample_metadata.loc[( df_all_sample_metadata['Year'] >= 2001 ) & ( df_all_sample_metadata['Year'] <= 2014 ), 'Year group'] = \"2001-2014\"\n",
        "\n",
        "# Catagorize samples collected in 2015-2019\n",
        "df_all_sample_metadata.loc[( df_all_sample_metadata['Year'] >= 2015 ) & ( df_all_sample_metadata['Year'] <= 2019 ), 'Year group'] = \"2015-2019\"\n",
        "\n",
        "# Convert 'Year' column into integer format\n",
        "df_all_sample_metadata['Year'] = pd.to_numeric(df_all_sample_metadata['Year'], downcast = 'integer')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5KQtxHnjd7yF"
      },
      "source": [
        "Next, we adjust the function to count samples within each group. As an addition, the resulting table will provide the proportions and counts for both year groups combined in each cell value."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {
        "id": "GnzE_uF87mG4"
      },
      "outputs": [],
      "source": [
        "def my_agg(x):\n",
        "    \"\"\"\n",
        "    Aggregate function to calculate proportions of resistance and counts for each drug in two year groups.\n",
        "\n",
        "    Parameters:\n",
        "    - x: DataFrame\n",
        "\n",
        "    Returns:\n",
        "    - pd.Series: Series containing proportions and counts of resistance for each drug in two year groups.\n",
        "    \"\"\"\n",
        "\n",
        "    # Initialize an ordered dictionary to store drug names and their corresponding proportions and counts\n",
        "    names = collections.OrderedDict()\n",
        "\n",
        "    # Iterate through each drug in drm_dict\n",
        "    for drug in drm_dict:\n",
        "        # Calculate counts and proportions for Year Group 1 (2001-2014)\n",
        "        n_year_group_1 = np.count_nonzero((x[drug] != 'Undetermined') & (x['Year group'] == '2001-2014'))\n",
        "        proportion_year_group_1 = (\n",
        "            np.count_nonzero((x[drug] == 'Resistant') & (x['Year group'] == '2001-2014')) /\n",
        "            np.count_nonzero((x[drug] != 'Undetermined') & (x['Year group'] == '2001-2014'))\n",
        "        ) if n_year_group_1 != 0 else np.nan\n",
        "\n",
        "        # Calculate counts and proportions for Year Group 2 (2015-2019)\n",
        "        n_year_group_2 = np.count_nonzero((x[drug] != 'Undetermined') & (x['Year group'] == '2015-2019'))\n",
        "        proportion_year_group_2 = (\n",
        "            np.count_nonzero((x[drug] == 'Resistant') & (x['Year group'] == '2015-2019')) /\n",
        "            np.count_nonzero((x[drug] != 'Undetermined') & (x['Year group'] == '2015-2019'))\n",
        "        ) if n_year_group_2 != 0 else np.nan\n",
        "\n",
        "        # Store drug name, proportions, and counts in the dictionary\n",
        "        names[drug] = f\"{proportion_year_group_1:.2f} (n={n_year_group_1})   {proportion_year_group_2:.2f} (n={n_year_group_2})\"\n",
        "\n",
        "    # Return the result as a pandas Series\n",
        "    return pd.Series(names)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "e9HUUAydfPEd"
      },
      "source": [
        "Let's create the table in a similar way to before."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 585
        },
        "id": "nuPMFWrRQqsX",
        "outputId": "fd0c6edd-9b25-41e0-e748-d21725d683d2"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-6e5083df-1fa0-45dd-a1fe-1dbd4364f0b2\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Associated with resistance to</th>\n",
              "      <th>South America</th>\n",
              "      <th>Africa - West</th>\n",
              "      <th>Africa - Central</th>\n",
              "      <th>Africa - Northeast</th>\n",
              "      <th>Africa - East</th>\n",
              "      <th>Asia - South - East</th>\n",
              "      <th>Asia - South - Far East</th>\n",
              "      <th>Asia - Southeast - West</th>\n",
              "      <th>Asia - Southeast - East</th>\n",
              "      <th>Oceania - New Guinea</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Marker</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>crt 76T</th>\n",
              "      <td>Chloroquine</td>\n",
              "      <td>1.00 (n=44)   1.00 (n=111)</td>\n",
              "      <td>0.42 (n=2500)   0.19 (n=3008)</td>\n",
              "      <td>0.65 (n=292)   0.52 (n=105)</td>\n",
              "      <td>0.19 (n=78)   0.61 (n=79)</td>\n",
              "      <td>0.18 (n=1244)   0.05 (n=19)</td>\n",
              "      <td>nan (n=0)   0.31 (n=217)</td>\n",
              "      <td>0.93 (n=74)   0.94 (n=1252)</td>\n",
              "      <td>1.00 (n=1158)   0.99 (n=713)</td>\n",
              "      <td>0.94 (n=1427)   0.95 (n=2232)</td>\n",
              "      <td>0.99 (n=195)   0.91 (n=138)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>dhfr 108N</th>\n",
              "      <td>Pyrimethamine</td>\n",
              "      <td>0.93 (n=44)   0.53 (n=110)</td>\n",
              "      <td>0.85 (n=2547)   0.92 (n=2902)</td>\n",
              "      <td>1.00 (n=383)   0.99 (n=134)</td>\n",
              "      <td>0.99 (n=86)   0.96 (n=84)</td>\n",
              "      <td>0.99 (n=1346)   1.00 (n=19)</td>\n",
              "      <td>nan (n=0)   0.64 (n=201)</td>\n",
              "      <td>1.00 (n=77)   1.00 (n=1292)</td>\n",
              "      <td>1.00 (n=1160)   1.00 (n=716)</td>\n",
              "      <td>0.99 (n=1438)   0.99 (n=2240)</td>\n",
              "      <td>0.99 (n=200)   0.98 (n=133)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>dhps 437G</th>\n",
              "      <td>Sulfadoxine</td>\n",
              "      <td>0.35 (n=43)   0.69 (n=111)</td>\n",
              "      <td>0.75 (n=2513)   0.84 (n=2866)</td>\n",
              "      <td>0.97 (n=371)   0.96 (n=130)</td>\n",
              "      <td>0.99 (n=85)   0.64 (n=77)</td>\n",
              "      <td>0.92 (n=1265)   1.00 (n=19)</td>\n",
              "      <td>nan (n=0)   0.08 (n=220)</td>\n",
              "      <td>0.97 (n=63)   0.89 (n=1228)</td>\n",
              "      <td>1.00 (n=1160)   0.99 (n=715)</td>\n",
              "      <td>0.88 (n=1377)   0.81 (n=2226)</td>\n",
              "      <td>0.61 (n=197)   0.81 (n=136)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>mdr1 2+ copies</th>\n",
              "      <td>Mefloquine</td>\n",
              "      <td>0.00 (n=44)   0.00 (n=114)</td>\n",
              "      <td>0.00 (n=2735)   0.00 (n=2644)</td>\n",
              "      <td>0.00 (n=349)   0.00 (n=129)</td>\n",
              "      <td>0.00 (n=86)   0.00 (n=37)</td>\n",
              "      <td>0.00 (n=1339)   0.00 (n=14)</td>\n",
              "      <td>nan (n=0)   0.00 (n=164)</td>\n",
              "      <td>0.00 (n=77)   0.00 (n=1191)</td>\n",
              "      <td>0.43 (n=1128)   0.06 (n=654)</td>\n",
              "      <td>0.12 (n=1334)   0.01 (n=2122)</td>\n",
              "      <td>0.01 (n=198)   0.02 (n=116)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>kelch13 WHO list</th>\n",
              "      <td>Artemisinin</td>\n",
              "      <td>0.00 (n=44)   0.00 (n=114)</td>\n",
              "      <td>0.00 (n=2781)   0.00 (n=2668)</td>\n",
              "      <td>0.00 (n=374)   0.00 (n=131)</td>\n",
              "      <td>0.00 (n=84)   0.00 (n=60)</td>\n",
              "      <td>0.00 (n=1341)   0.00 (n=20)</td>\n",
              "      <td>nan (n=0)   0.00 (n=189)</td>\n",
              "      <td>0.00 (n=78)   0.00 (n=1263)</td>\n",
              "      <td>0.28 (n=1103)   0.49 (n=665)</td>\n",
              "      <td>0.48 (n=1290)   0.64 (n=2179)</td>\n",
              "      <td>0.00 (n=199)   0.02 (n=106)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>plasmepsin 2-3 2+ copies</th>\n",
              "      <td>Piperaquine</td>\n",
              "      <td>0.00 (n=44)   0.00 (n=114)</td>\n",
              "      <td>0.00 (n=2769)   0.00 (n=2558)</td>\n",
              "      <td>0.00 (n=385)   0.00 (n=134)</td>\n",
              "      <td>0.00 (n=86)   0.00 (n=34)</td>\n",
              "      <td>0.00 (n=1337)   0.00 (n=19)</td>\n",
              "      <td>nan (n=0)   0.00 (n=172)</td>\n",
              "      <td>0.00 (n=78)   0.00 (n=1194)</td>\n",
              "      <td>0.00 (n=1139)   0.00 (n=651)</td>\n",
              "      <td>0.19 (n=1296)   0.49 (n=2128)</td>\n",
              "      <td>0.00 (n=201)   0.00 (n=97)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>dhfr triple mutant</th>\n",
              "      <td>SP (treatment)</td>\n",
              "      <td>0.00 (n=44)   0.00 (n=114)</td>\n",
              "      <td>0.76 (n=2425)   0.83 (n=2657)</td>\n",
              "      <td>0.83 (n=319)   0.89 (n=121)</td>\n",
              "      <td>0.95 (n=82)   0.28 (n=85)</td>\n",
              "      <td>0.87 (n=1235)   1.00 (n=18)</td>\n",
              "      <td>nan (n=0)   0.01 (n=230)</td>\n",
              "      <td>0.43 (n=65)   0.46 (n=1147)</td>\n",
              "      <td>0.89 (n=1123)   0.81 (n=710)</td>\n",
              "      <td>0.91 (n=1390)   0.86 (n=2223)</td>\n",
              "      <td>0.00 (n=201)   0.00 (n=138)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>dhfr and dhps sextuple mutant</th>\n",
              "      <td>SP (IPTp)</td>\n",
              "      <td>0.00 (n=44)   0.00 (n=114)</td>\n",
              "      <td>0.00 (n=2898)   0.00 (n=3183)</td>\n",
              "      <td>0.01 (n=379)   0.02 (n=131)</td>\n",
              "      <td>0.00 (n=84)   0.05 (n=86)</td>\n",
              "      <td>0.10 (n=1271)   0.00 (n=19)</td>\n",
              "      <td>nan (n=0)   0.00 (n=233)</td>\n",
              "      <td>0.19 (n=68)   0.13 (n=1149)</td>\n",
              "      <td>0.81 (n=981)   0.78 (n=676)</td>\n",
              "      <td>0.18 (n=1037)   0.09 (n=1016)</td>\n",
              "      <td>0.00 (n=201)   0.00 (n=140)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>kelch13 and mdr1</th>\n",
              "      <td>AS-MQ</td>\n",
              "      <td>0.00 (n=44)   0.00 (n=114)</td>\n",
              "      <td>0.00 (n=2848)   0.00 (n=2919)</td>\n",
              "      <td>0.00 (n=385)   0.00 (n=134)</td>\n",
              "      <td>0.00 (n=86)   0.00 (n=64)</td>\n",
              "      <td>0.00 (n=1356)   0.00 (n=20)</td>\n",
              "      <td>nan (n=0)   0.00 (n=203)</td>\n",
              "      <td>0.00 (n=78)   0.00 (n=1276)</td>\n",
              "      <td>0.15 (n=1121)   0.03 (n=677)</td>\n",
              "      <td>0.09 (n=1338)   0.01 (n=2151)</td>\n",
              "      <td>0.00 (n=201)   0.00 (n=123)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>kelch13 and plasmepsin 2-3</th>\n",
              "      <td>DHA-PPQ</td>\n",
              "      <td>0.00 (n=44)   0.00 (n=114)</td>\n",
              "      <td>0.00 (n=2839)   0.00 (n=2836)</td>\n",
              "      <td>0.00 (n=386)   0.00 (n=134)</td>\n",
              "      <td>0.00 (n=86)   0.00 (n=62)</td>\n",
              "      <td>0.00 (n=1350)   0.00 (n=20)</td>\n",
              "      <td>nan (n=0)   0.00 (n=199)</td>\n",
              "      <td>0.00 (n=78)   0.00 (n=1277)</td>\n",
              "      <td>0.00 (n=1150)   0.00 (n=679)</td>\n",
              "      <td>0.16 (n=1302)   0.45 (n=2134)</td>\n",
              "      <td>0.00 (n=201)   0.00 (n=115)</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6e5083df-1fa0-45dd-a1fe-1dbd4364f0b2')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-6e5083df-1fa0-45dd-a1fe-1dbd4364f0b2 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-6e5083df-1fa0-45dd-a1fe-1dbd4364f0b2');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-0f4c9cb3-fa6a-4261-b480-b8e143b22dcc\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-0f4c9cb3-fa6a-4261-b480-b8e143b22dcc')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-0f4c9cb3-fa6a-4261-b480-b8e143b22dcc button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "                              Associated with resistance to  \\\n",
              "Marker                                                        \n",
              "crt 76T                                         Chloroquine   \n",
              "dhfr 108N                                     Pyrimethamine   \n",
              "dhps 437G                                       Sulfadoxine   \n",
              "mdr1 2+ copies                                   Mefloquine   \n",
              "kelch13 WHO list                                Artemisinin   \n",
              "plasmepsin 2-3 2+ copies                        Piperaquine   \n",
              "dhfr triple mutant                           SP (treatment)   \n",
              "dhfr and dhps sextuple mutant                     SP (IPTp)   \n",
              "kelch13 and mdr1                                      AS-MQ   \n",
              "kelch13 and plasmepsin 2-3                          DHA-PPQ   \n",
              "\n",
              "                                            South America  \\\n",
              "Marker                                                      \n",
              "crt 76T                        1.00 (n=44)   1.00 (n=111)   \n",
              "dhfr 108N                      0.93 (n=44)   0.53 (n=110)   \n",
              "dhps 437G                      0.35 (n=43)   0.69 (n=111)   \n",
              "mdr1 2+ copies                 0.00 (n=44)   0.00 (n=114)   \n",
              "kelch13 WHO list               0.00 (n=44)   0.00 (n=114)   \n",
              "plasmepsin 2-3 2+ copies       0.00 (n=44)   0.00 (n=114)   \n",
              "dhfr triple mutant             0.00 (n=44)   0.00 (n=114)   \n",
              "dhfr and dhps sextuple mutant  0.00 (n=44)   0.00 (n=114)   \n",
              "kelch13 and mdr1               0.00 (n=44)   0.00 (n=114)   \n",
              "kelch13 and plasmepsin 2-3     0.00 (n=44)   0.00 (n=114)   \n",
              "\n",
              "                                               Africa - West  \\\n",
              "Marker                                                         \n",
              "crt 76T                        0.42 (n=2500)   0.19 (n=3008)   \n",
              "dhfr 108N                      0.85 (n=2547)   0.92 (n=2902)   \n",
              "dhps 437G                      0.75 (n=2513)   0.84 (n=2866)   \n",
              "mdr1 2+ copies                 0.00 (n=2735)   0.00 (n=2644)   \n",
              "kelch13 WHO list               0.00 (n=2781)   0.00 (n=2668)   \n",
              "plasmepsin 2-3 2+ copies       0.00 (n=2769)   0.00 (n=2558)   \n",
              "dhfr triple mutant             0.76 (n=2425)   0.83 (n=2657)   \n",
              "dhfr and dhps sextuple mutant  0.00 (n=2898)   0.00 (n=3183)   \n",
              "kelch13 and mdr1               0.00 (n=2848)   0.00 (n=2919)   \n",
              "kelch13 and plasmepsin 2-3     0.00 (n=2839)   0.00 (n=2836)   \n",
              "\n",
              "                                          Africa - Central  \\\n",
              "Marker                                                       \n",
              "crt 76T                        0.65 (n=292)   0.52 (n=105)   \n",
              "dhfr 108N                      1.00 (n=383)   0.99 (n=134)   \n",
              "dhps 437G                      0.97 (n=371)   0.96 (n=130)   \n",
              "mdr1 2+ copies                 0.00 (n=349)   0.00 (n=129)   \n",
              "kelch13 WHO list               0.00 (n=374)   0.00 (n=131)   \n",
              "plasmepsin 2-3 2+ copies       0.00 (n=385)   0.00 (n=134)   \n",
              "dhfr triple mutant             0.83 (n=319)   0.89 (n=121)   \n",
              "dhfr and dhps sextuple mutant  0.01 (n=379)   0.02 (n=131)   \n",
              "kelch13 and mdr1               0.00 (n=385)   0.00 (n=134)   \n",
              "kelch13 and plasmepsin 2-3     0.00 (n=386)   0.00 (n=134)   \n",
              "\n",
              "                                      Africa - Northeast  \\\n",
              "Marker                                                     \n",
              "crt 76T                        0.19 (n=78)   0.61 (n=79)   \n",
              "dhfr 108N                      0.99 (n=86)   0.96 (n=84)   \n",
              "dhps 437G                      0.99 (n=85)   0.64 (n=77)   \n",
              "mdr1 2+ copies                 0.00 (n=86)   0.00 (n=37)   \n",
              "kelch13 WHO list               0.00 (n=84)   0.00 (n=60)   \n",
              "plasmepsin 2-3 2+ copies       0.00 (n=86)   0.00 (n=34)   \n",
              "dhfr triple mutant             0.95 (n=82)   0.28 (n=85)   \n",
              "dhfr and dhps sextuple mutant  0.00 (n=84)   0.05 (n=86)   \n",
              "kelch13 and mdr1               0.00 (n=86)   0.00 (n=64)   \n",
              "kelch13 and plasmepsin 2-3     0.00 (n=86)   0.00 (n=62)   \n",
              "\n",
              "                                             Africa - East  \\\n",
              "Marker                                                       \n",
              "crt 76T                        0.18 (n=1244)   0.05 (n=19)   \n",
              "dhfr 108N                      0.99 (n=1346)   1.00 (n=19)   \n",
              "dhps 437G                      0.92 (n=1265)   1.00 (n=19)   \n",
              "mdr1 2+ copies                 0.00 (n=1339)   0.00 (n=14)   \n",
              "kelch13 WHO list               0.00 (n=1341)   0.00 (n=20)   \n",
              "plasmepsin 2-3 2+ copies       0.00 (n=1337)   0.00 (n=19)   \n",
              "dhfr triple mutant             0.87 (n=1235)   1.00 (n=18)   \n",
              "dhfr and dhps sextuple mutant  0.10 (n=1271)   0.00 (n=19)   \n",
              "kelch13 and mdr1               0.00 (n=1356)   0.00 (n=20)   \n",
              "kelch13 and plasmepsin 2-3     0.00 (n=1350)   0.00 (n=20)   \n",
              "\n",
              "                                    Asia - South - East  \\\n",
              "Marker                                                    \n",
              "crt 76T                        nan (n=0)   0.31 (n=217)   \n",
              "dhfr 108N                      nan (n=0)   0.64 (n=201)   \n",
              "dhps 437G                      nan (n=0)   0.08 (n=220)   \n",
              "mdr1 2+ copies                 nan (n=0)   0.00 (n=164)   \n",
              "kelch13 WHO list               nan (n=0)   0.00 (n=189)   \n",
              "plasmepsin 2-3 2+ copies       nan (n=0)   0.00 (n=172)   \n",
              "dhfr triple mutant             nan (n=0)   0.01 (n=230)   \n",
              "dhfr and dhps sextuple mutant  nan (n=0)   0.00 (n=233)   \n",
              "kelch13 and mdr1               nan (n=0)   0.00 (n=203)   \n",
              "kelch13 and plasmepsin 2-3     nan (n=0)   0.00 (n=199)   \n",
              "\n",
              "                                   Asia - South - Far East  \\\n",
              "Marker                                                       \n",
              "crt 76T                        0.93 (n=74)   0.94 (n=1252)   \n",
              "dhfr 108N                      1.00 (n=77)   1.00 (n=1292)   \n",
              "dhps 437G                      0.97 (n=63)   0.89 (n=1228)   \n",
              "mdr1 2+ copies                 0.00 (n=77)   0.00 (n=1191)   \n",
              "kelch13 WHO list               0.00 (n=78)   0.00 (n=1263)   \n",
              "plasmepsin 2-3 2+ copies       0.00 (n=78)   0.00 (n=1194)   \n",
              "dhfr triple mutant             0.43 (n=65)   0.46 (n=1147)   \n",
              "dhfr and dhps sextuple mutant  0.19 (n=68)   0.13 (n=1149)   \n",
              "kelch13 and mdr1               0.00 (n=78)   0.00 (n=1276)   \n",
              "kelch13 and plasmepsin 2-3     0.00 (n=78)   0.00 (n=1277)   \n",
              "\n",
              "                                    Asia - Southeast - West  \\\n",
              "Marker                                                        \n",
              "crt 76T                        1.00 (n=1158)   0.99 (n=713)   \n",
              "dhfr 108N                      1.00 (n=1160)   1.00 (n=716)   \n",
              "dhps 437G                      1.00 (n=1160)   0.99 (n=715)   \n",
              "mdr1 2+ copies                 0.43 (n=1128)   0.06 (n=654)   \n",
              "kelch13 WHO list               0.28 (n=1103)   0.49 (n=665)   \n",
              "plasmepsin 2-3 2+ copies       0.00 (n=1139)   0.00 (n=651)   \n",
              "dhfr triple mutant             0.89 (n=1123)   0.81 (n=710)   \n",
              "dhfr and dhps sextuple mutant   0.81 (n=981)   0.78 (n=676)   \n",
              "kelch13 and mdr1               0.15 (n=1121)   0.03 (n=677)   \n",
              "kelch13 and plasmepsin 2-3     0.00 (n=1150)   0.00 (n=679)   \n",
              "\n",
              "                                     Asia - Southeast - East  \\\n",
              "Marker                                                         \n",
              "crt 76T                        0.94 (n=1427)   0.95 (n=2232)   \n",
              "dhfr 108N                      0.99 (n=1438)   0.99 (n=2240)   \n",
              "dhps 437G                      0.88 (n=1377)   0.81 (n=2226)   \n",
              "mdr1 2+ copies                 0.12 (n=1334)   0.01 (n=2122)   \n",
              "kelch13 WHO list               0.48 (n=1290)   0.64 (n=2179)   \n",
              "plasmepsin 2-3 2+ copies       0.19 (n=1296)   0.49 (n=2128)   \n",
              "dhfr triple mutant             0.91 (n=1390)   0.86 (n=2223)   \n",
              "dhfr and dhps sextuple mutant  0.18 (n=1037)   0.09 (n=1016)   \n",
              "kelch13 and mdr1               0.09 (n=1338)   0.01 (n=2151)   \n",
              "kelch13 and plasmepsin 2-3     0.16 (n=1302)   0.45 (n=2134)   \n",
              "\n",
              "                                      Oceania - New Guinea  \n",
              "Marker                                                      \n",
              "crt 76T                        0.99 (n=195)   0.91 (n=138)  \n",
              "dhfr 108N                      0.99 (n=200)   0.98 (n=133)  \n",
              "dhps 437G                      0.61 (n=197)   0.81 (n=136)  \n",
              "mdr1 2+ copies                 0.01 (n=198)   0.02 (n=116)  \n",
              "kelch13 WHO list               0.00 (n=199)   0.02 (n=106)  \n",
              "plasmepsin 2-3 2+ copies        0.00 (n=201)   0.00 (n=97)  \n",
              "dhfr triple mutant             0.00 (n=201)   0.00 (n=138)  \n",
              "dhfr and dhps sextuple mutant  0.00 (n=201)   0.00 (n=140)  \n",
              "kelch13 and mdr1               0.00 (n=201)   0.00 (n=123)  \n",
              "kelch13 and plasmepsin 2-3     0.00 (n=201)   0.00 (n=115)  "
            ]
          },
          "execution_count": 19,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Create a table summarizing proportions and counts of resistance for each drug in two year groups by population\n",
        "\n",
        "df_drm_table_with_n = (\n",
        "    df_all_sample_metadata\n",
        "    .groupby('Population')\n",
        "    .apply(my_agg)\n",
        "    .rename_axis(None)\n",
        "    .transpose()\n",
        "    .loc[:, populations.keys()]\n",
        "    .reset_index()\n",
        ")\n",
        "\n",
        "# Customize the index to include gene and mutation information for each drug\n",
        "df_drm_table_with_n.index = [\"%s %s\" % (drm_dict[drug]['gene'], drm_dict[drug]['mutation']) for drug in drm_dict]\n",
        "df_drm_table_with_n.index.names = ['Marker']\n",
        "\n",
        "# Rename the columns to indicate their association with drug resistance and sample counts\n",
        "df_drm_table_with_n.rename(columns={'index': 'Associated with resistance to'}, inplace=True)\n",
        "\n",
        "# Update column names to include population names\n",
        "for population in populations:\n",
        "    new_column_name = populations[population]\n",
        "    df_drm_table_with_n.rename(columns={population: new_column_name}, inplace=True)\n",
        "\n",
        "# Display the final drug resistance table with sample counts\n",
        "df_drm_table_with_n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "K6PAWgMk70l7"
      },
      "source": [
        "Save table as:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "dVDvCqt07ubK"
      },
      "outputs": [],
      "source": [
        "df_drm_table_with_n.to_excel(\"DRM_before_after_table.xlsx\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-pPLQHK175Dd"
      },
      "source": [
        "## CRT 72-76 haplotypes table"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zAaR3IdbiBjH"
      },
      "source": [
        "The final table will feature crt mutations in codons 72-76 which were associated with a particular resistance profile, with the crt 76T allele a very reliable marker of chloroquine resistance.\n",
        "\n",
        "Let's look at the distribution of haplotypes with T and K alleles."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Hkc3sWQe79mK",
        "outputId": "390cc3c2-db6a-42e1-f4d2-b0781adee2ed"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "crt_76[K]  crt_72-76   \n",
              "-          -                 26\n",
              "K          -                  3\n",
              "           CVMNK           5784\n",
              "           CVMNK,CVMKK        1\n",
              "           CVMNK,YVMNK        1\n",
              "K,P        !,CVIEP            1\n",
              "K,T        CVIET,CVMNK        7\n",
              "           CVMNK,CVIDT        9\n",
              "           CVMNK,CVIDT*       1\n",
              "           CVMNK,CVIET      410\n",
              "           CVMNK,CVIET*       5\n",
              "           CVMNK,CVMNT        1\n",
              "           CVMNK,SVMNT*       4\n",
              "T          -                  2\n",
              "           CIIET,CVIET        1\n",
              "           CVIDT            535\n",
              "           CVIDT,!            1\n",
              "           CVIDT,CVIET       33\n",
              "           CVIDT,CVIET*      16\n",
              "           CVIET           8252\n",
              "           CVIET,!            1\n",
              "           CVIET,CIIET        1\n",
              "           CVIET,CVI*         2\n",
              "           CVIET,CVIDT       33\n",
              "           CVIET,CVIDT*      15\n",
              "           CVIET,CVMNT        1\n",
              "           CVIET,SVMNT*       1\n",
              "           CVIET,YVIET        3\n",
              "           CVMET            105\n",
              "           CVMET,CVMNT        2\n",
              "           CVMNT             33\n",
              "           CVMNT,CVMET        2\n",
              "           SVMNT            337\n",
              "           SVMNT,CVMET*       1\n",
              "           SVMNT,CVMNT        1\n",
              "           YVIET              2\n",
              "T,K        CVIDK,CVIET        1\n",
              "           CVIDT,CVMNK        7\n",
              "           CVIET,!            3\n",
              "           CVIET,CVIDK        1\n",
              "           CVIET,CVIEK        4\n",
              "           CVIET,CVMNK      519\n",
              "           CVIET,CVMNK*      17\n",
              "           CVMNK,CVIET       12\n",
              "           CVMNT,CVI*         1\n",
              "           SVMNT,CVMNK*       4\n",
              "T,Q*       CVIET,!*           1\n",
              "dtype: int64"
            ]
          },
          "execution_count": 20,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Rename haplotype column for a better representation\n",
        "df_all_sample_metadata['crt_72-76'] = df_all_sample_metadata['crt_72-76[CVMNK]']\n",
        "\n",
        "# Group and count haplotypes\n",
        "df_all_sample_metadata.groupby(['crt_76[K]', 'crt_72-76']).size()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7LTOnT1mAxrK"
      },
      "source": [
        "We can see that there are heterozygous and ambiguous calls which we will exclude from the frequency estimation. Only exception is 72-76T haplotypes that contains heterozygous and ambiguous calls, since 76T call is reliable marker and they represent a bigger group."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "dHhBLgvisLO0",
        "outputId": "049f1759-27c8-4cea-ca17-e8d531d5d289"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "crt_76[K]  crt_72-76\n",
              "K          CVMNK        5784\n",
              "T          CVIDT         535\n",
              "           CVIET        8252\n",
              "           CVMET         105\n",
              "           CVMNT          33\n",
              "           Other T       116\n",
              "           SVMNT         337\n",
              "           YVIET           2\n",
              "dtype: int64"
            ]
          },
          "execution_count": 21,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Set a group name for 76T haplotypes that contains ambiguous characters\n",
        "df_all_sample_metadata.loc[\n",
        "    df_all_sample_metadata['crt_72-76'].str.contains(',|-|X|\\?') & ( df_all_sample_metadata['crt_76[K]'] == 'T' ),\n",
        "    'crt_72-76'\n",
        "] = 'Other T'\n",
        "\n",
        "# Exclude het calls and haplotypes that contains ambiguous characters\n",
        "analysis_samples = (\n",
        "    ~df_all_sample_metadata['crt_72-76'].astype(str).str.contains(',|-|X|\\?'))\n",
        "\n",
        "# Look at the counts of haplotypes with 76K and 76T alleles\n",
        "df_all_sample_metadata.loc[analysis_samples].groupby(['crt_76[K]', 'crt_72-76']).size()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "SjsjzZRjozZp"
      },
      "source": [
        "Now we can find the distribution of haplotypes in the populations."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 332
        },
        "id": "dBJNRPDfsYwL",
        "outputId": "9e7ccf86-c6f3-45ba-8850-be4576327bfd"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-02745862-3b76-498b-8cc7-bac6a867bad8\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Population</th>\n",
              "      <th>SA</th>\n",
              "      <th>AF-W</th>\n",
              "      <th>AF-C</th>\n",
              "      <th>AF-NE</th>\n",
              "      <th>AF-E</th>\n",
              "      <th>AS-S-E</th>\n",
              "      <th>AS-S-FE</th>\n",
              "      <th>AS-SE-W</th>\n",
              "      <th>AS-SE-E</th>\n",
              "      <th>OC-NG</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>crt_72-76</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>CVIDT</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>535</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CVIET</th>\n",
              "      <td>0</td>\n",
              "      <td>1622</td>\n",
              "      <td>244</td>\n",
              "      <td>63</td>\n",
              "      <td>329</td>\n",
              "      <td>60</td>\n",
              "      <td>1241</td>\n",
              "      <td>1858</td>\n",
              "      <td>2835</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CVMET</th>\n",
              "      <td>105</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CVMNK</th>\n",
              "      <td>0</td>\n",
              "      <td>4035</td>\n",
              "      <td>152</td>\n",
              "      <td>94</td>\n",
              "      <td>1059</td>\n",
              "      <td>150</td>\n",
              "      <td>83</td>\n",
              "      <td>10</td>\n",
              "      <td>187</td>\n",
              "      <td>14</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CVMNT</th>\n",
              "      <td>33</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Other T</th>\n",
              "      <td>5</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>3</td>\n",
              "      <td>105</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>SVMNT</th>\n",
              "      <td>12</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>7</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>318</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>YVIET</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-02745862-3b76-498b-8cc7-bac6a867bad8')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-02745862-3b76-498b-8cc7-bac6a867bad8 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-02745862-3b76-498b-8cc7-bac6a867bad8');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-d9de668b-bfd4-4445-98f3-299a077710bd\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-d9de668b-bfd4-4445-98f3-299a077710bd')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-d9de668b-bfd4-4445-98f3-299a077710bd button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "Population   SA  AF-W  AF-C  AF-NE  AF-E  AS-S-E  AS-S-FE  AS-SE-W  AS-SE-E  \\\n",
              "crt_72-76                                                                     \n",
              "CVIDT         0     0     0      0     0       0        0        0      535   \n",
              "CVIET         0  1622   244     63   329      60     1241     1858     2835   \n",
              "CVMET       105     0     0      0     0       0        0        0        0   \n",
              "CVMNK         0  4035   152     94  1059     150       83       10      187   \n",
              "CVMNT        33     0     0      0     0       0        0        0        0   \n",
              "Other T       5     0     0      0     0       0        2        3      105   \n",
              "SVMNT        12     0     0      0     0       7        0        0        0   \n",
              "YVIET         0     0     0      0     0       0        0        0        2   \n",
              "\n",
              "Population  OC-NG  \n",
              "crt_72-76          \n",
              "CVIDT           0  \n",
              "CVIET           0  \n",
              "CVMET           0  \n",
              "CVMNK          14  \n",
              "CVMNT           0  \n",
              "Other T         1  \n",
              "SVMNT         318  \n",
              "YVIET           0  "
            ]
          },
          "execution_count": 22,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Create a pivot table with haplotypes as index, populations as columns, and counts as values\n",
        "df_crt = pd.pivot_table(\n",
        "    df_all_sample_metadata.loc[analysis_samples],\n",
        "    index='crt_72-76',\n",
        "    columns='Population',\n",
        "    values='Study',\n",
        "    aggfunc=len,\n",
        "    fill_value=0\n",
        ").loc[:, populations.keys()]\n",
        "\n",
        "df_crt"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "luQHgSttp943"
      },
      "source": [
        "As a last step we want to sort the order of the haplotypes. The wild type haplotype will be shown first, followed by other haplotypes sorted by penalty score and then by total number of haplotypes in descending order.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 23,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 332
        },
        "id": "HuvXqbebszXs",
        "outputId": "06c47605-c38f-42d4-fc64-58dfa12aa586"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-e9ddc884-7540-4a0a-888d-30eed8800180\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Population</th>\n",
              "      <th>SA</th>\n",
              "      <th>AF-W</th>\n",
              "      <th>AF-C</th>\n",
              "      <th>AF-NE</th>\n",
              "      <th>AF-E</th>\n",
              "      <th>AS-S-E</th>\n",
              "      <th>AS-S-FE</th>\n",
              "      <th>AS-SE-W</th>\n",
              "      <th>AS-SE-E</th>\n",
              "      <th>OC-NG</th>\n",
              "      <th>row_total</th>\n",
              "      <th>hamming_dist</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>crt_72-76</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>CVMNK</th>\n",
              "      <td>0.000</td>\n",
              "      <td>0.713</td>\n",
              "      <td>0.384</td>\n",
              "      <td>0.599</td>\n",
              "      <td>0.763</td>\n",
              "      <td>0.691</td>\n",
              "      <td>0.063</td>\n",
              "      <td>0.005</td>\n",
              "      <td>0.051</td>\n",
              "      <td>0.042</td>\n",
              "      <td>0.381</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CVMNT</th>\n",
              "      <td>0.213</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.002</td>\n",
              "      <td>0.04</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>SVMNT</th>\n",
              "      <td>0.077</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.032</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.955</td>\n",
              "      <td>0.022</td>\n",
              "      <td>0.08</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CVMET</th>\n",
              "      <td>0.677</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.007</td>\n",
              "      <td>0.08</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CVIET</th>\n",
              "      <td>0.000</td>\n",
              "      <td>0.287</td>\n",
              "      <td>0.616</td>\n",
              "      <td>0.401</td>\n",
              "      <td>0.237</td>\n",
              "      <td>0.276</td>\n",
              "      <td>0.936</td>\n",
              "      <td>0.993</td>\n",
              "      <td>0.774</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.544</td>\n",
              "      <td>0.12</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CVIDT</th>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.146</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.035</td>\n",
              "      <td>0.12</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>YVIET</th>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.001</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.16</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Other T</th>\n",
              "      <td>0.032</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.002</td>\n",
              "      <td>0.002</td>\n",
              "      <td>0.029</td>\n",
              "      <td>0.003</td>\n",
              "      <td>0.008</td>\n",
              "      <td>0.40</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-e9ddc884-7540-4a0a-888d-30eed8800180')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-e9ddc884-7540-4a0a-888d-30eed8800180 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-e9ddc884-7540-4a0a-888d-30eed8800180');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-35dd59c9-43db-4c68-9b9c-7a7ca354aae6\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-35dd59c9-43db-4c68-9b9c-7a7ca354aae6')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-35dd59c9-43db-4c68-9b9c-7a7ca354aae6 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "Population     SA   AF-W   AF-C  AF-NE   AF-E  AS-S-E  AS-S-FE  AS-SE-W  \\\n",
              "crt_72-76                                                                 \n",
              "CVMNK       0.000  0.713  0.384  0.599  0.763   0.691    0.063    0.005   \n",
              "CVMNT       0.213  0.000  0.000  0.000  0.000   0.000    0.000    0.000   \n",
              "SVMNT       0.077  0.000  0.000  0.000  0.000   0.032    0.000    0.000   \n",
              "CVMET       0.677  0.000  0.000  0.000  0.000   0.000    0.000    0.000   \n",
              "CVIET       0.000  0.287  0.616  0.401  0.237   0.276    0.936    0.993   \n",
              "CVIDT       0.000  0.000  0.000  0.000  0.000   0.000    0.000    0.000   \n",
              "YVIET       0.000  0.000  0.000  0.000  0.000   0.000    0.000    0.000   \n",
              "Other T     0.032  0.000  0.000  0.000  0.000   0.000    0.002    0.002   \n",
              "\n",
              "Population  AS-SE-E  OC-NG  row_total  hamming_dist  \n",
              "crt_72-76                                            \n",
              "CVMNK         0.051  0.042      0.381          0.00  \n",
              "CVMNT         0.000  0.000      0.002          0.04  \n",
              "SVMNT         0.000  0.955      0.022          0.08  \n",
              "CVMET         0.000  0.000      0.007          0.08  \n",
              "CVIET         0.774  0.000      0.544          0.12  \n",
              "CVIDT         0.146  0.000      0.035          0.12  \n",
              "YVIET         0.001  0.000      0.000          0.16  \n",
              "Other T       0.029  0.003      0.008          0.40  "
            ]
          },
          "execution_count": 23,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Add a row_total column representing the total count for each haplotype\n",
        "df_crt['row_total'] = df_crt.sum(1)\n",
        "\n",
        "# Sort the table based on penalty score\n",
        "df_crt['hamming_dist'] = [10 if x in ['Other T'] else scipy.spatial.distance.hamming(list('CVMNK'), list(str(x))) * len('CVMNK') for x in df_crt.index]\n",
        "sort_order = df_crt.sort_values(['hamming_dist', 'row_total'], ascending=[True, False]).index\n",
        "\n",
        "# Get the numbers of samples for each population\n",
        "numbers_of_samples = df_crt.sum(0)\n",
        "\n",
        "# Return proportions\n",
        "df_crt_final = round(df_crt.loc[sort_order].divide(list(numbers_of_samples)),3)\n",
        "df_crt_final"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "RIFYrQKksRvh"
      },
      "source": [
        "We will add the population sizes to the column names and remove the columns used for sorting."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 24,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 332
        },
        "id": "HkUdJPvstv1g",
        "outputId": "7da361c6-4ccc-4101-d895-76c8cc06bd98"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-20ce915e-eee0-4077-b2b7-cc27c4bd817b\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Population</th>\n",
              "      <th>SA (n=155)</th>\n",
              "      <th>AF-W (n=5657)</th>\n",
              "      <th>AF-C (n=396)</th>\n",
              "      <th>AF-NE (n=157)</th>\n",
              "      <th>AF-E (n=1388)</th>\n",
              "      <th>AS-S-E (n=217)</th>\n",
              "      <th>AS-S-FE (n=1326)</th>\n",
              "      <th>AS-SE-W (n=1871)</th>\n",
              "      <th>AS-SE-E (n=3664)</th>\n",
              "      <th>OC-NG (n=333)</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>crt_72-76</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>CVMNK</th>\n",
              "      <td>0.000</td>\n",
              "      <td>0.713</td>\n",
              "      <td>0.384</td>\n",
              "      <td>0.599</td>\n",
              "      <td>0.763</td>\n",
              "      <td>0.691</td>\n",
              "      <td>0.063</td>\n",
              "      <td>0.005</td>\n",
              "      <td>0.051</td>\n",
              "      <td>0.042</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CVMNT</th>\n",
              "      <td>0.213</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>SVMNT</th>\n",
              "      <td>0.077</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.032</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.955</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CVMET</th>\n",
              "      <td>0.677</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CVIET</th>\n",
              "      <td>0.000</td>\n",
              "      <td>0.287</td>\n",
              "      <td>0.616</td>\n",
              "      <td>0.401</td>\n",
              "      <td>0.237</td>\n",
              "      <td>0.276</td>\n",
              "      <td>0.936</td>\n",
              "      <td>0.993</td>\n",
              "      <td>0.774</td>\n",
              "      <td>0.000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CVIDT</th>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.146</td>\n",
              "      <td>0.000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>YVIET</th>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.001</td>\n",
              "      <td>0.000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Other T</th>\n",
              "      <td>0.032</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.000</td>\n",
              "      <td>0.002</td>\n",
              "      <td>0.002</td>\n",
              "      <td>0.029</td>\n",
              "      <td>0.003</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-20ce915e-eee0-4077-b2b7-cc27c4bd817b')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-20ce915e-eee0-4077-b2b7-cc27c4bd817b button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-20ce915e-eee0-4077-b2b7-cc27c4bd817b');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-da480244-6d61-41c4-8fa1-54d0ef9c0627\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-da480244-6d61-41c4-8fa1-54d0ef9c0627')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-da480244-6d61-41c4-8fa1-54d0ef9c0627 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "Population  SA (n=155)  AF-W (n=5657)  AF-C (n=396)  AF-NE (n=157)  \\\n",
              "crt_72-76                                                            \n",
              "CVMNK            0.000          0.713         0.384          0.599   \n",
              "CVMNT            0.213          0.000         0.000          0.000   \n",
              "SVMNT            0.077          0.000         0.000          0.000   \n",
              "CVMET            0.677          0.000         0.000          0.000   \n",
              "CVIET            0.000          0.287         0.616          0.401   \n",
              "CVIDT            0.000          0.000         0.000          0.000   \n",
              "YVIET            0.000          0.000         0.000          0.000   \n",
              "Other T          0.032          0.000         0.000          0.000   \n",
              "\n",
              "Population  AF-E (n=1388)  AS-S-E (n=217)  AS-S-FE (n=1326)  AS-SE-W (n=1871)  \\\n",
              "crt_72-76                                                                       \n",
              "CVMNK               0.763           0.691             0.063             0.005   \n",
              "CVMNT               0.000           0.000             0.000             0.000   \n",
              "SVMNT               0.000           0.032             0.000             0.000   \n",
              "CVMET               0.000           0.000             0.000             0.000   \n",
              "CVIET               0.237           0.276             0.936             0.993   \n",
              "CVIDT               0.000           0.000             0.000             0.000   \n",
              "YVIET               0.000           0.000             0.000             0.000   \n",
              "Other T             0.000           0.000             0.002             0.002   \n",
              "\n",
              "Population  AS-SE-E (n=3664)  OC-NG (n=333)  \n",
              "crt_72-76                                    \n",
              "CVMNK                  0.051          0.042  \n",
              "CVMNT                  0.000          0.000  \n",
              "SVMNT                  0.000          0.955  \n",
              "CVMET                  0.000          0.000  \n",
              "CVIET                  0.774          0.000  \n",
              "CVIDT                  0.146          0.000  \n",
              "YVIET                  0.001          0.000  \n",
              "Other T                0.029          0.003  "
            ]
          },
          "execution_count": 24,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Create final column names with population names and sample sizes\n",
        "final_column_names = collections.OrderedDict()\n",
        "for column_name in list(df_crt_final):\n",
        "    final_column_names[column_name] = \"%s (n=%d)\" % (column_name, numbers_of_samples[column_name])\n",
        "\n",
        "# Rename columns in the DataFrame\n",
        "df_crt_final = df_crt_final.rename(columns=final_column_names)\n",
        "\n",
        "# Drop row_total and hamming_dist columns\n",
        "df_crt_final = df_crt_final.drop(['row_total (n=15164)', 'hamming_dist (n=25)'], axis=1)\n",
        "df_crt_final"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2s2iVViw8Rug"
      },
      "source": [
        "Save the table as:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "bWcXWY6q8GVJ"
      },
      "outputs": [],
      "source": [
        "df_crt.to_excel(\"table_crt_haplotypes.xlsx\")"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "toc_visible": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
