
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Summarise hrp2 and hrp3 Deletions &#8212; MalariaGEN parasite data user guide</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Visualise hrp2 and hrp3 Deletions" href="visualise_hrp2_hrp3_deletions.html" />
    <link rel="prev" title="Plot Map of Chloroquine Resistance" href="chloroquine-resistance-map.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">MalariaGEN parasite data user guide</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../landing-page.html">
   MalariaGEN parasite data user guide
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Malaria Parasite Data Access
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../cloud_data_access/parasite-cloud-data-access.html">
   <strong>
    Malaria Parasite Data Access
   </strong>
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Pf7
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../Data_access.html">
   <strong>
    Pf7 Data Access
   </strong>
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../summarise-data.html">
   <strong>
    Pf7 Analysis Guides:
   </strong>
   Summarise Pf7 Data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../summarise_data/sample-summary-table.html">
     Summarise Samples by Administrative Division
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../summarise_data/sample-collection-over-time.html">
     Visualising sample collections over time
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../summarise_data/barplot-samples-by-country.html">
     Barplot of Samples by Country
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../summarise_data/variant-summary-table.html">
     Summarise Genetic Variants
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../summarise_data/global-sample-map.html">
     Plot Global Map of Samples
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../evolution-malaria-control.html">
   <strong>
    Pf7 Analysis Guides:
   </strong>
   Study Evolution Against Malaria Control Measures
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="drug-resistance-summary.html">
     Drug Resistance Frequencies Across Pf7
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="chloroquine-resistance-map.html">
     Plot Map of Chloroquine Resistance
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Summarise
     <em>
      hrp2
     </em>
     and
     <em>
      hrp3
     </em>
     Deletions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="visualise_hrp2_hrp3_deletions.html">
     Visualise
     <em>
      hrp2
     </em>
     and
     <em>
      hrp3
     </em>
     Deletions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="csp-cterm-analysis.html">
     Analyse csp C-Terminal Variation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../popgen.html">
   <strong>
    Pf7 Analysis Guides:
   </strong>
   Population Structure and Selection
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../popgen/PCoA.html">
     Principal Coordinate Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../popgen/geographic-genetic-differentiation.html">
     Geographic patterns of population differentiation and gene flow
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../popgen/neighbour-joining-tree-swga.html">
     Neighbour Joining Tree for sWGA vs. gDNA
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../popgen/proportion-eba175.html">
     Plot
     <em>
      eba175
     </em>
     C Allele Proportions
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../api.html">
   <strong>
    Pf7 API
   </strong>
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Pv4
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../pv4/Data_access.html">
   <strong>
    Pv4 Data Access
   </strong>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../pv4/api.html">
   <strong>
    Pv4 API
   </strong>
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/pf7/evolution_malaria_control/summarise_hrp2_hrp3_deletions.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
                onclick="printPdf(this)" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/malariagen/parasite-data"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/malariagen/parasite-data/issues/new?title=Issue%20on%20page%20%2Fpf7/evolution_malaria_control/summarise_hrp2_hrp3_deletions.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/malariagen/parasite-data/master?urlpath=lab/tree/docs/pf7/evolution_malaria_control/summarise_hrp2_hrp3_deletions.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        <a class="colab-button" href="https://colab.research.google.com/github/malariagen/parasite-data/blob/master/docs/pf7/evolution_malaria_control/summarise_hrp2_hrp3_deletions.ipynb"><button type="button" class="btn btn-secondary topbarbtn"
                title="Launch Colab" data-toggle="tooltip" data-placement="left"><img class="colab-button-logo"
                    src="../../_static/images/logo_colab.png"
                    alt="Interact on Colab">Colab</button></a>
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#access-pf7-data">
   Access Pf7 Data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#hrp2-and-hrp3-deletions">
   <em>
    hrp2
   </em>
   and
   <em>
    hrp3
   </em>
   Deletions
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#dataset-exploration">
   Dataset Exploration
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#summary-by-deletion-types-table">
   Summary by deletion types table
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#save-the-table">
     Save the table
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#summary-by-breakpoints">
   Summary by breakpoints
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     Save the table
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#summary-by-country">
   Summary by country
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id2">
     Save the table
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Summarise hrp2 and hrp3 Deletions</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#access-pf7-data">
   Access Pf7 Data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#hrp2-and-hrp3-deletions">
   <em>
    hrp2
   </em>
   and
   <em>
    hrp3
   </em>
   Deletions
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#dataset-exploration">
   Dataset Exploration
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#summary-by-deletion-types-table">
   Summary by deletion types table
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#save-the-table">
     Save the table
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#summary-by-breakpoints">
   Summary by breakpoints
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     Save the table
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#summary-by-country">
   Summary by country
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id2">
     Save the table
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            
              <div>
                
  <div class="section" id="summarise-hrp2-and-hrp3-deletions">
<h1>Summarise <em>hrp2</em> and <em>hrp3</em> Deletions<a class="headerlink" href="#summarise-hrp2-and-hrp3-deletions" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>This notebook will recreate two supplementary display items from the <a class="reference external" href="https://wellcomeopenresearch.org/articles/8-22">Pf7 paper</a> relating to summarising deletions in <em>hrp2</em> and <em>hrp3</em>:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://figshare.com/ndownloader/files/38478530">Supplementary Table 6</a>: frequency of <em>hrp2</em> and <em>hrp3</em> deletions by country</p></li>
<li><p><a class="reference external" href="https://figshare.com/ndownloader/files/38478530">Supplementary Table 7</a>: summary of <em>hrp2</em> and <em>hrp3</em> deletion breakpoints</p></li>
</ul>
<p><em>hrp2</em> and <em>hrp3</em> are genes located in subtelomeric regions of the genome with very high levels of natural variation. Deletion in those genes can cause failure of rapid diagnostic tests and is therefore important to monitor.</p>
<p>Deletion is a genetic event in which a segment of DNA is entirely removed or missing. In this context, ‘breakpoints’ denote specific locations on the chromosome where such deletions take place.</p>
<p><strong>This notebook should take approximately 1 minute to run.</strong></p>
</div>
<div class="section" id="setup">
<h2>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">¶</a></h2>
<p>Install and import the malariagen Python package:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>!pip install -q --no-warn-conflicts malariagen_data
import malariagen_data
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">148.5/148.5 kB</span> <span class=" -Color -Color-Red">3.1 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">3.1/3.1 MB</span> <span class=" -Color -Color-Red">13.1 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">52.8/52.8 MB</span> <span class=" -Color -Color-Red">7.0 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">10.2/10.2 MB</span> <span class=" -Color -Color-Red">43.8 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">4.0/4.0 MB</span> <span class=" -Color -Color-Red">28.2 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">302.5/302.5 kB</span> <span class=" -Color -Color-Red">14.3 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">138.7/138.7 kB</span> <span class=" -Color -Color-Red">2.6 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">20.9/20.9 MB</span> <span class=" -Color -Color-Red">8.6 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">8.1/8.1 MB</span> <span class=" -Color -Color-Red">65.3 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">206.9/206.9 kB</span> <span class=" -Color -Color-Red">14.9 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
?25h  Preparing metadata (setup.py) ... ?25l?25hdone
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">7.7/7.7 MB</span> <span class=" -Color -Color-Red">22.8 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">1.6/1.6 MB</span> <span class=" -Color -Color-Red">67.2 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
?25h  Building wheel for asciitree (setup.py) ... ?25l?25hdone
</pre></div>
</div>
</div>
</div>
<p>Import required python libraries that are installed at colab by default.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">from</span> <span class="nn">google.colab</span> <span class="kn">import</span> <span class="n">drive</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="access-pf7-data">
<h2>Access Pf7 Data<a class="headerlink" href="#access-pf7-data" title="Permalink to this headline">¶</a></h2>
<p>We use the malariagen data package to load the release data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">release_data</span> <span class="o">=</span> <span class="n">malariagen_data</span><span class="o">.</span><span class="n">Pf7</span><span class="p">()</span>
<span class="n">df_samples</span> <span class="o">=</span> <span class="n">release_data</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="hrp2-and-hrp3-deletions">
<h2><em>hrp2</em> and <em>hrp3</em> Deletions<a class="headerlink" href="#hrp2-and-hrp3-deletions" title="Permalink to this headline">¶</a></h2>
<p>We also need to access Pf7 samples <em>hrp2</em> and <em>hrp3</em> deletion status (including breakpoint and deletion type).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Fetch the details of hrp calls from this MalariaGEN.net file</span>
<span class="n">hrp_calls_fn</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;https://www.malariagen.net/wp-content/uploads/2024/01/hrp_calls_pf7.tsv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Let’s get a quick overview of this dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># print the shape</span>
<span class="nb">print</span><span class="p">(</span><span class="n">hrp_calls_fn</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="c1"># display first 5 rows of the dataset</span>
<span class="n">hrp_calls_fn</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(16203, 7)
</pre></div>
</div>
<div class="output text_html">
  <div id="df-b4e54d1a-81cd-44a3-888b-b652f521460c" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Sample</th>
      <th>HRP2</th>
      <th>HRP3</th>
      <th>HRP2_breakpoint</th>
      <th>HRP3_breakpoint</th>
      <th>HRP2_deletion_type</th>
      <th>HRP3_deletion_type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>FP0008-C</td>
      <td>nodel</td>
      <td>nodel</td>
      <td>-</td>
      <td>-</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>FP0009-C</td>
      <td>nodel</td>
      <td>nodel</td>
      <td>-</td>
      <td>-</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>FP0010-CW</td>
      <td>uncallable</td>
      <td>uncallable</td>
      <td>-</td>
      <td>-</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>FP0011-CW</td>
      <td>uncallable</td>
      <td>uncallable</td>
      <td>-</td>
      <td>-</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>FP0012-CW</td>
      <td>nodel</td>
      <td>nodel</td>
      <td>-</td>
      <td>-</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-b4e54d1a-81cd-44a3-888b-b652f521460c')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-b4e54d1a-81cd-44a3-888b-b652f521460c button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-b4e54d1a-81cd-44a3-888b-b652f521460c');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-37624ef7-5a8e-45e1-ba66-eb0210d5ecbc">
  <button class="colab-df-quickchart" onclick="quickchart('df-37624ef7-5a8e-45e1-ba66-eb0210d5ecbc')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-37624ef7-5a8e-45e1-ba66-eb0210d5ecbc button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
<p>We will continue to explore the dataset in the next section.</p>
</div>
<div class="section" id="dataset-exploration">
<h2>Dataset Exploration<a class="headerlink" href="#dataset-exploration" title="Permalink to this headline">¶</a></h2>
<p>We begin by merging the deletion status information with sample metadata which will allow us to perform geospatial analyses, such as identifying countries or specific regions that exhibit a higher susceptibility to <em>hrp2</em> and <em>hrp3</em> deletions.</p>
<p>One thing to note is that deletion data is exclusively curated for samples that have passed the quality control (QC) criteria. Consequently, the resulting merged dataset will exclusively contain these high-quality samples with deletion information.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Merge df_samples with hrp_calls_fn</span>
<span class="n">df_samples_hrp</span> <span class="o">=</span>  <span class="n">df_samples</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">hrp_calls_fn</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df_samples_hrp</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">df_samples_hrp</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(16203, 23)
</pre></div>
</div>
<div class="output text_html">
  <div id="df-a8bf9750-f96b-44dc-8df8-7ee5ea1459ab" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Sample</th>
      <th>Study</th>
      <th>Country</th>
      <th>Admin level 1</th>
      <th>Country latitude</th>
      <th>Country longitude</th>
      <th>Admin level 1 latitude</th>
      <th>Admin level 1 longitude</th>
      <th>Year</th>
      <th>ENA</th>
      <th>...</th>
      <th>QC pass</th>
      <th>Exclusion reason</th>
      <th>Sample type</th>
      <th>Sample was in Pf6</th>
      <th>HRP2</th>
      <th>HRP3</th>
      <th>HRP2_breakpoint</th>
      <th>HRP3_breakpoint</th>
      <th>HRP2_deletion_type</th>
      <th>HRP3_deletion_type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>FP0008-C</td>
      <td>1147-PF-MR-CONWAY</td>
      <td>Mauritania</td>
      <td>Hodh el Gharbi</td>
      <td>20.265149</td>
      <td>-10.337093</td>
      <td>16.565426</td>
      <td>-9.832345</td>
      <td>2014.0</td>
      <td>ERR1081237</td>
      <td>...</td>
      <td>True</td>
      <td>Analysis_set</td>
      <td>gDNA</td>
      <td>True</td>
      <td>nodel</td>
      <td>nodel</td>
      <td>-</td>
      <td>-</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>FP0009-C</td>
      <td>1147-PF-MR-CONWAY</td>
      <td>Mauritania</td>
      <td>Hodh el Gharbi</td>
      <td>20.265149</td>
      <td>-10.337093</td>
      <td>16.565426</td>
      <td>-9.832345</td>
      <td>2014.0</td>
      <td>ERR1081238</td>
      <td>...</td>
      <td>True</td>
      <td>Analysis_set</td>
      <td>gDNA</td>
      <td>True</td>
      <td>nodel</td>
      <td>nodel</td>
      <td>-</td>
      <td>-</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>FP0010-CW</td>
      <td>1147-PF-MR-CONWAY</td>
      <td>Mauritania</td>
      <td>Hodh el Gharbi</td>
      <td>20.265149</td>
      <td>-10.337093</td>
      <td>16.565426</td>
      <td>-9.832345</td>
      <td>2014.0</td>
      <td>ERR2889621</td>
      <td>...</td>
      <td>True</td>
      <td>Analysis_set</td>
      <td>sWGA</td>
      <td>False</td>
      <td>uncallable</td>
      <td>uncallable</td>
      <td>-</td>
      <td>-</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>3 rows × 23 columns</p>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-a8bf9750-f96b-44dc-8df8-7ee5ea1459ab')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-a8bf9750-f96b-44dc-8df8-7ee5ea1459ab button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-a8bf9750-f96b-44dc-8df8-7ee5ea1459ab');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-69f2dff7-fafa-4c6c-9e9b-f1fde578ec14">
  <button class="colab-df-quickchart" onclick="quickchart('df-69f2dff7-fafa-4c6c-9e9b-f1fde578ec14')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-69f2dff7-fafa-4c6c-9e9b-f1fde578ec14 button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
<p>What are the deletion types and breakpoints in the <em>hrp2</em> gene, and how many samples exhibit these deletions?</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Count and sort deletion types and breakpoints in highest to lowest occurrence</span>
<span class="n">df_samples_hrp</span><span class="p">[</span>
 <span class="n">df_samples_hrp</span><span class="p">[</span><span class="s1">&#39;HRP2&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;del&#39;</span>
<span class="p">]</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;HRP2_deletion_type&#39;</span><span class="p">,</span> <span class="s1">&#39;HRP2_breakpoint&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>HRP2_deletion_type  HRP2_breakpoint    
Telomere healing    Pf3D7_08_v3:1374986    6
                    Pf3D7_08_v3:1374462    5
                    Pf3D7_08_v3:1374932    2
                    Pf3D7_08_v3:1373732    1
                    Pf3D7_08_v3:1374280    1
dtype: int64
</pre></div>
</div>
</div>
</div>
<p>What are the deletion types and breakpoints in the <em>hrp3</em> gene, and how many samples exhibit these deletions?</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Count and sort deletion types and breakpoints in highest to lowest occurrence</span>
<span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
<span class="n">df_samples_hrp</span><span class="p">[</span>
<span class="n">df_samples_hrp</span><span class="p">[</span><span class="s1">&#39;HRP3&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;del&#39;</span>
<span class="p">]</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;HRP3_deletion_type&#39;</span><span class="p">,</span> <span class="s1">&#39;HRP3_breakpoint&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s2">&quot;Samples_with_deletion&quot;</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-d4df894b-bf9f-456c-8bc9-39b0793b9701" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>Samples_with_deletion</th>
    </tr>
    <tr>
      <th>HRP3_deletion_type</th>
      <th>HRP3_breakpoint</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Chromosome 11 recombination</th>
      <th>Pf3D7_13_v3:2800004-2807159</th>
      <td>151</td>
    </tr>
    <tr>
      <th>Chromosome 5 recombination</th>
      <th>Pf3D7_13_v3:2835587-2835612</th>
      <td>21</td>
    </tr>
    <tr>
      <th rowspan="14" valign="top">Telomere healing</th>
      <th>Pf3D7_13_v3:2830952</th>
      <td>7</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2837145</th>
      <td>7</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2837392</th>
      <td>3</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2838654</th>
      <td>2</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2811525</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2812344</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2815249</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2822480</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2823645</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2832080</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2834604</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2835532</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2841024</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2841120</th>
      <td>1</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-d4df894b-bf9f-456c-8bc9-39b0793b9701')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-d4df894b-bf9f-456c-8bc9-39b0793b9701 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-d4df894b-bf9f-456c-8bc9-39b0793b9701');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-a7cae7e7-cacf-4334-9c71-3fb35d228c9a">
  <button class="colab-df-quickchart" onclick="quickchart('df-a7cae7e7-cacf-4334-9c71-3fb35d228c9a')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-a7cae7e7-cacf-4334-9c71-3fb35d228c9a button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
<p>Note that we can expand the previous table with more categories (population, country and admin-level 1).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Group the HRP3 deletions by breakpoint, population, country, and admin level 1.</span>
<span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="n">df_samples_hrp</span><span class="p">[</span>
   <span class="n">df_samples_hrp</span><span class="p">[</span><span class="s1">&#39;HRP3&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;del&#39;</span>
    <span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;HRP3_deletion_type&#39;</span><span class="p">,</span> <span class="s1">&#39;HRP3_breakpoint&#39;</span><span class="p">,</span> <span class="s1">&#39;Population&#39;</span><span class="p">,</span> <span class="s1">&#39;Country&#39;</span><span class="p">,</span> <span class="s1">&#39;Admin level 1&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s2">&quot;Samples_with_deletion&quot;</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-9f6efb9c-ff33-480c-89d2-8570cfbd4503" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th>Samples_with_deletion</th>
    </tr>
    <tr>
      <th>HRP3_deletion_type</th>
      <th>HRP3_breakpoint</th>
      <th>Population</th>
      <th>Country</th>
      <th>Admin level 1</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="17" valign="top">Chromosome 11 recombination</th>
      <th rowspan="17" valign="top">Pf3D7_13_v3:2800004-2807159</th>
      <th rowspan="3" valign="top">AF-NE</th>
      <th rowspan="2" valign="top">Ethiopia</th>
      <th>Amhara</th>
      <td>6</td>
    </tr>
    <tr>
      <th>Oromia</th>
      <td>3</td>
    </tr>
    <tr>
      <th>Sudan</th>
      <th>Kassala</th>
      <td>5</td>
    </tr>
    <tr>
      <th rowspan="5" valign="top">AF-W</th>
      <th rowspan="2" valign="top">Gambia</th>
      <th>Upper River</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Western</th>
      <td>2</td>
    </tr>
    <tr>
      <th>Ghana</th>
      <th>Upper East</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Mali</th>
      <th>Kayes</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Senegal</th>
      <th>Dakar</th>
      <td>6</td>
    </tr>
    <tr>
      <th>AS-S-FE</th>
      <th>Bangladesh</th>
      <th>Chittagong</th>
      <td>1</td>
    </tr>
    <tr>
      <th rowspan="4" valign="top">AS-SE-E</th>
      <th>Cambodia</th>
      <th>Ratanakiri</th>
      <td>3</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">Laos</th>
      <th>Champasak</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Salavan</th>
      <td>13</td>
    </tr>
    <tr>
      <th>Vietnam</th>
      <th>Binh Phuoc</th>
      <td>2</td>
    </tr>
    <tr>
      <th>AS-SE-W</th>
      <th>Thailand</th>
      <th>Tak</th>
      <td>1</td>
    </tr>
    <tr>
      <th>OC-NG</th>
      <th>Indonesia</th>
      <th>Papua</th>
      <td>40</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">SA</th>
      <th>Colombia</th>
      <th>Cauca</th>
      <td>50</td>
    </tr>
    <tr>
      <th>Peru</th>
      <th>Loreto</th>
      <td>15</td>
    </tr>
    <tr>
      <th rowspan="4" valign="top">Chromosome 5 recombination</th>
      <th rowspan="4" valign="top">Pf3D7_13_v3:2835587-2835612</th>
      <th rowspan="4" valign="top">AS-SE-E</th>
      <th rowspan="3" valign="top">Cambodia</th>
      <th>Battambang</th>
      <td>4</td>
    </tr>
    <tr>
      <th>Pailin</th>
      <td>3</td>
    </tr>
    <tr>
      <th>Pursat</th>
      <td>13</td>
    </tr>
    <tr>
      <th>Vietnam</th>
      <th>Binh Phuoc</th>
      <td>1</td>
    </tr>
    <tr>
      <th rowspan="15" valign="top">Telomere healing</th>
      <th>Pf3D7_13_v3:2811525</th>
      <th>AS-S-E</th>
      <th>India</th>
      <th>West Bengal</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2812344</th>
      <th>AF-NE</th>
      <th>Sudan</th>
      <th>Kassala</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2815249</th>
      <th>AF-E</th>
      <th>Tanzania</th>
      <th>Tanga</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2822480</th>
      <th>AF-W</th>
      <th>Ghana</th>
      <th>Brong Ahafo</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2823645</th>
      <th>AF-E</th>
      <th>Kenya</th>
      <th>Kilifi</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2830952</th>
      <th>AS-SE-E</th>
      <th>Cambodia</th>
      <th>Battambang</th>
      <td>7</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2832080</th>
      <th>AF-C</th>
      <th>Democratic Republic of the Congo</th>
      <th>Kinshasa</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2834604</th>
      <th>AS-SE-E</th>
      <th>Vietnam</th>
      <th>Binh Phuoc</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2835532</th>
      <th>AS-SE-W</th>
      <th>Thailand</th>
      <th>Tak</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2837145</th>
      <th>AS-SE-E</th>
      <th>Vietnam</th>
      <th>Binh Phuoc</th>
      <td>7</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">Pf3D7_13_v3:2837392</th>
      <th rowspan="2" valign="top">AS-SE-E</th>
      <th>Cambodia</th>
      <th>Ratanakiri</th>
      <td>2</td>
    </tr>
    <tr>
      <th>Laos</th>
      <th>Attapeu</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2838654</th>
      <th>OC-NG</th>
      <th>Indonesia</th>
      <th>Papua</th>
      <td>2</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2841024</th>
      <th>AS-SE-W</th>
      <th>Thailand</th>
      <th>Tak</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2841120</th>
      <th>OC-NG</th>
      <th>Indonesia</th>
      <th>Papua</th>
      <td>1</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-9f6efb9c-ff33-480c-89d2-8570cfbd4503')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-9f6efb9c-ff33-480c-89d2-8570cfbd4503 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-9f6efb9c-ff33-480c-89d2-8570cfbd4503');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-16a093e5-56b3-49f5-a0dd-244b3a5c7dd8">
  <button class="colab-df-quickchart" onclick="quickchart('df-16a093e5-56b3-49f5-a0dd-244b3a5c7dd8')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-16a093e5-56b3-49f5-a0dd-244b3a5c7dd8 button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
<p>We can note that breakpoints in Chromosome 11 and 5 recombination have been observed in multiple in-country sampling locations.</p>
<p>In the following sections, we will create various summary tables.</p>
</div>
<div class="section" id="summary-by-deletion-types-table">
<h2>Summary by deletion types table<a class="headerlink" href="#summary-by-deletion-types-table" title="Permalink to this headline">¶</a></h2>
<p>The initial summary table will show the number of distinct breakpoints, countries and samples for distinct deletion types within the <em>hrp2</em> and <em>hrp3</em> genes.</p>
<p>To achieve this, we need a function to count number of distinct unique breakpoints, countries and samples for each deletion type.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">hrp2_agg</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function counts number of distinct unique breakpoints, countries and samples for each deletion type in hrp2.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create an ordered dictionary to store the information.</span>
    <span class="n">summary</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">()</span>

    <span class="c1"># Calculate the number of distinct breakpoints.</span>
    <span class="n">summary</span><span class="p">[</span><span class="s1">&#39;Distinct Breakpoints&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;HRP2_breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

    <span class="c1"># Calculate the number of unique countries.</span>
    <span class="n">summary</span><span class="p">[</span><span class="s1">&#39;Countries&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;Country&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

    <span class="c1"># Calculate the total number of samples with deletions.</span>
    <span class="n">summary</span><span class="p">[</span><span class="s1">&#39;Samples with Deletion&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

    <span class="c1"># Return the summary information as a Pandas Series.</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">summary</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">hrp2_agg</span></code> function can be adapted for <em>hrp3</em> gene deletions as well.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">hrp3_agg</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function counts number of distinct unique breakpoints, countries and samples for each deletion type in hrp3.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create an ordered dictionary to store the information.</span>
    <span class="n">summary</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">()</span>

    <span class="c1"># Calculate the number of distinct breakpoints.</span>
    <span class="n">summary</span><span class="p">[</span><span class="s1">&#39;Distinct Breakpoints&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;HRP3_breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

    <span class="c1"># Calculate the number of unique countries.</span>
    <span class="n">summary</span><span class="p">[</span><span class="s1">&#39;Countries&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;Country&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

    <span class="c1"># Calculate the total number of samples with deletions.</span>
    <span class="n">summary</span><span class="p">[</span><span class="s1">&#39;Samples with Deletion&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

    <span class="c1"># Return the summary information as a Pandas Series.</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">summary</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a summary of deletion types for hrp2 and hrp3 genes</span>

<span class="c1"># Filter samples with hrp2 deletions</span>
<span class="c1"># Group by deletion type, also include gene name</span>
<span class="c1"># Apply hrp2_agg function to find numbers</span>
<span class="n">hrp2_agg_df</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">df_samples_hrp</span><span class="p">[</span><span class="n">df_samples_hrp</span><span class="p">[</span><span class="s1">&#39;HRP2&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;del&#39;</span><span class="p">]</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">Gene</span><span class="o">=</span><span class="s1">&#39;hrp2&#39;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;HRP2_deletion_type&#39;</span><span class="p">:</span> <span class="s1">&#39;Deletion type&#39;</span><span class="p">})</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;Deletion type&#39;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">hrp2_agg</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># Filter samples with hrp3 deletions</span>
<span class="c1"># Group by deletion type, also include gene name</span>
<span class="c1"># Apply hrp2_agg function to find numbers</span>
<span class="n">hrp3_agg_df</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">df_samples_hrp</span><span class="p">[</span>
        <span class="n">df_samples_hrp</span><span class="p">[</span><span class="s1">&#39;QC pass&#39;</span><span class="p">]</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">df_samples_hrp</span><span class="p">[</span><span class="s1">&#39;HRP3&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;del&#39;</span><span class="p">)</span>
    <span class="p">]</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">Gene</span><span class="o">=</span><span class="s1">&#39;hrp3&#39;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;HRP3_deletion_type&#39;</span><span class="p">:</span> <span class="s1">&#39;Deletion type&#39;</span><span class="p">})</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;Deletion type&#39;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">hrp3_agg</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># Concatenate the results to create a summary of deletion types</span>
<span class="n">df_deletion_types_summary</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">hrp2_agg_df</span><span class="p">,</span> <span class="n">hrp3_agg_df</span><span class="p">])</span>

<span class="c1"># Display the summary</span>
<span class="n">df_deletion_types_summary</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-7c9a8a52-673a-4831-a1fa-512e7f688350" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>Distinct Breakpoints</th>
      <th>Countries</th>
      <th>Samples with Deletion</th>
    </tr>
    <tr>
      <th>Gene</th>
      <th>Deletion type</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>hrp2</th>
      <th>Telomere healing</th>
      <td>5</td>
      <td>4</td>
      <td>15</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">hrp3</th>
      <th>Chromosome 11 recombination</th>
      <td>1</td>
      <td>14</td>
      <td>151</td>
    </tr>
    <tr>
      <th>Chromosome 5 recombination</th>
      <td>1</td>
      <td>2</td>
      <td>21</td>
    </tr>
    <tr>
      <th>Telomere healing</th>
      <td>14</td>
      <td>11</td>
      <td>29</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-7c9a8a52-673a-4831-a1fa-512e7f688350')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-7c9a8a52-673a-4831-a1fa-512e7f688350 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-7c9a8a52-673a-4831-a1fa-512e7f688350');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-787921e0-9de2-4bbb-8233-9cea1cb95064">
  <button class="colab-df-quickchart" onclick="quickchart('df-787921e0-9de2-4bbb-8233-9cea1cb95064')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-787921e0-9de2-4bbb-8233-9cea1cb95064 button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
<div class="section" id="save-the-table">
<h3>Save the table<a class="headerlink" href="#save-the-table" title="Permalink to this headline">¶</a></h3>
<p>You will need to authorise Google Colab access to Google Drive.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">drive</span><span class="o">.</span><span class="n">mount</span><span class="p">(</span><span class="s1">&#39;/content/drive&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Mounted at /content/drive
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># This will send the file to your Google Drive, where you can download it from if needed</span>
<span class="c1"># Change the file path if you wish to send the file to a specific location</span>
<span class="c1"># Change the file name if you wish to call it something else</span>

<span class="n">df_deletion_types_summary</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="s1">&#39;/content/drive/My Drive/HRP_deletion_types_summary_unformatted.xlsx&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="summary-by-breakpoints">
<h2>Summary by breakpoints<a class="headerlink" href="#summary-by-breakpoints" title="Permalink to this headline">¶</a></h2>
<p>The second summary table will feature breakpoints, associated countries, and sample counts for distinct deletion types within the hrp2 and hrp3 genes.</p>
<p>This requires a function to aggregate countries and sample counts for each unique deletion type at specific breakpoints.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">breakpoint_agg</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function aggregates countries and sample counts for each unique deletion type at specific breakpoints.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">names</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">()</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;Countries&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">((</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;Country&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;Samples with deletion&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">names</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate summary statistics for hrp2</span>
<span class="c1"># Group samples with hrp3 deletions by gene name, deletion type, and breakpoint</span>
<span class="c1"># Apply breakpoint_agg function</span>

<span class="n">hrp2_breakpoints_summary</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">df_samples_hrp</span><span class="p">[</span><span class="n">df_samples_hrp</span><span class="p">[</span><span class="s1">&#39;HRP2&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;del&#39;</span><span class="p">]</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">Gene</span><span class="o">=</span><span class="s1">&#39;hrp2&#39;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;HRP2_deletion_type&#39;</span><span class="p">:</span> <span class="s1">&#39;Deletion type&#39;</span><span class="p">})</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;Deletion type&#39;</span><span class="p">,</span> <span class="s1">&#39;HRP2_breakpoint&#39;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">breakpoint_agg</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># Calculate summary statistics for hrp3</span>
<span class="c1"># Group samples with hrp3 deletions by gene name, deletion type, and breakpoint</span>
<span class="c1"># Apply breakpoint_agg function</span>

<span class="n">hrp3_breakpoints_summary</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">df_samples_hrp</span><span class="p">[</span><span class="n">df_samples_hrp</span><span class="p">[</span><span class="s1">&#39;HRP3&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;del&#39;</span><span class="p">]</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">Gene</span><span class="o">=</span><span class="s1">&#39;hrp3&#39;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;HRP3_deletion_type&#39;</span><span class="p">:</span> <span class="s1">&#39;Deletion type&#39;</span><span class="p">})</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;Deletion type&#39;</span><span class="p">,</span> <span class="s1">&#39;HRP3_breakpoint&#39;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">breakpoint_agg</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># Combine the summary statistics for HRP2 and HRP3 genes</span>
<span class="n">df_breakpoints_summary</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">hrp2_breakpoints_summary</span><span class="p">,</span> <span class="n">hrp3_breakpoints_summary</span><span class="p">])</span>

<span class="c1"># Display the combined summary</span>
<span class="n">df_breakpoints_summary</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-918d7957-745a-4b0f-bb8b-8dda1fd3bfd2" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th></th>
      <th>Countries</th>
      <th>Samples with deletion</th>
    </tr>
    <tr>
      <th>Gene</th>
      <th>Deletion type</th>
      <th>HRP2_breakpoint</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="5" valign="top">hrp2</th>
      <th rowspan="5" valign="top">Telomere healing</th>
      <th>Pf3D7_08_v3:1373732</th>
      <td>Cambodia</td>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_08_v3:1374280</th>
      <td>Sudan</td>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_08_v3:1374462</th>
      <td>Indonesia</td>
      <td>5</td>
    </tr>
    <tr>
      <th>Pf3D7_08_v3:1374932</th>
      <td>Peru</td>
      <td>2</td>
    </tr>
    <tr>
      <th>Pf3D7_08_v3:1374986</th>
      <td>Peru</td>
      <td>6</td>
    </tr>
    <tr>
      <th rowspan="16" valign="top">hrp3</th>
      <th>Chromosome 11 recombination</th>
      <th>Pf3D7_13_v3:2800004-2807159</th>
      <td>Thailand, Ghana, Indonesia, Peru, Bangladesh, ...</td>
      <td>151</td>
    </tr>
    <tr>
      <th>Chromosome 5 recombination</th>
      <th>Pf3D7_13_v3:2835587-2835612</th>
      <td>Cambodia, Vietnam</td>
      <td>21</td>
    </tr>
    <tr>
      <th rowspan="14" valign="top">Telomere healing</th>
      <th>Pf3D7_13_v3:2811525</th>
      <td>India</td>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2812344</th>
      <td>Sudan</td>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2815249</th>
      <td>Tanzania</td>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2822480</th>
      <td>Ghana</td>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2823645</th>
      <td>Kenya</td>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2830952</th>
      <td>Cambodia</td>
      <td>7</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2832080</th>
      <td>Democratic Republic of the Congo</td>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2834604</th>
      <td>Vietnam</td>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2835532</th>
      <td>Thailand</td>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2837145</th>
      <td>Vietnam</td>
      <td>7</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2837392</th>
      <td>Cambodia, Laos</td>
      <td>3</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2838654</th>
      <td>Indonesia</td>
      <td>2</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2841024</th>
      <td>Thailand</td>
      <td>1</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3:2841120</th>
      <td>Indonesia</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-918d7957-745a-4b0f-bb8b-8dda1fd3bfd2')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-918d7957-745a-4b0f-bb8b-8dda1fd3bfd2 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-918d7957-745a-4b0f-bb8b-8dda1fd3bfd2');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-0b73bc3d-30ec-4071-9854-8dbe7048df15">
  <button class="colab-df-quickchart" onclick="quickchart('df-0b73bc3d-30ec-4071-9854-8dbe7048df15')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-0b73bc3d-30ec-4071-9854-8dbe7048df15 button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
<p><strong>Table Legend. Summary of hrp2 and hrp3 deletion breakpoints.</strong> Telomere healing refers to the process whereby the end of a chromosome is deleted and a telomere repeat sequence attached to the breakpoint. Chromosome 11 recombination refers to a new hybrid chromosome being created by a recombination between chromosome 13 and 11 at a cluster of rRNA genes that appear to have orthologous copies on both chromosomes. Chromosome 5 recombination refers to a recombination between chromosome 13 and an inverted section of the middle of chromosome 5 containing the gene mdr1. For telomere healing an exact breakpoint position is given but for recombination events it is only possible to give a region in which the recombination has occurred.</p>
<div class="section" id="id1">
<h3>Save the table<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># This will send the file to your Google Drive, where you can download it from if needed</span>
<span class="c1"># Change the file path if you wish to send the file to a specific location</span>
<span class="c1"># Change the file name if you wish to call it something else</span>

<span class="n">df_breakpoints_summary</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="s1">&#39;/content/drive/My Drive/HRP_breakpoints_summary_unformatted.xlsx&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="summary-by-country">
<h2>Summary by country<a class="headerlink" href="#summary-by-country" title="Permalink to this headline">¶</a></h2>
<p>The final summary table will display the exact number and incidence of deletions within <em>hrp2</em> and <em>hrp3</em> in each country.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define an aggregation function to calculate proportions and counts for HRP2 and HRP3 genes.</span>
<span class="k">def</span> <span class="nf">proportion_agg</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function counts the number of deletions, the incidence of deletions, and more for both genes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">names</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">()</span>

    <span class="c1"># Calculate the number of HRP2 deletions</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;hrp2 calls&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;HRP2&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;uncallable&#39;</span><span class="p">)</span>

    <span class="c1"># Calculate the incidence of HRP2 deletions</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;</span><span class="si">% hr</span><span class="s1">p2 deletions&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;HRP2&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;del&#39;</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;HRP2&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;uncallable&#39;</span><span class="p">))</span><span class="o">*</span><span class="mi">100</span><span class="p">))</span>

    <span class="c1"># Calculate the number of HRP3 deletions</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;hrp3 calls&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;HRP3&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;uncallable&#39;</span><span class="p">)</span>

    <span class="c1"># Calculate the percentage of HRP3 deletions</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;</span><span class="si">% hr</span><span class="s1">p3 deletions&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;HRP3&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;del&#39;</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;HRP3&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;uncallable&#39;</span><span class="p">))</span><span class="o">*</span><span class="mi">100</span><span class="p">)</span>

    <span class="c1"># Calculate the number of deletions for both HRP2 and HRP3</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;hrp2 and hrp3 calls&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">((</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;HRP2&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;uncallable&#39;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;HRP3&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;uncallable&#39;</span><span class="p">))</span>

    <span class="c1"># Calculate the incidence of deletions for both HRP2 and HRP3</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;</span><span class="si">% hr</span><span class="s1">p2 and hrp3 deletions&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span>
        <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;HRP2&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;del&#39;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;HRP3&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;del&#39;</span><span class="p">)</span>
    <span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span>
        <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;HRP2&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;uncallable&#39;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;HRP3&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;uncallable&#39;</span><span class="p">)</span>
    <span class="p">))</span><span class="o">*</span><span class="mi">100</span><span class="p">)</span>
    <span class="c1"># Convert the values to integers</span>
    <span class="n">names</span> <span class="o">=</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">names</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">names</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a table to summarize HRP2 and HRP3 gene information by country</span>

<span class="c1"># Calculate proportions and counts for HRP2 and HRP3 genes by country</span>
<span class="n">df_hrp_by_country_table</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">df_samples_hrp</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Country&#39;</span><span class="p">)</span>  <span class="c1"># Group the data by country</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">proportion_agg</span><span class="p">)</span>  <span class="c1"># Apply the proportion aggregation function to calculate statistics</span>
    <span class="o">.</span><span class="n">transpose</span><span class="p">()</span>  <span class="c1"># Transpose the table for better readability</span>
<span class="p">)</span>

<span class="c1"># Calculate sample counts for each country</span>
<span class="n">hrp_by_country_n</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">df_samples_hrp</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Country&#39;</span><span class="p">)</span>  <span class="c1"># Group the data by country</span>
    <span class="o">.</span><span class="n">size</span><span class="p">()</span>  <span class="c1"># Calculate the number of samples in each country</span>
<span class="p">)</span>

<span class="c1"># Rename columns to include sample counts next to country names</span>
<span class="k">for</span> <span class="n">country</span> <span class="ow">in</span> <span class="n">df_samples_hrp</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df_samples_hrp</span><span class="p">[</span><span class="s1">&#39;QC pass&#39;</span><span class="p">],</span> <span class="s1">&#39;Country&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
    <span class="n">new_column_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">country</span><span class="si">}</span><span class="s1"> (n=</span><span class="si">{</span><span class="n">hrp_by_country_n</span><span class="p">[</span><span class="n">country</span><span class="p">]</span><span class="si">:</span><span class="s1">,</span><span class="si">}</span><span class="s1">)&#39;</span>
    <span class="n">df_hrp_by_country_table</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="n">country</span><span class="p">:</span> <span class="n">new_column_name</span><span class="p">},</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Transpose the table to have countries as rows and statistics as columns</span>
<span class="n">df_hrp_by_country_table</span> <span class="o">=</span> <span class="n">df_hrp_by_country_table</span><span class="o">.</span><span class="n">T</span>

<span class="c1"># Display the table with summarized information</span>
<span class="n">df_hrp_by_country_table</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-24d9f613-197b-4a71-bf94-ebeeefb0a57c" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>hrp2 calls</th>
      <th>% hrp2 deletions</th>
      <th>hrp3 calls</th>
      <th>% hrp3 deletions</th>
      <th>hrp2 and hrp3 calls</th>
      <th>% hrp2 and hrp3 deletions</th>
    </tr>
    <tr>
      <th>Country</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Bangladesh (n=1,310)</th>
      <td>939</td>
      <td>0</td>
      <td>850</td>
      <td>0</td>
      <td>819</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Benin (n=150)</th>
      <td>110</td>
      <td>0</td>
      <td>104</td>
      <td>0</td>
      <td>100</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Burkina Faso (n=57)</th>
      <td>43</td>
      <td>0</td>
      <td>32</td>
      <td>0</td>
      <td>32</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Cambodia (n=1,267)</th>
      <td>1109</td>
      <td>0</td>
      <td>1091</td>
      <td>3</td>
      <td>1064</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Cameroon (n=264)</th>
      <td>244</td>
      <td>0</td>
      <td>240</td>
      <td>0</td>
      <td>235</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Colombia (n=135)</th>
      <td>124</td>
      <td>0</td>
      <td>123</td>
      <td>41</td>
      <td>118</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Côte d'Ivoire (n=71)</th>
      <td>70</td>
      <td>0</td>
      <td>71</td>
      <td>0</td>
      <td>70</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Democratic Republic of the Congo (n=520)</th>
      <td>413</td>
      <td>0</td>
      <td>392</td>
      <td>0</td>
      <td>385</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Ethiopia (n=21)</th>
      <td>20</td>
      <td>0</td>
      <td>20</td>
      <td>45</td>
      <td>20</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Gabon (n=55)</th>
      <td>34</td>
      <td>0</td>
      <td>38</td>
      <td>0</td>
      <td>32</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Gambia (n=863)</th>
      <td>517</td>
      <td>0</td>
      <td>467</td>
      <td>1</td>
      <td>460</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Ghana (n=3,131)</th>
      <td>1529</td>
      <td>0</td>
      <td>1448</td>
      <td>0</td>
      <td>1343</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Guinea (n=151)</th>
      <td>121</td>
      <td>0</td>
      <td>119</td>
      <td>0</td>
      <td>119</td>
      <td>0</td>
    </tr>
    <tr>
      <th>India (n=300)</th>
      <td>75</td>
      <td>0</td>
      <td>70</td>
      <td>1</td>
      <td>68</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Indonesia (n=121)</th>
      <td>117</td>
      <td>4</td>
      <td>117</td>
      <td>37</td>
      <td>115</td>
      <td>2</td>
    </tr>
    <tr>
      <th>Kenya (n=690)</th>
      <td>660</td>
      <td>0</td>
      <td>647</td>
      <td>0</td>
      <td>645</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Laos (n=991)</th>
      <td>773</td>
      <td>0</td>
      <td>717</td>
      <td>2</td>
      <td>669</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Madagascar (n=24)</th>
      <td>24</td>
      <td>0</td>
      <td>22</td>
      <td>0</td>
      <td>22</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Malawi (n=265)</th>
      <td>265</td>
      <td>0</td>
      <td>264</td>
      <td>0</td>
      <td>264</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Mali (n=1,167)</th>
      <td>709</td>
      <td>0</td>
      <td>691</td>
      <td>0</td>
      <td>669</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Mauritania (n=92)</th>
      <td>79</td>
      <td>0</td>
      <td>81</td>
      <td>0</td>
      <td>79</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Mozambique (n=34)</th>
      <td>10</td>
      <td>0</td>
      <td>10</td>
      <td>0</td>
      <td>6</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Myanmar (n=985)</th>
      <td>645</td>
      <td>0</td>
      <td>606</td>
      <td>0</td>
      <td>585</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Nigeria (n=110)</th>
      <td>34</td>
      <td>0</td>
      <td>30</td>
      <td>0</td>
      <td>30</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Papua New Guinea (n=221)</th>
      <td>118</td>
      <td>0</td>
      <td>106</td>
      <td>0</td>
      <td>106</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Peru (n=21)</th>
      <td>21</td>
      <td>38</td>
      <td>20</td>
      <td>75</td>
      <td>20</td>
      <td>30</td>
    </tr>
    <tr>
      <th>Senegal (n=150)</th>
      <td>142</td>
      <td>0</td>
      <td>141</td>
      <td>4</td>
      <td>138</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Sudan (n=76)</th>
      <td>7</td>
      <td>14</td>
      <td>7</td>
      <td>86</td>
      <td>7</td>
      <td>14</td>
    </tr>
    <tr>
      <th>Tanzania (n=589)</th>
      <td>452</td>
      <td>0</td>
      <td>470</td>
      <td>0</td>
      <td>440</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Thailand (n=954)</th>
      <td>855</td>
      <td>0</td>
      <td>846</td>
      <td>0</td>
      <td>823</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Uganda (n=12)</th>
      <td>12</td>
      <td>0</td>
      <td>12</td>
      <td>0</td>
      <td>12</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Venezuela (n=2)</th>
      <td>2</td>
      <td>0</td>
      <td>2</td>
      <td>0</td>
      <td>2</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Vietnam (n=1,404)</th>
      <td>762</td>
      <td>0</td>
      <td>740</td>
      <td>1</td>
      <td>670</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-24d9f613-197b-4a71-bf94-ebeeefb0a57c')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-24d9f613-197b-4a71-bf94-ebeeefb0a57c button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-24d9f613-197b-4a71-bf94-ebeeefb0a57c');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-b257b9a5-afe1-4d81-80f1-771ab6fd3103">
  <button class="colab-df-quickchart" onclick="quickchart('df-b257b9a5-afe1-4d81-80f1-771ab6fd3103')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-b257b9a5-afe1-4d81-80f1-771ab6fd3103 button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
<p><strong>Table Legend. Frequency of HRP2 and HRP3 deletions by country.</strong> n=number of QC pass samples. Calls columns show number of samples for which an unambiguous deletion genotype (deleted or non-deleted) could be assigned.</p>
<div class="section" id="id2">
<h3>Save the table<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># This will send the file to your Google Drive, where you can download it from if needed</span>
<span class="c1"># Change the file path if you wish to send the file to a specific location</span>
<span class="c1"># Change the file name if you wish to call it something else</span>

<span class="n">df_hrp_by_country_table</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="s1">&#39;/content/drive/My Drive/HRP_country_summary_unformatted.xlsx&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./pf7/evolution_malaria_control"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="chloroquine-resistance-map.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Plot Map of Chloroquine Resistance</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="visualise_hrp2_hrp3_deletions.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Visualise <em>hrp2</em> and <em>hrp3</em> Deletions</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By MalariaGEN<br/>
    
        &copy; Copyright 2023.<br/>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>