
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Plot eba175 C Allele Proportions &#8212; MalariaGEN parasite data user guide</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Pf7 API" href="../api.html" />
    <link rel="prev" title="Neighbour Joining Tree for sWGA vs. gDNA" href="neighbour-joining-tree-swga.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">MalariaGEN parasite data user guide</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../landing-page.html">
   MalariaGEN parasite data user guide
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Pf7
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../Data_access.html">
   Pf7 Data Access
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../useful-how-tos.html">
   <em>
    <em>
     Pf7 Analysis Guides
    </em>
    :
   </em>
   * Useful Genomics How-Tos
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../useful_how_tos/igv-notebook.html">
     Visualizing Genomic Regions of Pf7 Samples with IGV Plots
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../summarise-data.html">
   <em>
    <em>
     Pf7 Analysis Guides
    </em>
    :
   </em>
   * Summarise Pf7 Data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../summarise_data/sample-summary-table.html">
     Summarise Samples by Administrative Division
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../summarise_data/sample-collection-over-time.html">
     Visualising sample collections over time
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../summarise_data/barplot-samples-by-country.html">
     Barplot of Samples by Country
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../summarise_data/variant-summary-table.html">
     Summarise Genetic Variants
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../summarise_data/global-sample-map.html">
     Plot Global Map of Samples
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../evolution-malaria-control.html">
   *
   <em>
    Pf7 Analysis Guides
   </em>
   :**Study Evolution Against Malaria Control Measures
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../evolution_malaria_control/drug-resistance-summary.html">
     Drug Resistance Frequencies Across Pf7
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../evolution_malaria_control/chloroquine-resistance-map.html">
     Plot Map of Chloroquine Resistance
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../evolution_malaria_control/summarise_hrp2_hrp3_deletions.html">
     Summarise
     <em>
      hrp2
     </em>
     and
     <em>
      hrp3
     </em>
     Deletions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../evolution_malaria_control/visualise_hrp2_hrp3_deletions.html">
     Visualise
     <em>
      hrp2
     </em>
     and
     <em>
      hrp3
     </em>
     Deletions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../evolution_malaria_control/csp-cterm-analysis.html">
     Analyse csp C-Terminal Variation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../popgen.html">
   <em>
    <em>
     Pf7 Analysis Guides
    </em>
    :
   </em>
   * Population Structure and Selection
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="PCoA.html">
     Principal Coordinate Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="geographic-genetic-differentiation.html">
     Geographic patterns of population differentiation and gene flow
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="neighbour-joining-tree-swga.html">
     Neighbour Joining Tree for sWGA vs. gDNA
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Plot
     <em>
      eba175
     </em>
     C Allele Proportions
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../api.html">
   Pf7 API
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Pv4
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../pv4/Data_access.html">
   Pv4 Data Access
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../pv4/api.html">
   Pv4 API
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/pf7/popgen/proportion-eba175.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
                onclick="printPdf(this)" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/malariagen/parasite-data"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/malariagen/parasite-data/issues/new?title=Issue%20on%20page%20%2Fpf7/popgen/proportion-eba175.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/malariagen/parasite-data/master?urlpath=lab/tree/docs/pf7/popgen/proportion-eba175.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        <a class="colab-button" href="https://colab.research.google.com/github/malariagen/parasite-data/blob/master/docs/pf7/popgen/proportion-eba175.ipynb"><button type="button" class="btn btn-secondary topbarbtn"
                title="Launch Colab" data-toggle="tooltip" data-placement="left"><img class="colab-button-logo"
                    src="../../_static/images/logo_colab.png"
                    alt="Interact on Colab">Colab</button></a>
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#access-pf7-data">
   Access Pf7 Data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#combine-data-into-a-single-dataframe">
   Combine data into a single dataframe
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#distribution-of-allelic-types-by-population">
   Distribution of allelic types by population
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#make-the-figure">
   Make the figure
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#save-the-figure">
   Save the figure
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#distribution-of-allelic-types-at-a-more-granular-level">
   Distribution of allelic types at a more granular level
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Plot eba175 C Allele Proportions</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#access-pf7-data">
   Access Pf7 Data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#combine-data-into-a-single-dataframe">
   Combine data into a single dataframe
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#distribution-of-allelic-types-by-population">
   Distribution of allelic types by population
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#make-the-figure">
   Make the figure
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#save-the-figure">
   Save the figure
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#distribution-of-allelic-types-at-a-more-granular-level">
   Distribution of allelic types at a more granular level
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            
              <div>
                
  <div class="section" id="plot-eba175-c-allele-proportions">
<h1>Plot <em>eba175</em> C Allele Proportions<a class="headerlink" href="#plot-eba175-c-allele-proportions" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>This notebook recreates <a class="reference external" href="https://figshare.com/ndownloader/files/38478530">Supplement Figure 11</a> from the Pf7 paper - a bar plot which shows the proportions of the <em>eba175</em> C allele in the major subpopulations.</p>
<p>Eba175 is a vaccine candidate gene with two distinct different allelic forms, known as the F- and C-types. Investigating the prevalence of the <em>eba175</em> C allele can help assess whether <em>eba175</em> experiences balancing selection, likely influenced by negative frequency-dependent selection driven by interactions with the human immune system.</p>
<p><strong>This notebook should take approximately two minutes to run.</strong></p>
</div>
<div class="section" id="setup">
<h2>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">¶</a></h2>
<p>Install the malariagen Python package:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>pip<span class="w"> </span>install<span class="w"> </span>-q<span class="w"> </span>--no-warn-conflicts<span class="w"> </span>malariagen_data
<span class="kn">import</span> <span class="nn">malariagen_data</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">133.2/133.2 kB</span> <span class=" -Color -Color-Red">2.8 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">3.1/3.1 MB</span> <span class=" -Color -Color-Red">42.4 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">10.4/10.4 MB</span> <span class=" -Color -Color-Red">77.1 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">3.6/3.6 MB</span> <span class=" -Color -Color-Red">57.4 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">302.5/302.5 kB</span> <span class=" -Color -Color-Red">15.8 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">138.7/138.7 kB</span> <span class=" -Color -Color-Red">10.8 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">20.9/20.9 MB</span> <span class=" -Color -Color-Red">48.5 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">8.1/8.1 MB</span> <span class=" -Color -Color-Red">96.2 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">206.9/206.9 kB</span> <span class=" -Color -Color-Red">20.5 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">233.6/233.6 kB</span> <span class=" -Color -Color-Red">22.0 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
?25h  Preparing metadata (setup.py) ... ?25l?25hdone
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">6.7/6.7 MB</span> <span class=" -Color -Color-Red">60.2 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">1.6/1.6 MB</span> <span class=" -Color -Color-Red">34.6 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
?25h  Building wheel for asciitree (setup.py) ... ?25l?25hdone
</pre></div>
</div>
</div>
</div>
<p>Import required python libraries that are installed at colab by default.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">from</span> <span class="nn">google.colab</span> <span class="kn">import</span> <span class="n">drive</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="access-pf7-data">
<h2>Access Pf7 Data<a class="headerlink" href="#access-pf7-data" title="Permalink to this headline">¶</a></h2>
<p>We use the malariagen data package to load sample data and metadata.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">release_data</span> <span class="o">=</span> <span class="n">malariagen_data</span><span class="o">.</span><span class="n">Pf7</span><span class="p">()</span>
<span class="n">sample_metadata</span> <span class="o">=</span> <span class="n">release_data</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>For this figure, we also need to load the <em>eba175</em> gene allele frequencies in major populations.</p>
<p>This data is available as a .txt file from a web link. We can directly read the data into a dataframe using the read_csv function in pandas!</p>
<p>If you click the left sidebar folder icon, you should see this file appear.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Read in calls file direct from MalariaGEN website</span>
<span class="n">df_eba</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
    <span class="s1">&#39;https://www.malariagen.net/sites/default/files/Pf7_eba175_callset.txt&#39;</span>
    <span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span>
    <span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span>
    <span class="p">,</span> <span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
<span class="c1"># Check shape (should contain 16,203 as this is the number of QC pass samples in Pf7)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df_eba</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="c1"># View the new dataframe stucture</span>
<span class="n">df_eba</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(16203, 3)
</pre></div>
</div>
<div class="output text_html">
  <div id="df-43c90cff-62ed-45ad-a913-b34c6603a5ce" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>f_frac</th>
      <th>c_frac</th>
      <th>eba175_call</th>
    </tr>
    <tr>
      <th>Sample</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>FP0008-C</th>
      <td>0.863946</td>
      <td>0.000000</td>
      <td>F</td>
    </tr>
    <tr>
      <th>FP0009-C</th>
      <td>0.003988</td>
      <td>0.627346</td>
      <td>Mixed</td>
    </tr>
    <tr>
      <th>FP0010-CW</th>
      <td>0.666667</td>
      <td>0.000000</td>
      <td>F</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-43c90cff-62ed-45ad-a913-b34c6603a5ce')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-43c90cff-62ed-45ad-a913-b34c6603a5ce button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-43c90cff-62ed-45ad-a913-b34c6603a5ce');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-20ef65b5-66bb-472b-b923-77d35abb3a3f">
  <button class="colab-df-quickchart" onclick="quickchart('df-20ef65b5-66bb-472b-b923-77d35abb3a3f')"
            title="Suggest charts."
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-20ef65b5-66bb-472b-b923-77d35abb3a3f button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
</div>
<div class="section" id="combine-data-into-a-single-dataframe">
<h2>Combine data into a single dataframe<a class="headerlink" href="#combine-data-into-a-single-dataframe" title="Permalink to this headline">¶</a></h2>
<p>In order to create the figure, it’s convenient to work with combined data.</p>
<p>We retain all samples from sample_metadata which pass QC (n = 16,203).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Merge the two dataframes on identical Sample IDs</span>
<span class="n">df_calls</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">sample_metadata</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">sample_metadata</span><span class="p">[</span><span class="s1">&#39;QC pass&#39;</span><span class="p">]]</span>
    <span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">df_eba</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">)</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df_calls</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="c1"># View the new dataframe stucture</span>
<span class="n">df_calls</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(16203, 20)
</pre></div>
</div>
<div class="output text_html">
  <div id="df-ca2002b5-1530-4834-bfef-e6b5c57b93dc" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Sample</th>
      <th>Study</th>
      <th>Country</th>
      <th>Admin level 1</th>
      <th>Country latitude</th>
      <th>Country longitude</th>
      <th>Admin level 1 latitude</th>
      <th>Admin level 1 longitude</th>
      <th>Year</th>
      <th>ENA</th>
      <th>All samples same case</th>
      <th>Population</th>
      <th>% callable</th>
      <th>QC pass</th>
      <th>Exclusion reason</th>
      <th>Sample type</th>
      <th>Sample was in Pf6</th>
      <th>f_frac</th>
      <th>c_frac</th>
      <th>eba175_call</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>FP0008-C</td>
      <td>1147-PF-MR-CONWAY</td>
      <td>Mauritania</td>
      <td>Hodh el Gharbi</td>
      <td>20.265149</td>
      <td>-10.337093</td>
      <td>16.565426</td>
      <td>-9.832345</td>
      <td>2014.0</td>
      <td>ERR1081237</td>
      <td>FP0008-C</td>
      <td>AF-W</td>
      <td>82.16</td>
      <td>True</td>
      <td>Analysis_set</td>
      <td>gDNA</td>
      <td>True</td>
      <td>0.863946</td>
      <td>0.000000</td>
      <td>F</td>
    </tr>
    <tr>
      <th>1</th>
      <td>FP0009-C</td>
      <td>1147-PF-MR-CONWAY</td>
      <td>Mauritania</td>
      <td>Hodh el Gharbi</td>
      <td>20.265149</td>
      <td>-10.337093</td>
      <td>16.565426</td>
      <td>-9.832345</td>
      <td>2014.0</td>
      <td>ERR1081238</td>
      <td>FP0009-C</td>
      <td>AF-W</td>
      <td>88.85</td>
      <td>True</td>
      <td>Analysis_set</td>
      <td>gDNA</td>
      <td>True</td>
      <td>0.003988</td>
      <td>0.627346</td>
      <td>Mixed</td>
    </tr>
    <tr>
      <th>2</th>
      <td>FP0010-CW</td>
      <td>1147-PF-MR-CONWAY</td>
      <td>Mauritania</td>
      <td>Hodh el Gharbi</td>
      <td>20.265149</td>
      <td>-10.337093</td>
      <td>16.565426</td>
      <td>-9.832345</td>
      <td>2014.0</td>
      <td>ERR2889621</td>
      <td>FP0010-CW</td>
      <td>AF-W</td>
      <td>86.46</td>
      <td>True</td>
      <td>Analysis_set</td>
      <td>sWGA</td>
      <td>False</td>
      <td>0.666667</td>
      <td>0.000000</td>
      <td>F</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-ca2002b5-1530-4834-bfef-e6b5c57b93dc')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-ca2002b5-1530-4834-bfef-e6b5c57b93dc button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-ca2002b5-1530-4834-bfef-e6b5c57b93dc');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-c869b796-44cd-467a-ae76-5b189bd11e8f">
  <button class="colab-df-quickchart" onclick="quickchart('df-c869b796-44cd-467a-ae76-5b189bd11e8f')"
            title="Suggest charts."
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-c869b796-44cd-467a-ae76-5b189bd11e8f button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
</div>
<div class="section" id="distribution-of-allelic-types-by-population">
<h2>Distribution of allelic types by population<a class="headerlink" href="#distribution-of-allelic-types-by-population" title="Permalink to this headline">¶</a></h2>
<p>First, we check the frequency of allelic types in the overall dataset before checking in each population.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">num_F</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">df_calls</span><span class="p">[</span><span class="s1">&#39;eba175_call&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;F&#39;</span><span class="p">)</span>
<span class="n">num_C</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">df_calls</span><span class="p">[</span><span class="s1">&#39;eba175_call&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;C&#39;</span><span class="p">)</span>
<span class="n">proportion_C</span> <span class="o">=</span> <span class="n">num_C</span> <span class="o">/</span> <span class="p">(</span><span class="n">num_C</span> <span class="o">+</span> <span class="n">num_F</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;We see </span><span class="si">{</span><span class="n">num_F</span><span class="si">:</span><span class="s2">,</span><span class="si">}</span><span class="s2"> samples with the F allele and </span><span class="si">{</span><span class="n">num_C</span><span class="si">:</span><span class="s2">,</span><span class="si">}</span><span class="s2"> samples with the C allele, </span><span class="se">\</span>
<span class="s2">meaning that overall </span><span class="si">{</span><span class="n">proportion_C</span><span class="si">:</span><span class="s2">.0%</span><span class="si">}</span><span class="s2"> of samples that have an unambiguous call have the rarer C allele.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>We see 7,380 samples with the F allele and 3,364 samples with the C allele, meaning that overall 31% of samples that have an unambiguous call have the rarer C allele.
</pre></div>
</div>
</div>
</div>
<p>Now, we will examine individual population allelic frequencies to generate the corresponding figure.</p>
<p>First we define the populations (for more details on populations, please visit the <a class="reference external" href="https://figshare.com/ndownloader/files/38478530">Supplement</a>).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define populations in an ordered dictionary</span>
<span class="n">populations</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">()</span>
<span class="n">populations</span><span class="p">[</span><span class="s1">&#39;SA&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;South America&#39;</span>
<span class="n">populations</span><span class="p">[</span><span class="s1">&#39;AF-W&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;West Africa&#39;</span>
<span class="n">populations</span><span class="p">[</span><span class="s1">&#39;AF-C&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Central Africa&#39;</span>
<span class="n">populations</span><span class="p">[</span><span class="s1">&#39;AF-NE&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Northeast Africa&#39;</span>
<span class="n">populations</span><span class="p">[</span><span class="s1">&#39;AF-E&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;East Africa&#39;</span>
<span class="n">populations</span><span class="p">[</span><span class="s1">&#39;AS-S-E&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Eastern South Asia&#39;</span>
<span class="n">populations</span><span class="p">[</span><span class="s1">&#39;AS-S-FE&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Far-eastern South Asia&#39;</span>
<span class="n">populations</span><span class="p">[</span><span class="s1">&#39;AS-SE-W&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Western Southeast Asia&#39;</span>
<span class="n">populations</span><span class="p">[</span><span class="s1">&#39;AS-SE-E&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Eastern Southeast Asia&#39;</span>
<span class="n">populations</span><span class="p">[</span><span class="s1">&#39;OC-NG&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Oceania&#39;</span>
</pre></div>
</div>
</div>
</div>
<p>Next, we need a function to aggregate allele frequencies in each population</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to aggregate population frequencies</span>
<span class="k">def</span> <span class="nf">my_agg</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
    <span class="n">names</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">()</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;Number F&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;eba175_call&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;F&#39;</span><span class="p">)</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;Number C&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;eba175_call&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;C&#39;</span><span class="p">)</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;Number F or C&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">names</span><span class="p">[</span><span class="s1">&#39;Number F&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">names</span><span class="p">[</span><span class="s1">&#39;Number C&#39;</span><span class="p">]</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;Proportion C&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">names</span><span class="p">[</span><span class="s1">&#39;Number C&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">names</span><span class="p">[</span><span class="s1">&#39;Number F or C&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">names</span><span class="p">[</span><span class="s1">&#39;Number F or C&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">names</span><span class="p">)</span>
<span class="c1"># Summary table with population aggregates</span>
<span class="n">df_pop_summary</span> <span class="o">=</span> <span class="n">df_calls</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Population&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">my_agg</span><span class="p">)</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">populations</span><span class="o">.</span><span class="n">keys</span><span class="p">()]</span>
<span class="n">df_pop_summary</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-6698e5a5-2700-457e-9e40-300d007a2e61" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Number F</th>
      <th>Number C</th>
      <th>Number F or C</th>
      <th>Proportion C</th>
    </tr>
    <tr>
      <th>Population</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>SA</th>
      <td>28.0</td>
      <td>44.0</td>
      <td>72.0</td>
      <td>0.611111</td>
    </tr>
    <tr>
      <th>AF-W</th>
      <td>3111.0</td>
      <td>762.0</td>
      <td>3873.0</td>
      <td>0.196747</td>
    </tr>
    <tr>
      <th>AF-C</th>
      <td>191.0</td>
      <td>30.0</td>
      <td>221.0</td>
      <td>0.135747</td>
    </tr>
    <tr>
      <th>AF-NE</th>
      <td>45.0</td>
      <td>48.0</td>
      <td>93.0</td>
      <td>0.516129</td>
    </tr>
    <tr>
      <th>AF-E</th>
      <td>751.0</td>
      <td>210.0</td>
      <td>961.0</td>
      <td>0.218522</td>
    </tr>
    <tr>
      <th>AS-S-E</th>
      <td>90.0</td>
      <td>15.0</td>
      <td>105.0</td>
      <td>0.142857</td>
    </tr>
    <tr>
      <th>AS-S-FE</th>
      <td>712.0</td>
      <td>343.0</td>
      <td>1055.0</td>
      <td>0.325118</td>
    </tr>
    <tr>
      <th>AS-SE-W</th>
      <td>702.0</td>
      <td>741.0</td>
      <td>1443.0</td>
      <td>0.513514</td>
    </tr>
    <tr>
      <th>AS-SE-E</th>
      <td>1655.0</td>
      <td>1056.0</td>
      <td>2711.0</td>
      <td>0.389524</td>
    </tr>
    <tr>
      <th>OC-NG</th>
      <td>95.0</td>
      <td>115.0</td>
      <td>210.0</td>
      <td>0.547619</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-6698e5a5-2700-457e-9e40-300d007a2e61')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-6698e5a5-2700-457e-9e40-300d007a2e61 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-6698e5a5-2700-457e-9e40-300d007a2e61');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-836b1768-a13b-4540-8173-21b3885073b7">
  <button class="colab-df-quickchart" onclick="quickchart('df-836b1768-a13b-4540-8173-21b3885073b7')"
            title="Suggest charts."
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-836b1768-a13b-4540-8173-21b3885073b7 button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
</div>
<div class="section" id="make-the-figure">
<h2>Make the figure<a class="headerlink" href="#make-the-figure" title="Permalink to this headline">¶</a></h2>
<p>First, we set-up population-specific colour-codes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create an ordered dictionary which maps the codes for major sub-populations -from west to east- to a colour code.</span>
<span class="n">population_colours</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">()</span>
<span class="n">population_colours</span><span class="p">[</span><span class="s1">&#39;SA&#39;</span><span class="p">]</span>       <span class="o">=</span> <span class="s2">&quot;#4daf4a&quot;</span>
<span class="n">population_colours</span><span class="p">[</span><span class="s1">&#39;AF-W&#39;</span><span class="p">]</span>     <span class="o">=</span> <span class="s2">&quot;#e31a1c&quot;</span>
<span class="n">population_colours</span><span class="p">[</span><span class="s1">&#39;AF-C&#39;</span><span class="p">]</span>     <span class="o">=</span> <span class="s2">&quot;#fd8d3c&quot;</span>
<span class="n">population_colours</span><span class="p">[</span><span class="s1">&#39;AF-NE&#39;</span><span class="p">]</span>    <span class="o">=</span> <span class="s2">&quot;#bb8129&quot;</span>
<span class="n">population_colours</span><span class="p">[</span><span class="s1">&#39;AF-E&#39;</span><span class="p">]</span>     <span class="o">=</span> <span class="s2">&quot;#fecc5c&quot;</span>
<span class="n">population_colours</span><span class="p">[</span><span class="s1">&#39;AS-S-E&#39;</span><span class="p">]</span>  <span class="o">=</span> <span class="s2">&quot;#dfc0eb&quot;</span>
<span class="n">population_colours</span><span class="p">[</span><span class="s1">&#39;AS-S-FE&#39;</span><span class="p">]</span>  <span class="o">=</span> <span class="s2">&quot;#984ea3&quot;</span>
<span class="n">population_colours</span><span class="p">[</span><span class="s1">&#39;AS-SE-W&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;#9ecae1&quot;</span>
<span class="n">population_colours</span><span class="p">[</span><span class="s1">&#39;AS-SE-E&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;#3182bd&quot;</span>
<span class="n">population_colours</span><span class="p">[</span><span class="s1">&#39;OC-NG&#39;</span><span class="p">]</span>    <span class="o">=</span> <span class="s2">&quot;#f781bf&quot;</span>
</pre></div>
</div>
</div>
</div>
<p>We are ready to make the bar plot using df_pop_summary data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Set the figure size</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="c1"># Set the chart variables</span>
<span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">df_pop_summary</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">df_pop_summary</span><span class="p">[</span><span class="s1">&#39;Proportion C&#39;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">population_colours</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
<span class="c1"># Set the labels</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">df_pop_summary</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Population&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Proportion of C allele&#39;</span><span class="p">)</span>
<span class="c1"># Hide the right spine (axis line) of the plot</span>
<span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="c1"># Hide the top spine (axis line) of the plot</span>
<span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;top&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="c1"># Set the y-axis limits</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="c1"># Add a horizontal line showing %50 proportion</span>
<span class="n">ax</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;lightgray&#39;</span><span class="p">)</span>
<span class="c1"># Adjust figure to ensure everything fits nicely</span>
<span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;ipython-input-10-1f8c815b50e5&gt;:6: UserWarning: FixedFormatter should only be used together with FixedLocator
  ax.set_xticklabels(df_pop_summary.index, rotation=90)
</pre></div>
</div>
<img alt="../../_images/proportion-eba175_29_1.png" src="../../_images/proportion-eba175_29_1.png" />
</div>
</div>
<p><strong>Figure legend:</strong> Proportion of C allele of <em>eba175</em> in different major subpopulations. Horizontal line shows 50% proportion.</p>
</div>
<div class="section" id="save-the-figure">
<h2>Save the figure<a class="headerlink" href="#save-the-figure" title="Permalink to this headline">¶</a></h2>
<p>We can output this to a location in Google Drive</p>
<p>First we need to connect Google Drive by running the following:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># You will need to authorise Google Colab access to Google Drive</span>
<span class="n">drive</span><span class="o">.</span><span class="n">mount</span><span class="p">(</span><span class="s1">&#39;/content/drive&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Mounted at /content/drive
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># This will send the file to your Google Drive, where you can download it from if needed</span>
<span class="c1"># Change the file path if you wish to send the file to a specific location</span>
<span class="c1"># Change the file name if you wish to call it something else</span>

<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;/content/drive/My Drive/Proportion_of_C_allele_by_population_20210720.pdf&#39;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;/content/drive/My Drive/Proportion_of_C_allele_by_population_20210720.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">480</span><span class="p">)</span> <span class="c1"># increase the dpi for higher resolution</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="distribution-of-allelic-types-at-a-more-granular-level">
<h2>Distribution of allelic types at a more granular level<a class="headerlink" href="#distribution-of-allelic-types-at-a-more-granular-level" title="Permalink to this headline">¶</a></h2>
<p>Here we look at the population characteristics organized by country, first-level administrative division, and year.</p>
<p>Then, we identify key summary statistics, minimum and maximum proportions of C allele, within each year and country.</p>
<p>This table was not included in the Pf7 manuscript.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Group the df_calls by variables (Population, Country, Admin level 1, and Year)</span>
<span class="c1"># and apply the custom aggregation function &#39;my_agg&#39; to calculate the allele frequencies</span>
<span class="n">df_admin1_year_summary</span> <span class="o">=</span> <span class="n">df_calls</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Population&#39;</span><span class="p">,</span> <span class="s1">&#39;Country&#39;</span><span class="p">,</span> <span class="s1">&#39;Admin level 1&#39;</span><span class="p">,</span> <span class="s1">&#39;Year&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">my_agg</span><span class="p">)</span>
<span class="c1"># Print the shape of the resulting DataFrame</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df_admin1_year_summary</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="c1"># Calculate and print the maximum and minimum proportions of the C-type allele</span>
<span class="c1"># considering n = number of samples with an F or C type allele called</span>
<span class="c1"># min and max C allele proportions where n &gt;= 30</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Max proporiton C (n &gt; 30) = </span><span class="si">{</span><span class="n">df_admin1_year_summary</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df_admin1_year_summary</span><span class="p">[</span><span class="s1">&#39;Number F or C&#39;</span><span class="p">]</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="mi">30</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;Proportion C&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Min proporiton C (n &gt; 30) = </span><span class="si">{</span><span class="n">df_admin1_year_summary</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df_admin1_year_summary</span><span class="p">[</span><span class="s1">&#39;Number F or C&#39;</span><span class="p">]</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="mi">30</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;Proportion C&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="c1"># display 200 rows of resulting df where n &gt; 30</span>
<span class="n">pd</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">display</span><span class="o">.</span><span class="n">max_rows</span> <span class="o">=</span> <span class="mi">200</span>
<span class="n">df_admin1_year_summary</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df_admin1_year_summary</span><span class="p">[</span><span class="s1">&#39;Number F or C&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">30</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(297, 4)
Max proporiton C (n &gt; 30) = 0.8048780487804879
Min proporiton C (n &gt; 30) = 0.018518518518518517
</pre></div>
</div>
<div class="output text_html">
  <div id="df-81fa6b17-7202-4412-aff9-12a2bbc833f5" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th>Number F</th>
      <th>Number C</th>
      <th>Number F or C</th>
      <th>Proportion C</th>
    </tr>
    <tr>
      <th>Population</th>
      <th>Country</th>
      <th>Admin level 1</th>
      <th>Year</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="3" valign="top">AF-C</th>
      <th rowspan="3" valign="top">Democratic Republic of the Congo</th>
      <th rowspan="3" valign="top">Kinshasa</th>
      <th>2012.0</th>
      <td>84.0</td>
      <td>16.0</td>
      <td>100.0</td>
      <td>0.160000</td>
    </tr>
    <tr>
      <th>2013.0</th>
      <td>51.0</td>
      <td>13.0</td>
      <td>64.0</td>
      <td>0.203125</td>
    </tr>
    <tr>
      <th>2014.0</th>
      <td>31.0</td>
      <td>1.0</td>
      <td>32.0</td>
      <td>0.031250</td>
    </tr>
    <tr>
      <th rowspan="11" valign="top">AF-E</th>
      <th rowspan="6" valign="top">Kenya</th>
      <th rowspan="6" valign="top">Kilifi</th>
      <th>1996.0</th>
      <td>31.0</td>
      <td>9.0</td>
      <td>40.0</td>
      <td>0.225000</td>
    </tr>
    <tr>
      <th>2007.0</th>
      <td>42.0</td>
      <td>8.0</td>
      <td>50.0</td>
      <td>0.160000</td>
    </tr>
    <tr>
      <th>2009.0</th>
      <td>34.0</td>
      <td>5.0</td>
      <td>39.0</td>
      <td>0.128205</td>
    </tr>
    <tr>
      <th>2010.0</th>
      <td>47.0</td>
      <td>19.0</td>
      <td>66.0</td>
      <td>0.287879</td>
    </tr>
    <tr>
      <th>2011.0</th>
      <td>39.0</td>
      <td>8.0</td>
      <td>47.0</td>
      <td>0.170213</td>
    </tr>
    <tr>
      <th>2012.0</th>
      <td>49.0</td>
      <td>16.0</td>
      <td>65.0</td>
      <td>0.246154</td>
    </tr>
    <tr>
      <th>Malawi</th>
      <th>Chikwawa</th>
      <th>2011.0</th>
      <td>109.0</td>
      <td>23.0</td>
      <td>132.0</td>
      <td>0.174242</td>
    </tr>
    <tr>
      <th rowspan="4" valign="top">Tanzania</th>
      <th>Kigoma</th>
      <th>2014.0</th>
      <td>54.0</td>
      <td>28.0</td>
      <td>82.0</td>
      <td>0.341463</td>
    </tr>
    <tr>
      <th>Lindi</th>
      <th>2013.0</th>
      <td>30.0</td>
      <td>4.0</td>
      <td>34.0</td>
      <td>0.117647</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">Tanga</th>
      <th>2013.0</th>
      <td>75.0</td>
      <td>15.0</td>
      <td>90.0</td>
      <td>0.166667</td>
    </tr>
    <tr>
      <th>2014.0</th>
      <td>70.0</td>
      <td>27.0</td>
      <td>97.0</td>
      <td>0.278351</td>
    </tr>
    <tr>
      <th>AF-NE</th>
      <th>Sudan</th>
      <th>Khartoum</th>
      <th>2017.0</th>
      <td>17.0</td>
      <td>26.0</td>
      <td>43.0</td>
      <td>0.604651</td>
    </tr>
    <tr>
      <th rowspan="36" valign="top">AF-W</th>
      <th>Benin</th>
      <th>Littoral</th>
      <th>2016.0</th>
      <td>53.0</td>
      <td>1.0</td>
      <td>54.0</td>
      <td>0.018519</td>
    </tr>
    <tr>
      <th>Burkina Faso</th>
      <th>Haut-Bassins</th>
      <th>2008.0</th>
      <td>36.0</td>
      <td>2.0</td>
      <td>38.0</td>
      <td>0.052632</td>
    </tr>
    <tr>
      <th>Cameroon</th>
      <th>Sud-Ouest</th>
      <th>2013.0</th>
      <td>95.0</td>
      <td>26.0</td>
      <td>121.0</td>
      <td>0.214876</td>
    </tr>
    <tr>
      <th>Côte d'Ivoire</th>
      <th>Abidjan</th>
      <th>2013.0</th>
      <td>38.0</td>
      <td>2.0</td>
      <td>40.0</td>
      <td>0.050000</td>
    </tr>
    <tr>
      <th>Gabon</th>
      <th>Wouleu-Ntem</th>
      <th>2014.0</th>
      <td>31.0</td>
      <td>8.0</td>
      <td>39.0</td>
      <td>0.205128</td>
    </tr>
    <tr>
      <th rowspan="8" valign="top">Gambia</th>
      <th>North Bank</th>
      <th>1984.0</th>
      <td>53.0</td>
      <td>11.0</td>
      <td>64.0</td>
      <td>0.171875</td>
    </tr>
    <tr>
      <th rowspan="5" valign="top">Upper River</th>
      <th>2013.0</th>
      <td>25.0</td>
      <td>14.0</td>
      <td>39.0</td>
      <td>0.358974</td>
    </tr>
    <tr>
      <th>2014.0</th>
      <td>106.0</td>
      <td>62.0</td>
      <td>168.0</td>
      <td>0.369048</td>
    </tr>
    <tr>
      <th>2015.0</th>
      <td>33.0</td>
      <td>8.0</td>
      <td>41.0</td>
      <td>0.195122</td>
    </tr>
    <tr>
      <th>2016.0</th>
      <td>37.0</td>
      <td>13.0</td>
      <td>50.0</td>
      <td>0.260000</td>
    </tr>
    <tr>
      <th>2017.0</th>
      <td>36.0</td>
      <td>8.0</td>
      <td>44.0</td>
      <td>0.181818</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">Western</th>
      <th>2008.0</th>
      <td>43.0</td>
      <td>19.0</td>
      <td>62.0</td>
      <td>0.306452</td>
    </tr>
    <tr>
      <th>2015.0</th>
      <td>54.0</td>
      <td>26.0</td>
      <td>80.0</td>
      <td>0.325000</td>
    </tr>
    <tr>
      <th rowspan="13" valign="top">Ghana</th>
      <th rowspan="2" valign="top">Ashanti</th>
      <th>2014.0</th>
      <td>87.0</td>
      <td>18.0</td>
      <td>105.0</td>
      <td>0.171429</td>
    </tr>
    <tr>
      <th>2015.0</th>
      <td>61.0</td>
      <td>13.0</td>
      <td>74.0</td>
      <td>0.175676</td>
    </tr>
    <tr>
      <th>Central</th>
      <th>2014.0</th>
      <td>51.0</td>
      <td>12.0</td>
      <td>63.0</td>
      <td>0.190476</td>
    </tr>
    <tr>
      <th>Greater Accra</th>
      <th>2018.0</th>
      <td>64.0</td>
      <td>32.0</td>
      <td>96.0</td>
      <td>0.333333</td>
    </tr>
    <tr>
      <th rowspan="9" valign="top">Upper East</th>
      <th>2009.0</th>
      <td>51.0</td>
      <td>8.0</td>
      <td>59.0</td>
      <td>0.135593</td>
    </tr>
    <tr>
      <th>2010.0</th>
      <td>94.0</td>
      <td>5.0</td>
      <td>99.0</td>
      <td>0.050505</td>
    </tr>
    <tr>
      <th>2011.0</th>
      <td>45.0</td>
      <td>3.0</td>
      <td>48.0</td>
      <td>0.062500</td>
    </tr>
    <tr>
      <th>2013.0</th>
      <td>111.0</td>
      <td>8.0</td>
      <td>119.0</td>
      <td>0.067227</td>
    </tr>
    <tr>
      <th>2014.0</th>
      <td>22.0</td>
      <td>12.0</td>
      <td>34.0</td>
      <td>0.352941</td>
    </tr>
    <tr>
      <th>2015.0</th>
      <td>121.0</td>
      <td>33.0</td>
      <td>154.0</td>
      <td>0.214286</td>
    </tr>
    <tr>
      <th>2016.0</th>
      <td>110.0</td>
      <td>44.0</td>
      <td>154.0</td>
      <td>0.285714</td>
    </tr>
    <tr>
      <th>2017.0</th>
      <td>319.0</td>
      <td>64.0</td>
      <td>383.0</td>
      <td>0.167102</td>
    </tr>
    <tr>
      <th>2018.0</th>
      <td>356.0</td>
      <td>73.0</td>
      <td>429.0</td>
      <td>0.170163</td>
    </tr>
    <tr>
      <th>Guinea</th>
      <th>Nzerekore</th>
      <th>2011.0</th>
      <td>44.0</td>
      <td>11.0</td>
      <td>55.0</td>
      <td>0.200000</td>
    </tr>
    <tr>
      <th rowspan="6" valign="top">Mali</th>
      <th>Bamako</th>
      <th>2013.0</th>
      <td>80.0</td>
      <td>13.0</td>
      <td>93.0</td>
      <td>0.139785</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">Kayes</th>
      <th>2015.0</th>
      <td>35.0</td>
      <td>15.0</td>
      <td>50.0</td>
      <td>0.300000</td>
    </tr>
    <tr>
      <th>2016.0</th>
      <td>27.0</td>
      <td>12.0</td>
      <td>39.0</td>
      <td>0.307692</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">Koulikoro</th>
      <th>2013.0</th>
      <td>49.0</td>
      <td>11.0</td>
      <td>60.0</td>
      <td>0.183333</td>
    </tr>
    <tr>
      <th>2014.0</th>
      <td>40.0</td>
      <td>1.0</td>
      <td>41.0</td>
      <td>0.024390</td>
    </tr>
    <tr>
      <th>2016.0</th>
      <td>139.0</td>
      <td>37.0</td>
      <td>176.0</td>
      <td>0.210227</td>
    </tr>
    <tr>
      <th>Nigeria</th>
      <th>Lagos</th>
      <th>2017.0</th>
      <td>24.0</td>
      <td>14.0</td>
      <td>38.0</td>
      <td>0.368421</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">Senegal</th>
      <th>Dakar</th>
      <th>2013.0</th>
      <td>31.0</td>
      <td>13.0</td>
      <td>44.0</td>
      <td>0.295455</td>
    </tr>
    <tr>
      <th>Sedhiou</th>
      <th>2015.0</th>
      <td>43.0</td>
      <td>7.0</td>
      <td>50.0</td>
      <td>0.140000</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">AS-S-E</th>
      <th rowspan="2" valign="top">India</th>
      <th>Odisha</th>
      <th>2017.0</th>
      <td>28.0</td>
      <td>6.0</td>
      <td>34.0</td>
      <td>0.176471</td>
    </tr>
    <tr>
      <th>West Bengal</th>
      <th>2017.0</th>
      <td>40.0</td>
      <td>5.0</td>
      <td>45.0</td>
      <td>0.111111</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">AS-S-FE</th>
      <th rowspan="3" valign="top">Bangladesh</th>
      <th rowspan="3" valign="top">Chittagong</th>
      <th>2012.0</th>
      <td>23.0</td>
      <td>8.0</td>
      <td>31.0</td>
      <td>0.258065</td>
    </tr>
    <tr>
      <th>2015.0</th>
      <td>234.0</td>
      <td>132.0</td>
      <td>366.0</td>
      <td>0.360656</td>
    </tr>
    <tr>
      <th>2016.0</th>
      <td>408.0</td>
      <td>194.0</td>
      <td>602.0</td>
      <td>0.322259</td>
    </tr>
    <tr>
      <th rowspan="27" valign="top">AS-SE-E</th>
      <th rowspan="9" valign="top">Cambodia</th>
      <th>Pailin</th>
      <th>2011.0</th>
      <td>26.0</td>
      <td>12.0</td>
      <td>38.0</td>
      <td>0.315789</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">Preah Vihear</th>
      <th>2011.0</th>
      <td>49.0</td>
      <td>10.0</td>
      <td>59.0</td>
      <td>0.169492</td>
    </tr>
    <tr>
      <th>2012.0</th>
      <td>24.0</td>
      <td>8.0</td>
      <td>32.0</td>
      <td>0.250000</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">Pursat</th>
      <th>2010.0</th>
      <td>34.0</td>
      <td>49.0</td>
      <td>83.0</td>
      <td>0.590361</td>
    </tr>
    <tr>
      <th>2011.0</th>
      <td>30.0</td>
      <td>47.0</td>
      <td>77.0</td>
      <td>0.610390</td>
    </tr>
    <tr>
      <th>2013.0</th>
      <td>8.0</td>
      <td>27.0</td>
      <td>35.0</td>
      <td>0.771429</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">Ratanakiri</th>
      <th>2010.0</th>
      <td>23.0</td>
      <td>9.0</td>
      <td>32.0</td>
      <td>0.281250</td>
    </tr>
    <tr>
      <th>2011.0</th>
      <td>38.0</td>
      <td>20.0</td>
      <td>58.0</td>
      <td>0.344828</td>
    </tr>
    <tr>
      <th>2016.0</th>
      <td>17.0</td>
      <td>37.0</td>
      <td>54.0</td>
      <td>0.685185</td>
    </tr>
    <tr>
      <th rowspan="9" valign="top">Laos</th>
      <th rowspan="3" valign="top">Attapeu</th>
      <th>2011.0</th>
      <td>23.0</td>
      <td>13.0</td>
      <td>36.0</td>
      <td>0.361111</td>
    </tr>
    <tr>
      <th>2017.0</th>
      <td>39.0</td>
      <td>19.0</td>
      <td>58.0</td>
      <td>0.327586</td>
    </tr>
    <tr>
      <th>2018.0</th>
      <td>25.0</td>
      <td>20.0</td>
      <td>45.0</td>
      <td>0.444444</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">Champasak</th>
      <th>2017.0</th>
      <td>66.0</td>
      <td>15.0</td>
      <td>81.0</td>
      <td>0.185185</td>
    </tr>
    <tr>
      <th>2018.0</th>
      <td>54.0</td>
      <td>44.0</td>
      <td>98.0</td>
      <td>0.448980</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">Salavan</th>
      <th>2017.0</th>
      <td>71.0</td>
      <td>25.0</td>
      <td>96.0</td>
      <td>0.260417</td>
    </tr>
    <tr>
      <th>2018.0</th>
      <td>36.0</td>
      <td>4.0</td>
      <td>40.0</td>
      <td>0.100000</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">Savannakhet</th>
      <th>2017.0</th>
      <td>153.0</td>
      <td>66.0</td>
      <td>219.0</td>
      <td>0.301370</td>
    </tr>
    <tr>
      <th>2018.0</th>
      <td>59.0</td>
      <td>54.0</td>
      <td>113.0</td>
      <td>0.477876</td>
    </tr>
    <tr>
      <th rowspan="9" valign="top">Vietnam</th>
      <th rowspan="5" valign="top">Binh Phuoc</th>
      <th>2010.0</th>
      <td>33.0</td>
      <td>23.0</td>
      <td>56.0</td>
      <td>0.410714</td>
    </tr>
    <tr>
      <th>2011.0</th>
      <td>53.0</td>
      <td>21.0</td>
      <td>74.0</td>
      <td>0.283784</td>
    </tr>
    <tr>
      <th>2015.0</th>
      <td>24.0</td>
      <td>8.0</td>
      <td>32.0</td>
      <td>0.250000</td>
    </tr>
    <tr>
      <th>2017.0</th>
      <td>21.0</td>
      <td>19.0</td>
      <td>40.0</td>
      <td>0.475000</td>
    </tr>
    <tr>
      <th>2018.0</th>
      <td>85.0</td>
      <td>83.0</td>
      <td>168.0</td>
      <td>0.494048</td>
    </tr>
    <tr>
      <th>Dak Lak</th>
      <th>2017.0</th>
      <td>46.0</td>
      <td>41.0</td>
      <td>87.0</td>
      <td>0.471264</td>
    </tr>
    <tr>
      <th>Dak Nong</th>
      <th>2017.0</th>
      <td>18.0</td>
      <td>23.0</td>
      <td>41.0</td>
      <td>0.560976</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">Gia Lai</th>
      <th>2017.0</th>
      <td>104.0</td>
      <td>54.0</td>
      <td>158.0</td>
      <td>0.341772</td>
    </tr>
    <tr>
      <th>2018.0</th>
      <td>75.0</td>
      <td>32.0</td>
      <td>107.0</td>
      <td>0.299065</td>
    </tr>
    <tr>
      <th rowspan="13" valign="top">AS-SE-W</th>
      <th rowspan="5" valign="top">Myanmar</th>
      <th>Bago</th>
      <th>2012.0</th>
      <td>15.0</td>
      <td>21.0</td>
      <td>36.0</td>
      <td>0.583333</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">Kayin</th>
      <th>2016.0</th>
      <td>156.0</td>
      <td>182.0</td>
      <td>338.0</td>
      <td>0.538462</td>
    </tr>
    <tr>
      <th>2017.0</th>
      <td>117.0</td>
      <td>121.0</td>
      <td>238.0</td>
      <td>0.508403</td>
    </tr>
    <tr>
      <th>Mandalay</th>
      <th>2013.0</th>
      <td>14.0</td>
      <td>32.0</td>
      <td>46.0</td>
      <td>0.695652</td>
    </tr>
    <tr>
      <th>Tanintharyi</th>
      <th>2011.0</th>
      <td>8.0</td>
      <td>33.0</td>
      <td>41.0</td>
      <td>0.804878</td>
    </tr>
    <tr>
      <th rowspan="8" valign="top">Thailand</th>
      <th rowspan="8" valign="top">Tak</th>
      <th>2002.0</th>
      <td>27.0</td>
      <td>13.0</td>
      <td>40.0</td>
      <td>0.325000</td>
    </tr>
    <tr>
      <th>2003.0</th>
      <td>33.0</td>
      <td>27.0</td>
      <td>60.0</td>
      <td>0.450000</td>
    </tr>
    <tr>
      <th>2004.0</th>
      <td>31.0</td>
      <td>24.0</td>
      <td>55.0</td>
      <td>0.436364</td>
    </tr>
    <tr>
      <th>2005.0</th>
      <td>22.0</td>
      <td>15.0</td>
      <td>37.0</td>
      <td>0.405405</td>
    </tr>
    <tr>
      <th>2008.0</th>
      <td>109.0</td>
      <td>69.0</td>
      <td>178.0</td>
      <td>0.387640</td>
    </tr>
    <tr>
      <th>2011.0</th>
      <td>18.0</td>
      <td>18.0</td>
      <td>36.0</td>
      <td>0.500000</td>
    </tr>
    <tr>
      <th>2012.0</th>
      <td>34.0</td>
      <td>32.0</td>
      <td>66.0</td>
      <td>0.484848</td>
    </tr>
    <tr>
      <th>2013.0</th>
      <td>25.0</td>
      <td>48.0</td>
      <td>73.0</td>
      <td>0.657534</td>
    </tr>
    <tr>
      <th>OC-NG</th>
      <th>Papua New Guinea</th>
      <th>East Sepik</th>
      <th>2017.0</th>
      <td>30.0</td>
      <td>24.0</td>
      <td>54.0</td>
      <td>0.444444</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-81fa6b17-7202-4412-aff9-12a2bbc833f5')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-81fa6b17-7202-4412-aff9-12a2bbc833f5 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-81fa6b17-7202-4412-aff9-12a2bbc833f5');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-d190ef86-ef81-4bae-b1c4-82c104de7e32">
  <button class="colab-df-quickchart" onclick="quickchart('df-d190ef86-ef81-4bae-b1c4-82c104de7e32')"
            title="Suggest charts."
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-d190ef86-ef81-4bae-b1c4-82c104de7e32 button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>
    </div>
  </div>
</div></div>
</div>
<p><strong>Distribution of allelic types in each country, first-level administrative division and year.</strong>
This table presents a breakdown of longitudinal and geospatial data concerning the distribution of allelic types (F and C) across 97 combinations of first-level administrative divisions and years for which over 30 samples have either an F or a C type called. The fact that we see both types present in all 97 combinations of admin division and year adds weight to the argument that <em>eba175</em> is under balancing selection.</p>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./pf7/popgen"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="neighbour-joining-tree-swga.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Neighbour Joining Tree for sWGA vs. gDNA</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../api.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Pf7 API</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By MalariaGEN<br/>
    
        &copy; Copyright 2023.<br/>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>