
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Explore hrp2 and hrp3 deletion breakpoints &#8212; MalariaGEN parasite data user guide</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Drug resistance in Pf7 vs Pf8" href="Drug_Resistance_Pf7_Pf8_Concordance.html" />
    <link rel="prev" title="Plot copy number variation (CNV) trends" href="cnv_fig.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">MalariaGEN parasite data user guide</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../landing-page.html">
   MalariaGEN parasite data user guide
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Malaria Parasite Google Cloud Data Access
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../cloud_data_access/parasite-cloud-data-access.html">
   <strong>
    Malaria Parasite Google Cloud Data Access
   </strong>
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Pf8
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../Data_access.html">
   <strong>
    Pf8 Data Access
   </strong>
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../useful-how-tos.html">
   <strong>
    Pf8 Analysis Guides:
   </strong>
   Useful Genomics How-Tos
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../useful_genomics_howtos/igv_notebook.html">
     Visualizing genomic regions of Pf8 samples with IGV-notebook
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../summarise-data.html">
   <strong>
    Pf8 Analysis Guides:
   </strong>
   Summarise Pf8 Data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../summarise_data/table_sample_summary.html">
     Create sample location summary table
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../summarise_data/Visualising_sample_collections_over_time.html">
     Visualising sample collections over time
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../summarise_data/SamplesByCountry_Barplot.html">
     Plot sample collection per country
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../summarise_data/table_variants_summary.html">
     Summarise genetic variant types
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../summarise_data/figure1_samplemap.html">
     Create map of sample collection locations
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../evolution-malaria-control.html">
   <strong>
    Pf8 Analysis Guides:
   </strong>
   Study Evolution Against Malaria Control Measures
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="table_dr_summary.html">
     Plot drug resistance frequencies
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="drug_resistance_trends.html">
     Explore drug resistance trends in Africa
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="cnv_fig.html">
     Plot copy number variation (CNV) trends
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Explore
     <em>
      hrp2
     </em>
     and
     <em>
      hrp3
     </em>
     deletion breakpoints
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="Drug_Resistance_Pf7_Pf8_Concordance.html">
     Drug resistance in Pf7 vs Pf8
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../popgen.html">
   <strong>
    Pf8 Analysis Guides:
   </strong>
   Population Structure and Selection
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../popgen/PCoA.html">
     Principal Coordinate Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../popgen/njt_pf8_anjl.html">
     Create a Neighbour Joining Tree
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../popgen/Pf8_fws_figure_public_notebook.html">
     Explore Fws across populations
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../api.html">
   <strong>
    Pf8 API
   </strong>
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Pf7
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../pf7/Data_access.html">
   <strong>
    Pf7 Data Access
   </strong>
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../pf7/summarise-data.html">
   <strong>
    Pf7 Analysis Guides:
   </strong>
   Summarise Pf7 Data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pf7/summarise_data/sample-summary-table.html">
     Summarise Samples by Administrative Division
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pf7/summarise_data/sample-collection-over-time.html">
     Visualising sample collections over time
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pf7/summarise_data/barplot-samples-by-country.html">
     Barplot of Samples by Country
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pf7/summarise_data/variant-summary-table.html">
     Summarise Genetic Variants
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pf7/summarise_data/global-sample-map.html">
     Plot Global Map of Samples
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../pf7/evolution-malaria-control.html">
   <strong>
    Pf7 Analysis Guides:
   </strong>
   Study Evolution Against Malaria Control Measures
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pf7/evolution_malaria_control/drug-resistance-summary.html">
     Drug Resistance Frequencies Across Pf7
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pf7/evolution_malaria_control/chloroquine-resistance-map.html">
     Plot Map of Chloroquine Resistance
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pf7/evolution_malaria_control/summarise_hrp2_hrp3_deletions.html">
     Summarise
     <em>
      hrp2
     </em>
     and
     <em>
      hrp3
     </em>
     Deletions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pf7/evolution_malaria_control/visualise_hrp2_hrp3_deletions.html">
     Visualise
     <em>
      hrp2
     </em>
     and
     <em>
      hrp3
     </em>
     Deletions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pf7/evolution_malaria_control/csp-cterm-analysis.html">
     Analyse csp C-Terminal Variation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../pf7/popgen.html">
   <strong>
    Pf7 Analysis Guides:
   </strong>
   Population Structure and Selection
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pf7/popgen/PCoA.html">
     Principal Coordinate Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pf7/popgen/geographic-genetic-differentiation.html">
     Geographic patterns of population differentiation and gene flow
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pf7/popgen/neighbour-joining-tree-swga.html">
     Neighbour Joining Tree for sWGA vs. gDNA
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pf7/popgen/proportion-eba175.html">
     Plot
     <em>
      eba175
     </em>
     C Allele Proportions
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../pf7/api.html">
   <strong>
    Pf7 API
   </strong>
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Pv4
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../pv4/Data_access.html">
   <strong>
    Pv4 Data Access
   </strong>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../pv4/api.html">
   <strong>
    Pv4 API
   </strong>
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/pf8/evolution_malaria_control/Figure_HRPDeletion.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
                onclick="printPdf(this)" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/malariagen/parasite-data"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/malariagen/parasite-data/issues/new?title=Issue%20on%20page%20%2Fpf8/evolution_malaria_control/Figure_HRPDeletion.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/malariagen/parasite-data/master?urlpath=lab/tree/docs/pf8/evolution_malaria_control/Figure_HRPDeletion.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        <a class="colab-button" href="https://colab.research.google.com/github/malariagen/parasite-data/blob/master/docs/pf8/evolution_malaria_control/Figure_HRPDeletion.ipynb"><button type="button" class="btn btn-secondary topbarbtn"
                title="Launch Colab" data-toggle="tooltip" data-placement="left"><img class="colab-button-logo"
                    src="../../_static/images/logo_colab.png"
                    alt="Interact on Colab">Colab</button></a>
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#access-pf8-data">
   Access Pf8 Data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#hrp2-hrp3-deletions">
   <em>
    hrp2
   </em>
   &amp;
   <em>
    hrp3
   </em>
   Deletions
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#d7-reference-genome-annotation">
   3D7 Reference Genome Annotation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#figure-preparation">
   Figure Preparation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#create-figure">
   Create Figure
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#save-figure">
     Save Figure
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Explore hrp2 and hrp3 deletion breakpoints</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#access-pf8-data">
   Access Pf8 Data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#hrp2-hrp3-deletions">
   <em>
    hrp2
   </em>
   &amp;
   <em>
    hrp3
   </em>
   Deletions
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#d7-reference-genome-annotation">
   3D7 Reference Genome Annotation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#figure-preparation">
   Figure Preparation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#create-figure">
   Create Figure
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#save-figure">
     Save Figure
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            
              <div>
                
  <div class="section" id="explore-hrp2-and-hrp3-deletion-breakpoints">
<h1>Explore <em>hrp2</em> and <em>hrp3</em> deletion breakpoints<a class="headerlink" href="#explore-hrp2-and-hrp3-deletion-breakpoints" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>This notebook will create a figure of histidine-rich protein (<em>hrp</em>) 2 and <em>hrp3</em> deletion breakpoints.</p>
<p><em>hrp2</em> and <em>hrp3</em> are genes located in subtelomeric regions of the genome with very high levels of natural variation. Deletion in those genes can cause failure of rapid diagnostic tests and is therefore important to monitor.</p>
<p>Deletion is a genetic event in which a segment of DNA is entirely removed or missing. In this context, ‘breakpoints’ denote specific locations on the chromosome where such deletions take place.</p>
<p><strong>This notebook should take approximately two minutes to run.</strong></p>
</div>
<div class="section" id="setup">
<h2>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">¶</a></h2>
<p>Install and import the malariagen Python package:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>pip<span class="w"> </span>install<span class="w"> </span>malariagen_data<span class="w"> </span>-q<span class="w"> </span>--no-warn-conflicts
<span class="kn">import</span><span class="w"> </span><span class="nn">malariagen_data</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  Installing build dependencies ... ?25l?25hdone
  Getting requirements to build wheel ... ?25l?25hdone
  Preparing metadata (pyproject.toml) ... ?25l?25hdone
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">4.0/4.0 MB</span> <span class=" -Color -Color-Red">33.3 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
?25h  Preparing metadata (setup.py) ... ?25l?25hdone
  Preparing metadata (setup.py) ... ?25l?25hdone
  Preparing metadata (setup.py) ... ?25l?25hdone
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">71.7/71.7 kB</span> <span class=" -Color -Color-Red">5.1 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">775.9/775.9 kB</span> <span class=" -Color -Color-Red">33.6 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">25.9/25.9 MB</span> <span class=" -Color -Color-Red">66.8 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">8.7/8.7 MB</span> <span class=" -Color -Color-Red">93.8 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">210.6/210.6 kB</span> <span class=" -Color -Color-Red">14.3 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">6.3/6.3 MB</span> <span class=" -Color -Color-Red">80.2 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">3.3/3.3 MB</span> <span class=" -Color -Color-Red">64.3 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">8.0/8.0 MB</span> <span class=" -Color -Color-Red">86.1 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">78.3/78.3 kB</span> <span class=" -Color -Color-Red">5.4 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">101.7/101.7 kB</span> <span class=" -Color -Color-Red">7.6 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">8.9/8.9 MB</span> <span class=" -Color -Color-Red">93.8 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">228.0/228.0 kB</span> <span class=" -Color -Color-Red">16.4 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">13.4/13.4 MB</span> <span class=" -Color -Color-Red">92.7 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">1.6/1.6 MB</span> <span class=" -Color -Color-Red">60.1 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
?25h  Building wheel for malariagen_data (pyproject.toml) ... ?25l?25hdone
  Building wheel for dash-cytoscape (setup.py) ... ?25l?25hdone
  Building wheel for stringcase (setup.py) ... ?25l?25hdone
  Building wheel for asciitree (setup.py) ... ?25l?25hdone
</pre></div>
</div>
</div>
</div>
<p>Import required python libraries that are installed at colab by default.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">collections</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">google.colab</span><span class="w"> </span><span class="kn">import</span> <span class="n">drive</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="access-pf8-data">
<h2>Access Pf8 Data<a class="headerlink" href="#access-pf8-data" title="Permalink to this headline">¶</a></h2>
<p>We use the malariagen data package to load the release data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">release_data</span> <span class="o">=</span> <span class="n">malariagen_data</span><span class="o">.</span><span class="n">Pf8</span><span class="p">()</span>
<span class="n">df_samples</span> <span class="o">=</span> <span class="n">release_data</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="hrp2-hrp3-deletions">
<h2><em>hrp2</em> &amp; <em>hrp3</em> Deletions<a class="headerlink" href="#hrp2-hrp3-deletions" title="Permalink to this headline">¶</a></h2>
<p>We additionally require list of deletion and breakpoint locations within the <em>hrp2</em> and <em>hrp3</em> across 24,409 QC-pass samples. We can access this data along with other copy-number variation (CNV) calls from <a class="reference external" href="https://pf8-release.cog.sanger.ac.uk/Pf8_cnv_calls.tsv">Sanger cloud storage</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Read data directly from url</span>
<span class="n">hrp_calls_fn</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;https://pf8-release.cog.sanger.ac.uk/Pf8_cnv_calls.tsv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> 

<span class="c1"># Print the shape and first rows</span>
<span class="nb">print</span><span class="p">(</span><span class="n">hrp_calls_fn</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">hrp_calls_fn</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(24409, 29)
</pre></div>
</div>
<div class="output text_html">
  <div id="df-20c60bcb-31a5-4616-83fc-125439041e03" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Sample</th>
      <th>CRT_uncurated_coverage_only</th>
      <th>CRT_curated_coverage_only</th>
      <th>CRT_breakpoint</th>
      <th>CRT_faceaway_only</th>
      <th>CRT_final_amplification_call</th>
      <th>GCH1_uncurated_coverage_only</th>
      <th>GCH1_curated_coverage_only</th>
      <th>GCH1_breakpoint</th>
      <th>GCH1_faceaway_only</th>
      <th>...</th>
      <th>PM2_PM3_faceaway_only</th>
      <th>PM2_PM3_final_amplification_call</th>
      <th>HRP2_uncurated_coverage_only</th>
      <th>HRP2_breakpoint</th>
      <th>HRP2_deletion_type</th>
      <th>HRP2_final_deletion_call</th>
      <th>HRP3_uncurated_coverage_only</th>
      <th>HRP3_breakpoint</th>
      <th>HRP3_deletion_type</th>
      <th>HRP3_final_deletion_call</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>FP0008-C</td>
      <td>0</td>
      <td>0</td>
      <td>-</td>
      <td>-1</td>
      <td>0</td>
      <td>-1</td>
      <td>-1</td>
      <td>-</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>-</td>
      <td>-</td>
      <td>0</td>
      <td>0</td>
      <td>-</td>
      <td>-</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>FP0009-C</td>
      <td>0</td>
      <td>0</td>
      <td>-</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>-</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>-</td>
      <td>-</td>
      <td>0</td>
      <td>0</td>
      <td>-</td>
      <td>-</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>FP0010-CW</td>
      <td>-1</td>
      <td>-1</td>
      <td>-</td>
      <td>0</td>
      <td>0</td>
      <td>-1</td>
      <td>-1</td>
      <td>-</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>-1</td>
      <td>-</td>
      <td>-</td>
      <td>-1</td>
      <td>-1</td>
      <td>-</td>
      <td>-</td>
      <td>-1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>FP0011-CW</td>
      <td>-1</td>
      <td>-1</td>
      <td>-</td>
      <td>-1</td>
      <td>-1</td>
      <td>-1</td>
      <td>-1</td>
      <td>-</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>-1</td>
      <td>-</td>
      <td>-</td>
      <td>-1</td>
      <td>-1</td>
      <td>-</td>
      <td>-</td>
      <td>-1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>FP0012-CW</td>
      <td>-1</td>
      <td>-1</td>
      <td>-</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>-</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>-1</td>
      <td>-</td>
      <td>-</td>
      <td>-1</td>
      <td>0</td>
      <td>-</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 29 columns</p>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-20c60bcb-31a5-4616-83fc-125439041e03')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-20c60bcb-31a5-4616-83fc-125439041e03 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-20c60bcb-31a5-4616-83fc-125439041e03');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-52f3f7cc-542a-45a3-9604-59b84d8b971c">
  <button class="colab-df-quickchart" onclick="quickchart('df-52f3f7cc-542a-45a3-9604-59b84d8b971c')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-52f3f7cc-542a-45a3-9604-59b84d8b971c button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>

    </div>
  </div>
</div></div>
</div>
<p>Now, let’s merge <code class="docutils literal notranslate"><span class="pre">hrp_calls_fn</span></code> with <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> which contains various metadata of Pf8 samples.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Merge df_samples with hrp_calls_fn</span>
<span class="n">df_samples</span> <span class="o">=</span>  <span class="n">df_samples</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">hrp_calls_fn</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="d7-reference-genome-annotation">
<h2>3D7 Reference Genome Annotation<a class="headerlink" href="#d7-reference-genome-annotation" title="Permalink to this headline">¶</a></h2>
<p>We would like to know where breakpoints occur in the genome, such as whether they fall within protein-coding regions or mRNA regions.</p>
<p>To facilitate this, we will use the 3D7 reference genome annotation. This data is in a tabular format where each row specifies a genomic feature, such as an exon, mRNA, or protein-coding gene, along with its corresponding coordinates (<code class="docutils literal notranslate"><span class="pre">start</span></code> and <code class="docutils literal notranslate"><span class="pre">end</span></code> columns). For more information about the annotation data format, refer to this <a class="reference external" href="https://en.wikipedia.org/wiki/General_feature_format#:~:text=In%20bioinformatics%2C%20the%20general%20feature,.gff%20%2C%20.gff3">wiki page</a>.</p>
<p>This dataset is available through the malariagen_data package.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_gff</span><span class="o">=</span> <span class="n">release_data</span><span class="o">.</span><span class="n">genome_features</span><span class="p">()</span>

<span class="c1"># print first rows</span>
<span class="n">df_gff</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-195459e2-d535-44c6-b53f-f59ea0317678" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>contig</th>
      <th>source</th>
      <th>type</th>
      <th>start</th>
      <th>end</th>
      <th>score</th>
      <th>strand</th>
      <th>phase</th>
      <th>ID</th>
      <th>Parent</th>
      <th>Name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>protein_coding_gene</td>
      <td>624510</td>
      <td>626292</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1314600</td>
      <td>NaN</td>
      <td>LipL1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>mRNA</td>
      <td>624510</td>
      <td>626292</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1314600.1</td>
      <td>PF3D7_1314600</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>624510</td>
      <td>626292</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1314600.1-E1</td>
      <td>PF3D7_1314600.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>624785</td>
      <td>626011</td>
      <td>NaN</td>
      <td>+</td>
      <td>0.0</td>
      <td>PF3D7_1314600.1-p1-CDS1</td>
      <td>PF3D7_1314600.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>five_prime_UTR</td>
      <td>624510</td>
      <td>624784</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>utr_PF3D7_1314600.1_1</td>
      <td>PF3D7_1314600.1</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-195459e2-d535-44c6-b53f-f59ea0317678')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-195459e2-d535-44c6-b53f-f59ea0317678 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-195459e2-d535-44c6-b53f-f59ea0317678');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-de2be48b-f76b-490b-9e0a-a59244f3c71a">
  <button class="colab-df-quickchart" onclick="quickchart('df-de2be48b-f76b-490b-9e0a-a59244f3c71a')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-de2be48b-f76b-490b-9e0a-a59244f3c71a button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>

    </div>
  </div>
</div></div>
</div>
</div>
<div class="section" id="figure-preparation">
<h2>Figure Preparation<a class="headerlink" href="#figure-preparation" title="Permalink to this headline">¶</a></h2>
<p>We need to find the start and end positions of chromosomes (‘contig’ column in the dataframe) to draw gene annotations in the figure that we are going to create.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Find start and end positions for each chromosome by grouping chromosome coordinates</span>
<span class="n">df_chroms</span> <span class="o">=</span> <span class="n">df_gff</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;contig&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span><span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="s1">&#39;max&#39;</span><span class="p">})</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="c1"># Set &#39;chrom&#39; as the index</span>
<span class="n">df_chroms</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;contig&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">df_chroms</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-4167807d-c594-4a9f-bc47-dd41c136e189" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>start</th>
      <th>end</th>
    </tr>
    <tr>
      <th>contig</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Pf3D7_01_v3</th>
      <td>29510</td>
      <td>614893</td>
    </tr>
    <tr>
      <th>Pf3D7_02_v3</th>
      <td>25232</td>
      <td>923648</td>
    </tr>
    <tr>
      <th>Pf3D7_03_v3</th>
      <td>36965</td>
      <td>1038254</td>
    </tr>
    <tr>
      <th>Pf3D7_04_v3</th>
      <td>28706</td>
      <td>1180226</td>
    </tr>
    <tr>
      <th>Pf3D7_05_v3</th>
      <td>20929</td>
      <td>1342964</td>
    </tr>
    <tr>
      <th>Pf3D7_06_v3</th>
      <td>653</td>
      <td>1382627</td>
    </tr>
    <tr>
      <th>Pf3D7_07_v3</th>
      <td>20307</td>
      <td>1426234</td>
    </tr>
    <tr>
      <th>Pf3D7_08_v3</th>
      <td>21361</td>
      <td>1443449</td>
    </tr>
    <tr>
      <th>Pf3D7_09_v3</th>
      <td>20080</td>
      <td>1503336</td>
    </tr>
    <tr>
      <th>Pf3D7_10_v3</th>
      <td>28490</td>
      <td>1649948</td>
    </tr>
    <tr>
      <th>Pf3D7_11_v3</th>
      <td>24160</td>
      <td>2035886</td>
    </tr>
    <tr>
      <th>Pf3D7_12_v3</th>
      <td>16973</td>
      <td>2248962</td>
    </tr>
    <tr>
      <th>Pf3D7_13_v3</th>
      <td>21364</td>
      <td>2892340</td>
    </tr>
    <tr>
      <th>Pf3D7_14_v3</th>
      <td>1393</td>
      <td>3291501</td>
    </tr>
    <tr>
      <th>Pf3D7_API_v3</th>
      <td>1</td>
      <td>34225</td>
    </tr>
    <tr>
      <th>Pf3D7_MIT_v3</th>
      <td>3</td>
      <td>5954</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-4167807d-c594-4a9f-bc47-dd41c136e189')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-4167807d-c594-4a9f-bc47-dd41c136e189 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-4167807d-c594-4a9f-bc47-dd41c136e189');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-676f97a0-d098-4306-b94c-35c438a005a8">
  <button class="colab-df-quickchart" onclick="quickchart('df-676f97a0-d098-4306-b94c-35c438a005a8')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-676f97a0-d098-4306-b94c-35c438a005a8 button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>

  <div id="id_615c09b0-0c9f-49cf-bb50-f256bb37ac17">
    <style>
      .colab-df-generate {
        background-color: #E8F0FE;
        border: none;
        border-radius: 50%;
        cursor: pointer;
        display: none;
        fill: #1967D2;
        height: 32px;
        padding: 0 0 0 0;
        width: 32px;
      }

      .colab-df-generate:hover {
        background-color: #E2EBFA;
        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
        fill: #174EA6;
      }

      [theme=dark] .colab-df-generate {
        background-color: #3B4455;
        fill: #D2E3FC;
      }

      [theme=dark] .colab-df-generate:hover {
        background-color: #434B5C;
        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
        fill: #FFFFFF;
      }
    </style>
    <button class="colab-df-generate" onclick="generateWithVariable('df_chroms')"
            title="Generate code using this dataframe."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
       width="24px">
    <path d="M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z"/>
  </svg>
    </button>
    <script>
      (() => {
      const buttonEl =
        document.querySelector('#id_615c09b0-0c9f-49cf-bb50-f256bb37ac17 button.colab-df-generate');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      buttonEl.onclick = () => {
        google.colab.notebook.generateWithVariable('df_chroms');
      }
      })();
    </script>
  </div>

    </div>
  </div>
</div></div>
</div>
<p>The next question is: How many samples have deletions in each country?</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># This function returns samples with deletion for each country.</span>

<span class="k">def</span><span class="w"> </span><span class="nf">breakpoint_agg</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
    <span class="n">names</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">()</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;Countries&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="n">countries</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># Loop over each country</span>
    <span class="c1"># Count non-zero samples (with deletion)</span>
    <span class="k">for</span> <span class="n">country</span> <span class="ow">in</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Country&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
        <span class="n">countries</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">country</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;Country&#39;</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">country</span><span class="p">)</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
    <span class="c1"># Join together country name and number</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;Countries&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">countries</span><span class="p">)</span>
    <span class="n">names</span><span class="p">[</span><span class="s1">&#39;Samples with deletion&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">names</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We will apply <code class="docutils literal notranslate"><span class="pre">breakpoint_agg</span></code> function to <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> separately for <em>hrp2</em> and <em>hrp3</em>.</p>
<p>Additionally, we will find genomic coordinates at the edges for mapping in the figure.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Group samples by Deletion type and HRP3_breakpoint</span>
<span class="c1"># Apply breakpoint_agg to count samples with deletion in each country</span>
<span class="n">df_hrp2</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">df_samples</span><span class="p">[</span>
        <span class="n">df_samples</span><span class="p">[</span><span class="s1">&#39;QC pass&#39;</span><span class="p">]</span>
        <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_samples</span><span class="p">[</span><span class="s1">&#39;HRP2_final_deletion_call&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span> <span class="p">)</span>
    <span class="p">]</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">Gene</span><span class="o">=</span><span class="s1">&#39;$hrp2$&#39;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;HRP2_deletion_type&#39;</span><span class="p">:</span> <span class="s1">&#39;Deletion type&#39;</span><span class="p">})</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Deletion type&#39;</span><span class="p">,</span> <span class="s1">&#39;HRP2_breakpoint&#39;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">breakpoint_agg</span><span class="p">,</span> <span class="n">include_groups</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="p">)</span>
<span class="c1"># Seperate coordinate value from chromosome</span>
<span class="n">df_hrp2</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_hrp2</span><span class="p">[</span><span class="s1">&#39;HRP2_breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]))</span>

<span class="c1"># Print min and max coordinates</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;HRP2 min breakpoint = </span><span class="si">{</span><span class="n">df_hrp2</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;HRP2 max breakpoint = </span><span class="si">{</span><span class="n">df_hrp2</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">df_hrp2</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>HRP2 min breakpoint = 1373732
HRP2 max breakpoint = 1374986
</pre></div>
</div>
<div class="output text_html">
  <div id="df-ae1370e9-26ec-4d21-8aad-0b2cc9a74fc0" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Deletion type</th>
      <th>HRP2_breakpoint</th>
      <th>Countries</th>
      <th>Samples with deletion</th>
      <th>breakpoint</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Telomere healing</td>
      <td>Pf3D7_08_v3:1373732</td>
      <td>Cambodia (1)</td>
      <td>1</td>
      <td>1373732</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Telomere healing</td>
      <td>Pf3D7_08_v3:1374280</td>
      <td>Sudan (1)</td>
      <td>1</td>
      <td>1374280</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Telomere healing</td>
      <td>Pf3D7_08_v3:1374462</td>
      <td>Indonesia (5)</td>
      <td>5</td>
      <td>1374462</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Telomere healing</td>
      <td>Pf3D7_08_v3:1374932</td>
      <td>Peru (2)</td>
      <td>2</td>
      <td>1374932</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Telomere healing</td>
      <td>Pf3D7_08_v3:1374986</td>
      <td>Peru (6)</td>
      <td>6</td>
      <td>1374986</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-ae1370e9-26ec-4d21-8aad-0b2cc9a74fc0')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-ae1370e9-26ec-4d21-8aad-0b2cc9a74fc0 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-ae1370e9-26ec-4d21-8aad-0b2cc9a74fc0');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-a91b54d8-b4a2-419c-9d5d-26168cd4e35a">
  <button class="colab-df-quickchart" onclick="quickchart('df-a91b54d8-b4a2-419c-9d5d-26168cd4e35a')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-a91b54d8-b4a2-419c-9d5d-26168cd4e35a button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>

  <div id="id_f9a285d0-3b46-4c38-b593-f7104a425472">
    <style>
      .colab-df-generate {
        background-color: #E8F0FE;
        border: none;
        border-radius: 50%;
        cursor: pointer;
        display: none;
        fill: #1967D2;
        height: 32px;
        padding: 0 0 0 0;
        width: 32px;
      }

      .colab-df-generate:hover {
        background-color: #E2EBFA;
        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
        fill: #174EA6;
      }

      [theme=dark] .colab-df-generate {
        background-color: #3B4455;
        fill: #D2E3FC;
      }

      [theme=dark] .colab-df-generate:hover {
        background-color: #434B5C;
        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
        fill: #FFFFFF;
      }
    </style>
    <button class="colab-df-generate" onclick="generateWithVariable('df_hrp2')"
            title="Generate code using this dataframe."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
       width="24px">
    <path d="M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z"/>
  </svg>
    </button>
    <script>
      (() => {
      const buttonEl =
        document.querySelector('#id_f9a285d0-3b46-4c38-b593-f7104a425472 button.colab-df-generate');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      buttonEl.onclick = () => {
        google.colab.notebook.generateWithVariable('df_hrp2');
      }
      })();
    </script>
  </div>

    </div>
  </div>
</div></div>
</div>
<p>A repeat of the same look-up for <em>hrp3</em>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Group samples by Deletion type and HRP3_breakpoint</span>
<span class="c1"># Apply breakpoint_agg to count samples with deletion in each country</span>
<span class="n">df_hrp3</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">df_samples</span><span class="p">[</span>
        <span class="n">df_samples</span><span class="p">[</span><span class="s1">&#39;QC pass&#39;</span><span class="p">]</span>
        <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_samples</span><span class="p">[</span><span class="s1">&#39;HRP3_final_deletion_call&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span> <span class="p">)</span>
    <span class="p">]</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;HRP3_deletion_type&#39;</span><span class="p">:</span> <span class="s1">&#39;Deletion type&#39;</span><span class="p">})</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Deletion type&#39;</span><span class="p">,</span> <span class="s1">&#39;HRP3_breakpoint&#39;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">breakpoint_agg</span><span class="p">,</span> <span class="n">include_groups</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="p">)</span>
<span class="c1"># Seperate coordinate value from chromosome</span>
<span class="n">df_hrp3</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_hrp3</span><span class="p">[</span><span class="s1">&#39;HRP3_breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>

<span class="c1"># Print min and max coordinates</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;HRP3 min breakpoint = </span><span class="si">{</span><span class="n">df_hrp3</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;HRP3 max breakpoint = </span><span class="si">{</span><span class="n">df_hrp3</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">df_hrp3</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>HRP3 min breakpoint = 2800004-2807159
HRP3 max breakpoint = 2841120
</pre></div>
</div>
<div class="output text_html">
  <div id="df-fd304d2e-8966-46a3-9b9e-2088260b089f" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Deletion type</th>
      <th>HRP3_breakpoint</th>
      <th>Countries</th>
      <th>Samples with deletion</th>
      <th>breakpoint</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Chromosome 11 recombination</td>
      <td>Pf3D7_13_v3:2800004-2807159</td>
      <td>Thailand (1), Ghana (3), Indonesia (38), Peru ...</td>
      <td>162</td>
      <td>2800004-2807159</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Chromosome 5 recombination</td>
      <td>Pf3D7_13_v3:2835587-2835612</td>
      <td>Cambodia (20), Vietnam (1)</td>
      <td>21</td>
      <td>2835587-2835612</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2811525</td>
      <td>India (1)</td>
      <td>1</td>
      <td>2811525</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2812344</td>
      <td>Sudan (1)</td>
      <td>1</td>
      <td>2812344</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2815249</td>
      <td>Tanzania (1)</td>
      <td>1</td>
      <td>2815249</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2822480</td>
      <td>Ghana (1)</td>
      <td>1</td>
      <td>2822480</td>
    </tr>
    <tr>
      <th>6</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2823645</td>
      <td>Kenya (1)</td>
      <td>1</td>
      <td>2823645</td>
    </tr>
    <tr>
      <th>7</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2830952</td>
      <td>Cambodia (7)</td>
      <td>7</td>
      <td>2830952</td>
    </tr>
    <tr>
      <th>8</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2834604</td>
      <td>Vietnam (1)</td>
      <td>1</td>
      <td>2834604</td>
    </tr>
    <tr>
      <th>9</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2835532</td>
      <td>Thailand (1)</td>
      <td>1</td>
      <td>2835532</td>
    </tr>
    <tr>
      <th>10</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2835899</td>
      <td>Ghana (1)</td>
      <td>1</td>
      <td>2835899</td>
    </tr>
    <tr>
      <th>11</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2837144</td>
      <td>Vietnam (8)</td>
      <td>8</td>
      <td>2837144</td>
    </tr>
    <tr>
      <th>12</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2837392</td>
      <td>Cambodia (2), Laos (1)</td>
      <td>3</td>
      <td>2837392</td>
    </tr>
    <tr>
      <th>13</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2838654</td>
      <td>Indonesia (2)</td>
      <td>2</td>
      <td>2838654</td>
    </tr>
    <tr>
      <th>14</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2840859</td>
      <td>Gambia (2)</td>
      <td>2</td>
      <td>2840859</td>
    </tr>
    <tr>
      <th>15</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2841024</td>
      <td>Thailand (1)</td>
      <td>1</td>
      <td>2841024</td>
    </tr>
    <tr>
      <th>16</th>
      <td>Telomere healing</td>
      <td>Pf3D7_13_v3:2841120</td>
      <td>Indonesia (1)</td>
      <td>1</td>
      <td>2841120</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-fd304d2e-8966-46a3-9b9e-2088260b089f')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-fd304d2e-8966-46a3-9b9e-2088260b089f button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-fd304d2e-8966-46a3-9b9e-2088260b089f');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-913fda83-407b-4b56-bd85-5c62c3ecaebd">
  <button class="colab-df-quickchart" onclick="quickchart('df-913fda83-407b-4b56-bd85-5c62c3ecaebd')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-913fda83-407b-4b56-bd85-5c62c3ecaebd button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>

  <div id="id_575a564d-9cd5-4f4e-a798-89daae1e9213">
    <style>
      .colab-df-generate {
        background-color: #E8F0FE;
        border: none;
        border-radius: 50%;
        cursor: pointer;
        display: none;
        fill: #1967D2;
        height: 32px;
        padding: 0 0 0 0;
        width: 32px;
      }

      .colab-df-generate:hover {
        background-color: #E2EBFA;
        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
        fill: #174EA6;
      }

      [theme=dark] .colab-df-generate {
        background-color: #3B4455;
        fill: #D2E3FC;
      }

      [theme=dark] .colab-df-generate:hover {
        background-color: #434B5C;
        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
        fill: #FFFFFF;
      }
    </style>
    <button class="colab-df-generate" onclick="generateWithVariable('df_hrp3')"
            title="Generate code using this dataframe."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
       width="24px">
    <path d="M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z"/>
  </svg>
    </button>
    <script>
      (() => {
      const buttonEl =
        document.querySelector('#id_575a564d-9cd5-4f4e-a798-89daae1e9213 button.colab-df-generate');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      buttonEl.onclick = () => {
        google.colab.notebook.generateWithVariable('df_hrp3');
      }
      })();
    </script>
  </div>

    </div>
  </div>
</div></div>
</div>
<p>It seems that Chromosome 11 recombination breakpoints are observed in multiple countries which motivates us to print the full list of countries.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Locate the first row by using the index</span>
<span class="n">df_hrp3</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;Countries&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;Thailand (1), Ghana (3), Indonesia (38), Peru (15), Bangladesh (1), Vietnam (2), Colombia (50), Ethiopia (9), Senegal (6), Laos (19), Cambodia (3), Sudan (6), Mali (1), Gambia (7), Kenya (1)&#39;
</pre></div>
</div>
</div>
</div>
<p>Given the fact that many of these events result in the deletion of other genes in addition to <em>hrp2</em> and <em>hrp3</em>.</p>
<p>We could have a look at which genes are present within the range of breakpoints before including them in the plot.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Genes in hrp2 breakpoints</span>
<span class="n">df_gff</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;contig&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Pf3D7_08_v3&#39;</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">1375500</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">1364000</span> <span class="p">)</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-d828440d-0631-4464-be1c-a8f5516001a6" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>contig</th>
      <th>source</th>
      <th>type</th>
      <th>start</th>
      <th>end</th>
      <th>score</th>
      <th>strand</th>
      <th>phase</th>
      <th>ID</th>
      <th>Parent</th>
      <th>Name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>17090</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>protein_coding_gene</td>
      <td>1373212</td>
      <td>1376988</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>PF3D7_0831800</td>
      <td>NaN</td>
      <td>HRP2</td>
    </tr>
    <tr>
      <th>17091</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>mRNA</td>
      <td>1373212</td>
      <td>1376988</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>PF3D7_0831800.1</td>
      <td>PF3D7_0831800</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17092</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>1373212</td>
      <td>1375084</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>exon_PF3D7_0831800.1-E2</td>
      <td>PF3D7_0831800.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17093</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>1375231</td>
      <td>1376988</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>exon_PF3D7_0831800.1-E1</td>
      <td>PF3D7_0831800.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17094</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>1374236</td>
      <td>1375084</td>
      <td>NaN</td>
      <td>-</td>
      <td>0.0</td>
      <td>PF3D7_0831800.1-p1-CDS2</td>
      <td>PF3D7_0831800.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17095</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>1375231</td>
      <td>1375299</td>
      <td>NaN</td>
      <td>-</td>
      <td>0.0</td>
      <td>PF3D7_0831800.1-p1-CDS1</td>
      <td>PF3D7_0831800.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17096</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>three_prime_UTR</td>
      <td>1373212</td>
      <td>1374235</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>utr_PF3D7_0831800.1_1</td>
      <td>PF3D7_0831800.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17097</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>five_prime_UTR</td>
      <td>1375300</td>
      <td>1376988</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>utr_PF3D7_0831800.1_2</td>
      <td>PF3D7_0831800.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>27865</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>protein_coding_gene</td>
      <td>1364640</td>
      <td>1369862</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>PF3D7_0831700</td>
      <td>NaN</td>
      <td>HSP70x</td>
    </tr>
    <tr>
      <th>27866</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>mRNA</td>
      <td>1364640</td>
      <td>1369862</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>PF3D7_0831700.1</td>
      <td>PF3D7_0831700</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>27867</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>1364640</td>
      <td>1367640</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>exon_PF3D7_0831700.1-E2</td>
      <td>PF3D7_0831700.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>27868</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>1368649</td>
      <td>1369862</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>exon_PF3D7_0831700.1-E1</td>
      <td>PF3D7_0831700.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>27869</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>1365467</td>
      <td>1367506</td>
      <td>NaN</td>
      <td>-</td>
      <td>0.0</td>
      <td>PF3D7_0831700.1-p1-CDS1</td>
      <td>PF3D7_0831700.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>27870</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>three_prime_UTR</td>
      <td>1364640</td>
      <td>1365466</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>utr_PF3D7_0831700.1_1</td>
      <td>PF3D7_0831700.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>27871</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>five_prime_UTR</td>
      <td>1367507</td>
      <td>1367640</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>utr_PF3D7_0831700.1_2</td>
      <td>PF3D7_0831700.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>27872</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>five_prime_UTR</td>
      <td>1368649</td>
      <td>1369862</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>utr_PF3D7_0831700.1_3</td>
      <td>PF3D7_0831700.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28984</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>pseudogene</td>
      <td>1371847</td>
      <td>1372720</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_0831750</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28985</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>pseudogenic_transcript</td>
      <td>1371847</td>
      <td>1372720</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_0831750.1</td>
      <td>PF3D7_0831750</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28986</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>1371847</td>
      <td>1372100</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_0831750.1-E1</td>
      <td>PF3D7_0831750.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28987</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>1372103</td>
      <td>1372223</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_0831750.1-E2</td>
      <td>PF3D7_0831750.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28988</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>1372225</td>
      <td>1372291</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_0831750.1-E3</td>
      <td>PF3D7_0831750.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28989</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>1372294</td>
      <td>1372577</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_0831750.1-E4</td>
      <td>PF3D7_0831750.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28990</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>1372579</td>
      <td>1372667</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_0831750.1-E5</td>
      <td>PF3D7_0831750.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28991</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>1372669</td>
      <td>1372720</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_0831750.1-E6</td>
      <td>PF3D7_0831750.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28992</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>1371847</td>
      <td>1372100</td>
      <td>NaN</td>
      <td>+</td>
      <td>0.0</td>
      <td>PF3D7_0831750.1-p1-CDS1</td>
      <td>PF3D7_0831750.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28993</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>1372103</td>
      <td>1372223</td>
      <td>NaN</td>
      <td>+</td>
      <td>1.0</td>
      <td>PF3D7_0831750.1-p1-CDS2</td>
      <td>PF3D7_0831750.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28994</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>1372225</td>
      <td>1372291</td>
      <td>NaN</td>
      <td>+</td>
      <td>0.0</td>
      <td>PF3D7_0831750.1-p1-CDS3</td>
      <td>PF3D7_0831750.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28995</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>1372294</td>
      <td>1372403</td>
      <td>NaN</td>
      <td>+</td>
      <td>2.0</td>
      <td>PF3D7_0831750.1-p1-CDS4</td>
      <td>PF3D7_0831750.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28996</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>three_prime_UTR</td>
      <td>1372404</td>
      <td>1372577</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>utr_PF3D7_0831750.1_1</td>
      <td>PF3D7_0831750.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28997</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>three_prime_UTR</td>
      <td>1372579</td>
      <td>1372667</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>utr_PF3D7_0831750.1_2</td>
      <td>PF3D7_0831750.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28998</th>
      <td>Pf3D7_08_v3</td>
      <td>VEuPathDB</td>
      <td>three_prime_UTR</td>
      <td>1372669</td>
      <td>1372720</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>utr_PF3D7_0831750.1_3</td>
      <td>PF3D7_0831750.1</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-d828440d-0631-4464-be1c-a8f5516001a6')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-d828440d-0631-4464-be1c-a8f5516001a6 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-d828440d-0631-4464-be1c-a8f5516001a6');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-cd9bfe1b-4e95-4d3f-9b4d-6ce24be2906c">
  <button class="colab-df-quickchart" onclick="quickchart('df-cd9bfe1b-4e95-4d3f-9b4d-6ce24be2906c')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-cd9bfe1b-4e95-4d3f-9b4d-6ce24be2906c button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>

    </div>
  </div>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Genes in hrp3 breakpoints</span>
<span class="n">pd</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">display</span><span class="o">.</span><span class="n">max_rows</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">df_gff</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;contig&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Pf3D7_13_v3&#39;</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">2845000</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">2795000</span> <span class="p">)</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-c71898c8-0f4a-4653-b7ee-a6e09cf72872" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>contig</th>
      <th>source</th>
      <th>type</th>
      <th>start</th>
      <th>end</th>
      <th>score</th>
      <th>strand</th>
      <th>phase</th>
      <th>ID</th>
      <th>Parent</th>
      <th>Name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>17986</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>pseudogene</td>
      <td>2811706</td>
      <td>2820270</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1371600</td>
      <td>NaN</td>
      <td>EBL1</td>
    </tr>
    <tr>
      <th>17987</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>pseudogenic_transcript</td>
      <td>2811706</td>
      <td>2820270</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1371600.1</td>
      <td>PF3D7_1371600</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17988</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2811706</td>
      <td>2812263</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1371600.1-E1</td>
      <td>PF3D7_1371600.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17989</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2812266</td>
      <td>2819628</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1371600.1-E2</td>
      <td>PF3D7_1371600.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17990</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2819764</td>
      <td>2819851</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1371600.1-E3</td>
      <td>PF3D7_1371600.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17991</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2820015</td>
      <td>2820088</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1371600.1-E4</td>
      <td>PF3D7_1371600.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17992</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2820227</td>
      <td>2820270</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1371600.1-E5</td>
      <td>PF3D7_1371600.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17993</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>2811706</td>
      <td>2812263</td>
      <td>NaN</td>
      <td>+</td>
      <td>0.0</td>
      <td>PF3D7_1371600.1-p1-CDS1</td>
      <td>PF3D7_1371600.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17994</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>2812266</td>
      <td>2815526</td>
      <td>NaN</td>
      <td>+</td>
      <td>0.0</td>
      <td>PF3D7_1371600.1-p1-CDS2</td>
      <td>PF3D7_1371600.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17995</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>three_prime_UTR</td>
      <td>2815527</td>
      <td>2819628</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>utr_PF3D7_1371600.1_1</td>
      <td>PF3D7_1371600.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17996</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>three_prime_UTR</td>
      <td>2819764</td>
      <td>2819851</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>utr_PF3D7_1371600.1_2</td>
      <td>PF3D7_1371600.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17997</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>three_prime_UTR</td>
      <td>2820015</td>
      <td>2820088</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>utr_PF3D7_1371600.1_3</td>
      <td>PF3D7_1371600.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17998</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>three_prime_UTR</td>
      <td>2820227</td>
      <td>2820270</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>utr_PF3D7_1371600.1_4</td>
      <td>PF3D7_1371600.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28226</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>protein_coding_gene</td>
      <td>2835756</td>
      <td>2839580</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1372100</td>
      <td>NaN</td>
      <td>GEXP04</td>
    </tr>
    <tr>
      <th>28227</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>mRNA</td>
      <td>2835756</td>
      <td>2839580</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1372100.1</td>
      <td>PF3D7_1372100</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28228</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2835756</td>
      <td>2837136</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1372100.1-E1</td>
      <td>PF3D7_1372100.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28229</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2837313</td>
      <td>2839580</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1372100.1-E2</td>
      <td>PF3D7_1372100.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28230</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>2837053</td>
      <td>2837136</td>
      <td>NaN</td>
      <td>+</td>
      <td>0.0</td>
      <td>PF3D7_1372100.1-p1-CDS1</td>
      <td>PF3D7_1372100.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28231</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>2837313</td>
      <td>2839058</td>
      <td>NaN</td>
      <td>+</td>
      <td>0.0</td>
      <td>PF3D7_1372100.1-p1-CDS2</td>
      <td>PF3D7_1372100.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28232</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>five_prime_UTR</td>
      <td>2835756</td>
      <td>2837052</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>utr_PF3D7_1372100.1_1</td>
      <td>PF3D7_1372100.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28233</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>three_prime_UTR</td>
      <td>2839059</td>
      <td>2839580</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>utr_PF3D7_1372100.1_2</td>
      <td>PF3D7_1372100.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>30629</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>ncRNA_gene</td>
      <td>2796119</td>
      <td>2797144</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1370800</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>30630</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>ncRNA</td>
      <td>2796119</td>
      <td>2797144</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1370800.1</td>
      <td>PF3D7_1370800</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>30631</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2796119</td>
      <td>2797144</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1370800.1-E1</td>
      <td>PF3D7_1370800.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>31628</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>ncRNA_gene</td>
      <td>2797507</td>
      <td>2798103</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1370900</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>31629</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>ncRNA</td>
      <td>2797507</td>
      <td>2798103</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1370900.1</td>
      <td>PF3D7_1370900</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>31630</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2797507</td>
      <td>2798103</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1370900.1-E1</td>
      <td>PF3D7_1370900.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>34729</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>protein_coding_gene</td>
      <td>2808200</td>
      <td>2810256</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>PF3D7_1371500</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>34730</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>mRNA</td>
      <td>2808200</td>
      <td>2810256</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>PF3D7_1371500.1</td>
      <td>PF3D7_1371500</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>34731</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2808200</td>
      <td>2809000</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>exon_PF3D7_1371500.1-E2</td>
      <td>PF3D7_1371500.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>34732</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2809154</td>
      <td>2810256</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>exon_PF3D7_1371500.1-E1</td>
      <td>PF3D7_1371500.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>34733</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>2808563</td>
      <td>2809000</td>
      <td>NaN</td>
      <td>-</td>
      <td>0.0</td>
      <td>PF3D7_1371500.1-p1-CDS2</td>
      <td>PF3D7_1371500.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>34734</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>2809154</td>
      <td>2809222</td>
      <td>NaN</td>
      <td>-</td>
      <td>0.0</td>
      <td>PF3D7_1371500.1-p1-CDS1</td>
      <td>PF3D7_1371500.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>34735</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>three_prime_UTR</td>
      <td>2808200</td>
      <td>2808562</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>utr_PF3D7_1371500.1_1</td>
      <td>PF3D7_1371500.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>34736</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>five_prime_UTR</td>
      <td>2809223</td>
      <td>2810256</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>utr_PF3D7_1371500.1_2</td>
      <td>PF3D7_1371500.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>35981</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>protein_coding_gene</td>
      <td>2840236</td>
      <td>2842840</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>PF3D7_1372200</td>
      <td>NaN</td>
      <td>HRPIII</td>
    </tr>
    <tr>
      <th>35982</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>mRNA</td>
      <td>2840236</td>
      <td>2842840</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>PF3D7_1372200.1</td>
      <td>PF3D7_1372200</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>35983</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2840236</td>
      <td>2841485</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>exon_PF3D7_1372200.1-E3</td>
      <td>PF3D7_1372200.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>35984</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2841635</td>
      <td>2841716</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>exon_PF3D7_1372200.1-E2</td>
      <td>PF3D7_1372200.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>35985</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2842024</td>
      <td>2842840</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>exon_PF3D7_1372200.1-E1</td>
      <td>PF3D7_1372200.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>35986</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>2840727</td>
      <td>2841485</td>
      <td>NaN</td>
      <td>-</td>
      <td>0.0</td>
      <td>PF3D7_1372200.1-p1-CDS2</td>
      <td>PF3D7_1372200.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>35987</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>2841635</td>
      <td>2841703</td>
      <td>NaN</td>
      <td>-</td>
      <td>0.0</td>
      <td>PF3D7_1372200.1-p1-CDS1</td>
      <td>PF3D7_1372200.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>35988</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>three_prime_UTR</td>
      <td>2840236</td>
      <td>2840726</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>utr_PF3D7_1372200.1_1</td>
      <td>PF3D7_1372200.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>35989</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>five_prime_UTR</td>
      <td>2841704</td>
      <td>2841716</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>utr_PF3D7_1372200.1_2</td>
      <td>PF3D7_1372200.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>35990</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>five_prime_UTR</td>
      <td>2842024</td>
      <td>2842840</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>utr_PF3D7_1372200.1_3</td>
      <td>PF3D7_1372200.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>36301</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>ncRNA_gene</td>
      <td>2802945</td>
      <td>2807159</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1371300</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>36302</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>rRNA</td>
      <td>2802945</td>
      <td>2807159</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1371300.1</td>
      <td>PF3D7_1371300</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>36303</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2802945</td>
      <td>2807159</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1371300.1-E1</td>
      <td>PF3D7_1371300.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>37550</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>protein_coding_gene</td>
      <td>2823251</td>
      <td>2825967</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>PF3D7_1371800</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>37551</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>mRNA</td>
      <td>2823251</td>
      <td>2825967</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>PF3D7_1371800.1</td>
      <td>PF3D7_1371800</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>37552</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2823251</td>
      <td>2825552</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>exon_PF3D7_1371800.1-E2</td>
      <td>PF3D7_1371800.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>37553</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2825781</td>
      <td>2825967</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>exon_PF3D7_1371800.1-E1</td>
      <td>PF3D7_1371800.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>37554</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>2824302</td>
      <td>2825552</td>
      <td>NaN</td>
      <td>-</td>
      <td>0.0</td>
      <td>PF3D7_1371800.1-p1-CDS2</td>
      <td>PF3D7_1371800.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>37555</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>2825781</td>
      <td>2825852</td>
      <td>NaN</td>
      <td>-</td>
      <td>0.0</td>
      <td>PF3D7_1371800.1-p1-CDS1</td>
      <td>PF3D7_1371800.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>37556</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>three_prime_UTR</td>
      <td>2823251</td>
      <td>2824301</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>utr_PF3D7_1371800.1_1</td>
      <td>PF3D7_1371800.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>37557</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>five_prime_UTR</td>
      <td>2825853</td>
      <td>2825967</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>utr_PF3D7_1371800.1_2</td>
      <td>PF3D7_1371800.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>37701</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>protein_coding_gene</td>
      <td>2832623</td>
      <td>2835439</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1372000</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>37702</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>mRNA</td>
      <td>2832623</td>
      <td>2835439</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1372000.1</td>
      <td>PF3D7_1372000</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>37703</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2832623</td>
      <td>2833086</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1372000.1-E1</td>
      <td>PF3D7_1372000.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>37704</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2833204</td>
      <td>2835439</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1372000.1-E2</td>
      <td>PF3D7_1372000.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>37705</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>2832952</td>
      <td>2833086</td>
      <td>NaN</td>
      <td>+</td>
      <td>0.0</td>
      <td>PF3D7_1372000.1-p1-CDS1</td>
      <td>PF3D7_1372000.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>37706</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>2833204</td>
      <td>2834322</td>
      <td>NaN</td>
      <td>+</td>
      <td>0.0</td>
      <td>PF3D7_1372000.1-p1-CDS2</td>
      <td>PF3D7_1372000.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>37707</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>five_prime_UTR</td>
      <td>2832623</td>
      <td>2832951</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>utr_PF3D7_1372000.1_1</td>
      <td>PF3D7_1372000.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>37708</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>three_prime_UTR</td>
      <td>2834323</td>
      <td>2835439</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>utr_PF3D7_1372000.1_2</td>
      <td>PF3D7_1372000.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>41004</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>protein_coding_gene</td>
      <td>2821078</td>
      <td>2824446</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1371700</td>
      <td>NaN</td>
      <td>FIKK13</td>
    </tr>
    <tr>
      <th>41005</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>mRNA</td>
      <td>2821078</td>
      <td>2824446</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1371700.1</td>
      <td>PF3D7_1371700</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>41006</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2821078</td>
      <td>2821173</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1371700.1-E1</td>
      <td>PF3D7_1371700.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>41007</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2821278</td>
      <td>2822786</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1371700.1-E2</td>
      <td>PF3D7_1371700.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>41008</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2823212</td>
      <td>2824446</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1371700.1-E3</td>
      <td>PF3D7_1371700.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>41009</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>2821078</td>
      <td>2821173</td>
      <td>NaN</td>
      <td>+</td>
      <td>0.0</td>
      <td>PF3D7_1371700.1-p1-CDS1</td>
      <td>PF3D7_1371700.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>41010</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>2821278</td>
      <td>2822786</td>
      <td>NaN</td>
      <td>+</td>
      <td>0.0</td>
      <td>PF3D7_1371700.1-p1-CDS2</td>
      <td>PF3D7_1371700.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>41011</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>2823212</td>
      <td>2823292</td>
      <td>NaN</td>
      <td>+</td>
      <td>0.0</td>
      <td>PF3D7_1371700.1-p1-CDS3</td>
      <td>PF3D7_1371700.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>41012</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>three_prime_UTR</td>
      <td>2823293</td>
      <td>2824446</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>utr_PF3D7_1371700.1_1</td>
      <td>PF3D7_1371700.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>41067</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>ncRNA_gene</td>
      <td>2800004</td>
      <td>2802154</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1371000</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>41068</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>rRNA</td>
      <td>2800004</td>
      <td>2802154</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1371000.1</td>
      <td>PF3D7_1371000</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>41069</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2800004</td>
      <td>2802154</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1371000.1-E1</td>
      <td>PF3D7_1371000.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>42330</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>protein_coding_gene</td>
      <td>2843157</td>
      <td>2847557</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1372300</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>42331</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>mRNA</td>
      <td>2843157</td>
      <td>2847557</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1372300.1</td>
      <td>PF3D7_1372300</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>42332</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2843157</td>
      <td>2845850</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1372300.1-E1</td>
      <td>PF3D7_1372300.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>42336</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>five_prime_UTR</td>
      <td>2843157</td>
      <td>2845766</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>utr_PF3D7_1372300.1_1</td>
      <td>PF3D7_1372300.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>44509</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>protein_coding_gene</td>
      <td>2829530</td>
      <td>2830830</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1371900</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>44510</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>mRNA</td>
      <td>2829530</td>
      <td>2830830</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1371900.1</td>
      <td>PF3D7_1371900</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>44511</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2829530</td>
      <td>2829927</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1371900.1-E1</td>
      <td>PF3D7_1371900.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>44512</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2830109</td>
      <td>2830830</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1371900.1-E2</td>
      <td>PF3D7_1371900.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>44513</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>2829856</td>
      <td>2829927</td>
      <td>NaN</td>
      <td>+</td>
      <td>0.0</td>
      <td>PF3D7_1371900.1-p1-CDS1</td>
      <td>PF3D7_1371900.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>44514</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>CDS</td>
      <td>2830109</td>
      <td>2830669</td>
      <td>NaN</td>
      <td>+</td>
      <td>0.0</td>
      <td>PF3D7_1371900.1-p1-CDS2</td>
      <td>PF3D7_1371900.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>44515</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>five_prime_UTR</td>
      <td>2829530</td>
      <td>2829855</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>utr_PF3D7_1371900.1_1</td>
      <td>PF3D7_1371900.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>44516</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>three_prime_UTR</td>
      <td>2830670</td>
      <td>2830830</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>utr_PF3D7_1371900.1_2</td>
      <td>PF3D7_1371900.1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>46143</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>ncRNA_gene</td>
      <td>2802527</td>
      <td>2802686</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1371200</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>46144</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>rRNA</td>
      <td>2802527</td>
      <td>2802686</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>PF3D7_1371200.1</td>
      <td>PF3D7_1371200</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>46145</th>
      <td>Pf3D7_13_v3</td>
      <td>VEuPathDB</td>
      <td>exon</td>
      <td>2802527</td>
      <td>2802686</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>exon_PF3D7_1371200.1-E1</td>
      <td>PF3D7_1371200.1</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">

  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-c71898c8-0f4a-4653-b7ee-a6e09cf72872')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>

  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

    <script>
      const buttonEl =
        document.querySelector('#df-c71898c8-0f4a-4653-b7ee-a6e09cf72872 button.colab-df-convert');
      buttonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';

      async function convertToInteractive(key) {
        const element = document.querySelector('#df-c71898c8-0f4a-4653-b7ee-a6e09cf72872');
        const dataTable =
          await google.colab.kernel.invokeFunction('convertToInteractive',
                                                    [key], {});
        if (!dataTable) return;

        const docLinkHtml = 'Like what you see? Visit the ' +
          '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
          + ' to learn more about interactive tables.';
        element.innerHTML = '';
        dataTable['output_type'] = 'display_data';
        await google.colab.output.renderOutput(dataTable, element);
        const docLink = document.createElement('div');
        docLink.innerHTML = docLinkHtml;
        element.appendChild(docLink);
      }
    </script>
  </div>


<div id="df-5f8cbaa4-3e16-4c93-97cf-12a6f80ce8df">
  <button class="colab-df-quickchart" onclick="quickchart('df-5f8cbaa4-3e16-4c93-97cf-12a6f80ce8df')"
            title="Suggest charts"
            style="display:none;">

<svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
     width="24px">
    <g>
        <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"/>
    </g>
</svg>
  </button>

<style>
  .colab-df-quickchart {
      --bg-color: #E8F0FE;
      --fill-color: #1967D2;
      --hover-bg-color: #E2EBFA;
      --hover-fill-color: #174EA6;
      --disabled-fill-color: #AAA;
      --disabled-bg-color: #DDD;
  }

  [theme=dark] .colab-df-quickchart {
      --bg-color: #3B4455;
      --fill-color: #D2E3FC;
      --hover-bg-color: #434B5C;
      --hover-fill-color: #FFFFFF;
      --disabled-bg-color: #3B4455;
      --disabled-fill-color: #666;
  }

  .colab-df-quickchart {
    background-color: var(--bg-color);
    border: none;
    border-radius: 50%;
    cursor: pointer;
    display: none;
    fill: var(--fill-color);
    height: 32px;
    padding: 0;
    width: 32px;
  }

  .colab-df-quickchart:hover {
    background-color: var(--hover-bg-color);
    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    fill: var(--button-hover-fill-color);
  }

  .colab-df-quickchart-complete:disabled,
  .colab-df-quickchart-complete:disabled:hover {
    background-color: var(--disabled-bg-color);
    fill: var(--disabled-fill-color);
    box-shadow: none;
  }

  .colab-df-spinner {
    border: 2px solid var(--fill-color);
    border-color: transparent;
    border-bottom-color: var(--fill-color);
    animation:
      spin 1s steps(1) infinite;
  }

  @keyframes spin {
    0% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
      border-left-color: var(--fill-color);
    }
    20% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    30% {
      border-color: transparent;
      border-left-color: var(--fill-color);
      border-top-color: var(--fill-color);
      border-right-color: var(--fill-color);
    }
    40% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-top-color: var(--fill-color);
    }
    60% {
      border-color: transparent;
      border-right-color: var(--fill-color);
    }
    80% {
      border-color: transparent;
      border-right-color: var(--fill-color);
      border-bottom-color: var(--fill-color);
    }
    90% {
      border-color: transparent;
      border-bottom-color: var(--fill-color);
    }
  }
</style>

  <script>
    async function quickchart(key) {
      const quickchartButtonEl =
        document.querySelector('#' + key + ' button');
      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.
      quickchartButtonEl.classList.add('colab-df-spinner');
      try {
        const charts = await google.colab.kernel.invokeFunction(
            'suggestCharts', [key], {});
      } catch (error) {
        console.error('Error during call to suggestCharts:', error);
      }
      quickchartButtonEl.classList.remove('colab-df-spinner');
      quickchartButtonEl.classList.add('colab-df-quickchart-complete');
    }
    (() => {
      let quickchartButtonEl =
        document.querySelector('#df-5f8cbaa4-3e16-4c93-97cf-12a6f80ce8df button');
      quickchartButtonEl.style.display =
        google.colab.kernel.accessAllowed ? 'block' : 'none';
    })();
  </script>
</div>

    </div>
  </div>
</div></div>
</div>
</div>
<div class="section" id="create-figure">
<h2>Create Figure<a class="headerlink" href="#create-figure" title="Permalink to this headline">¶</a></h2>
<p>This intricate figure serves to map deletion breakpoints in <em>hrp2</em> and <em>hrp3</em> across various countries. The x-axis of the figure displays the deletion breakpoints, while the y-axis shows the countries along with the number of breakpoints they exhibit.</p>
<p>The figure comprises two sections: <em>hrp2</em> (positioned at the top, subplots 1-5) and <em>hrp3</em> (located at the bottom, subplots 6-12), consisting of a total of 12 subplots.</p>
<p>We start by defining distinct colour codes for each genomic region that we will annotate.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a dictionary with distinct colour codes</span>
<span class="n">figure_colours</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">()</span>
<span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_8_13&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;#377eb8&#39;</span>
<span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_11&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;#984ea3&#39;</span>
<span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_5&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;#ff7f00&#39;</span>
<span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;similar_sequence&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;#ffff33&#39;</span>
<span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;hrp_genes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;#e41a1c&#39;</span>
<span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;rrna_genes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;#4daf4a&#39;</span>
<span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;other_genes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;black&#39;</span>
<span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;pseudogenes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;grey&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Full figure</span>

<span class="c1"># The figure will consist of 12 subplots, each with varying sizes.</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">13</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="n">gridspec_kw</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;height_ratios&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">12</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">60</span><span class="p">,</span> <span class="mi">24</span><span class="p">,</span> <span class="mi">12</span><span class="p">,</span> <span class="mi">12</span><span class="p">],</span> <span class="s1">&#39;hspace&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">})</span>

<span class="c1">### HRP2 (upper panel)</span>

<span class="c1"># Set the minimum and maximum positions for hrp2</span>

<span class="n">min_pos</span> <span class="o">=</span> <span class="mi">1371500</span>
<span class="n">max_pos</span> <span class="o">=</span> <span class="mi">1375500</span>

<span class="c1">## Subplot 1: Chromosome Title and Region</span>

<span class="c1"># Set the title</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Pf3D7_08_v3&#39;</span><span class="p">)</span>
<span class="c1"># Don&#39;t display other properties here</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_08_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_08_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">]))</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_facecolor</span><span class="p">(</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_8_13&#39;</span><span class="p">])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">axvspan</span><span class="p">(</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>

<span class="c1">## Subplot 2: Chromosome Boundaries</span>

<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_08_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_08_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">]))</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_08_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">]),</span> <span class="n">min_pos</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_08_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">]),</span> <span class="n">max_pos</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;left&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># Subplot 3: Deletion Breakpoints by Country</span>

<span class="c1"># Create lists for y-axis labels and positions</span>
<span class="n">ylabels</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">yposes</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">ypos</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df_hrp2</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="n">yposes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ypos</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]),</span> <span class="p">(</span><span class="n">ypos</span><span class="p">,</span> <span class="n">ypos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_8_13&#39;</span><span class="p">])</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="mi">50</span><span class="p">,</span> <span class="n">ypos</span> <span class="p">,</span> <span class="s1">&#39;$(GGGTT[T/C]A)_n$&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">bbox</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">boxstyle</span><span class="o">=</span><span class="s1">&#39;round,pad=0.3&#39;</span><span class="p">))</span>
    <span class="n">ylabels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Countries&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">yposes</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">ylabels</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">yposes</span><span class="p">)</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">yposes</span><span class="p">)</span><span class="o">+</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>

<span class="c1">## Subplot 4: Genes on the x-axis</span>

<span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

<span class="k">for</span> <span class="n">ix</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df_gff</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;contig&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Pf3D7_08_v3&#39;</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_pos</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_pos</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;CDS&#39;</span> <span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">1373212</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">1376988</span><span class="p">:</span>
        <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;hrp_genes&#39;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;other_genes&#39;</span><span class="p">]</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]),</span> <span class="p">(</span><span class="mf">0.75</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">)</span>

<span class="k">for</span> <span class="n">ix</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df_gff</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;contig&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Pf3D7_08_v3&#39;</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_pos</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_pos</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;pseudogene&#39;</span> <span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]),</span> <span class="p">(</span><span class="mf">0.75</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;pseudogenes&#39;</span><span class="p">])</span>

<span class="c1"># Add gene names for hrp2 panel</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span> <span class="mi">1375299</span> <span class="o">+</span> <span class="mi">1374236</span> <span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="s1">&#39;$</span><span class="si">{hrp2}</span><span class="s1">$&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span> <span class="mi">1371847</span> <span class="o">+</span> <span class="mi">1372720</span> <span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="s1">&#39;PF3D7_0831750 (pseudogene)&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Genes&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>
<span class="c1"># Add a connection line for the gene</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="mi">1_375_084</span><span class="p">,</span> <span class="mi">1_375_231</span><span class="p">),</span> <span class="p">(</span><span class="mf">0.75</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">),</span> <span class="n">c</span> <span class="o">=</span> <span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;hrp_genes&#39;</span><span class="p">])</span>

<span class="c1">## Subplot 5: Ticks for Genomic Coordinates</span>
<span class="c1"># Set the x-axis limits to cover the specified genomic region between &#39;min_pos&#39; and &#39;max_pos&#39;</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">)</span>
<span class="c1"># Hide the left and right spines to create a cleaner appearance</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;left&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="c1"># Configure the x-axis ticks at specific positions</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([</span><span class="mi">1372000</span><span class="p">,</span> <span class="mi">1373000</span><span class="p">,</span> <span class="mi">1374000</span><span class="p">,</span> <span class="mi">1375000</span><span class="p">])</span>
<span class="c1"># Label the x-axis ticks with corresponding values</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">([</span><span class="s2">&quot;1.372 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;1.373 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;1.374 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;1.375 Mbp&quot;</span><span class="p">])</span>
<span class="c1"># Remove y-axis ticks to maintain a clean look</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="c1"># Position x-axis ticks at the top</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">tick_top</span><span class="p">()</span>
<span class="c1"># Adjust the direction of x-axis ticks and add padding</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">direction</span><span class="o">=</span><span class="s2">&quot;in&quot;</span><span class="p">,</span> <span class="n">pad</span><span class="o">=-</span><span class="mi">20</span><span class="p">)</span>

<span class="c1">### HRP3</span>

<span class="c1"># Set the minimum and maximum positions for hrp3</span>
<span class="n">min_pos</span> <span class="o">=</span> <span class="mi">2797000</span>
<span class="n">max_pos</span> <span class="o">=</span> <span class="mi">2845000</span>

<span class="c1">## Subplot 6: Chromosome Name as Title and</span>
<span class="c1"># Set the title</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Pf3D7_13_v3&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_13_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_13_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">]))</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">set_facecolor</span><span class="p">(</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_8_13&#39;</span><span class="p">])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">axvspan</span><span class="p">(</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>

<span class="c1">## Subplot 7: Chromosome Boundaries</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="c1"># Set the x-axis limits to cover the specified genomic region between the start and end positions of &#39;Pf3D7_13_v3&#39;</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_13_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_13_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">]))</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="c1"># y-axis will only have one annotation</span>
<span class="c1"># Create lines to mark the start and end positions of the chromosome region with black color</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_13_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">]),</span> <span class="n">min_pos</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">df_chroms</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Pf3D7_13_v3&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">]),</span> <span class="n">max_pos</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
<span class="c1"># Hide the left and right spines to create a cleaner appearance</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;left&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>


<span class="c1">## Subplot 8 and 9: Deletion Breakpoints by Country</span>

<span class="c1"># Initialize lists for labels and positions</span>
<span class="n">ylabels_7</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">yposes_7</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">ylabels_8</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">yposes_8</span> <span class="o">=</span> <span class="p">[]</span>

<span class="c1"># How many countries have Chromosome 11 recombination breakpoints?</span>
<span class="n">multicountry_label</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">df_hrp3</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Countries&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">))</span><span class="si">}</span><span class="s1"> countries (</span><span class="si">{</span><span class="n">df_hrp3</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Samples with deletion&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">)&#39;</span>

<span class="c1"># Iterate through df_hrp3 sorted by breakpoints</span>
<span class="k">for</span> <span class="n">ypos</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df_hrp3</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>

    <span class="c1"># Check the deletion type and apply different plotting styles accordingly</span>
    <span class="c1"># Telomere healing</span>
    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Deletion type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Telomere healing&#39;</span><span class="p">:</span>
        <span class="nb">breakpoint</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">])</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">min_pos</span><span class="p">,</span> <span class="nb">breakpoint</span><span class="p">),</span> <span class="p">(</span><span class="n">ypos</span><span class="p">,</span> <span class="n">ypos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_8_13&#39;</span><span class="p">])</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="nb">breakpoint</span> <span class="o">+</span> <span class="mi">600</span><span class="p">,</span> <span class="n">ypos</span> <span class="p">,</span> <span class="s1">&#39;$(GGGTT[T/C]A)_n$&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">bbox</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">boxstyle</span><span class="o">=</span><span class="s1">&#39;round,pad=0.3&#39;</span><span class="p">))</span>
        <span class="n">ylabels_8</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Countries&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">yposes_8</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ypos</span><span class="p">)</span>

    <span class="c1"># Chromosome 5 recombination</span>
    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Deletion type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Chromosome 5 recombination&#39;</span><span class="p">:</span>
        <span class="n">breakpoint_start</span><span class="p">,</span> <span class="n">breakpoint_end</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)]</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">breakpoint_start</span><span class="p">),</span> <span class="p">(</span><span class="n">ypos</span><span class="p">,</span> <span class="n">ypos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_8_13&#39;</span><span class="p">])</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">breakpoint_end</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">),</span> <span class="p">(</span><span class="n">ypos</span><span class="p">,</span> <span class="n">ypos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_5&#39;</span><span class="p">])</span>
        <span class="n">ylabels_7</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Countries&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">yposes_7</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ypos</span><span class="p">)</span>

    <span class="c1"># Chromosome 11 recombination</span>
    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Deletion type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Chromosome 11 recombination&#39;</span><span class="p">:</span>
        <span class="n">breakpoint_start</span><span class="p">,</span> <span class="n">breakpoint_end</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;breakpoint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)]</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">breakpoint_start</span><span class="p">),</span> <span class="p">(</span><span class="n">ypos</span><span class="p">,</span> <span class="n">ypos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_8_13&#39;</span><span class="p">])</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">breakpoint_end</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">),</span> <span class="p">(</span><span class="n">ypos</span><span class="p">,</span> <span class="n">ypos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_11&#39;</span><span class="p">])</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">breakpoint_start</span><span class="p">,</span> <span class="n">breakpoint_end</span><span class="p">),</span> <span class="p">(</span><span class="n">ypos</span><span class="p">,</span> <span class="n">ypos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;similar_sequence&#39;</span><span class="p">])</span>
        <span class="n">yposes_7</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ypos</span><span class="p">)</span>
        <span class="n">ylabels_7</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">multicountry_label</span><span class="p">)</span>

<span class="c1"># Add a line to seperate Chrom 5 and 11, this may need to be adjusted manually in the future</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="nb">max</span><span class="p">(</span><span class="n">yposes_7</span><span class="p">)</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="c1"># Set the ticks and labels for Subplot 8</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">yposes_7</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">ylabels_7</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">yposes_7</span><span class="p">)</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">yposes_7</span><span class="p">)</span><span class="o">+</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>

<span class="c1"># Set the ticks and labels for Subplot 9</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">yposes_8</span><span class="p">)</span>  <span class="c1"># Ensure yposes_8 has the same length as ylabels_8</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">ylabels_8</span><span class="p">)</span>  <span class="c1"># Ensure the number of labels matches the number of ticks</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">yposes_8</span><span class="p">)</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">yposes_8</span><span class="p">)</span><span class="o">+</span><span class="mf">0.5</span><span class="p">)</span>  <span class="c1"># Use yposes_8 for ylim</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>

<span class="c1">## Subplot 10: Gene annotations on the x-axis</span>

<span class="n">bar_pos</span> <span class="o">=</span> <span class="mf">0.9</span>
<span class="n">text_pos</span> <span class="o">=</span> <span class="mf">0.8</span>

<span class="c1"># Set the x-axis and y-axis limits</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

<span class="c1"># Group CDS regions and draw connection lines by using proximity of breakpoints.</span>
<span class="c1"># Previously we were using polypeptide type to draw connection lines, but this does not exist in the new GFF</span>
<span class="n">prev_end</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">connections</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">ix</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df_gff</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;contig&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Pf3D7_13_v3&#39;</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_pos</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_pos</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;CDS&#39;</span> <span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">2840727</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">2841703</span><span class="p">:</span>
        <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;hrp_genes&#39;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;other_genes&#39;</span><span class="p">]</span>
    <span class="c1">#print(row.id)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]),</span> <span class="p">(</span><span class="n">bar_pos</span><span class="p">,</span> <span class="n">bar_pos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">prev_end</span> <span class="ow">and</span> <span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">prev_end</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">500</span> <span class="ow">and</span> <span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">prev_end</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">connections</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">prev_end</span><span class="p">,</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">],</span> <span class="n">color</span><span class="p">))</span>
    <span class="n">prev_end</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span>

<span class="c1"># Plot connection lines between breakpoints that are in same CDS region and within 500 base-diameter</span>
<span class="c1"># These lines will be thinner, so lw=2</span>
<span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">color</span> <span class="ow">in</span> <span class="n">connections</span><span class="p">:</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">),</span> <span class="p">(</span><span class="n">bar_pos</span><span class="p">,</span> <span class="n">bar_pos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">)</span>

<span class="c1"># Plot rRNA annotations</span>
<span class="k">for</span> <span class="n">ix</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df_gff</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;contig&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Pf3D7_13_v3&#39;</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_pos</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_pos</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;rRNA&#39;</span> <span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]),</span> <span class="p">(</span><span class="n">bar_pos</span><span class="p">,</span> <span class="n">bar_pos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;rrna_genes&#39;</span><span class="p">])</span>

<span class="c1"># Plot pseudogene annotations</span>
<span class="k">for</span> <span class="n">ix</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df_gff</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;contig&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Pf3D7_13_v3&#39;</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_pos</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_pos</span> <span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span> <span class="n">df_gff</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;pseudogene&#39;</span> <span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]),</span> <span class="p">(</span><span class="n">bar_pos</span><span class="p">,</span> <span class="n">bar_pos</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;pseudogenes&#39;</span><span class="p">])</span>

<span class="c1"># Annotate specific gene positions with labels</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span> <span class="mi">2800004</span> <span class="o">+</span> <span class="mi">2802154</span> <span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;18S rRNA&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span> <span class="mi">2802527</span> <span class="o">+</span> <span class="mi">2802686</span> <span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;5.8S rRNA&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span> <span class="mi">2802945</span> <span class="o">+</span> <span class="mi">2807159</span> <span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;28S rRNA&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span> <span class="mi">2808200</span> <span class="o">+</span> <span class="mi">2809700</span> <span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;PF3D7_1371500&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span> <span class="mi">2811706</span> <span class="o">+</span> <span class="mi">2820270</span> <span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;$</span><span class="si">{ebl1}</span><span class="s1">$</span><span class="se">\n</span><span class="s1">(pseudogene)&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span> <span class="mi">2821078</span> <span class="o">+</span> <span class="mi">2823292</span> <span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;$</span><span class="si">{fikk13}</span><span class="s1">$&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span> <span class="mi">2824302</span> <span class="o">+</span> <span class="mi">2825852</span> <span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;PF3D7_1371800&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span> <span class="mi">2829856</span> <span class="o">+</span> <span class="mi">2830669</span> <span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;PF3D7_1371900&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span> <span class="mi">2832952</span> <span class="o">+</span> <span class="mi">2834322</span> <span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;PF3D7_1372000&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span> <span class="mi">2837053</span> <span class="o">+</span> <span class="mi">2839058</span> <span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;$</span><span class="si">{gexp04}</span><span class="s1">$&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span> <span class="mi">2840727</span> <span class="o">+</span> <span class="mi">2841703</span> <span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">text_pos</span><span class="p">,</span> <span class="s1">&#39;$</span><span class="si">{hrp3}</span><span class="s1">$&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="c1"># Set y-axis ticks, customize x-axis labels, and add x-axis label and y-axis label</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Genomic coordinates&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Genes&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>

<span class="c1">## Subplot 11: Ticks (x-axis) for Genomic Coordinates</span>
<span class="c1"># Hide left, right, and bottom spines</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_pos</span><span class="p">,</span> <span class="n">max_pos</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;left&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([</span><span class="mi">2800000</span><span class="p">,</span> <span class="mi">2810000</span><span class="p">,</span> <span class="mi">2820000</span><span class="p">,</span> <span class="mi">2830000</span><span class="p">,</span> <span class="mi">2840000</span><span class="p">])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">([</span><span class="s2">&quot;2.80 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;2.81 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;2.82 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;2.83 Mbp&quot;</span><span class="p">,</span> <span class="s2">&quot;2.84 Mbp&quot;</span><span class="p">])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">tick_top</span><span class="p">()</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">direction</span><span class="o">=</span><span class="s2">&quot;in&quot;</span><span class="p">,</span> <span class="n">pad</span><span class="o">=-</span><span class="mi">20</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="n">min_pos</span> <span class="o">+</span> <span class="n">max_pos</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">,</span> <span class="s1">&#39;Genomic coordinates&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>

<span class="c1">## Subplot 12: Legend</span>
<span class="c1"># Hide left, right, and bottom spines</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;left&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;bottom&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="c1"># Hide both x and y-axis ticks</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="c1"># Create legend elements with color bars and labels</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">),</span> <span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_5&#39;</span><span class="p">])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.12</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s1">&#39;Chr 5 sequence&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">),</span> <span class="p">(</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;chr_11&#39;</span><span class="p">])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.12</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="s1">&#39;Chr 11 sequence&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="mf">0.38</span><span class="p">,</span> <span class="mf">0.48</span><span class="p">),</span> <span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">solid_capstyle</span><span class="o">=</span><span class="s1">&#39;butt&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">figure_colours</span><span class="p">[</span><span class="s1">&#39;similar_sequence&#39;</span><span class="p">])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.50</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s1">&#39;Highly similar sequence on chr 13 and chr 11&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">.405</span><span class="p">,</span> <span class="mf">.206</span><span class="p">,</span> <span class="s1">&#39;$(GGGTT[T/C]A)_n$&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">bbox</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">boxstyle</span><span class="o">=</span><span class="s1">&#39;round,pad=0.3&#39;</span><span class="p">))</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.50</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="s1">&#39;Telomeric repeat sequence&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>
<span class="c1"># Set the limits for this subplot</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

<span class="c1"># Ensure the figure layout is tidy</span>
<span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/Figure_HRPDeletion_37_0.png" src="../../_images/Figure_HRPDeletion_37_0.png" />
</div>
</div>
<p><strong>Figure Legend: <em>hrp2</em> and <em>hrp3</em> deletion breakpoints.</strong> We see five different breakpoints resulting in the deletion of <em>hrp2</em>. Four of these are within exon 2 of the gene whereas the fifth is found between <em>hrp2</em> and the pseudogene PF3D7_0831750. For all five events we see evidence of telomeric healing from reads that contain part Pf3D7_08_v3 sequence and part telomeric repeat sequence (GGGTTCA/GGGTTTA). We see 17 different breakpoints resulting in the deletion of hrp3. For 15 of these we see evidence of telomeric healing. Note that many of these events result in the deletion of other genes in addition to <em>hrp3</em>. For twenty samples from Cambodia and a single sample from Vietnam we see evidence of a recombination with chromosome 5 which results in a hybrid chromosome comprising mostly chromosome 13 sequence but a small inverted section of an internal portion of chromosome 5 containing the gene <em>mdr1</em>. We also see evidence of a recombination with chromosome 11 which results in a hybrid chromosome comprising mostly chromosome 13 sequence but also a section of the 3’ end of chromosome 11. This is the most common deletion type, being seen in 162 samples from 15 different countries. Because the recombination occurs between highly similar sequences of a set of three orthologous ribosomal RNA genes found on both chromosomes, it is not possible to identify the exact breakpoints.</p>
<div class="section" id="save-figure">
<h3>Save Figure<a class="headerlink" href="#save-figure" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># You will need to authorise Google Colab access to Google Drive</span>
<span class="n">drive</span><span class="o">.</span><span class="n">mount</span><span class="p">(</span><span class="s1">&#39;/content/drive&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Mounted at /content/drive
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># This will send the file to your Google Drive, where you can download it from if needed</span>
<span class="c1"># Change the file path if you wish to send the file to a specific location</span>
<span class="c1"># Change the file name if you wish to call it something else</span>

<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;/content/drive/My Drive/HRP_Deletions_Figure.pdf&#39;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;/content/drive/My Drive/HRP_Deletions_Figure.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">480</span><span class="p">)</span> <span class="c1"># increase the dpi for higher resolution</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./pf8/evolution_malaria_control"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="cnv_fig.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Plot copy number variation (CNV) trends</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="Drug_Resistance_Pf7_Pf8_Concordance.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Drug resistance in Pf7 vs Pf8</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By MalariaGEN<br/>
    
        &copy; Copyright 2023.<br/>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>